{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import time\n",
    "import os\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_remove_null_nan_empty_entries\n",
    "from utils.column_names_reader import get_col_names_from_gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging the downloaded dataset\n",
    "def clean_df(data_filepath):\n",
    "    df = pd.read_csv(data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "    df = df.drop_duplicates(subset=\"snp_id\", keep=\"first\")\n",
    "    return df\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_0.txt\"\n",
    "df_0 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_1.txt\"\n",
    "df_1 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_2.txt\"\n",
    "df_2 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_3.txt\"\n",
    "df_3 = clean_df(data_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5140350, 14)\n",
      "Number of NAN rows removed: 151569\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "(4988781, 14)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# removing duplicate data points and variants with no protein variants\n",
    "df = pd.concat([df_0, df_1, df_2, df_3], ignore_index=True)\n",
    "df = df.drop_duplicates(subset=\"snp_id\", keep=\"first\")\n",
    "print(df.shape)\n",
    "df = filter_remove_null_nan_empty_entries(df, a_col_name=\"variations\") # NNE: null, nan or empty variants\n",
    "print(df.shape)\n",
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4988781\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "genes                                                                  \n",
       "TTN:7273,TTN-AS1:100506866                                                 9381\n",
       "MUC16:94025                                                                8341\n",
       "MUC4:4585                                                                  5782\n",
       "TTN:7273                                                                   5151\n",
       "AHNAK2:113146                                                              4726\n",
       "                                                                           ... \n",
       "CASP5:838,LOC124902742:124902742                                              1\n",
       "SLX1B:79008,BOLA2:552900,SLX1B-SULT1A4:100526831,BOLA2-SMG1P6:107282092       1\n",
       "RASSF7:8045,LMNTD2:256329,LMNTD2-AS1:692247                                   1\n",
       "CASP4:837,LOC124902813:124902813                                              1\n",
       "EDA:1896,MIR676:100500887                                                     1\n",
       "Length: 27296, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parsing gene-symbols\n",
    "initially_cleaned_df = df.copy()\n",
    "initially_cleaned_df[\"snp_id\"] = initially_cleaned_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n",
    "dbsnp_ids = initially_cleaned_df[\"snp_id\"].unique().tolist()\n",
    "print(len(dbsnp_ids))\n",
    "initially_cleaned_df[[\"genes\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NP_116165.1:p.Glu2535Asp,NP_116165.1:p.Glu2535Asp,NP_001305082.1:p.Glu2247Asp,NP_001305082.1:p.Glu2247Asp,NP_001269877.1:p.Glu2353Asp,NP_001269877.1:p.Glu2353Asp,NP_001309181.1:p.Glu2497Asp,NP_001309181.1:p.Glu2497Asp,NP_001309183.1:p.Glu2316Asp,NP_001309183.1:p.Glu2316Asp,NP_001305083.1:p.Glu2353Asp,NP_001305083.1:p.Glu2353Asp,NP_001309187.1:p.Glu2316Asp,NP_001309187.1:p.Glu2316Asp'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "initially_cleaned_df[initially_cleaned_df[\"snp_id\"]==\"rs1935\"][\"variations\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>genes</th>\n",
       "      <th>mafs</th>\n",
       "      <th>variations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1491004261</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128176888</td>\n",
       "      <td>NC_000002.12:128176887:T:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non_coding_transcript_variant,coding_sequence_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>UGGT1:56886</td>\n",
       "      <td>GnomAD_exomes:C=0.000004/1,TOPMED:C=0.000008/2...</td>\n",
       "      <td>NP_064505.1:p.Ile1205Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1491004253</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>5:53042133</td>\n",
       "      <td>NC_000005.10:53042132:G:A,NC_000005.10:5304213...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 07:59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>ITGA2:3673</td>\n",
       "      <td>GnomAD_exomes:T=0.000004/1,TOPMED:A=0.000004/1...</td>\n",
       "      <td>NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1491004226</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>20:33663605</td>\n",
       "      <td>NC_000020.11:33663604:C:G,NC_000020.11:3366360...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 17:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,5_prime_UTR_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>NECAB3:63941,C20orf144:128864</td>\n",
       "      <td>ALFA:T=0./0</td>\n",
       "      <td>NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1491003662</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:26946984</td>\n",
       "      <td>NC_000002.12:26946983:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 05:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>DPYSL5:56896</td>\n",
       "      <td>GnomAD:A=0.000007/1,TOPMED:A=0.000004/1,ALFA:A...</td>\n",
       "      <td>NP_064519.2:p.Gly562Ser,NP_001240653.1:p.Gly56...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1491003598</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269218</td>\n",
       "      <td>NC_000001.11:205269217:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>TMCC2:9911</td>\n",
       "      <td>GnomAD_exomes:G=0.000004/1,TOMMO:G=0.000035/1,...</td>\n",
       "      <td>NP_055673.2:p.Lys339Arg,NP_001229854.1:p.Lys26...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1491003486</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:9262263</td>\n",
       "      <td>NC_000001.11:9262262:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 16:11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant,5_pri...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>H6PD:9563</td>\n",
       "      <td>GnomAD_exomes:A=0.000012/3,TOPMED:A=0.000004/1...</td>\n",
       "      <td>NP_004276.2:p.Ser317Asn,NP_001269516.1:p.Ser32...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1491003469</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:210683706</td>\n",
       "      <td>NC_000001.11:210683705:C:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 04:45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>KCNH1:3756</td>\n",
       "      <td>GnomAD:G=0.000036/5,GnomAD_exomes:G=0.000016/4...</td>\n",
       "      <td>NP_002229.1:p.Asp822His,NP_758872.1:p.Asp849His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1491003245</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:105881549</td>\n",
       "      <td>NC_000002.12:105881548:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 20:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,intron_variant,missens...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>NCK2:8440</td>\n",
       "      <td>GnomAD:T=0.000007/1,GnomAD_exomes:T=0.000004/1...</td>\n",
       "      <td>NP_003572.2:p.Arg150Trp,NP_001004720.1:p.Arg15...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1491002990</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6:122447250</td>\n",
       "      <td>NC_000006.12:122447249:G:C,NC_000006.12:122447...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 10:49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>SERINC1:57515</td>\n",
       "      <td>GnomAD:C=0.000007/1,TOPMED:C=0.000008/2,ALFA:T...</td>\n",
       "      <td>NP_065806.1:p.Pro289Arg,NP_065806.1:p.Pro289Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1491002863</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>9:2047295</td>\n",
       "      <td>NC_000009.12:2047294:G:A,NC_000009.12:2047294:G:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 10:50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,genic_upstream_transcript_var...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>SMARCA2:6595</td>\n",
       "      <td>GnomAD:C=0./0,ALFA:C=0./0</td>\n",
       "      <td>NP_003061.3:p.Arg286Gln,NP_003061.3:p.Arg286Pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1491002848</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>5:67165100</td>\n",
       "      <td>NC_000005.10:67165099:C:G,NC_000005.10:6716509...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 08:05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>MAST4:375449</td>\n",
       "      <td>GnomAD:T=0./0,GnomAD_exomes:T=0.000005/1,ALFA:...</td>\n",
       "      <td>NP_055998.1:p.Ser1785Cys,NP_055998.1:p.Ser1785...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1491002485</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7:138106626</td>\n",
       "      <td>NC_000007.14:138106625:T:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 13:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>AKR1D1:6718</td>\n",
       "      <td>GnomAD_exomes:C=0.000008/2,TOPMED:C=0.000008/2...</td>\n",
       "      <td>NP_005980.1:p.Phe200Leu,NP_001177836.1:p.Phe20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1491002341</td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>3:119393478</td>\n",
       "      <td>NC_000003.12:119393477:C:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 02:38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>ARHGAP31:57514</td>\n",
       "      <td>GnomAD:G=0.000014/2,GnomAD_exomes:G=0.000008/2...</td>\n",
       "      <td>NP_065805.2:p.Ser298Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1491002177</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>13:109140496</td>\n",
       "      <td>NC_000013.11:109140495:C:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 19:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,upstream_transcript_variant,2...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>MYO16:23026,LOC124903208:124903208</td>\n",
       "      <td>GnomAD:G=0.000007/1,TOPMED:G=0.000004/1,ALFA:G...</td>\n",
       "      <td>NP_001185879.1:p.Asp1428Glu,NP_055826.1:p.Asp1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1491002060</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>21:45992191</td>\n",
       "      <td>NC_000021.9:45992190:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 17:39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>COL6A1:1291</td>\n",
       "      <td>GnomAD_exomes:A=0.000004/1,TOPMED:A=0.000004/1...</td>\n",
       "      <td>NP_001839.2:p.Gly404Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs1491001982</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7:149730272</td>\n",
       "      <td>NC_000007.14:149730271:T:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 13:10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant,genic...</td>\n",
       "      <td>by-frequency,by-alfa</td>\n",
       "      <td>KRBA1:84626</td>\n",
       "      <td>ALFA:C=0./0</td>\n",
       "      <td>NP_115923.2:p.Ser646Pro,NP_001277116.1:p.Ser68...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs1491001877</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11:4385428</td>\n",
       "      <td>NC_000011.10:4385427:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 13:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa</td>\n",
       "      <td>TRIM21:6737</td>\n",
       "      <td>TOPMED:A=0.000015/4,ALFA:A=0.000071/1</td>\n",
       "      <td>NP_003132.2:p.Leu429Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs1491001822</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19:54576083</td>\n",
       "      <td>NC_000019.10:54576082:G:A,NC_000019.10:5457608...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 02:29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>LILRA2:11027</td>\n",
       "      <td>GnomAD_exomes:A=0.000004/1,TOPMED:A=0.000004/1...</td>\n",
       "      <td>NP_001124389.2:p.Ser410Asn,NP_001124389.2:p.Se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs1491000935</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:21980608</td>\n",
       "      <td>NC_000016.10:21980607:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 22:57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>UQCRC2:7385,PDZD9:255762</td>\n",
       "      <td>GnomAD_exomes:A=0.000004/1,ALFA:A=0.000111/1</td>\n",
       "      <td>NP_003357.2:p.Val396Ile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs1491000702</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>X:50193125</td>\n",
       "      <td>NC_000023.11:50193124:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 03:42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>AKAP4:8852</td>\n",
       "      <td>GnomAD:T=0.000019/2,GnomAD_exomes:T=0.000005/1...</td>\n",
       "      <td>NP_003877.2:p.Gly530Arg,NP_647450.1:p.Gly521Arg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          snp_id           acc        chrpos  \\\n",
       "0   rs1491004261  NC_000002.12   2:128176888   \n",
       "1   rs1491004253  NC_000005.10    5:53042133   \n",
       "2   rs1491004226  NC_000020.11   20:33663605   \n",
       "3   rs1491003662  NC_000002.12    2:26946984   \n",
       "4   rs1491003598  NC_000001.11   1:205269218   \n",
       "5   rs1491003486  NC_000001.11     1:9262263   \n",
       "6   rs1491003469  NC_000001.11   1:210683706   \n",
       "7   rs1491003245  NC_000002.12   2:105881549   \n",
       "8   rs1491002990  NC_000006.12   6:122447250   \n",
       "9   rs1491002863  NC_000009.12     9:2047295   \n",
       "10  rs1491002848  NC_000005.10    5:67165100   \n",
       "11  rs1491002485  NC_000007.14   7:138106626   \n",
       "12  rs1491002341  NC_000003.12   3:119393478   \n",
       "13  rs1491002177  NC_000013.11  13:109140496   \n",
       "14  rs1491002060   NC_000021.9   21:45992191   \n",
       "15  rs1491001982  NC_000007.14   7:149730272   \n",
       "16  rs1491001877  NC_000011.10    11:4385428   \n",
       "17  rs1491001822  NC_000019.10   19:54576083   \n",
       "18  rs1491000935  NC_000016.10   16:21980608   \n",
       "19  rs1491000702  NC_000023.11    X:50193125   \n",
       "\n",
       "                                                 spdi  tax_id snp_class  \\\n",
       "0                          NC_000002.12:128176887:T:C    9606       snv   \n",
       "1   NC_000005.10:53042132:G:A,NC_000005.10:5304213...    9606       snv   \n",
       "2   NC_000020.11:33663604:C:G,NC_000020.11:3366360...    9606       snv   \n",
       "3                           NC_000002.12:26946983:G:A    9606       snv   \n",
       "4                          NC_000001.11:205269217:A:G    9606       snv   \n",
       "5                            NC_000001.11:9262262:G:A    9606       snv   \n",
       "6                          NC_000001.11:210683705:C:G    9606       snv   \n",
       "7                          NC_000002.12:105881548:C:T    9606       snv   \n",
       "8   NC_000006.12:122447249:G:C,NC_000006.12:122447...    9606       snv   \n",
       "9   NC_000009.12:2047294:G:A,NC_000009.12:2047294:G:C    9606       snv   \n",
       "10  NC_000005.10:67165099:C:G,NC_000005.10:6716509...    9606       snv   \n",
       "11                         NC_000007.14:138106625:T:C    9606       snv   \n",
       "12                         NC_000003.12:119393477:C:G    9606       snv   \n",
       "13                         NC_000013.11:109140495:C:G    9606       snv   \n",
       "14                           NC_000021.9:45992190:G:A    9606       snv   \n",
       "15                         NC_000007.14:149730271:T:C    9606       snv   \n",
       "16                           NC_000011.10:4385427:G:A    9606       snv   \n",
       "17  NC_000019.10:54576082:G:A,NC_000019.10:5457608...    9606       snv   \n",
       "18                          NC_000016.10:21980607:G:A    9606       snv   \n",
       "19                          NC_000023.11:50193124:C:T    9606       snv   \n",
       "\n",
       "         create_date       update_date clinical_significance  \\\n",
       "0   2017/11/09 09:55  2022/10/12 21:08                   NaN   \n",
       "1   2017/11/09 09:55  2022/10/13 07:59                   NaN   \n",
       "2   2017/11/09 09:55  2022/10/13 17:00                   NaN   \n",
       "3   2017/11/09 09:55  2022/10/17 05:04                   NaN   \n",
       "4   2017/11/09 09:55  2022/10/12 18:08                   NaN   \n",
       "5   2017/11/09 09:55  2022/10/12 16:11                   NaN   \n",
       "6   2017/11/09 09:55  2022/10/17 04:45                   NaN   \n",
       "7   2017/11/09 09:55  2022/10/12 20:41                   NaN   \n",
       "8   2017/11/09 09:55  2022/10/13 10:49                   NaN   \n",
       "9   2017/11/09 09:55  2022/10/16 10:50                   NaN   \n",
       "10  2017/11/09 09:55  2022/10/17 08:05                   NaN   \n",
       "11  2017/11/09 09:55  2022/10/13 13:04                   NaN   \n",
       "12  2017/11/09 09:55  2022/10/13 02:38                   NaN   \n",
       "13  2017/11/09 09:55  2022/10/16 19:33                   NaN   \n",
       "14  2017/11/09 09:55  2022/10/13 17:39                   NaN   \n",
       "15  2017/11/09 09:55  2022/10/13 13:10                   NaN   \n",
       "16  2017/11/09 09:55  2022/10/16 13:47                   NaN   \n",
       "17  2017/11/09 09:55  2022/10/17 02:29                   NaN   \n",
       "18  2017/11/09 09:55  2022/10/16 22:57                   NaN   \n",
       "19  2017/11/09 09:55  2022/10/17 03:42                   NaN   \n",
       "\n",
       "                                            fxn_class  \\\n",
       "0   non_coding_transcript_variant,coding_sequence_...   \n",
       "1   coding_sequence_variant,non_coding_transcript_...   \n",
       "2   genic_upstream_transcript_variant,5_prime_UTR_...   \n",
       "3            coding_sequence_variant,missense_variant   \n",
       "4            coding_sequence_variant,missense_variant   \n",
       "5   coding_sequence_variant,missense_variant,5_pri...   \n",
       "6            coding_sequence_variant,missense_variant   \n",
       "7   coding_sequence_variant,intron_variant,missens...   \n",
       "8            coding_sequence_variant,missense_variant   \n",
       "9   missense_variant,genic_upstream_transcript_var...   \n",
       "10  coding_sequence_variant,genic_downstream_trans...   \n",
       "11           missense_variant,coding_sequence_variant   \n",
       "12           missense_variant,coding_sequence_variant   \n",
       "13  missense_variant,upstream_transcript_variant,2...   \n",
       "14           coding_sequence_variant,missense_variant   \n",
       "15  missense_variant,coding_sequence_variant,genic...   \n",
       "16           missense_variant,coding_sequence_variant   \n",
       "17  genic_downstream_transcript_variant,missense_v...   \n",
       "18  genic_downstream_transcript_variant,missense_v...   \n",
       "19           coding_sequence_variant,missense_variant   \n",
       "\n",
       "                          validated                               genes  \\\n",
       "0   by-frequency,by-alfa,by-cluster                         UGGT1:56886   \n",
       "1   by-frequency,by-alfa,by-cluster                          ITGA2:3673   \n",
       "2   by-frequency,by-alfa,by-cluster       NECAB3:63941,C20orf144:128864   \n",
       "3   by-frequency,by-alfa,by-cluster                        DPYSL5:56896   \n",
       "4   by-frequency,by-alfa,by-cluster                          TMCC2:9911   \n",
       "5   by-frequency,by-alfa,by-cluster                           H6PD:9563   \n",
       "6   by-frequency,by-alfa,by-cluster                          KCNH1:3756   \n",
       "7   by-frequency,by-alfa,by-cluster                           NCK2:8440   \n",
       "8   by-frequency,by-alfa,by-cluster                       SERINC1:57515   \n",
       "9   by-frequency,by-alfa,by-cluster                        SMARCA2:6595   \n",
       "10  by-frequency,by-alfa,by-cluster                        MAST4:375449   \n",
       "11  by-frequency,by-alfa,by-cluster                         AKR1D1:6718   \n",
       "12  by-frequency,by-alfa,by-cluster                      ARHGAP31:57514   \n",
       "13  by-frequency,by-alfa,by-cluster  MYO16:23026,LOC124903208:124903208   \n",
       "14  by-frequency,by-alfa,by-cluster                         COL6A1:1291   \n",
       "15             by-frequency,by-alfa                         KRBA1:84626   \n",
       "16             by-frequency,by-alfa                         TRIM21:6737   \n",
       "17  by-frequency,by-alfa,by-cluster                        LILRA2:11027   \n",
       "18  by-frequency,by-alfa,by-cluster            UQCRC2:7385,PDZD9:255762   \n",
       "19  by-frequency,by-alfa,by-cluster                          AKAP4:8852   \n",
       "\n",
       "                                                 mafs  \\\n",
       "0   GnomAD_exomes:C=0.000004/1,TOPMED:C=0.000008/2...   \n",
       "1   GnomAD_exomes:T=0.000004/1,TOPMED:A=0.000004/1...   \n",
       "2                                         ALFA:T=0./0   \n",
       "3   GnomAD:A=0.000007/1,TOPMED:A=0.000004/1,ALFA:A...   \n",
       "4   GnomAD_exomes:G=0.000004/1,TOMMO:G=0.000035/1,...   \n",
       "5   GnomAD_exomes:A=0.000012/3,TOPMED:A=0.000004/1...   \n",
       "6   GnomAD:G=0.000036/5,GnomAD_exomes:G=0.000016/4...   \n",
       "7   GnomAD:T=0.000007/1,GnomAD_exomes:T=0.000004/1...   \n",
       "8   GnomAD:C=0.000007/1,TOPMED:C=0.000008/2,ALFA:T...   \n",
       "9                           GnomAD:C=0./0,ALFA:C=0./0   \n",
       "10  GnomAD:T=0./0,GnomAD_exomes:T=0.000005/1,ALFA:...   \n",
       "11  GnomAD_exomes:C=0.000008/2,TOPMED:C=0.000008/2...   \n",
       "12  GnomAD:G=0.000014/2,GnomAD_exomes:G=0.000008/2...   \n",
       "13  GnomAD:G=0.000007/1,TOPMED:G=0.000004/1,ALFA:G...   \n",
       "14  GnomAD_exomes:A=0.000004/1,TOPMED:A=0.000004/1...   \n",
       "15                                        ALFA:C=0./0   \n",
       "16              TOPMED:A=0.000015/4,ALFA:A=0.000071/1   \n",
       "17  GnomAD_exomes:A=0.000004/1,TOPMED:A=0.000004/1...   \n",
       "18       GnomAD_exomes:A=0.000004/1,ALFA:A=0.000111/1   \n",
       "19  GnomAD:T=0.000019/2,GnomAD_exomes:T=0.000005/1...   \n",
       "\n",
       "                                           variations  \n",
       "0                            NP_064505.1:p.Ile1205Thr  \n",
       "1       NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys  \n",
       "2       NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val  \n",
       "3   NP_064519.2:p.Gly562Ser,NP_001240653.1:p.Gly56...  \n",
       "4   NP_055673.2:p.Lys339Arg,NP_001229854.1:p.Lys26...  \n",
       "5   NP_004276.2:p.Ser317Asn,NP_001269516.1:p.Ser32...  \n",
       "6     NP_002229.1:p.Asp822His,NP_758872.1:p.Asp849His  \n",
       "7   NP_003572.2:p.Arg150Trp,NP_001004720.1:p.Arg15...  \n",
       "8     NP_065806.1:p.Pro289Arg,NP_065806.1:p.Pro289Gln  \n",
       "9   NP_003061.3:p.Arg286Gln,NP_003061.3:p.Arg286Pr...  \n",
       "10  NP_055998.1:p.Ser1785Cys,NP_055998.1:p.Ser1785...  \n",
       "11  NP_005980.1:p.Phe200Leu,NP_001177836.1:p.Phe20...  \n",
       "12                            NP_065805.2:p.Ser298Cys  \n",
       "13  NP_001185879.1:p.Asp1428Glu,NP_055826.1:p.Asp1...  \n",
       "14                            NP_001839.2:p.Gly404Arg  \n",
       "15  NP_115923.2:p.Ser646Pro,NP_001277116.1:p.Ser68...  \n",
       "16                            NP_003132.2:p.Leu429Phe  \n",
       "17  NP_001124389.2:p.Ser410Asn,NP_001124389.2:p.Se...  \n",
       "18                            NP_003357.2:p.Val396Ile  \n",
       "19    NP_003877.2:p.Gly530Arg,NP_647450.1:p.Gly521Arg  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "initially_cleaned_df.head(20)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading and filtering ALFA data as it is big"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 (19889, 21)\n",
      "1 (5336, 21)\n",
      "2 (9148, 21)\n",
      "3 (13846, 21)\n",
      "4 (10780, 21)\n",
      "5 (11830, 21)\n",
      "6 (13104, 21)\n",
      "7 (9094, 21)\n",
      "8 (13853, 21)\n",
      "9 (6234, 21)\n",
      "10 (10527, 21)\n",
      "11 (9397, 21)\n",
      "12 (11384, 21)\n",
      "13 (14070, 21)\n",
      "14 (9227, 21)\n",
      "15 (2668, 21)\n",
      "16 (11260, 21)\n",
      "17 (2431, 21)\n",
      "18 (2125, 21)\n",
      "19 (5695, 21)\n",
      "20 (3971, 21)\n",
      "21 (1675, 21)\n",
      "22 (3139, 21)\n",
      "23 (3918, 21)\n",
      "24 (361, 21)\n",
      "25 (3480, 21)\n",
      "26 (3361, 21)\n",
      "27 (5891, 21)\n",
      "28 (6887, 21)\n",
      "29 (654, 21)\n",
      "30 (4283, 21)\n",
      "31 (1847, 21)\n",
      "32 (711, 21)\n",
      "33 (13125, 21)\n",
      "34 (7633, 21)\n",
      "35 (6052, 21)\n",
      "36 (4362, 21)\n",
      "37 (0, 21)\n",
      "38 (0, 21)\n",
      "39 (4314, 21)\n",
      "40 (8112, 21)\n",
      "41 (24622, 21)\n",
      "42 (19463, 21)\n",
      "43 (18059, 21)\n",
      "44 (13846, 21)\n",
      "45 (2027, 21)\n",
      "46 (4562, 21)\n",
      "47 (8178, 21)\n",
      "48 (4510, 21)\n",
      "49 (4567, 21)\n",
      "50 (7363, 21)\n",
      "51 (6500, 21)\n",
      "52 (3521, 21)\n",
      "53 (450, 21)\n",
      "54 (1052, 21)\n",
      "55 (5528, 21)\n",
      "56 (9333, 21)\n",
      "57 (12048, 21)\n",
      "58 (9030, 21)\n",
      "59 (4372, 21)\n",
      "60 (4553, 21)\n",
      "61 (3738, 21)\n",
      "62 (3485, 21)\n",
      "63 (7571, 21)\n",
      "64 (12840, 21)\n",
      "65 (6017, 21)\n",
      "66 (5233, 21)\n",
      "67 (7201, 21)\n",
      "68 (3100, 21)\n",
      "69 (2704, 21)\n",
      "70 (9736, 21)\n",
      "71 (5993, 21)\n",
      "72 (1499, 21)\n",
      "73 (731, 21)\n",
      "74 (6472, 21)\n",
      "75 (235, 21)\n",
      "76 (3432, 21)\n",
      "77 (5674, 21)\n",
      "78 (3236, 21)\n",
      "79 (15826, 21)\n",
      "80 (5564, 21)\n",
      "81 (5560, 21)\n",
      "82 (2457, 21)\n",
      "83 (5947, 21)\n",
      "84 (1438, 21)\n",
      "85 (6652, 21)\n",
      "86 (4817, 21)\n",
      "87 (2747, 21)\n",
      "88 (238, 21)\n",
      "89 (7093, 21)\n",
      "90 (471, 21)\n",
      "91 (4036, 21)\n",
      "92 (3266, 21)\n",
      "93 (2301, 21)\n",
      "94 (6819, 21)\n",
      "95 (13209, 21)\n",
      "96 (746, 21)\n",
      "97 (856, 21)\n",
      "98 (2035, 21)\n",
      "99 (6993, 21)\n",
      "100 (1117, 21)\n",
      "101 (0, 21)\n",
      "102 (10401, 21)\n",
      "103 (6999, 21)\n",
      "104 (4498, 21)\n",
      "105 (2972, 21)\n",
      "106 (6721, 21)\n",
      "107 (6810, 21)\n",
      "108 (358, 21)\n",
      "109 (3578, 21)\n",
      "110 (1793, 21)\n",
      "111 (702, 21)\n",
      "112 (5613, 21)\n",
      "113 (7803, 21)\n",
      "114 (4162, 21)\n",
      "115 (1148, 21)\n",
      "116 (2148, 21)\n",
      "117 (1015, 21)\n",
      "118 (1657, 21)\n",
      "119 (6763, 21)\n",
      "120 (395, 21)\n",
      "121 (4429, 21)\n",
      "122 (4799, 21)\n",
      "123 (4038, 21)\n",
      "124 (7452, 21)\n",
      "125 (5999, 21)\n",
      "126 (4429, 21)\n",
      "127 (22075, 21)\n",
      "128 (2664, 21)\n",
      "129 (3951, 21)\n",
      "130 (2825, 21)\n",
      "131 (4595, 21)\n",
      "132 (963, 21)\n",
      "133 (5131, 21)\n",
      "134 (8055, 21)\n",
      "135 (4142, 21)\n",
      "136 (6050, 21)\n",
      "137 (4174, 21)\n",
      "138 (3531, 21)\n",
      "139 (9193, 21)\n",
      "140 (10843, 21)\n",
      "141 (3204, 21)\n",
      "142 (4349, 21)\n",
      "143 (5507, 21)\n",
      "144 (10392, 21)\n",
      "145 (7381, 21)\n",
      "146 (8970, 21)\n",
      "147 (4270, 21)\n",
      "148 (2604, 21)\n",
      "149 (970, 21)\n",
      "150 (8912, 21)\n",
      "151 (6227, 21)\n",
      "152 (6977, 21)\n",
      "153 (2305, 21)\n",
      "154 (199, 21)\n",
      "155 (1730, 21)\n",
      "156 (1819, 21)\n",
      "157 (1896, 21)\n",
      "158 (3231, 21)\n",
      "159 (10774, 21)\n",
      "160 (5758, 21)\n",
      "161 (10143, 21)\n",
      "162 (19883, 21)\n",
      "163 (23359, 21)\n",
      "164 (8876, 21)\n",
      "165 (8323, 21)\n",
      "166 (181, 21)\n",
      "167 (3718, 21)\n",
      "168 (1977, 21)\n",
      "169 (2315, 21)\n",
      "170 (2159, 21)\n",
      "171 (1507, 21)\n",
      "172 (1378, 21)\n",
      "173 (329, 21)\n",
      "174 (135, 21)\n",
      "175 (1178, 21)\n",
      "176 (328, 21)\n",
      "177 (905, 21)\n",
      "178 (4515, 21)\n",
      "179 (5713, 21)\n",
      "180 (593, 21)\n",
      "181 (1407, 21)\n",
      "182 (3350, 21)\n",
      "183 (8983, 21)\n",
      "184 (249, 21)\n",
      "185 (4851, 21)\n",
      "186 (11598, 21)\n",
      "187 (7453, 21)\n",
      "188 (9697, 21)\n",
      "189 (6038, 21)\n",
      "190 (2278, 21)\n",
      "191 (4943, 21)\n",
      "192 (4838, 21)\n",
      "193 (1205, 21)\n",
      "194 (4616, 21)\n",
      "195 (4692, 21)\n",
      "196 (3112, 21)\n",
      "197 (3204, 21)\n",
      "198 (516, 21)\n",
      "199 (2044, 21)\n",
      "200 (4345, 21)\n",
      "201 (2917, 21)\n",
      "202 (694, 21)\n",
      "203 (2508, 21)\n",
      "204 (3813, 21)\n",
      "205 (11335, 21)\n",
      "206 (2191, 21)\n",
      "207 (2636, 21)\n",
      "208 (12044, 21)\n",
      "209 (11233, 21)\n",
      "210 (11617, 21)\n",
      "211 (7183, 21)\n",
      "212 (6186, 21)\n",
      "213 (1508, 21)\n",
      "214 (3268, 21)\n",
      "215 (4030, 21)\n",
      "216 (823, 21)\n",
      "217 (1969, 21)\n",
      "218 (2275, 21)\n",
      "219 (529, 21)\n",
      "220 (0, 21)\n",
      "221 (1986, 21)\n",
      "222 (6845, 21)\n",
      "223 (2431, 21)\n",
      "224 (1339, 21)\n",
      "225 (4844, 21)\n",
      "226 (2312, 21)\n",
      "227 (6743, 21)\n",
      "228 (0, 21)\n",
      "229 (630, 21)\n",
      "230 (695, 21)\n",
      "231 (3172, 21)\n",
      "232 (6062, 21)\n",
      "233 (4033, 21)\n",
      "234 (5921, 21)\n",
      "235 (4250, 21)\n",
      "236 (4519, 21)\n",
      "237 (6060, 21)\n",
      "238 (3220, 21)\n",
      "239 (1387, 21)\n",
      "240 (947, 21)\n",
      "241 (4065, 21)\n",
      "242 (3145, 21)\n",
      "243 (3273, 21)\n",
      "244 (4197, 21)\n",
      "245 (4242, 21)\n",
      "246 (700, 21)\n",
      "247 (3246, 21)\n",
      "248 (4559, 21)\n",
      "249 (2746, 21)\n",
      "250 (2861, 21)\n",
      "251 (424, 21)\n",
      "252 (168, 21)\n",
      "253 (2302, 21)\n",
      "254 (2627, 21)\n",
      "255 (3628, 21)\n",
      "256 (1781, 21)\n",
      "257 (4525, 21)\n",
      "258 (5828, 21)\n",
      "259 (2805, 21)\n",
      "260 (379, 21)\n",
      "261 (2500, 21)\n",
      "262 (2942, 21)\n",
      "263 (1124, 21)\n",
      "264 (1689, 21)\n",
      "265 (1644, 21)\n",
      "266 (0, 21)\n",
      "267 (5348, 21)\n",
      "268 (7053, 21)\n",
      "269 (6471, 21)\n",
      "270 (2127, 21)\n",
      "271 (2534, 21)\n",
      "272 (2462, 21)\n",
      "273 (894, 21)\n",
      "274 (5579, 21)\n",
      "275 (553, 21)\n",
      "276 (339, 21)\n",
      "277 (801, 21)\n",
      "278 (328, 21)\n",
      "279 (2536, 21)\n",
      "280 (4296, 21)\n",
      "281 (7560, 21)\n",
      "282 (4272, 21)\n",
      "283 (2561, 21)\n",
      "284 (255, 21)\n",
      "285 (530, 21)\n",
      "286 (4147, 21)\n",
      "287 (3073, 21)\n",
      "288 (1416, 21)\n",
      "289 (2557, 21)\n",
      "290 (3092, 21)\n",
      "291 (6008, 21)\n",
      "292 (6072, 21)\n",
      "293 (6042, 21)\n",
      "294 (3557, 21)\n",
      "295 (1794, 21)\n",
      "296 (820, 21)\n",
      "297 (3302, 21)\n",
      "298 (4167, 21)\n",
      "299 (2655, 21)\n",
      "300 (1163, 21)\n",
      "301 (853, 21)\n",
      "302 (644, 21)\n",
      "303 (2264, 21)\n",
      "304 (4256, 21)\n",
      "305 (2615, 21)\n",
      "306 (2286, 21)\n",
      "307 (2717, 21)\n",
      "308 (1284, 21)\n",
      "309 (4233, 21)\n",
      "310 (6526, 21)\n",
      "311 (4490, 21)\n",
      "312 (7209, 21)\n",
      "313 (30017, 21)\n",
      "314 (2440, 21)\n",
      "315 (5460, 21)\n",
      "316 (10149, 21)\n",
      "317 (3152, 21)\n",
      "318 (4033, 21)\n",
      "319 (2915, 21)\n",
      "320 (593, 21)\n",
      "321 (2868, 21)\n",
      "322 (4239, 21)\n",
      "323 (2214, 21)\n",
      "324 (11801, 21)\n",
      "325 (10009, 21)\n",
      "326 (3476, 21)\n",
      "327 (2187, 21)\n",
      "328 (4114, 21)\n",
      "329 (4768, 21)\n",
      "330 (2609, 21)\n",
      "331 (3014, 21)\n",
      "332 (1092, 21)\n",
      "333 (9276, 21)\n",
      "334 (10093, 21)\n",
      "335 (31801, 21)\n",
      "336 (10445, 21)\n",
      "337 (6281, 21)\n",
      "338 (15836, 21)\n",
      "339 (7399, 21)\n",
      "340 (3864, 21)\n",
      "341 (4774, 21)\n",
      "342 (3245, 21)\n",
      "343 (5138, 21)\n",
      "344 (444, 21)\n",
      "345 (2787, 21)\n",
      "346 (364, 21)\n",
      "347 (3378, 21)\n",
      "348 (4975, 21)\n",
      "349 (1135, 21)\n",
      "350 (1727, 21)\n",
      "351 (4146, 21)\n",
      "352 (3599, 21)\n",
      "353 (5609, 21)\n",
      "354 (350, 21)\n",
      "355 (1803, 21)\n",
      "356 (3183, 21)\n",
      "357 (286, 21)\n",
      "358 (3705, 21)\n",
      "359 (6136, 21)\n",
      "360 (3448, 21)\n",
      "361 (5676, 21)\n",
      "362 (1785, 21)\n",
      "363 (1898, 21)\n",
      "364 (2662, 21)\n",
      "365 (3935, 21)\n",
      "366 (5156, 21)\n",
      "367 (4438, 21)\n",
      "368 (3787, 21)\n",
      "369 (3244, 21)\n",
      "370 (4392, 21)\n",
      "371 (4718, 21)\n",
      "372 (7360, 21)\n",
      "373 (3527, 21)\n",
      "374 (8717, 21)\n",
      "375 (1587, 21)\n",
      "376 (1463, 21)\n",
      "377 (4924, 21)\n",
      "378 (7729, 21)\n",
      "379 (5816, 21)\n",
      "380 (12989, 21)\n",
      "381 (1888, 21)\n",
      "382 (2747, 21)\n",
      "383 (838, 21)\n",
      "384 (1985, 21)\n",
      "385 (2554, 21)\n",
      "386 (5963, 21)\n",
      "387 (4360, 21)\n",
      "388 (4572, 21)\n",
      "389 (3024, 21)\n",
      "390 (2456, 21)\n",
      "391 (3037, 21)\n",
      "392 (4883, 21)\n",
      "393 (5124, 21)\n",
      "394 (6199, 21)\n",
      "395 (1047, 21)\n",
      "396 (2602, 21)\n",
      "397 (473, 21)\n",
      "398 (261, 21)\n",
      "399 (4331, 21)\n",
      "400 (101, 21)\n",
      "401 (4042, 21)\n",
      "402 (7741, 21)\n",
      "403 (2578, 21)\n",
      "404 (4115, 21)\n",
      "405 (672, 21)\n",
      "406 (4398, 21)\n",
      "407 (7138, 21)\n",
      "408 (3738, 21)\n",
      "409 (8299, 21)\n",
      "410 (27282, 21)\n",
      "411 (5131, 21)\n",
      "412 (5907, 21)\n",
      "413 (1590, 21)\n",
      "414 (1478, 21)\n",
      "415 (3563, 21)\n",
      "416 (3033, 21)\n",
      "417 (3268, 21)\n",
      "418 (7717, 21)\n",
      "419 (4180, 21)\n",
      "420 (4657, 21)\n",
      "421 (3604, 21)\n",
      "422 (8557, 21)\n",
      "423 (8876, 21)\n",
      "424 (1004, 21)\n",
      "425 (14799, 21)\n",
      "426 (3448, 21)\n",
      "427 (3185, 21)\n",
      "428 (3060, 21)\n",
      "429 (5703, 21)\n",
      "430 (304, 21)\n",
      "431 (3800, 21)\n",
      "432 (3442, 21)\n",
      "433 (8052, 21)\n",
      "434 (2168, 21)\n",
      "435 (1886, 21)\n",
      "436 (5820, 21)\n",
      "437 (7339, 21)\n",
      "438 (6342, 21)\n",
      "439 (6402, 21)\n",
      "440 (3794, 21)\n",
      "441 (1523, 21)\n",
      "442 (6121, 21)\n",
      "443 (4323, 21)\n",
      "444 (2548, 21)\n",
      "445 (3047, 21)\n",
      "446 (1244, 21)\n",
      "447 (4077, 21)\n",
      "448 (1931, 21)\n",
      "449 (2691, 21)\n",
      "450 (1131, 21)\n",
      "451 (4509, 21)\n",
      "452 (2437, 21)\n",
      "453 (2351, 21)\n",
      "454 (2222, 21)\n",
      "455 (1468, 21)\n",
      "456 (967, 21)\n",
      "457 (3221, 21)\n",
      "458 (1518, 21)\n",
      "459 (2734, 21)\n",
      "460 (3811, 21)\n",
      "461 (5834, 21)\n",
      "462 (3337, 21)\n",
      "463 (2075, 21)\n",
      "464 (3382, 21)\n",
      "465 (1632, 21)\n",
      "466 (1182, 21)\n",
      "467 (3390, 21)\n",
      "468 (3526, 21)\n",
      "469 (3262, 21)\n",
      "470 (1061, 21)\n",
      "471 (4667, 21)\n",
      "472 (679, 21)\n",
      "473 (1623, 21)\n",
      "474 (4300, 21)\n",
      "475 (30807, 21)\n",
      "476 (3812, 21)\n",
      "477 (3492, 21)\n",
      "478 (4797, 21)\n",
      "479 (1019, 21)\n",
      "480 (857, 21)\n",
      "481 (4321, 21)\n",
      "482 (1698, 21)\n",
      "483 (6221, 21)\n",
      "484 (1852, 21)\n",
      "485 (308, 21)\n",
      "486 (2288, 21)\n",
      "487 (2285, 21)\n",
      "488 (16944, 21)\n",
      "489 (5786, 21)\n",
      "490 (1988, 21)\n",
      "491 (0, 21)\n",
      "492 (374, 21)\n",
      "493 (3504, 21)\n",
      "494 (4112, 21)\n",
      "495 (1970, 21)\n",
      "496 (4738, 21)\n",
      "497 (1890, 21)\n",
      "498 (2657, 21)\n",
      "499 (5032, 21)\n",
      "500 (2478, 21)\n",
      "501 (8108, 21)\n",
      "502 (7439, 21)\n",
      "503 (5449, 21)\n",
      "504 (3083, 21)\n",
      "505 (2240, 21)\n",
      "506 (7148, 21)\n",
      "507 (9395, 21)\n",
      "508 (2929, 21)\n",
      "509 (4331, 21)\n",
      "510 (8639, 21)\n",
      "511 (8356, 21)\n",
      "512 (16447, 21)\n",
      "513 (12908, 21)\n",
      "514 (23190, 21)\n",
      "515 (7020, 21)\n",
      "516 (1183, 21)\n",
      "517 (6352, 21)\n",
      "518 (528, 21)\n",
      "519 (4193, 21)\n",
      "520 (4282, 21)\n",
      "521 (4381, 21)\n",
      "522 (2985, 21)\n",
      "523 (4646, 21)\n",
      "524 (4621, 21)\n",
      "525 (3886, 21)\n",
      "526 (2551, 21)\n",
      "527 (1974, 21)\n",
      "528 (0, 21)\n",
      "529 (2864, 21)\n",
      "530 (6611, 21)\n",
      "531 (7901, 21)\n",
      "532 (1944, 21)\n",
      "533 (146, 21)\n",
      "534 (3393, 21)\n",
      "535 (2702, 21)\n",
      "536 (662, 21)\n",
      "537 (10029, 21)\n",
      "538 (11259, 21)\n",
      "539 (2555, 21)\n",
      "540 (4552, 21)\n",
      "541 (380, 21)\n",
      "542 (5232, 21)\n",
      "543 (5288, 21)\n",
      "544 (6387, 21)\n",
      "545 (9893, 21)\n",
      "546 (9599, 21)\n",
      "547 (15024, 21)\n",
      "548 (2191, 21)\n",
      "549 (739, 21)\n",
      "550 (4094, 21)\n",
      "551 (6246, 21)\n",
      "552 (4663, 21)\n",
      "553 (6408, 21)\n",
      "554 (5059, 21)\n",
      "555 (3671, 21)\n",
      "556 (4355, 21)\n",
      "557 (30058, 21)\n",
      "558 (11044, 21)\n",
      "559 (20506, 21)\n",
      "560 (7200, 21)\n",
      "561 (3566, 21)\n",
      "562 (6669, 21)\n",
      "563 (7141, 21)\n",
      "564 (1209, 21)\n",
      "565 (1274, 21)\n",
      "566 (1707, 21)\n",
      "567 (5364, 21)\n",
      "568 (4684, 21)\n",
      "569 (201, 21)\n",
      "570 (116, 21)\n",
      "571 (6913, 21)\n",
      "572 (10612, 21)\n",
      "573 (355, 21)\n",
      "574 (611, 21)\n",
      "575 (12198, 21)\n",
      "576 (12088, 21)\n",
      "577 (19239, 21)\n",
      "578 (31283, 21)\n",
      "579 (11290, 21)\n",
      "580 (8981, 21)\n",
      "581 (8986, 21)\n",
      "582 (6276, 21)\n",
      "583 (1960, 21)\n",
      "584 (3330, 21)\n",
      "585 (2012, 21)\n",
      "586 (2598, 21)\n",
      "587 (8576, 21)\n",
      "588 (783, 21)\n",
      "589 (2364, 21)\n",
      "590 (5816, 21)\n",
      "591 (6100, 21)\n",
      "592 (3110, 21)\n",
      "593 (6307, 21)\n",
      "594 (7063, 21)\n",
      "595 (12771, 21)\n",
      "596 (9429, 21)\n",
      "597 (4227, 21)\n",
      "598 (5442, 21)\n",
      "599 (3199, 21)\n",
      "600 (6389, 21)\n",
      "601 (6050, 21)\n",
      "602 (16681, 21)\n",
      "603 (8019, 21)\n",
      "604 (5283, 21)\n",
      "605 (2273, 21)\n",
      "606 (4206, 21)\n",
      "607 (1967, 21)\n",
      "608 (4356, 21)\n",
      "609 (3347, 21)\n",
      "610 (2974, 21)\n",
      "611 (0, 21)\n",
      "612 (1723, 21)\n",
      "613 (3441, 21)\n",
      "614 (4435, 21)\n",
      "615 (13322, 21)\n",
      "616 (17333, 21)\n",
      "617 (16603, 21)\n",
      "618 (16470, 21)\n",
      "619 (1076, 21)\n",
      "620 (3699, 21)\n",
      "621 (3551, 21)\n",
      "622 (5196, 21)\n",
      "623 (1124, 21)\n",
      "624 (2628, 21)\n",
      "625 (4098, 21)\n",
      "626 (2043, 21)\n",
      "627 (2187, 21)\n",
      "628 (1758, 21)\n",
      "629 (3371, 21)\n",
      "630 (4782, 21)\n",
      "631 (6067, 21)\n",
      "632 (5569, 21)\n",
      "633 (5485, 21)\n",
      "634 (11386, 21)\n",
      "635 (6891, 21)\n",
      "636 (3642, 21)\n",
      "637 (9957, 21)\n",
      "638 (14473, 21)\n",
      "639 (425, 21)\n",
      "640 (3528, 21)\n",
      "641 (10585, 21)\n",
      "642 (0, 21)\n",
      "643 (4200, 21)\n",
      "644 (4034, 21)\n",
      "645 (5885, 21)\n",
      "646 (3929, 21)\n",
      "647 (6679, 21)\n",
      "648 (3000, 21)\n",
      "649 (3800, 21)\n",
      "650 (4800, 21)\n",
      "651 (4624, 21)\n",
      "652 (3862, 21)\n",
      "653 (5345, 21)\n",
      "654 (8, 21)\n",
      "655 (1358, 21)\n",
      "656 (421, 21)\n",
      "657 (176, 21)\n",
      "658 (526, 21)\n",
      "659 (1629, 21)\n",
      "660 (1757, 21)\n",
      "661 (3688, 21)\n",
      "662 (106, 21)\n",
      "663 (692, 21)\n",
      "664 (446, 21)\n",
      "665 (275, 21)\n",
      "666 (1585, 21)\n",
      "667 (3434, 21)\n",
      "668 (3825, 21)\n",
      "669 (4891, 21)\n",
      "670 (1075, 21)\n",
      "671 (4582, 21)\n",
      "672 (7589, 21)\n",
      "673 (8814, 21)\n",
      "674 (19084, 21)\n",
      "675 (5371, 21)\n",
      "676 (802, 21)\n",
      "677 (4803, 21)\n",
      "678 (3601, 21)\n",
      "679 (3574, 21)\n",
      "680 (235, 21)\n",
      "681 (3180, 21)\n",
      "682 (1491, 21)\n",
      "683 (7161, 21)\n",
      "684 (4013, 21)\n",
      "685 (4366, 21)\n",
      "686 (5610, 21)\n",
      "687 (8092, 21)\n",
      "688 (4940, 21)\n",
      "689 (5418, 21)\n",
      "690 (13194, 21)\n",
      "691 (5401, 21)\n",
      "692 (2009, 21)\n",
      "693 (284, 21)\n",
      "694 (2976, 21)\n",
      "695 (5894, 21)\n",
      "696 (9336, 21)\n",
      "697 (2133, 21)\n",
      "698 (4541, 21)\n",
      "699 (17040, 21)\n",
      "700 (3212, 21)\n",
      "701 (3758, 21)\n",
      "702 (3553, 21)\n",
      "703 (4512, 21)\n",
      "704 (7189, 21)\n",
      "705 (6298, 21)\n",
      "706 (1049, 21)\n",
      "707 (23825, 21)\n",
      "708 (14516, 21)\n",
      "709 (4389, 21)\n",
      "710 (8465, 21)\n",
      "711 (3247, 21)\n",
      "712 (4789, 21)\n",
      "713 (3761, 21)\n",
      "714 (8456, 21)\n",
      "715 (9916, 21)\n",
      "716 (4362, 21)\n",
      "717 (10269, 21)\n",
      "718 (9282, 21)\n",
      "719 (7939, 21)\n",
      "720 (6592, 21)\n",
      "721 (4564, 21)\n",
      "722 (11217, 21)\n",
      "723 (4827, 21)\n",
      "724 (587, 21)\n",
      "725 (2605, 21)\n",
      "726 (16931, 21)\n",
      "727 (34636, 21)\n",
      "728 (14503, 21)\n",
      "729 (305, 21)\n",
      "730 (2055, 21)\n",
      "731 (7688, 21)\n",
      "732 (5603, 21)\n",
      "733 (5194, 21)\n",
      "734 (11838, 21)\n",
      "735 (7407, 21)\n",
      "736 (8160, 21)\n",
      "737 (22183, 21)\n",
      "738 (0, 21)\n",
      "739 (3242, 21)\n",
      "740 (5028, 21)\n",
      "741 (2876, 21)\n",
      "742 (9356, 21)\n",
      "743 (3480, 21)\n",
      "744 (248, 21)\n",
      "745 (9361, 21)\n",
      "746 (16626, 21)\n",
      "747 (11373, 21)\n",
      "748 (6537, 21)\n",
      "749 (2782, 21)\n",
      "750 (1270, 21)\n",
      "751 (4623, 21)\n",
      "752 (9352, 21)\n",
      "753 (3768, 21)\n",
      "754 (19615, 21)\n",
      "755 (13813, 21)\n",
      "756 (18941, 21)\n",
      "757 (24970, 21)\n",
      "758 (9833, 21)\n",
      "759 (3752, 21)\n",
      "760 (14674, 21)\n",
      "761 (3549, 21)\n",
      "762 (1474, 21)\n",
      "763 (15803, 21)\n",
      "764 (3603, 21)\n",
      "765 (13042, 21)\n",
      "766 (21875, 21)\n",
      "767 (22028, 21)\n",
      "768 (10817, 21)\n",
      "769 (12926, 21)\n",
      "770 (850, 21)\n",
      "771 (5947, 21)\n",
      "772 (7829, 21)\n",
      "773 (12056, 21)\n",
      "774 (4842, 21)\n",
      "775 (4723, 21)\n",
      "776 (6052, 21)\n",
      "777 (21092, 21)\n",
      "778 (14917, 21)\n",
      "779 (18270, 21)\n",
      "780 (4140, 21)\n",
      "781 (2702, 21)\n",
      "782 (4877, 21)\n",
      "783 (5633, 21)\n",
      "784 (2163, 21)\n",
      "785 (319, 21)\n",
      "786 (7023, 21)\n",
      "787 (2158, 21)\n",
      "788 (1338, 21)\n",
      "789 (5615, 21)\n",
      "790 (4190, 21)\n",
      "791 (0, 21)\n",
      "792 (548, 21)\n",
      "793 (6593, 21)\n",
      "794 (3656, 21)\n",
      "795 (1823, 21)\n",
      "796 (2184, 21)\n",
      "797 (4120, 21)\n",
      "798 (4044, 21)\n",
      "799 (2347, 21)\n",
      "800 (1072, 21)\n",
      "801 (1579, 21)\n",
      "802 (3248, 21)\n",
      "803 (2814, 21)\n",
      "804 (22265, 21)\n",
      "805 (20832, 21)\n",
      "806 (17619, 21)\n",
      "807 (29785, 21)\n",
      "808 (23551, 21)\n",
      "809 (18233, 21)\n",
      "810 (24529, 21)\n",
      "811 (10285, 21)\n",
      "812 (4400, 21)\n",
      "813 (1753, 21)\n",
      "814 (6033, 21)\n",
      "815 (20595, 21)\n",
      "816 (25295, 21)\n",
      "817 (19514, 21)\n",
      "818 (23963, 21)\n",
      "819 (26059, 21)\n",
      "820 (25819, 21)\n",
      "821 (27075, 21)\n",
      "822 (25376, 21)\n",
      "823 (11549, 21)\n",
      "824 (7549, 21)\n",
      "825 (1734, 21)\n",
      "826 (1577, 21)\n",
      "827 (2175, 21)\n",
      "828 (3957, 21)\n",
      "829 (4244, 21)\n",
      "830 (5742, 21)\n",
      "831 (0, 21)\n",
      "832 (5034, 21)\n",
      "833 (13837, 21)\n",
      "834 (8787, 21)\n",
      "835 (3412, 21)\n",
      "836 (10338, 21)\n",
      "837 (6466, 21)\n",
      "838 (6224, 21)\n",
      "839 (1774, 21)\n",
      "840 (4317, 21)\n",
      "841 (4636, 21)\n",
      "842 (19330, 21)\n",
      "843 (590, 21)\n",
      "844 (1541, 21)\n",
      "845 (1428, 21)\n",
      "846 (360, 21)\n",
      "847 (0, 21)\n",
      "848 (1805, 21)\n",
      "849 (4658, 21)\n",
      "850 (8318, 21)\n",
      "851 (5104, 21)\n",
      "852 (3260, 21)\n",
      "853 (12741, 21)\n",
      "854 (10482, 21)\n",
      "855 (678, 21)\n",
      "856 (8974, 21)\n",
      "857 (11474, 21)\n",
      "858 (10209, 21)\n",
      "859 (5792, 21)\n",
      "860 (15063, 21)\n",
      "861 (2051, 21)\n",
      "862 (14268, 21)\n",
      "863 (12667, 21)\n",
      "864 (8567, 21)\n",
      "865 (6816, 21)\n",
      "866 (12847, 21)\n",
      "867 (3445, 21)\n",
      "868 (3093, 21)\n",
      "869 (3654, 21)\n",
      "870 (4767, 21)\n",
      "871 (2580, 21)\n",
      "872 (1959, 21)\n",
      "873 (2635, 21)\n",
      "874 (2149, 21)\n",
      "875 (3447, 21)\n",
      "876 (1192, 21)\n",
      "877 (9698, 21)\n",
      "878 (5792, 21)\n",
      "879 (1436, 21)\n",
      "880 (1056, 21)\n",
      "881 (1609, 21)\n",
      "882 (6256, 21)\n",
      "883 (2031, 21)\n",
      "884 (2269, 21)\n",
      "885 (714, 21)\n",
      "886 (1713, 21)\n",
      "887 (367, 21)\n",
      "888 (626, 21)\n",
      "889 (371, 21)\n",
      "890 (5471, 21)\n",
      "891 (4205, 21)\n",
      "892 (4253, 21)\n",
      "893 (1944, 21)\n",
      "894 (3602, 21)\n",
      "895 (1937, 21)\n",
      "896 (993, 21)\n",
      "897 (4557, 21)\n",
      "898 (4171, 21)\n",
      "899 (2500, 21)\n",
      "900 (533, 21)\n",
      "901 (2253, 21)\n",
      "902 (12584, 21)\n",
      "903 (4200, 21)\n",
      "904 (608, 21)\n",
      "(4988568, 21)\n",
      "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
      "       'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698',\n",
      "       'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702',\n",
      "       'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14405</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69116</td>\n",
       "      <td>rs757299236</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:1</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14410</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69134</td>\n",
       "      <td>rs781394307</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:9</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:5</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14414</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69149</td>\n",
       "      <td>rs1458521218</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:3</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:1</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14419</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69184</td>\n",
       "      <td>rs1231643173</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14422</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69200</td>\n",
       "      <td>rs1453226491</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:5</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:5</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904165833</th>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>275107</td>\n",
       "      <td>rs772563100</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>5918:1</td>\n",
       "      <td>...</td>\n",
       "      <td>14:0</td>\n",
       "      <td>584:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>22:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>2218:0</td>\n",
       "      <td>592:0</td>\n",
       "      <td>36:0</td>\n",
       "      <td>8764:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904165847</th>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>275289</td>\n",
       "      <td>rs1469647256</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>16262:1</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>350:0</td>\n",
       "      <td>18:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>3936:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20586:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904165849</th>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>275302</td>\n",
       "      <td>rs151150114</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>22180:4</td>\n",
       "      <td>...</td>\n",
       "      <td>14:0</td>\n",
       "      <td>602:0</td>\n",
       "      <td>350:0</td>\n",
       "      <td>18:0</td>\n",
       "      <td>22:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>6154:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>36:0</td>\n",
       "      <td>29348:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904165850</th>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>275304</td>\n",
       "      <td>rs1334679620</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>16262:1</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>350:0</td>\n",
       "      <td>18:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>3936:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20586:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904165852</th>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>275308</td>\n",
       "      <td>rs11574606</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>23482:3160</td>\n",
       "      <td>...</td>\n",
       "      <td>1808:0</td>\n",
       "      <td>676:62</td>\n",
       "      <td>6:0</td>\n",
       "      <td>22:1</td>\n",
       "      <td>1188:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>5148:465</td>\n",
       "      <td>684:62</td>\n",
       "      <td>2996:0</td>\n",
       "      <td>32338:3688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4988568 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CHROM     POS            ID REF ALT QUAL FILTER INFO  \\\n",
       "14405        NC_000001.11   69116   rs757299236   G   A    .      .    .   \n",
       "14410        NC_000001.11   69134   rs781394307   A   G    .      .    .   \n",
       "14414        NC_000001.11   69149  rs1458521218   T   A    .      .    .   \n",
       "14419        NC_000001.11   69184  rs1231643173   G   A    .      .    .   \n",
       "14422        NC_000001.11   69200  rs1453226491   T   C    .      .    .   \n",
       "...                   ...     ...           ...  ..  ..  ...    ...  ...   \n",
       "904165833  NW_003571061.2  275107   rs772563100   T   C    .      .    .   \n",
       "904165847  NW_003571061.2  275289  rs1469647256   A   G    .      .    .   \n",
       "904165849  NW_003571061.2  275302   rs151150114   G   C    .      .    .   \n",
       "904165850  NW_003571061.2  275304  rs1334679620   A   G    .      .    .   \n",
       "904165852  NW_003571061.2  275308    rs11574606   G   A    .      .    .   \n",
       "\n",
       "          FORMAT SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699  \\\n",
       "14405      AN:AC       7618:0  ...         84:0       2708:0        146:0   \n",
       "14410      AN:AC       7618:9  ...         84:0       2708:0        146:0   \n",
       "14414      AN:AC       7618:3  ...         84:0       2708:0        146:1   \n",
       "14419      AN:AC       7618:0  ...         84:0       2708:0        146:0   \n",
       "14422      AN:AC       7618:0  ...         84:0       2708:5        146:0   \n",
       "...          ...          ...  ...          ...          ...          ...   \n",
       "904165833  AN:AC       5918:1  ...         14:0        584:0          0:0   \n",
       "904165847  AN:AC      16262:1  ...          0:0         20:0        350:0   \n",
       "904165849  AN:AC      22180:4  ...         14:0        602:0        350:0   \n",
       "904165850  AN:AC      16262:1  ...          0:0         20:0        350:0   \n",
       "904165852  AN:AC   23482:3160  ...       1808:0       676:62          6:0   \n",
       "\n",
       "          SAMN10492700 SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703  \\\n",
       "14405            610:0         24:0         94:0        470:1       2816:0   \n",
       "14410            610:5         24:0         94:0        470:0       2816:0   \n",
       "14414            610:0         24:0         94:0        470:0       2816:0   \n",
       "14419            610:0         24:0         94:0        470:0       2816:0   \n",
       "14422            610:0         24:0         94:0        470:0       2816:5   \n",
       "...                ...          ...          ...          ...          ...   \n",
       "904165833          0:0         22:0          0:0       2218:0        592:0   \n",
       "904165847         18:0          0:0          0:0       3936:0         20:0   \n",
       "904165849         18:0         22:0          0:0       6154:0        610:0   \n",
       "904165850         18:0          0:0          0:0       3936:0         20:0   \n",
       "904165852         22:1       1188:0          0:0     5148:465       684:62   \n",
       "\n",
       "          SAMN10492704 SAMN10492705  \n",
       "14405            108:0      11862:1  \n",
       "14410            108:0     11862:14  \n",
       "14414            108:0      11862:4  \n",
       "14419            108:0      11862:0  \n",
       "14422            108:0      11862:5  \n",
       "...                ...          ...  \n",
       "904165833         36:0       8764:1  \n",
       "904165847          0:0      20586:1  \n",
       "904165849         36:0      29348:4  \n",
       "904165850          0:0      20586:1  \n",
       "904165852       2996:0   32338:3688  \n",
       "\n",
       "[4988568 rows x 21 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do not run this unless you changed something before, it takes 1 hour\n",
    "# better use the following block\n",
    "from utils.column_names_reader import get_col_names_from_gzip\n",
    "filepath = home_dir+\"data/ALFA_population_freq/freq.vcf.gz\"\n",
    "col_names = get_col_names_from_gzip(filepath, \"#CHROM\")\n",
    "alfa_df_iterator = pd.read_csv(filepath, compression='gzip', chunksize=1000000, comment=\"#\", names=col_names, delim_whitespace=False, sep=\"\\t\")\n",
    "# alfa_df_iterator.__next__() # first df \n",
    "\n",
    "chunk_list = []\n",
    "for i, alfa_df in enumerate(alfa_df_iterator):\n",
    "    chunk = alfa_df[alfa_df[\"ID\"].isin(dbsnp_ids)]\n",
    "    chunk_list.append(chunk)\n",
    "    print(i, chunk.shape)\n",
    "    # if i==0: break\n",
    "\n",
    "filtered_alfa_df = pd.concat(chunk_list)  \n",
    "print(filtered_alfa_df.shape)\n",
    "print(filtered_alfa_df.columns)\n",
    "filtered_alfa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
      "       'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698',\n",
      "       'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702',\n",
      "       'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705'],\n",
      "      dtype='object')\n",
      "(4988568, 21)\n",
      "(4988568,)\n"
     ]
    }
   ],
   "source": [
    "filtered_alpha_filepath = home_dir+\"data/temp_popu_freq_data/alfa_filtered.tsv\"\n",
    "# filtered_alfa_df.to_csv(filtered_alpha_filepath, index=False, sep=\"\\t\")\n",
    "filtered_alfa_df = pd.read_csv(filtered_alpha_filepath, sep=\"\\t\")\n",
    "print(filtered_alfa_df.columns)\n",
    "print(filtered_alfa_df.shape)\n",
    "print(filtered_alfa_df[\"ID\"].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1497138</th>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>141374104</td>\n",
       "      <td>rs200348921</td>\n",
       "      <td>T</td>\n",
       "      <td>C,A,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>20198:13,1,2</td>\n",
       "      <td>...</td>\n",
       "      <td>112:0,0,0</td>\n",
       "      <td>3368:0,0,0</td>\n",
       "      <td>146:0,0,0</td>\n",
       "      <td>610:0,0,0</td>\n",
       "      <td>56:0,0,0</td>\n",
       "      <td>98:0,0,0</td>\n",
       "      <td>2780:9,0,0</td>\n",
       "      <td>3490:0,0,0</td>\n",
       "      <td>168:0,0,0</td>\n",
       "      <td>27490:22,1,2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                CHROM        POS           ID REF    ALT QUAL FILTER INFO  \\\n",
       "1497138  NC_000005.10  141374104  rs200348921   T  C,A,G    .      .    .   \n",
       "\n",
       "        FORMAT  SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699  \\\n",
       "1497138  AN:AC  20198:13,1,2  ...    112:0,0,0   3368:0,0,0    146:0,0,0   \n",
       "\n",
       "        SAMN10492700 SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703  \\\n",
       "1497138    610:0,0,0     56:0,0,0     98:0,0,0   2780:9,0,0   3490:0,0,0   \n",
       "\n",
       "        SAMN10492704  SAMN10492705  \n",
       "1497138    168:0,0,0  27490:22,1,2  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_alfa_df[filtered_alfa_df[\"ID\"]==\"rs200348921\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "C                           1664374\n",
       "G                           1652765\n",
       "A                            836569\n",
       "T                            834852\n",
       "TA                                1\n",
       "CCACTTCTTTAATCACCACCACCA          1\n",
       "CCAACCT                           1\n",
       "CCTT                              1\n",
       "ATAGCCTCTCCTCCTGTTCACGT           1\n",
       "GCCT                              1\n",
       "AT                                1\n",
       "ATCT                              1\n",
       "Name: REF, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_alfa_df[\"REF\"].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to MANE proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19120, 14)\n",
      "Index(['#NCBI_GeneID', 'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name',\n",
      "       'RefSeq_nuc', 'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot',\n",
      "       'MANE_status', 'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand'],\n",
      "      dtype='object')\n",
      "MANE Select           19062\n",
      "MANE Plus Clinical       58\n",
      "Name: MANE_status, dtype: int64\n",
      "CACNA1D         2\n",
      "CACNA1C         2\n",
      "DYNC2H1         2\n",
      "SLC39A14        2\n",
      "CLPB            2\n",
      "               ..\n",
      "ZNF432          1\n",
      "SAFB2           1\n",
      "DZIP3           1\n",
      "MARF1           1\n",
      "LOC122539214    1\n",
      "Name: symbol, Length: 19062, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>GeneID:8291</td>\n",
       "      <td>ENSG00000135636.16</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NM_003494.4</td>\n",
       "      <td>NP_003485.1</td>\n",
       "      <td>ENST00000258104.8</td>\n",
       "      <td>ENSP00000258104.3</td>\n",
       "      <td>MANE Plus Clinical</td>\n",
       "      <td>2</td>\n",
       "      <td>71453561</td>\n",
       "      <td>71686763</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5408</th>\n",
       "      <td>GeneID:8291</td>\n",
       "      <td>ENSG00000135636.16</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NM_001130987.2</td>\n",
       "      <td>NP_001124459.1</td>\n",
       "      <td>ENST00000410020.8</td>\n",
       "      <td>ENSP00000386881.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>2</td>\n",
       "      <td>71466699</td>\n",
       "      <td>71686763</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #NCBI_GeneID        Ensembl_Gene    HGNC_ID symbol       name  \\\n",
       "5407  GeneID:8291  ENSG00000135636.16  HGNC:3097   DYSF  dysferlin   \n",
       "5408  GeneID:8291  ENSG00000135636.16  HGNC:3097   DYSF  dysferlin   \n",
       "\n",
       "          RefSeq_nuc     RefSeq_prot        Ensembl_nuc       Ensembl_prot  \\\n",
       "5407     NM_003494.4     NP_003485.1  ENST00000258104.8  ENSP00000258104.3   \n",
       "5408  NM_001130987.2  NP_001124459.1  ENST00000410020.8  ENSP00000386881.3   \n",
       "\n",
       "             MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "5407  MANE Plus Clinical          2   71453561  71686763          +  \n",
       "5408         MANE Select          2   71466699  71686763          +  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loading MANE data\n",
    "mane_df = pd.read_csv(home_dir+\"data/refseq/MANE.GRCh38.v1.0.summary.txt.gz\", compression='gzip', delim_whitespace=False, sep=\"\\t\") # MANE refseq-protein mapping\n",
    "print(mane_df.shape)\n",
    "print(mane_df.columns)\n",
    "\n",
    "print(mane_df[\"MANE_status\"].value_counts())\n",
    "print(mane_df[\"symbol\"].value_counts()) # some gene can be mapped to multiple proteins because of MANE select and Plus clinical.\n",
    "\n",
    "mane_short_df = mane_df[[\"symbol\", \"RefSeq_prot\", \"MANE_status\"]]\n",
    "# mane_short_df.value_counts()\n",
    "mane_df[mane_df[\"symbol\"]==\"DYSF\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some dbSNP-ids are annotated to multiple genes. ie rs756964411 is mapped to PSMB7, NEK6, LOC100129034. \n",
    "# we map gene-symbols to the MANE gene-symbols, refseq proteins.\n",
    "# this block takes >1 hour to run, better use the saved data from the next block\n",
    "data = []\n",
    "for i, row in enumerate(initially_cleaned_df.itertuples(index=False)):\n",
    "    gene_symbols_list = [gene.split(\":\")[0] for gene in row.genes.split(\",\")]\n",
    "    mane_things = mane_short_df[mane_short_df[\"symbol\"].isin(gene_symbols_list)]\n",
    "    mane_genes = mane_things[\"symbol\"].unique().tolist()\n",
    "    mane_proteins = mane_things[\"RefSeq_prot\"].unique().tolist()\n",
    "\n",
    "    variations = row.variations.split(\",\")\n",
    "    mane_protein_variants = []\n",
    "    for variation in variations:\n",
    "        prot = variation.split(\":\")[0]\n",
    "        if prot in mane_proteins:\n",
    "            mane_protein_variants.append(variation)\n",
    "\n",
    "    datam = row._asdict()\n",
    "    datam[\"mane_genes\"] = \",\".join(mane_genes)\n",
    "    datam[\"mane_proteins\"] = \",\".join(mane_proteins)\n",
    "    datam[\"mane_protein_variants\"] = \",\".join(mane_protein_variants)\n",
    "\n",
    "    data.append(datam)\n",
    "    \n",
    "    if i%100000==0: print(i)\n",
    "    # if i==5: break\n",
    "\n",
    "variants_with_mane_df = pd.DataFrame(data) \n",
    "print(variants_with_mane_df.shape)\n",
    "print(variants_with_mane_df.columns)\n",
    "print(variants_with_mane_df[\"snp_id\"].unique().shape)\n",
    "variants_with_mane_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4988781, 17)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'mane_genes', 'mane_proteins',\n",
      "       'mane_protein_variants'],\n",
      "      dtype='object')\n",
      "(4988781,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>genes</th>\n",
       "      <th>mafs</th>\n",
       "      <th>variations</th>\n",
       "      <th>mane_genes</th>\n",
       "      <th>mane_proteins</th>\n",
       "      <th>mane_protein_variants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1491004261</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128176888</td>\n",
       "      <td>NC_000002.12:128176887:T:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non_coding_transcript_variant,coding_sequence_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>UGGT1:56886</td>\n",
       "      <td>GnomAD_exomes:C=0.000004/1,TOPMED:C=0.000008/2...</td>\n",
       "      <td>NP_064505.1:p.Ile1205Thr</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>NP_064505.1:p.Ile1205Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1491004253</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>5:53042133</td>\n",
       "      <td>NC_000005.10:53042132:G:A,NC_000005.10:5304213...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 07:59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>ITGA2:3673</td>\n",
       "      <td>GnomAD_exomes:T=0.000004/1,TOPMED:A=0.000004/1...</td>\n",
       "      <td>NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys</td>\n",
       "      <td>ITGA2</td>\n",
       "      <td>NP_002194.2</td>\n",
       "      <td>NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1491004226</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>20:33663605</td>\n",
       "      <td>NC_000020.11:33663604:C:G,NC_000020.11:3366360...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/13 17:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,5_prime_UTR_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>NECAB3:63941,C20orf144:128864</td>\n",
       "      <td>ALFA:T=0./0</td>\n",
       "      <td>NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val</td>\n",
       "      <td>NECAB3,C20orf144</td>\n",
       "      <td>NP_112509.3,NP_543015.1</td>\n",
       "      <td>NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1491003662</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:26946984</td>\n",
       "      <td>NC_000002.12:26946983:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 05:04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>DPYSL5:56896</td>\n",
       "      <td>GnomAD:A=0.000007/1,TOPMED:A=0.000004/1,ALFA:A...</td>\n",
       "      <td>NP_064519.2:p.Gly562Ser,NP_001240653.1:p.Gly56...</td>\n",
       "      <td>DPYSL5</td>\n",
       "      <td>NP_064519.2</td>\n",
       "      <td>NP_064519.2:p.Gly562Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1491003598</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269218</td>\n",
       "      <td>NC_000001.11:205269217:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>TMCC2:9911</td>\n",
       "      <td>GnomAD_exomes:G=0.000004/1,TOMMO:G=0.000035/1,...</td>\n",
       "      <td>NP_055673.2:p.Lys339Arg,NP_001229854.1:p.Lys26...</td>\n",
       "      <td>TMCC2</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>NP_055673.2:p.Lys339Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988776</th>\n",
       "      <td>rs334</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>NC_000011.10:5227001:T:A,NC_000011.10:5227001:...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/16 14:17</td>\n",
       "      <td>protective,other,pathogenic,benign,likely-beni...</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HBB:3043</td>\n",
       "      <td>1000Genomes:A=0.028732/144,ExAC:A=0.004384/532...</td>\n",
       "      <td>NP_000509.1:p.Glu7Val,NP_000509.1:p.Glu7Gly,NP...</td>\n",
       "      <td>HBB</td>\n",
       "      <td>NP_000509.1</td>\n",
       "      <td>NP_000509.1:p.Glu7Val,NP_000509.1:p.Glu7Gly,NP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988777</th>\n",
       "      <td>rs300</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19959376</td>\n",
       "      <td>NC_000008.11:19959375:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>benign,benign-likely-benign</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>LPL:4023</td>\n",
       "      <td>1000Genomes:G=0.003592/18,ExAC:G=0.001084/131,...</td>\n",
       "      <td>NP_000228.1:p.Thr379Ala</td>\n",
       "      <td>LPL</td>\n",
       "      <td>NP_000228.1</td>\n",
       "      <td>NP_000228.1:p.Thr379Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988778</th>\n",
       "      <td>rs298</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19959349</td>\n",
       "      <td>NC_000008.11:19959348:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>LPL:4023</td>\n",
       "      <td>ExAC:A=0.00005/6,GnomAD:A=0.000093/13,GnomAD_e...</td>\n",
       "      <td>NP_000228.1:p.Val370Met</td>\n",
       "      <td>LPL</td>\n",
       "      <td>NP_000228.1</td>\n",
       "      <td>NP_000228.1:p.Val370Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988779</th>\n",
       "      <td>rs268</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NC_000008.11:19956017:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>conflicting-interpretations-of-pathogenicity,u...</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>LPL:4023</td>\n",
       "      <td>1000Genomes:G=0.004685/23,ALSPAC:G=0.018682/72...</td>\n",
       "      <td>NP_000228.1:p.Asn318Ser</td>\n",
       "      <td>LPL</td>\n",
       "      <td>NP_000228.1</td>\n",
       "      <td>NP_000228.1:p.Asn318Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988780</th>\n",
       "      <td>rs47</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7:11541494</td>\n",
       "      <td>NC_000007.14:11541493:T:A,NC_000007.14:1154149...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 07:37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>THSD7A:221981</td>\n",
       "      <td>1000Genomes:T=0.164741/825,ALSPAC:T=0.248573/9...</td>\n",
       "      <td>NP_056019.1:p.Asn583Tyr,NP_056019.1:p.Asn583As...</td>\n",
       "      <td>THSD7A</td>\n",
       "      <td>NP_056019.1</td>\n",
       "      <td>NP_056019.1:p.Asn583Tyr,NP_056019.1:p.Asn583As...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4988781 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               snp_id           acc       chrpos  \\\n",
       "0        rs1491004261  NC_000002.12  2:128176888   \n",
       "1        rs1491004253  NC_000005.10   5:53042133   \n",
       "2        rs1491004226  NC_000020.11  20:33663605   \n",
       "3        rs1491003662  NC_000002.12   2:26946984   \n",
       "4        rs1491003598  NC_000001.11  1:205269218   \n",
       "...               ...           ...          ...   \n",
       "4988776         rs334  NC_000011.10   11:5227002   \n",
       "4988777         rs300  NC_000008.11   8:19959376   \n",
       "4988778         rs298  NC_000008.11   8:19959349   \n",
       "4988779         rs268  NC_000008.11   8:19956018   \n",
       "4988780          rs47  NC_000007.14   7:11541494   \n",
       "\n",
       "                                                      spdi  tax_id snp_class  \\\n",
       "0                               NC_000002.12:128176887:T:C    9606       snv   \n",
       "1        NC_000005.10:53042132:G:A,NC_000005.10:5304213...    9606       snv   \n",
       "2        NC_000020.11:33663604:C:G,NC_000020.11:3366360...    9606       snv   \n",
       "3                                NC_000002.12:26946983:G:A    9606       snv   \n",
       "4                               NC_000001.11:205269217:A:G    9606       snv   \n",
       "...                                                    ...     ...       ...   \n",
       "4988776  NC_000011.10:5227001:T:A,NC_000011.10:5227001:...    9606       snv   \n",
       "4988777                          NC_000008.11:19959375:A:G    9606       snv   \n",
       "4988778                          NC_000008.11:19959348:G:A    9606       snv   \n",
       "4988779                          NC_000008.11:19956017:A:G    9606       snv   \n",
       "4988780  NC_000007.14:11541493:T:A,NC_000007.14:1154149...    9606       snv   \n",
       "\n",
       "              create_date       update_date  \\\n",
       "0        2017/11/09 09:55  2022/10/12 21:08   \n",
       "1        2017/11/09 09:55  2022/10/13 07:59   \n",
       "2        2017/11/09 09:55  2022/10/13 17:00   \n",
       "3        2017/11/09 09:55  2022/10/17 05:04   \n",
       "4        2017/11/09 09:55  2022/10/12 18:08   \n",
       "...                   ...               ...   \n",
       "4988776  2000/09/19 17:02  2022/10/16 14:17   \n",
       "4988777  2000/09/19 17:02  2022/10/14 11:06   \n",
       "4988778  2000/09/19 17:02  2022/10/14 11:06   \n",
       "4988779  2000/09/19 17:02  2022/10/14 11:06   \n",
       "4988780  2000/09/19 17:02  2022/10/14 07:37   \n",
       "\n",
       "                                     clinical_significance  \\\n",
       "0                                                      NaN   \n",
       "1                                                      NaN   \n",
       "2                                                      NaN   \n",
       "3                                                      NaN   \n",
       "4                                                      NaN   \n",
       "...                                                    ...   \n",
       "4988776  protective,other,pathogenic,benign,likely-beni...   \n",
       "4988777                        benign,benign-likely-benign   \n",
       "4988778                                                NaN   \n",
       "4988779  conflicting-interpretations-of-pathogenicity,u...   \n",
       "4988780                                                NaN   \n",
       "\n",
       "                                                 fxn_class  \\\n",
       "0        non_coding_transcript_variant,coding_sequence_...   \n",
       "1        coding_sequence_variant,non_coding_transcript_...   \n",
       "2        genic_upstream_transcript_variant,5_prime_UTR_...   \n",
       "3                 coding_sequence_variant,missense_variant   \n",
       "4                 coding_sequence_variant,missense_variant   \n",
       "...                                                    ...   \n",
       "4988776           missense_variant,coding_sequence_variant   \n",
       "4988777           missense_variant,coding_sequence_variant   \n",
       "4988778           missense_variant,coding_sequence_variant   \n",
       "4988779           missense_variant,coding_sequence_variant   \n",
       "4988780           missense_variant,coding_sequence_variant   \n",
       "\n",
       "                               validated                          genes  \\\n",
       "0        by-frequency,by-alfa,by-cluster                    UGGT1:56886   \n",
       "1        by-frequency,by-alfa,by-cluster                     ITGA2:3673   \n",
       "2        by-frequency,by-alfa,by-cluster  NECAB3:63941,C20orf144:128864   \n",
       "3        by-frequency,by-alfa,by-cluster                   DPYSL5:56896   \n",
       "4        by-frequency,by-alfa,by-cluster                     TMCC2:9911   \n",
       "...                                  ...                            ...   \n",
       "4988776  by-frequency,by-alfa,by-cluster                       HBB:3043   \n",
       "4988777  by-frequency,by-alfa,by-cluster                       LPL:4023   \n",
       "4988778  by-frequency,by-alfa,by-cluster                       LPL:4023   \n",
       "4988779  by-frequency,by-alfa,by-cluster                       LPL:4023   \n",
       "4988780  by-frequency,by-alfa,by-cluster                  THSD7A:221981   \n",
       "\n",
       "                                                      mafs  \\\n",
       "0        GnomAD_exomes:C=0.000004/1,TOPMED:C=0.000008/2...   \n",
       "1        GnomAD_exomes:T=0.000004/1,TOPMED:A=0.000004/1...   \n",
       "2                                              ALFA:T=0./0   \n",
       "3        GnomAD:A=0.000007/1,TOPMED:A=0.000004/1,ALFA:A...   \n",
       "4        GnomAD_exomes:G=0.000004/1,TOMMO:G=0.000035/1,...   \n",
       "...                                                    ...   \n",
       "4988776  1000Genomes:A=0.028732/144,ExAC:A=0.004384/532...   \n",
       "4988777  1000Genomes:G=0.003592/18,ExAC:G=0.001084/131,...   \n",
       "4988778  ExAC:A=0.00005/6,GnomAD:A=0.000093/13,GnomAD_e...   \n",
       "4988779  1000Genomes:G=0.004685/23,ALSPAC:G=0.018682/72...   \n",
       "4988780  1000Genomes:T=0.164741/825,ALSPAC:T=0.248573/9...   \n",
       "\n",
       "                                                variations        mane_genes  \\\n",
       "0                                 NP_064505.1:p.Ile1205Thr             UGGT1   \n",
       "1            NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys             ITGA2   \n",
       "2            NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val  NECAB3,C20orf144   \n",
       "3        NP_064519.2:p.Gly562Ser,NP_001240653.1:p.Gly56...            DPYSL5   \n",
       "4        NP_055673.2:p.Lys339Arg,NP_001229854.1:p.Lys26...             TMCC2   \n",
       "...                                                    ...               ...   \n",
       "4988776  NP_000509.1:p.Glu7Val,NP_000509.1:p.Glu7Gly,NP...               HBB   \n",
       "4988777                            NP_000228.1:p.Thr379Ala               LPL   \n",
       "4988778                            NP_000228.1:p.Val370Met               LPL   \n",
       "4988779                            NP_000228.1:p.Asn318Ser               LPL   \n",
       "4988780  NP_056019.1:p.Asn583Tyr,NP_056019.1:p.Asn583As...            THSD7A   \n",
       "\n",
       "                   mane_proteins  \\\n",
       "0                    NP_064505.1   \n",
       "1                    NP_002194.2   \n",
       "2        NP_112509.3,NP_543015.1   \n",
       "3                    NP_064519.2   \n",
       "4                    NP_055673.2   \n",
       "...                          ...   \n",
       "4988776              NP_000509.1   \n",
       "4988777              NP_000228.1   \n",
       "4988778              NP_000228.1   \n",
       "4988779              NP_000228.1   \n",
       "4988780              NP_056019.1   \n",
       "\n",
       "                                     mane_protein_variants  \n",
       "0                                 NP_064505.1:p.Ile1205Thr  \n",
       "1            NP_002194.2:p.Trp69Ter,NP_002194.2:p.Trp69Cys  \n",
       "2            NP_543015.1:p.Ala67Gly,NP_543015.1:p.Ala67Val  \n",
       "3                                  NP_064519.2:p.Gly562Ser  \n",
       "4                                  NP_055673.2:p.Lys339Arg  \n",
       "...                                                    ...  \n",
       "4988776  NP_000509.1:p.Glu7Val,NP_000509.1:p.Glu7Gly,NP...  \n",
       "4988777                            NP_000228.1:p.Thr379Ala  \n",
       "4988778                            NP_000228.1:p.Val370Met  \n",
       "4988779                            NP_000228.1:p.Asn318Ser  \n",
       "4988780  NP_056019.1:p.Asn583Tyr,NP_056019.1:p.Asn583As...  \n",
       "\n",
       "[4988781 rows x 17 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath = home_dir+\"data/temp_popu_freq_data/variants_with_mane.tsv\"\n",
    "# variants_with_mane_df.to_csv(filepath, index=False, sep=\"\\t\")\n",
    "variants_with_mane_df = pd.read_csv(filepath, sep=\"\\t\")\n",
    "print(variants_with_mane_df.shape)\n",
    "print(variants_with_mane_df.columns)\n",
    "print(variants_with_mane_df[\"snp_id\"].unique().shape)\n",
    "variants_with_mane_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NAN rows removed: 174080\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "(4814701, 17)\n",
      "4814701\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'mane_genes', 'mane_proteins',\n",
      "       'mane_protein_variants'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "variants_with_refseq_prot_variants_df = variants_with_mane_df.copy()\n",
    "variants_with_refseq_prot_variants_df = filter_remove_null_nan_empty_entries(variants_with_refseq_prot_variants_df, a_col_name=\"mane_protein_variants\") # NNE: null, nan or empty variants\n",
    "print(variants_with_refseq_prot_variants_df.shape)\n",
    "print(variants_with_refseq_prot_variants_df[\"snp_id\"].unique().shape[0])\n",
    "print(variants_with_refseq_prot_variants_df.columns)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to ALFA population freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4814631, 38)\n",
      "(4814631,)\n"
     ]
    }
   ],
   "source": [
    "variants_with_popu_count_df = pd.merge(left=variants_with_refseq_prot_variants_df, right=filtered_alfa_df, left_on=\"snp_id\", right_on=\"ID\", how=\"inner\")\n",
    "print(variants_with_popu_count_df.shape)\n",
    "print(variants_with_popu_count_df[\"snp_id\"].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "variants_with_popu_count_df.to_csv(home_dir+\"data/temp_popu_freq_data/snps_with_mane_and_popu_count_raw.tsv\", index=False, sep=\"\\t\")\n",
    "# this is ~3gb, so saving a short form by selecting cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_dict = {\"snp_id\":\"snp_id\", \"CHROM\":\"chrom_acc_version\", \"POS\":\"chrom_pos_1indexed\", \"REF\":\"ref_allele\", \"ALT\":\"alt_alleles\", \n",
    "             \"mane_genes\":\"mane_genes\", \"mane_proteins\":\"mane_proteins\", \"mane_protein_variants\":\"mane_protein_variants\", \n",
    "             \"SAMN10492705\":\"alfa_popu_counts\"}\n",
    "\n",
    "variants_selected_cols_df = variants_with_popu_count_df[cols_dict.keys()].copy()\n",
    "variants_selected_cols_df.rename(columns=cols_dict, inplace=True)\n",
    "variants_selected_cols_df.to_csv(home_dir+\"data/temp_popu_freq_data/snps_with_mane_and_popu_count_raw_short.tsv\", index=False, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
