{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19120, 14)\n",
      "19062\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:1</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>NP_570602.2</td>\n",
       "      <td>ENST00000263100.8</td>\n",
       "      <td>ENSP00000263100.2</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>19</td>\n",
       "      <td>58345183</td>\n",
       "      <td>58353492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:2</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>NP_000005.3</td>\n",
       "      <td>ENST00000318602.12</td>\n",
       "      <td>ENSP00000323929.8</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>12</td>\n",
       "      <td>9067708</td>\n",
       "      <td>9115919</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:9</td>\n",
       "      <td>ENSG00000171428.15</td>\n",
       "      <td>HGNC:7645</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>NM_000662.8</td>\n",
       "      <td>NP_000653.3</td>\n",
       "      <td>ENST00000307719.9</td>\n",
       "      <td>ENSP00000307218.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18210109</td>\n",
       "      <td>18223689</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:10</td>\n",
       "      <td>ENSG00000156006.5</td>\n",
       "      <td>HGNC:7646</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>NM_000015.3</td>\n",
       "      <td>NP_000006.2</td>\n",
       "      <td>ENST00000286479.4</td>\n",
       "      <td>ENSP00000286479.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18391282</td>\n",
       "      <td>18401218</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:12</td>\n",
       "      <td>ENSG00000196136.18</td>\n",
       "      <td>HGNC:16</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>serpin family A member 3</td>\n",
       "      <td>NM_001085.5</td>\n",
       "      <td>NP_001076.2</td>\n",
       "      <td>ENST00000393078.5</td>\n",
       "      <td>ENSP00000376793.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>14</td>\n",
       "      <td>94612391</td>\n",
       "      <td>94624053</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #NCBI_GeneID        Ensembl_Gene    HGNC_ID    symbol  \\\n",
       "0     GeneID:1  ENSG00000121410.12     HGNC:5      A1BG   \n",
       "1     GeneID:2  ENSG00000175899.15     HGNC:7       A2M   \n",
       "2     GeneID:9  ENSG00000171428.15  HGNC:7645      NAT1   \n",
       "3    GeneID:10   ENSG00000156006.5  HGNC:7646      NAT2   \n",
       "4    GeneID:12  ENSG00000196136.18    HGNC:16  SERPINA3   \n",
       "\n",
       "                       name   RefSeq_nuc  RefSeq_prot         Ensembl_nuc  \\\n",
       "0    alpha-1-B glycoprotein  NM_130786.4  NP_570602.2   ENST00000263100.8   \n",
       "1     alpha-2-macroglobulin  NM_000014.6  NP_000005.3  ENST00000318602.12   \n",
       "2     N-acetyltransferase 1  NM_000662.8  NP_000653.3   ENST00000307719.9   \n",
       "3     N-acetyltransferase 2  NM_000015.3  NP_000006.2   ENST00000286479.4   \n",
       "4  serpin family A member 3  NM_001085.5  NP_001076.2   ENST00000393078.5   \n",
       "\n",
       "        Ensembl_prot  MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "0  ENSP00000263100.2  MANE Select         19   58345183  58353492          -  \n",
       "1  ENSP00000323929.8  MANE Select         12    9067708   9115919          -  \n",
       "2  ENSP00000307218.4  MANE Select          8   18210109  18223689          +  \n",
       "3  ENSP00000286479.3  MANE Select          8   18391282  18401218          +  \n",
       "4  ENSP00000376793.3  MANE Select         14   94612391  94624053          +  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2refseq_mapping_filepath = home_dir+\"data/refseq/MANE.GRCh38.v1.0.summary.txt.gz\" # \"symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\" # \"Symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/refseq/HumanLRGRefSeqGene.txt.gz\"  # \"Symbol\" is the Gene-symbol col\n",
    "\n",
    "gene_symbol_col_name = \"symbol\" if gene2refseq_mapping_filepath.__contains__(\"MANE\") else \"Symbol\"\n",
    "gene2refseq_mapping_df = pd.read_csv(gene2refseq_mapping_filepath, compression='gzip', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "\n",
    "print(gene2refseq_mapping_df.shape)\n",
    "print(len(gene2refseq_mapping_df[gene_symbol_col_name].unique()))\n",
    "gene2refseq_mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n",
      "Downloaded:  (668003, 14)\n",
      "Ater removing duplicate SNPs (actual downloaded):  (667303, 14)\n",
      "After removing empty protein associated variants:  (643316, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>genes</th>\n",
       "      <th>mafs</th>\n",
       "      <th>variations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1490996759</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860386</td>\n",
       "      <td>NC_000016.10:71860385:C:A,NC_000016.10:7186038...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant,missense_variant,stop_gain...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>ZNF821:55565</td>\n",
       "      <td>1000Genomes:A=0.000156/1,GnomAD_exomes:T=0.000...</td>\n",
       "      <td>NP_001188481.1:p.Glu291Ter,NP_001188481.1:p.Gl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1490983839</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11:1184843</td>\n",
       "      <td>NC_000011.10:1184842:C:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 14:23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>MUC5AC:4586</td>\n",
       "      <td>1000Genomes:G=0.000781/5,GnomAD:G=0.001191/167...</td>\n",
       "      <td>NP_001291288.1:p.Pro2233Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1490970779</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19:40192608</td>\n",
       "      <td>NC_000019.10:40192607:C:G,NC_000019.10:4019260...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 01:20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant,genic...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>MAP3K10:4294</td>\n",
       "      <td>1000Genomes:T=0.000156/1,TOPMED:G=0.000011/3,A...</td>\n",
       "      <td>NP_002437.2:p.Pro193Ala,NP_002437.2:p.Pro193Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1490958560</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:131264510</td>\n",
       "      <td>NC_000002.12:131264509:G:A,NC_000002.12:131264...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>POTEE:445582</td>\n",
       "      <td>1000Genomes:T=0.000156/1,TOPMED:A=0.000004/1,A...</td>\n",
       "      <td>NP_001077007.1:p.Ala1019Thr,NP_001077007.1:p.A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1490935890</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17:82781702</td>\n",
       "      <td>NC_000017.11:82781701:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 00:32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,upstream_tra...</td>\n",
       "      <td>by-frequency,by-cluster</td>\n",
       "      <td>TBCD:6904</td>\n",
       "      <td>1000Genomes:G=0.0002/1</td>\n",
       "      <td>NP_005984.3:p.Asp251Gly</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       snp_id           acc       chrpos  \\\n",
       "0  1490996759  NC_000016.10  16:71860386   \n",
       "1  1490983839  NC_000011.10   11:1184843   \n",
       "2  1490970779  NC_000019.10  19:40192608   \n",
       "3  1490958560  NC_000002.12  2:131264510   \n",
       "4  1490935890  NC_000017.11  17:82781702   \n",
       "\n",
       "                                                spdi  tax_id snp_class  \\\n",
       "0  NC_000016.10:71860385:C:A,NC_000016.10:7186038...    9606       snv   \n",
       "1                           NC_000011.10:1184842:C:G    9606       snv   \n",
       "2  NC_000019.10:40192607:C:G,NC_000019.10:4019260...    9606       snv   \n",
       "3  NC_000002.12:131264509:G:A,NC_000002.12:131264...    9606       snv   \n",
       "4                          NC_000017.11:82781701:A:G    9606       snv   \n",
       "\n",
       "        create_date       update_date clinical_significance  \\\n",
       "0  2017/11/09 09:55  2022/10/16 23:33                   NaN   \n",
       "1  2017/11/09 09:55  2022/10/16 14:23                   NaN   \n",
       "2  2017/11/09 09:55  2022/10/17 01:20                   NaN   \n",
       "3  2017/11/09 09:55  2022/10/12 21:10                   NaN   \n",
       "4  2017/11/09 09:55  2022/10/17 00:32                   NaN   \n",
       "\n",
       "                                           fxn_class  \\\n",
       "0  3_prime_UTR_variant,missense_variant,stop_gain...   \n",
       "1           missense_variant,coding_sequence_variant   \n",
       "2  missense_variant,coding_sequence_variant,genic...   \n",
       "3           coding_sequence_variant,missense_variant   \n",
       "4  genic_upstream_transcript_variant,upstream_tra...   \n",
       "\n",
       "                         validated         genes  \\\n",
       "0  by-frequency,by-alfa,by-cluster  ZNF821:55565   \n",
       "1  by-frequency,by-alfa,by-cluster   MUC5AC:4586   \n",
       "2  by-frequency,by-alfa,by-cluster  MAP3K10:4294   \n",
       "3  by-frequency,by-alfa,by-cluster  POTEE:445582   \n",
       "4          by-frequency,by-cluster     TBCD:6904   \n",
       "\n",
       "                                                mafs  \\\n",
       "0  1000Genomes:A=0.000156/1,GnomAD_exomes:T=0.000...   \n",
       "1  1000Genomes:G=0.000781/5,GnomAD:G=0.001191/167...   \n",
       "2  1000Genomes:T=0.000156/1,TOPMED:G=0.000011/3,A...   \n",
       "3  1000Genomes:T=0.000156/1,TOPMED:A=0.000004/1,A...   \n",
       "4                             1000Genomes:G=0.0002/1   \n",
       "\n",
       "                                          variations  \n",
       "0  NP_001188481.1:p.Glu291Ter,NP_001188481.1:p.Gl...  \n",
       "1                        NP_001291288.1:p.Pro2233Arg  \n",
       "2    NP_002437.2:p.Pro193Ala,NP_002437.2:p.Pro193Ser  \n",
       "3  NP_001077007.1:p.Ala1019Thr,NP_001077007.1:p.A...  \n",
       "4                            NP_005984.3:p.Asp251Gly  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "# dbsnp_data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseExtremelyRareVariants.txt\"\n",
    "dbsnp_data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseRareVariants.txt\"\n",
    "# dbsnp_data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseCommonVariants.txt\"\n",
    "dbsnp_missense_variants_df = pd.read_csv(dbsnp_data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "print(dbsnp_missense_variants_df.columns)\n",
    "print(\"Downloaded: \", dbsnp_missense_variants_df.shape)\n",
    "dbsnp_missense_variants_df.drop_duplicates(\"snp_id\", keep=\"first\", inplace=True, ignore_index=True)\n",
    "print(\"Ater removing duplicate SNPs (actual downloaded): \", dbsnp_missense_variants_df.shape) \n",
    "dbsnp_missense_variants_df = dbsnp_missense_variants_df[~pd.isna(dbsnp_missense_variants_df[\"variations\"])]\n",
    "print(\"After removing empty protein associated variants: \", dbsnp_missense_variants_df.shape)\n",
    "dbsnp_missense_variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14728, 14)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_on_create_date(create_date):\n",
    "    # print(datetime.datetime.strptime(\"2017/11/09 09:55\", '%Y/%m/%d %H:%M').date() >= datetime.date(2022, 1, 1))\n",
    "    return datetime.datetime.strptime(create_date, '%Y/%m/%d %H:%M').date() >= datetime.date(2017, 1, 1)\n",
    "dbsnp_missense_variants_filtered_on_date_df = dbsnp_missense_variants_df[dbsnp_missense_variants_df[\"create_date\"].apply(filter_on_create_date)]\n",
    "dbsnp_missense_variants_filtered_on_date_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>...</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1490996759</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860386</td>\n",
       "      <td>NC_000016.10:71860385:C:A,NC_000016.10:7186038...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant,missense_variant,stop_gain...</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 821</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1348312707</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71867974</td>\n",
       "      <td>NC_000016.10:71867973:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,5_prime_UTR_variant,up...</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 821</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>572277950</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860397</td>\n",
       "      <td>NC_000016.10:71860396:T:A,NC_000016.10:7186039...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2014/08/21 18:13</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant,missense_variant,coding_se...</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 821</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>368877515</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71861807</td>\n",
       "      <td>NC_000016.10:71861806:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2013/04/26 00:46</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,intron_variant,coding_sequenc...</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 821</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>200621442</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860152</td>\n",
       "      <td>NC_000016.10:71860151:G:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2012/05/04 13:09</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,3_prime_UTR_variant,mi...</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 821</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642694</th>\n",
       "      <td>61732893</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>X:154230695</td>\n",
       "      <td>NC_000023.11:154230694:C:A,NC_000023.11:154230...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2008/03/23 22:41</td>\n",
       "      <td>2022/10/17 05:27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>opsin 1, medium wave sensitive 2</td>\n",
       "      <td>NM_001048181.3</td>\n",
       "      <td>NP_001041646.1</td>\n",
       "      <td>ENST00000369929.8</td>\n",
       "      <td>ENSP00000358945.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>X</td>\n",
       "      <td>154219756</td>\n",
       "      <td>154233286</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642695</th>\n",
       "      <td>41281479</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:24068453</td>\n",
       "      <td>NC_000002.12:24068452:G:A,NC_000002.12:2406845...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2006/10/18 10:37</td>\n",
       "      <td>2022/10/17 07:25</td>\n",
       "      <td>benign</td>\n",
       "      <td>synonymous_variant,missense_variant,coding_seq...</td>\n",
       "      <td>...</td>\n",
       "      <td>splicing factor 3b subunit 6</td>\n",
       "      <td>NM_016047.4</td>\n",
       "      <td>NP_057131.1</td>\n",
       "      <td>ENST00000233468.5</td>\n",
       "      <td>ENSP00000233468.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>2</td>\n",
       "      <td>24067586</td>\n",
       "      <td>24076326</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642696</th>\n",
       "      <td>35617141</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19:46608950</td>\n",
       "      <td>NC_000019.10:46608949:C:G,NC_000019.10:4660894...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2006/03/15 11:47</td>\n",
       "      <td>2022/10/13 16:56</td>\n",
       "      <td>pathogenic,benign</td>\n",
       "      <td>synonymous_variant,missense_variant,coding_seq...</td>\n",
       "      <td>...</td>\n",
       "      <td>calmodulin 3</td>\n",
       "      <td>NM_005184.4</td>\n",
       "      <td>NP_005175.2</td>\n",
       "      <td>ENST00000291295.14</td>\n",
       "      <td>ENSP00000291295.8</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>19</td>\n",
       "      <td>46601330</td>\n",
       "      <td>46610782</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642697</th>\n",
       "      <td>3914902</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:12109988</td>\n",
       "      <td>NC_000008.11:12109987:A:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2002/10/10 09:21</td>\n",
       "      <td>2022/10/17 10:57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>zinc finger protein 705D</td>\n",
       "      <td>NM_001039615.3</td>\n",
       "      <td>NP_001034704.2</td>\n",
       "      <td>ENST00000400085.8</td>\n",
       "      <td>ENSP00000382957.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>12089338</td>\n",
       "      <td>12115516</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642698</th>\n",
       "      <td>2792441</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7:151208081</td>\n",
       "      <td>NC_000007.14:151208080:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2001/09/28 17:27</td>\n",
       "      <td>2022/10/17 10:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non_coding_transcript_variant,synonymous_varia...</td>\n",
       "      <td>...</td>\n",
       "      <td>H2B.K variant histone 1</td>\n",
       "      <td>NM_001369125.3</td>\n",
       "      <td>NP_001356054.2</td>\n",
       "      <td>ENST00000644661.2</td>\n",
       "      <td>ENSP00000495192.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>7</td>\n",
       "      <td>151207837</td>\n",
       "      <td>151210488</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>642699 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            snp_id           acc       chrpos  \\\n",
       "0       1490996759  NC_000016.10  16:71860386   \n",
       "1       1348312707  NC_000016.10  16:71867974   \n",
       "2        572277950  NC_000016.10  16:71860397   \n",
       "3        368877515  NC_000016.10  16:71861807   \n",
       "4        200621442  NC_000016.10  16:71860152   \n",
       "...            ...           ...          ...   \n",
       "642694    61732893  NC_000023.11  X:154230695   \n",
       "642695    41281479  NC_000002.12   2:24068453   \n",
       "642696    35617141  NC_000019.10  19:46608950   \n",
       "642697     3914902  NC_000008.11   8:12109988   \n",
       "642698     2792441  NC_000007.14  7:151208081   \n",
       "\n",
       "                                                     spdi  tax_id snp_class  \\\n",
       "0       NC_000016.10:71860385:C:A,NC_000016.10:7186038...    9606       snv   \n",
       "1                               NC_000016.10:71867973:G:A    9606       snv   \n",
       "2       NC_000016.10:71860396:T:A,NC_000016.10:7186039...    9606       snv   \n",
       "3                               NC_000016.10:71861806:C:T    9606       snv   \n",
       "4                               NC_000016.10:71860151:G:C    9606       snv   \n",
       "...                                                   ...     ...       ...   \n",
       "642694  NC_000023.11:154230694:C:A,NC_000023.11:154230...    9606       snv   \n",
       "642695  NC_000002.12:24068452:G:A,NC_000002.12:2406845...    9606       snv   \n",
       "642696  NC_000019.10:46608949:C:G,NC_000019.10:4660894...    9606       snv   \n",
       "642697                          NC_000008.11:12109987:A:T    9606       snv   \n",
       "642698                         NC_000007.14:151208080:G:A    9606       snv   \n",
       "\n",
       "             create_date       update_date clinical_significance  \\\n",
       "0       2017/11/09 09:55  2022/10/16 23:33                   NaN   \n",
       "1       2017/11/09 09:55  2022/10/16 23:33                   NaN   \n",
       "2       2014/08/21 18:13  2022/10/16 23:33                   NaN   \n",
       "3       2013/04/26 00:46  2022/10/16 23:33                   NaN   \n",
       "4       2012/05/04 13:09  2022/10/16 23:33                   NaN   \n",
       "...                  ...               ...                   ...   \n",
       "642694  2008/03/23 22:41  2022/10/17 05:27                   NaN   \n",
       "642695  2006/10/18 10:37  2022/10/17 07:25                benign   \n",
       "642696  2006/03/15 11:47  2022/10/13 16:56     pathogenic,benign   \n",
       "642697  2002/10/10 09:21  2022/10/17 10:57                   NaN   \n",
       "642698  2001/09/28 17:27  2022/10/17 10:47                   NaN   \n",
       "\n",
       "                                                fxn_class  ...  \\\n",
       "0       3_prime_UTR_variant,missense_variant,stop_gain...  ...   \n",
       "1       coding_sequence_variant,5_prime_UTR_variant,up...  ...   \n",
       "2       3_prime_UTR_variant,missense_variant,coding_se...  ...   \n",
       "3       missense_variant,intron_variant,coding_sequenc...  ...   \n",
       "4       coding_sequence_variant,3_prime_UTR_variant,mi...  ...   \n",
       "...                                                   ...  ...   \n",
       "642694           missense_variant,coding_sequence_variant  ...   \n",
       "642695  synonymous_variant,missense_variant,coding_seq...  ...   \n",
       "642696  synonymous_variant,missense_variant,coding_seq...  ...   \n",
       "642697           missense_variant,coding_sequence_variant  ...   \n",
       "642698  non_coding_transcript_variant,synonymous_varia...  ...   \n",
       "\n",
       "                                    name      RefSeq_nuc     RefSeq_prot  \\\n",
       "0                zinc finger protein 821  NM_001201552.2  NP_001188481.1   \n",
       "1                zinc finger protein 821  NM_001201552.2  NP_001188481.1   \n",
       "2                zinc finger protein 821  NM_001201552.2  NP_001188481.1   \n",
       "3                zinc finger protein 821  NM_001201552.2  NP_001188481.1   \n",
       "4                zinc finger protein 821  NM_001201552.2  NP_001188481.1   \n",
       "...                                  ...             ...             ...   \n",
       "642694  opsin 1, medium wave sensitive 2  NM_001048181.3  NP_001041646.1   \n",
       "642695      splicing factor 3b subunit 6     NM_016047.4     NP_057131.1   \n",
       "642696                      calmodulin 3     NM_005184.4     NP_005175.2   \n",
       "642697          zinc finger protein 705D  NM_001039615.3  NP_001034704.2   \n",
       "642698           H2B.K variant histone 1  NM_001369125.3  NP_001356054.2   \n",
       "\n",
       "               Ensembl_nuc       Ensembl_prot  MANE_status GRCh38_chr  \\\n",
       "0        ENST00000425432.6  ENSP00000398089.1  MANE Select         16   \n",
       "1        ENST00000425432.6  ENSP00000398089.1  MANE Select         16   \n",
       "2        ENST00000425432.6  ENSP00000398089.1  MANE Select         16   \n",
       "3        ENST00000425432.6  ENSP00000398089.1  MANE Select         16   \n",
       "4        ENST00000425432.6  ENSP00000398089.1  MANE Select         16   \n",
       "...                    ...                ...          ...        ...   \n",
       "642694   ENST00000369929.8  ENSP00000358945.4  MANE Select          X   \n",
       "642695   ENST00000233468.5  ENSP00000233468.4  MANE Select          2   \n",
       "642696  ENST00000291295.14  ENSP00000291295.8  MANE Select         19   \n",
       "642697   ENST00000400085.8  ENSP00000382957.3  MANE Select          8   \n",
       "642698   ENST00000644661.2  ENSP00000495192.1  MANE Select          7   \n",
       "\n",
       "        chr_start    chr_end chr_strand  \n",
       "0        71859685   71884182          -  \n",
       "1        71859685   71884182          -  \n",
       "2        71859685   71884182          -  \n",
       "3        71859685   71884182          -  \n",
       "4        71859685   71884182          -  \n",
       "...           ...        ...        ...  \n",
       "642694  154219756  154233286          +  \n",
       "642695   24067586   24076326          -  \n",
       "642696   46601330   46610782          +  \n",
       "642697   12089338   12115516          +  \n",
       "642698  151207837  151210488          -  \n",
       "\n",
       "[642699 rows x 29 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_missense_variants_df[\"gene_symbol\"] = dbsnp_missense_variants_df[\"genes\"].apply(lambda x: x.split(\",\")[0].split(\":\")[0])#HSFX4:101927685\"\n",
    "# gene2refseq_filtered_df[gene2refseq_filtered_df[\"Symbol\"]==\"ANK2\"] # 73 refseq transcripts and 73 protein isoforms\n",
    "merged_df = dbsnp_missense_variants_df.merge(gene2refseq_mapping_df, how=\"inner\", left_on=\"gene_symbol\", right_on=gene_symbol_col_name) # mapping clinvar mRNA transcripts to protein accession\n",
    "merged_df\n",
    "# merged_df[[\"snp_id\"]].value_counts()\n",
    "# merged_df[merged_df[\"snp_id\"]==369224273]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NP_001188481.1:p.Pro35Leu'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_protein_isoforms_to_canonical(x):\n",
    "    # return x[\"variations\"]\n",
    "    variations = x[\"variations\"].split(\",\")\n",
    "    # if len(variations) == 1: return \",\".join(variations)\n",
    "    new_variations = []\n",
    "    for var in variations:\n",
    "        if var.split(\":\")[0] == x[\"RefSeq_prot\"]:\n",
    "            new_variations.append(var)\n",
    "    return \",\".join(new_variations)\n",
    "\n",
    "# merged_df[[\"variations\", \"RefSeq_prot\"]].apply(xx, axis=1)\n",
    "merged_df[\"new_variations\"] = merged_df[[\"variations\", \"RefSeq_prot\"]].apply(filter_protein_isoforms_to_canonical, axis=1)\n",
    "merged_df[\"new_variations\"][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>...</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "      <th>new_variations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1490996759</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860386</td>\n",
       "      <td>NC_000016.10:71860385:C:A,NC_000016.10:7186038...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant,missense_variant,stop_gain...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001188481.1:p.Glu291Ter,NP_001188481.1:p.Gl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1348312707</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71867974</td>\n",
       "      <td>NC_000016.10:71867973:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,5_prime_UTR_variant,up...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001188481.1:p.Pro35Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>572277950</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860397</td>\n",
       "      <td>NC_000016.10:71860396:T:A,NC_000016.10:7186039...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2014/08/21 18:13</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant,missense_variant,coding_se...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001188481.1:p.Asn287Ile,NP_001188481.1:p.As...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>368877515</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71861807</td>\n",
       "      <td>NC_000016.10:71861806:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2013/04/26 00:46</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,intron_variant,coding_sequenc...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001188481.1:p.Ala185Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>200621442</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:71860152</td>\n",
       "      <td>NC_000016.10:71860151:G:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2012/05/04 13:09</td>\n",
       "      <td>2022/10/16 23:33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,3_prime_UTR_variant,mi...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001201552.2</td>\n",
       "      <td>NP_001188481.1</td>\n",
       "      <td>ENST00000425432.6</td>\n",
       "      <td>ENSP00000398089.1</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>71859685</td>\n",
       "      <td>71884182</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001188481.1:p.Gln369Glu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642693</th>\n",
       "      <td>41397847</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:173548</td>\n",
       "      <td>NC_000016.10:173547:T:A,NC_000016.10:173547:T:...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2006/11/30 15:22</td>\n",
       "      <td>2022/10/16 22:43</td>\n",
       "      <td>other,pathogenic</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_000517.6</td>\n",
       "      <td>NP_000508.1</td>\n",
       "      <td>ENST00000251595.11</td>\n",
       "      <td>ENSP00000251595.6</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>16</td>\n",
       "      <td>172876</td>\n",
       "      <td>173710</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_000508.1:p.Leu126Gln,NP_000508.1:p.Leu126Pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642694</th>\n",
       "      <td>61732893</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>X:154230695</td>\n",
       "      <td>NC_000023.11:154230694:C:A,NC_000023.11:154230...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2008/03/23 22:41</td>\n",
       "      <td>2022/10/17 05:27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001048181.3</td>\n",
       "      <td>NP_001041646.1</td>\n",
       "      <td>ENST00000369929.8</td>\n",
       "      <td>ENSP00000358945.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>X</td>\n",
       "      <td>154219756</td>\n",
       "      <td>154233286</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_001041646.1:p.Pro298Thr,NP_001041646.1:p.Pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642695</th>\n",
       "      <td>41281479</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:24068453</td>\n",
       "      <td>NC_000002.12:24068452:G:A,NC_000002.12:2406845...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2006/10/18 10:37</td>\n",
       "      <td>2022/10/17 07:25</td>\n",
       "      <td>benign</td>\n",
       "      <td>synonymous_variant,missense_variant,coding_seq...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_016047.4</td>\n",
       "      <td>NP_057131.1</td>\n",
       "      <td>ENST00000233468.5</td>\n",
       "      <td>ENSP00000233468.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>2</td>\n",
       "      <td>24067586</td>\n",
       "      <td>24076326</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_057131.1:p.Asn52Lys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642696</th>\n",
       "      <td>35617141</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19:46608950</td>\n",
       "      <td>NC_000019.10:46608949:C:G,NC_000019.10:4660894...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2006/03/15 11:47</td>\n",
       "      <td>2022/10/13 16:56</td>\n",
       "      <td>pathogenic,benign</td>\n",
       "      <td>synonymous_variant,missense_variant,coding_seq...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_005184.4</td>\n",
       "      <td>NP_005175.2</td>\n",
       "      <td>ENST00000291295.14</td>\n",
       "      <td>ENSP00000291295.8</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>19</td>\n",
       "      <td>46601330</td>\n",
       "      <td>46610782</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_005175.2:p.Asp130Glu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642697</th>\n",
       "      <td>3914902</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:12109988</td>\n",
       "      <td>NC_000008.11:12109987:A:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2002/10/10 09:21</td>\n",
       "      <td>2022/10/17 10:57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_001039615.3</td>\n",
       "      <td>NP_001034704.2</td>\n",
       "      <td>ENST00000400085.8</td>\n",
       "      <td>ENSP00000382957.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>12089338</td>\n",
       "      <td>12115516</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_001034704.2:p.Asp34Val</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>600069 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            snp_id           acc       chrpos  \\\n",
       "0       1490996759  NC_000016.10  16:71860386   \n",
       "1       1348312707  NC_000016.10  16:71867974   \n",
       "2        572277950  NC_000016.10  16:71860397   \n",
       "3        368877515  NC_000016.10  16:71861807   \n",
       "4        200621442  NC_000016.10  16:71860152   \n",
       "...            ...           ...          ...   \n",
       "642693    41397847  NC_000016.10    16:173548   \n",
       "642694    61732893  NC_000023.11  X:154230695   \n",
       "642695    41281479  NC_000002.12   2:24068453   \n",
       "642696    35617141  NC_000019.10  19:46608950   \n",
       "642697     3914902  NC_000008.11   8:12109988   \n",
       "\n",
       "                                                     spdi  tax_id snp_class  \\\n",
       "0       NC_000016.10:71860385:C:A,NC_000016.10:7186038...    9606       snv   \n",
       "1                               NC_000016.10:71867973:G:A    9606       snv   \n",
       "2       NC_000016.10:71860396:T:A,NC_000016.10:7186039...    9606       snv   \n",
       "3                               NC_000016.10:71861806:C:T    9606       snv   \n",
       "4                               NC_000016.10:71860151:G:C    9606       snv   \n",
       "...                                                   ...     ...       ...   \n",
       "642693  NC_000016.10:173547:T:A,NC_000016.10:173547:T:...    9606       snv   \n",
       "642694  NC_000023.11:154230694:C:A,NC_000023.11:154230...    9606       snv   \n",
       "642695  NC_000002.12:24068452:G:A,NC_000002.12:2406845...    9606       snv   \n",
       "642696  NC_000019.10:46608949:C:G,NC_000019.10:4660894...    9606       snv   \n",
       "642697                          NC_000008.11:12109987:A:T    9606       snv   \n",
       "\n",
       "             create_date       update_date clinical_significance  \\\n",
       "0       2017/11/09 09:55  2022/10/16 23:33                   NaN   \n",
       "1       2017/11/09 09:55  2022/10/16 23:33                   NaN   \n",
       "2       2014/08/21 18:13  2022/10/16 23:33                   NaN   \n",
       "3       2013/04/26 00:46  2022/10/16 23:33                   NaN   \n",
       "4       2012/05/04 13:09  2022/10/16 23:33                   NaN   \n",
       "...                  ...               ...                   ...   \n",
       "642693  2006/11/30 15:22  2022/10/16 22:43      other,pathogenic   \n",
       "642694  2008/03/23 22:41  2022/10/17 05:27                   NaN   \n",
       "642695  2006/10/18 10:37  2022/10/17 07:25                benign   \n",
       "642696  2006/03/15 11:47  2022/10/13 16:56     pathogenic,benign   \n",
       "642697  2002/10/10 09:21  2022/10/17 10:57                   NaN   \n",
       "\n",
       "                                                fxn_class  ...  \\\n",
       "0       3_prime_UTR_variant,missense_variant,stop_gain...  ...   \n",
       "1       coding_sequence_variant,5_prime_UTR_variant,up...  ...   \n",
       "2       3_prime_UTR_variant,missense_variant,coding_se...  ...   \n",
       "3       missense_variant,intron_variant,coding_sequenc...  ...   \n",
       "4       coding_sequence_variant,3_prime_UTR_variant,mi...  ...   \n",
       "...                                                   ...  ...   \n",
       "642693           missense_variant,coding_sequence_variant  ...   \n",
       "642694           missense_variant,coding_sequence_variant  ...   \n",
       "642695  synonymous_variant,missense_variant,coding_seq...  ...   \n",
       "642696  synonymous_variant,missense_variant,coding_seq...  ...   \n",
       "642697           missense_variant,coding_sequence_variant  ...   \n",
       "\n",
       "            RefSeq_nuc     RefSeq_prot         Ensembl_nuc       Ensembl_prot  \\\n",
       "0       NM_001201552.2  NP_001188481.1   ENST00000425432.6  ENSP00000398089.1   \n",
       "1       NM_001201552.2  NP_001188481.1   ENST00000425432.6  ENSP00000398089.1   \n",
       "2       NM_001201552.2  NP_001188481.1   ENST00000425432.6  ENSP00000398089.1   \n",
       "3       NM_001201552.2  NP_001188481.1   ENST00000425432.6  ENSP00000398089.1   \n",
       "4       NM_001201552.2  NP_001188481.1   ENST00000425432.6  ENSP00000398089.1   \n",
       "...                ...             ...                 ...                ...   \n",
       "642693     NM_000517.6     NP_000508.1  ENST00000251595.11  ENSP00000251595.6   \n",
       "642694  NM_001048181.3  NP_001041646.1   ENST00000369929.8  ENSP00000358945.4   \n",
       "642695     NM_016047.4     NP_057131.1   ENST00000233468.5  ENSP00000233468.4   \n",
       "642696     NM_005184.4     NP_005175.2  ENST00000291295.14  ENSP00000291295.8   \n",
       "642697  NM_001039615.3  NP_001034704.2   ENST00000400085.8  ENSP00000382957.3   \n",
       "\n",
       "        MANE_status GRCh38_chr  chr_start    chr_end chr_strand  \\\n",
       "0       MANE Select         16   71859685   71884182          -   \n",
       "1       MANE Select         16   71859685   71884182          -   \n",
       "2       MANE Select         16   71859685   71884182          -   \n",
       "3       MANE Select         16   71859685   71884182          -   \n",
       "4       MANE Select         16   71859685   71884182          -   \n",
       "...             ...        ...        ...        ...        ...   \n",
       "642693  MANE Select         16     172876     173710          +   \n",
       "642694  MANE Select          X  154219756  154233286          +   \n",
       "642695  MANE Select          2   24067586   24076326          -   \n",
       "642696  MANE Select         19   46601330   46610782          +   \n",
       "642697  MANE Select          8   12089338   12115516          +   \n",
       "\n",
       "                                           new_variations  \n",
       "0       NP_001188481.1:p.Glu291Ter,NP_001188481.1:p.Gl...  \n",
       "1                               NP_001188481.1:p.Pro35Leu  \n",
       "2       NP_001188481.1:p.Asn287Ile,NP_001188481.1:p.As...  \n",
       "3                              NP_001188481.1:p.Ala185Thr  \n",
       "4                              NP_001188481.1:p.Gln369Glu  \n",
       "...                                                   ...  \n",
       "642693  NP_000508.1:p.Leu126Gln,NP_000508.1:p.Leu126Pr...  \n",
       "642694  NP_001041646.1:p.Pro298Thr,NP_001041646.1:p.Pr...  \n",
       "642695                             NP_057131.1:p.Asn52Lys  \n",
       "642696                            NP_005175.2:p.Asp130Glu  \n",
       "642697                          NP_001034704.2:p.Asp34Val  \n",
       "\n",
       "[600069 rows x 30 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(merged_df[pd.isna(merged_df[\"new_variations\"])].index)\n",
    "print(merged_df[pd.isnull(merged_df[\"new_variations\"])].index)\n",
    "# merged_df[merged_df[\"new_variations\"] == \"\"]\n",
    "merged_df[merged_df[\"new_variations\"] != \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_filepath = home_dir+\"data/dbsnp/filtered/\" + dbsnp_data_filepath.split(\"/\")[-1]\n",
    "merged_df.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
