import sys
sys.path.append("../variant_effect_analysis")

import xml.etree.ElementTree as ET
tree = ET.parse('data/dbsnp/4314.xml')
root = tree.getroot()

namespace = "https://www.ncbi.nlm.nih.gov/SNP/docsum"
ns = u'{%s}' % namespace
nsl = len(ns)

import xmltodict, json

f = open('data/dbsnp/4314.xml', "r")
# x = '<?xml version="1.0" ?><ExchangeSet xmlns:xsi="https://www.w3.org/2001/XMLSchema-instance" xmlns="https://www.ncbi.nlm.nih.gov/SNP/docsum" xsi:schemaLocation="https://www.ncbi.nlm.nih.gov/SNP/docsum ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_eutils.xsd" ><DocumentSummary uid="4314"><SNP_ID>4314</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>T=0.001874/9</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>T=0.000259/1</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>T=0.00067/3</FREQ></MAF><MAF><STUDY>ExAC</STUDY><FREQ>T=0.000996/118</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>T=0.003159/443</FREQ></MAF><MAF><STUDY>GnomAD_exomes</STUDY><FREQ>T=0.000778/195</FREQ></MAF><MAF><STUDY>GoESP</STUDY><FREQ>T=0.003691/48</FREQ></MAF><MAF><STUDY>HapMap</STUDY><FREQ>T=0./0</FREQ></MAF><MAF><STUDY>PharmGKB</STUDY><FREQ>T=0.001845/2</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>T=0./0</FREQ></MAF><MAF><STUDY>ALFA</STUDY><FREQ>T=0.000357/31</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>benign</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>ACE</NAME><GENE_ID>1636</GENE_ID></GENE_E></GENES><ACC>NC_000017.11</ACC><CHR>17</CHR><HANDLE>HUGCELL_USP,AFFY,EVA,PHARMGKB_GERA,TOPMED,EXOME_CHIP,NHLBI-ESP,HUMAN_LONGEVITY,DEBNICK,1000G_HIGH_COVERAGE,SANFORD_IMAGENETICS,EGCUT_WGS,1000GENOMES,EVA_UK10K_ALSPAC,EVA_UK10K_TWINSUK,ILLUMINA,GNOMAD,KHV_HUMAN_GENOMES,EVA_EXAC</HANDLE><SPDI>NC_000017.11:63483942:C:G,NC_000017.11:63483942:C:T</SPDI><FXN_CLASS>genic_upstream_transcript_variant,coding_sequence_variant,missense_variant,upstream_transcript_variant</FXN_CLASS><VALIDATED>by-frequency,by-alfa,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000017.11:g.63483943C&gt;G,NC_000017.11:g.63483943C&gt;T,NC_000017.10:g.61561304C&gt;G,NC_000017.10:g.61561304C&gt;T,NG_011648.1:g.11871C&gt;G,NG_011648.1:g.11871C&gt;T,NM_000789.4:c.1681C&gt;G,NM_000789.4:c.1681C&gt;T,NM_000789.3:c.1681C&gt;G,NM_000789.3:c.1681C&gt;T,NM_001382701.1:c.829C&gt;G,NM_001382701.1:c.829C&gt;T,NM_001382700.1:c.1114C&gt;G,NM_001382700.1:c.1114C&gt;T,NP_000780.1:p.Arg561Gly,NP_000780.1:p.Arg561Trp,NP_001369630.1:p.Arg277Gly,NP_001369630.1:p.Arg277Trp,NP_001369629.1:p.Arg372Gly,NP_001369629.1:p.Arg372Trp|SEQ=[C/G/T]|LEN=1|GENE=ACE:1636</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>36</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2000/09/19 17:02</CREATEDATE><UPDATEDATE>2022/10/17 00:32</UPDATEDATE><SS>4703,84170039,342461117,465004561,491129086,491526053,780731042,783407426,1359206799,1635930929,1678924962,1692904921,1752234174,1917921222,1946462261,1959760336,2217954465,2742915523,2749834140,2951511563,2985101807,3021795184,3625716395,3627701066,3634682421,3640389731,3644692766,3652212699,3653873461,3682636982,3744447380,3744982683,3772480456,3820094431,3825139521,5040937494,5040937495,5303628690,5428486541,5496622615,5607591736,5624400581,5660375101,5847803558,5914485510,5951770526,5979511735</SS><ALLELE>B</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>17:63483943</CHRPOS><CHRPOS_PREV_ASSM>17:61561304</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>0000004314</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0063483943</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary></ExchangeSet>'
x = '<?xml version="1.0" ?><ExchangeSet xmlns:xsi="https://www.w3.org/2001/XMLSchema-instance" xmlns="https://www.ncbi.nlm.nih.gov/SNP/docsum" xsi:schemaLocation="https://www.ncbi.nlm.nih.gov/SNP/docsum ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_eutils.xsd" ><DocumentSummary uid="104893968"><SNP_ID>104893968</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>T=0./0</FREQ></MAF><MAF><STUDY>ALSPAC</STUDY><FREQ>T=0.001557/6</FREQ></MAF><MAF><STUDY>Estonian</STUDY><FREQ>T=0.000893/4</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>T=0.001305/183</FREQ></MAF><MAF><STUDY>MGP</STUDY><FREQ>T=0.007491/4</FREQ></MAF><MAF><STUDY>PAGE_STUDY</STUDY><FREQ>T=0.000839/66</FREQ></MAF><MAF><STUDY>SGDP_PRJ</STUDY><FREQ>C=0.5/1</FREQ></MAF><MAF><STUDY>TOPMED</STUDY><FREQ>T=0.001198/317</FREQ></MAF><MAF><STUDY>TWINSUK</STUDY><FREQ>T=0.000539/2</FREQ></MAF><MAF><STUDY>ALFA</STUDY><FREQ>T=0.001568/312</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE>pathogenic,conflicting-interpretations-of-pathogenicity,uncertain-significance</CLINICAL_SIGNIFICANCE><GENES><GENE_E><NAME>GUCA1A</NAME><GENE_ID>2978</GENE_ID></GENE_E><GENE_E><NAME>LOC118142757</NAME><GENE_ID>118142757</GENE_ID></GENE_E></GENES><ACC>NC_000006.12</ACC><CHR>6</CHR><HANDLE>HUGCELL_USP,OMIM-CURATED-RECORDS,ILLUMINA,SWEGEN,OMICIA,EVA_DECODE,EVA,AFFY,NHLBI-ESP,EVA_UK10K_TWINSUK,TOPMED,CLINSEQ_SNP,GNOMAD,EXOME_CHIP,PAGE_CC,HUMAN_LONGEVITY,EVA_EXAC,EGCUT_WGS,CSHL,1000GENOMES,NORTHRUP_AU,SGDP_PRJ,EVA_MGP,EVA_UK10K_ALSPAC</HANDLE><SPDI>NC_000006.12:42173761:C:G,NC_000006.12:42173761:C:T</SPDI><FXN_CLASS>missense_variant,coding_sequence_variant</FXN_CLASS><VALIDATED>by-frequency,by-alfa,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000006.12:g.42173762C&gt;G,NC_000006.12:g.42173762C&gt;T,NC_000006.11:g.42141500C&gt;G,NC_000006.11:g.42141500C&gt;T,NG_009938.1:g.23357C&gt;G,NG_009938.1:g.23357C&gt;T,NM_001384910.1:c.149C&gt;G,NM_001384910.1:c.149C&gt;T,NM_000409.5:c.149C&gt;G,NM_000409.5:c.149C&gt;T,NM_000409.4:c.149C&gt;G,NM_000409.4:c.149C&gt;T,NM_000409.3:c.149C&gt;G,NM_000409.3:c.149C&gt;T,NM_001319061.2:c.149C&gt;G,NM_001319061.2:c.149C&gt;T,NM_001319061.1:c.149C&gt;G,NM_001319061.1:c.149C&gt;T,NM_001319062.2:c.149C&gt;G,NM_001319062.2:c.149C&gt;T,NM_001319062.1:c.149C&gt;G,NM_001319062.1:c.149C&gt;T,NP_001371839.1:p.Pro50Arg,NP_001371839.1:p.Pro50Leu,NP_000400.2:p.Pro50Arg,NP_000400.2:p.Pro50Leu,NP_001305990.1:p.Pro50Arg,NP_001305990.1:p.Pro50Leu,NP_001305991.1:p.Pro50Arg,NP_001305991.1:p.Pro50Leu|SEQ=[C/G/T]|LEN=1|GENE=GUCA1A:2978,LOC118142757:118142757</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>132</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2010/05/28 14:55</CREATEDATE><UPDATEDATE>2022/10/13 09:39</UPDATEDATE><SS>244238549,256302425,342211988,488836569,491385827,491889629,780849304,783533079,1319874798,1592400855,1615446702,1658440735,1688315687,1688315688,1711129340,1752637193,1917805454,1946177936,1958904533,2283554544,2735781102,2747622001,2838274951,2985367854,2998962077,3022615969,3346933851,3625901932,3629534666,3635065138,3640772437,3644910995,3653132553,3654133172,3666853593,3717086558,3726342443,3744554064,3745365078,3771289994,3772858656,3824187281,3864449677,3983906904,3986344838,4700788776,5365216065,5465916418,5842155690,5848095689,5883559751,5968770168</SS><ALLELE>B</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>6:42173762</CHRPOS><CHRPOS_PREV_ASSM>6:42141500</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>0104893968</SNP_ID_SORT><CLINICAL_SORT>1</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0042173762</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary><DocumentSummary uid="1599696717"><SNP_ID>1599696717</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>C=0./0</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>C=0.000096/10</FREQ></MAF><MAF><STUDY>Korea1K</STUDY><FREQ>C=0./0</FREQ></MAF><MAF><STUDY>TOMMO</STUDY><FREQ>C=0./0</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE/><GENES><GENE_E><NAME>FCGBP</NAME><GENE_ID>8857</GENE_ID></GENE_E></GENES><ACC>NC_000019.10</ACC><CHR>19</CHR><HANDLE>YY_MCH,TOMMO_GENOMICS,EVA,KOGIC,1000G_HIGH_COVERAGE,GNOMAD</HANDLE><SPDI>NC_000019.10:39906231:C:G</SPDI><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000019.10:g.39906232C&gt;G,NW_009646206.1:g.165541C&gt;G,NC_000019.9:g.40396305C&gt;G,NM_003890.2:c.7092G&gt;C,NP_003881.2:p.Glu2364Asp|SEQ=[C/G]|LEN=1|GENE=FCGBP:8857</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>154</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2020/04/27 18:22</CREATEDATE><UPDATEDATE>2022/10/17 01:19</UPDATEDATE><SS>3981352567,4331248087,5237247399,5612985817,5786399962,5817603372</SS><ALLELE>S</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>19:39906232</CHRPOS><CHRPOS_PREV_ASSM>19:40396305</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>1599696717</SNP_ID_SORT><CLINICAL_SORT>0</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0039906232</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary><DocumentSummary uid="1599696710"><SNP_ID>1599696710</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0./0</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.00011/10</FREQ></MAF><MAF><STUDY>Korea1K</STUDY><FREQ>G=0./0</FREQ></MAF><MAF><STUDY>TOMMO</STUDY><FREQ>G=0.00004/1</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE/><GENES><GENE_E><NAME>FCGBP</NAME><GENE_ID>8857</GENE_ID></GENE_E></GENES><ACC>NC_000019.10</ACC><CHR>19</CHR><HANDLE>YY_MCH,TOMMO_GENOMICS,EVA,KOGIC,1000G_HIGH_COVERAGE,GNOMAD</HANDLE><SPDI>NC_000019.10:39906226:G:T</SPDI><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000019.10:g.39906227G&gt;T,NW_009646206.1:g.165536G&gt;T,NC_000019.9:g.40396300G&gt;T,NM_003890.2:c.7097C&gt;A,NP_003881.2:p.Ala2366Asp|SEQ=[G/T]|LEN=1|GENE=FCGBP:8857</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>154</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2020/04/27 18:22</CREATEDATE><UPDATEDATE>2022/10/17 01:19</UPDATEDATE><SS>3981352565,4331248085,5237247397,5612985815,5786399960,5817603370</SS><ALLELE>K</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>19:39906227</CHRPOS><CHRPOS_PREV_ASSM>19:40396300</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>1599696710</SNP_ID_SORT><CLINICAL_SORT>0</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0039906227</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary><DocumentSummary uid="1599696678"><SNP_ID>1599696678</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>G=0./0</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>G=0.00022/3</FREQ></MAF><MAF><STUDY>Korea1K</STUDY><FREQ>G=0./0</FREQ></MAF><MAF><STUDY>TOMMO</STUDY><FREQ>G=0./0</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE/><GENES><GENE_E><NAME>FCGBP</NAME><GENE_ID>8857</GENE_ID></GENE_E></GENES><ACC>NC_000019.10</ACC><CHR>19</CHR><HANDLE>YY_MCH,TOMMO_GENOMICS,EVA,KOGIC,1000G_HIGH_COVERAGE,GNOMAD</HANDLE><SPDI>NC_000019.10:39906199:G:T</SPDI><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000019.10:g.39906200G&gt;T,NW_009646206.1:g.165509G&gt;T,NC_000019.9:g.40396273G&gt;T,NM_003890.2:c.7124C&gt;A,NP_003881.2:p.Ala2375Asp|SEQ=[G/T]|LEN=1|GENE=FCGBP:8857</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>154</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2020/04/27 18:22</CREATEDATE><UPDATEDATE>2022/10/17 01:19</UPDATEDATE><SS>3981352559,4331248077,5237247391,5612985807,5786399953,5817603364</SS><ALLELE>K</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>19:39906200</CHRPOS><CHRPOS_PREV_ASSM>19:40396273</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>1599696678</SNP_ID_SORT><CLINICAL_SORT>0</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0039906200</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary><DocumentSummary uid="1410709689"><SNP_ID>1410709689</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>A=0./0</FREQ></MAF><MAF><STUDY>TOMMO</STUDY><FREQ>A=0.00295/67</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE/><GENES><GENE_E><NAME>BCLAF1P2</NAME><GENE_ID>731605</GENE_ID></GENE_E></GENES><ACC>NC_000016.10</ACC><CHR>16</CHR><HANDLE>YY_MCH,KHV_HUMAN_GENOMES,TOMMO_GENOMICS,GNOMAD,1000G_HIGH_COVERAGE,URBANLAB,EVA</HANDLE><SPDI>NC_000016.10:34268072:A:G,NC_000016.10:34268072:A:T</SPDI><FXN_CLASS>coding_sequence_variant,missense_variant</FXN_CLASS><VALIDATED>by-frequency,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000016.10:g.34268073A&gt;G,NC_000016.10:g.34268073A&gt;T,XM_047435041.1:c.577T&gt;C,XM_047435041.1:c.577T&gt;A,XP_047290997.1:p.Ser193Pro,XP_047290997.1:p.Ser193Thr|SEQ=[A/G/T]|LEN=1|GENE=BCLAF1P2:731605</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>152</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2017/11/09 09:55</CREATEDATE><UPDATEDATE>2022/10/17 00:33</UPDATEDATE><SS>3650501829,3819173090,3846373671,4299846344,4299846345,5603276007,5774064063,5815915183</SS><ALLELE>D</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>16:34268073</CHRPOS><CHRPOS_PREV_ASSM>16:-1</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>1410709689</SNP_ID_SORT><CLINICAL_SORT>0</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0034268073</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary><DocumentSummary uid="1320022007"><SNP_ID>1320022007</SNP_ID><ALLELE_ORIGIN/><GLOBAL_MAFS><MAF><STUDY>1000Genomes</STUDY><FREQ>T=0./0</FREQ></MAF><MAF><STUDY>GnomAD</STUDY><FREQ>T=0.000095/10</FREQ></MAF><MAF><STUDY>Korea1K</STUDY><FREQ>T=0.000573/1</FREQ></MAF><MAF><STUDY>TOMMO</STUDY><FREQ>T=0./0</FREQ></MAF><MAF><STUDY>ALFA</STUDY><FREQ>T=0.096442/1144</FREQ></MAF></GLOBAL_MAFS><GLOBAL_POPULATION/><GLOBAL_SAMPLESIZE>0</GLOBAL_SAMPLESIZE><SUSPECTED/><CLINICAL_SIGNIFICANCE/><GENES><GENE_E><NAME>FCGBP</NAME><GENE_ID>8857</GENE_ID></GENE_E></GENES><ACC>NC_000019.10</ACC><CHR>19</CHR><HANDLE>YY_MCH,TOMMO_GENOMICS,KOGIC,GNOMAD,EVA,1000G_HIGH_COVERAGE</HANDLE><SPDI>NC_000019.10:39906232:T:C</SPDI><FXN_CLASS>missense_variant,coding_sequence_variant</FXN_CLASS><VALIDATED>by-frequency,by-alfa,by-cluster</VALIDATED><DOCSUM>HGVS=NC_000019.10:g.39906233T&gt;C,NW_009646206.1:g.165542T&gt;C,NC_000019.9:g.40396306T&gt;C,NM_003890.2:c.7091A&gt;G,NP_003881.2:p.Glu2364Gly|SEQ=[T/C]|LEN=1|GENE=FCGBP:8857</DOCSUM><TAX_ID>9606</TAX_ID><ORIG_BUILD>154</ORIG_BUILD><UPD_BUILD>156</UPD_BUILD><CREATEDATE>2017/11/09 09:55</CREATEDATE><UPDATEDATE>2022/10/17 01:19</UPDATEDATE><SS>3981352568,4331248088,5237247400,5612985818,5786399963,5817603373</SS><ALLELE>Y</ALLELE><SNP_CLASS>snv</SNP_CLASS><CHRPOS>19:39906233</CHRPOS><CHRPOS_PREV_ASSM>19:40396306</CHRPOS_PREV_ASSM><TEXT/><SNP_ID_SORT>1320022007</SNP_ID_SORT><CLINICAL_SORT>0</CLINICAL_SORT><CITED_SORT/><CHRPOS_SORT>0039906233</CHRPOS_SORT><MERGED_SORT>0</MERGED_SORT></DocumentSummary></ExchangeSet>'

o = xmltodict.parse(x) #returns dict obj
docs = o["ExchangeSet"]["DocumentSummary"]

outs = []
for i, doc in enumerate(docs):
    genes = doc["GENES"]["GENE_E"] # can have multiples
    if isinstance(genes, list):
        genes = [f'{gene["NAME"]}:{gene["GENE_ID"]}' for gene in genes]
        genes = ",".join(genes)
    elif isinstance(genes, dict):
        genes = f'{genes["NAME"]}:{genes["GENE_ID"]}'
    # print(genes)

    mafs = doc["GLOBAL_MAFS"]["MAF"] # can have multiples
    mafs = [f'{maf["STUDY"]}:{maf["FREQ"]}' for maf in mafs]
    mafs = ",".join(mafs)
    # print(mafs)

    data = {
        "snp_id": doc["SNP_ID"],
        "acc": doc["ACC"],
        "chrpos": doc["CHRPOS"],
        "spdi": doc["SPDI"],
        "tax_id": doc["TAX_ID"],
        "snp_class": doc["SNP_CLASS"],
        "create_date": doc["CREATEDATE"],
        "update_date": doc["UPDATEDATE"],
        "clinical_significance": doc["CLINICAL_SIGNIFICANCE"],
        "fxn_class": doc["FXN_CLASS"],
        "validated": doc["VALIDATED"],
        "genes": genes,
        "mafs": mafs
    }
    outs.append(data)
    
    # break
print(outs)


# ns = {'namespace': 'https://www.ncbi.nlm.nih.gov/SNP/docsum'}

# docs = root.findall('namespace:DocumentSummary', ns)
# for doc in docs:
#     print(doc.find('namespace:SNP_ID', ns))
#     # for item in ch:
#     #     print(item.tag, item.text)