{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_remove_null_nan_empty_entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63738, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>status</th>\n",
       "      <th>RNA_nucleotide_accession.version</th>\n",
       "      <th>RNA_nucleotide_gi</th>\n",
       "      <th>protein_accession.version</th>\n",
       "      <th>protein_gi</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>genomic_nucleotide_gi</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>assembly</th>\n",
       "      <th>mature_peptide_accession.version</th>\n",
       "      <th>mature_peptide_gi</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>1653960645</td>\n",
       "      <td>NP_570602.2</td>\n",
       "      <td>21071030</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>568815579</td>\n",
       "      <td>58345182</td>\n",
       "      <td>58353491</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>1779421316</td>\n",
       "      <td>NP_000005.3</td>\n",
       "      <td>1779421317</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347423.2</td>\n",
       "      <td>1677538580</td>\n",
       "      <td>NP_001334352.2</td>\n",
       "      <td>1677538581</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347424.2</td>\n",
       "      <td>1779421277</td>\n",
       "      <td>NP_001334353.2</td>\n",
       "      <td>1779421278</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347425.2</td>\n",
       "      <td>1779421327</td>\n",
       "      <td>NP_001334354.2</td>\n",
       "      <td>1779421328</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  GeneID    status RNA_nucleotide_accession.version  \\\n",
       "0    9606       1  REVIEWED                      NM_130786.4   \n",
       "1    9606       2  REVIEWED                      NM_000014.6   \n",
       "2    9606       2  REVIEWED                   NM_001347423.2   \n",
       "3    9606       2  REVIEWED                   NM_001347424.2   \n",
       "4    9606       2  REVIEWED                   NM_001347425.2   \n",
       "\n",
       "   RNA_nucleotide_gi protein_accession.version  protein_gi  \\\n",
       "0         1653960645               NP_570602.2    21071030   \n",
       "1         1779421316               NP_000005.3  1779421317   \n",
       "2         1677538580            NP_001334352.2  1677538581   \n",
       "3         1779421277            NP_001334353.2  1779421278   \n",
       "4         1779421327            NP_001334354.2  1779421328   \n",
       "\n",
       "  genomic_nucleotide_accession.version  genomic_nucleotide_gi  \\\n",
       "0                         NC_000019.10              568815579   \n",
       "1                         NC_000012.12              568815586   \n",
       "2                         NC_000012.12              568815586   \n",
       "3                         NC_000012.12              568815586   \n",
       "4                         NC_000012.12              568815586   \n",
       "\n",
       "   start_position_on_the_genomic_accession  \\\n",
       "0                                 58345182   \n",
       "1                                  9067707   \n",
       "2                                  9067707   \n",
       "3                                  9067707   \n",
       "4                                  9067707   \n",
       "\n",
       "   end_position_on_the_genomic_accession orientation  \\\n",
       "0                               58353491           -   \n",
       "1                                9116228           -   \n",
       "2                                9116228           -   \n",
       "3                                9116228           -   \n",
       "4                                9116228           -   \n",
       "\n",
       "                                assembly mature_peptide_accession.version  \\\n",
       "0  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "1  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "2  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "3  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "4  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "\n",
       "  mature_peptide_gi Symbol  \n",
       "0                 -   A1BG  \n",
       "1                 -    A2M  \n",
       "2                 -    A2M  \n",
       "3                 -    A2M  \n",
       "4                 -    A2M  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2refseq_filtered_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\"\n",
    "gene2refseq_filtered_df = pd.read_csv(gene2refseq_filtered_filepath, compression='gzip', comment='#', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "print(gene2refseq_filtered_df.shape)\n",
    "gene2refseq_filtered_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4006, 16)\n",
      "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)',\n",
      "       'Clinical significance (Last reviewed)', 'Review status', 'Accession',\n",
      "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
      "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
      "       'Canonical SPDI', 'Unnamed: 15'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Clinical significance (Last reviewed)</th>\n",
       "      <th>Review status</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>AlleleID(s)</th>\n",
       "      <th>dbSNP ID</th>\n",
       "      <th>Canonical SPDI</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001394713.1(PERM1):c.2330T&gt;C (p.Val777Ala)</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>V663A, V755A, V777A</td>\n",
       "      <td>Renal tubular epithelial cell apoptosis|Neutro...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>VCV001320032</td>\n",
       "      <td>1</td>\n",
       "      <td>911595</td>\n",
       "      <td>1</td>\n",
       "      <td>976215</td>\n",
       "      <td>1320032</td>\n",
       "      <td>1310278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:976214:A:G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_080605.4(B3GALT6):c.2T&gt;C (p.Met1Thr)</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>M1T</td>\n",
       "      <td>Spondyloepimetaphyseal dysplasia with joint la...</td>\n",
       "      <td>Pathogenic(Last reviewed: Oct 17, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV001452181</td>\n",
       "      <td>1</td>\n",
       "      <td>1167660</td>\n",
       "      <td>1</td>\n",
       "      <td>1232280</td>\n",
       "      <td>1452181</td>\n",
       "      <td>1438938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1232279:T:C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_080605.4(B3GALT6):c.3G&gt;A (p.Met1Ile)</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>M1I</td>\n",
       "      <td>Spondyloepimetaphyseal dysplasia with joint la...</td>\n",
       "      <td>Pathogenic(Last reviewed: May 16, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV002151860</td>\n",
       "      <td>1</td>\n",
       "      <td>1167661</td>\n",
       "      <td>1</td>\n",
       "      <td>1232281</td>\n",
       "      <td>2151860</td>\n",
       "      <td>1882027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1232280:G:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001170535.3(ATAD3A):c.230T&gt;G (p.Leu77Arg)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>L77R</td>\n",
       "      <td>Pontocerebellar hypoplasia, hypotonia, and res...</td>\n",
       "      <td>Pathogenic(Last reviewed: Mar 8, 2022)</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>VCV000828051</td>\n",
       "      <td>1</td>\n",
       "      <td>1451416</td>\n",
       "      <td>1</td>\n",
       "      <td>1516036</td>\n",
       "      <td>828051</td>\n",
       "      <td>816386</td>\n",
       "      <td>rs1570319915</td>\n",
       "      <td>NC_000001.11:1516035:T:G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001170535.3(ATAD3A):c.1361A&gt;G (p.Asn454Ser)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>N375S, N454S, N502S</td>\n",
       "      <td>Harel-Yoon syndrome</td>\n",
       "      <td>Pathogenic/Likely pathogenic(Last reviewed: Oc...</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>VCV001685550</td>\n",
       "      <td>1</td>\n",
       "      <td>1463098</td>\n",
       "      <td>1</td>\n",
       "      <td>1527718</td>\n",
       "      <td>1685550</td>\n",
       "      <td>1677567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1527717:A:G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Name  Gene(s)  \\\n",
       "0   NM_001394713.1(PERM1):c.2330T>C (p.Val777Ala)    PERM1   \n",
       "1         NM_080605.4(B3GALT6):c.2T>C (p.Met1Thr)  B3GALT6   \n",
       "2         NM_080605.4(B3GALT6):c.3G>A (p.Met1Ile)  B3GALT6   \n",
       "3    NM_001170535.3(ATAD3A):c.230T>G (p.Leu77Arg)   ATAD3A   \n",
       "4  NM_001170535.3(ATAD3A):c.1361A>G (p.Asn454Ser)   ATAD3A   \n",
       "\n",
       "        Protein change                                       Condition(s)  \\\n",
       "0  V663A, V755A, V777A  Renal tubular epithelial cell apoptosis|Neutro...   \n",
       "1                  M1T  Spondyloepimetaphyseal dysplasia with joint la...   \n",
       "2                  M1I  Spondyloepimetaphyseal dysplasia with joint la...   \n",
       "3                 L77R  Pontocerebellar hypoplasia, hypotonia, and res...   \n",
       "4  N375S, N454S, N502S                                Harel-Yoon syndrome   \n",
       "\n",
       "               Clinical significance (Last reviewed)  \\\n",
       "0                                         Pathogenic   \n",
       "1            Pathogenic(Last reviewed: Oct 17, 2022)   \n",
       "2            Pathogenic(Last reviewed: May 16, 2022)   \n",
       "3             Pathogenic(Last reviewed: Mar 8, 2022)   \n",
       "4  Pathogenic/Likely pathogenic(Last reviewed: Oc...   \n",
       "\n",
       "                                       Review status     Accession  \\\n",
       "0                     no assertion criteria provided  VCV001320032   \n",
       "1                criteria provided, single submitter  VCV001452181   \n",
       "2                criteria provided, single submitter  VCV002151860   \n",
       "3                     no assertion criteria provided  VCV000828051   \n",
       "4  criteria provided, multiple submitters, no con...  VCV001685550   \n",
       "\n",
       "  GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \\\n",
       "0                1         911595                1         976215   \n",
       "1                1        1167660                1        1232280   \n",
       "2                1        1167661                1        1232281   \n",
       "3                1        1451416                1        1516036   \n",
       "4                1        1463098                1        1527718   \n",
       "\n",
       "   VariationID AlleleID(s)      dbSNP ID            Canonical SPDI  \\\n",
       "0      1320032     1310278           NaN   NC_000001.11:976214:A:G   \n",
       "1      1452181     1438938           NaN  NC_000001.11:1232279:T:C   \n",
       "2      2151860     1882027           NaN  NC_000001.11:1232280:G:A   \n",
       "3       828051      816386  rs1570319915  NC_000001.11:1516035:T:G   \n",
       "4      1685550     1677567           NaN  NC_000001.11:1527717:A:G   \n",
       "\n",
       "   Unnamed: 15  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_data_filepath = home_dir+\"data/clinvar/search_results/clinvar_HumanPathogenicMissenseVariants01012022To14022023.txt\"\n",
    "# clinvar_data_filepath = home_dir+\"data/clinvar/search_results/clinvar_HumanLikelyPathogenicMissenseVariants01012022To14022023.txt\"\n",
    "clinvar_hum_patho_missense_variants_df = pd.read_csv(clinvar_data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "print(clinvar_hum_patho_missense_variants_df.shape)\n",
    "print(clinvar_hum_patho_missense_variants_df.columns)\n",
    "clinvar_hum_patho_missense_variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NAN rows removed: 87\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(3919, 16)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_hum_patho_missense_variants_df = filter_remove_null_nan_empty_entries(clinvar_hum_patho_missense_variants_df, a_col_name=\"Protein change\")\n",
    "clinvar_hum_patho_missense_variants_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3916, 34)\n",
      "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)',\n",
      "       'Clinical significance (Last reviewed)', 'Review status', 'Accession',\n",
      "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
      "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
      "       'Canonical SPDI', 'Unnamed: 15', 'RNA', 'tax_id', 'GeneID', 'status',\n",
      "       'RNA_nucleotide_accession.version', 'RNA_nucleotide_gi',\n",
      "       'protein_accession.version', 'protein_gi',\n",
      "       'genomic_nucleotide_accession.version', 'genomic_nucleotide_gi',\n",
      "       'start_position_on_the_genomic_accession',\n",
      "       'end_position_on_the_genomic_accession', 'orientation', 'assembly',\n",
      "       'mature_peptide_accession.version', 'mature_peptide_gi', 'Symbol',\n",
      "       'Canonical SPDI chr'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "clinvar_hum_patho_missense_variants_df[\"RNA\"] = clinvar_hum_patho_missense_variants_df[\"Name\"].apply(lambda x: x.split(\"(\")[0]) # RNA transcript accession: NM_accession.version\n",
    "merged_df = clinvar_hum_patho_missense_variants_df.merge(gene2refseq_filtered_df, how=\"left\", left_on=\"RNA\", right_on=\"RNA_nucleotide_accession.version\") # mapping clinvar mRNA transcripts to protein accession\n",
    "\n",
    "# same gene can be present in X and Y. Removing chromosome X to Y wrongly mapped variations.\n",
    "merged_df[\"Canonical SPDI chr\"] = merged_df[\"Canonical SPDI\"].apply(lambda x:x.split(\":\")[0]) \n",
    "merged_df = merged_df[merged_df[\"Canonical SPDI chr\"] == merged_df[\"genomic_nucleotide_accession.version\"]] # not(!) in the condition will give wrongly mapped variations\n",
    "print(merged_df.shape)\n",
    "print(merged_df.columns)\n",
    "# merged_df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NM_001288953.1', 'NM_000074.2', 'NM_004646.3'}\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(3916, 34)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nm_ids_not_mapped = set(clinvar_hum_patho_missense_variants_df[\"RNA\"].unique()) - set(merged_df[\"RNA\"].unique())\n",
    "print(nm_ids_not_mapped)\n",
    "df_removed_NNE_protChange = filter_remove_null_nan_empty_entries(merged_df, a_col_name=\"Protein change\")\n",
    "df_removed_NNE_protAccVersion = filter_remove_null_nan_empty_entries(df_removed_NNE_protChange, a_col_name=\"protein_accession.version\")\n",
    "df_removed_NNE_protAccVersion.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_filepath = home_dir+\"data/clinvar/filtered/\" + clinvar_data_filepath.split(\"/\")[-1]\n",
    "df_removed_NNE_protAccVersion.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
