{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_remove_null_nan_empty_entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63738, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>status</th>\n",
       "      <th>RNA_nucleotide_accession.version</th>\n",
       "      <th>RNA_nucleotide_gi</th>\n",
       "      <th>protein_accession.version</th>\n",
       "      <th>protein_gi</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>genomic_nucleotide_gi</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>assembly</th>\n",
       "      <th>mature_peptide_accession.version</th>\n",
       "      <th>mature_peptide_gi</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>1653960645</td>\n",
       "      <td>NP_570602.2</td>\n",
       "      <td>21071030</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>568815579</td>\n",
       "      <td>58345182</td>\n",
       "      <td>58353491</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>1779421316</td>\n",
       "      <td>NP_000005.3</td>\n",
       "      <td>1779421317</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347423.2</td>\n",
       "      <td>1677538580</td>\n",
       "      <td>NP_001334352.2</td>\n",
       "      <td>1677538581</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347424.2</td>\n",
       "      <td>1779421277</td>\n",
       "      <td>NP_001334353.2</td>\n",
       "      <td>1779421278</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001347425.2</td>\n",
       "      <td>1779421327</td>\n",
       "      <td>NP_001334354.2</td>\n",
       "      <td>1779421328</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>568815586</td>\n",
       "      <td>9067707</td>\n",
       "      <td>9116228</td>\n",
       "      <td>-</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  GeneID    status RNA_nucleotide_accession.version  \\\n",
       "0    9606       1  REVIEWED                      NM_130786.4   \n",
       "1    9606       2  REVIEWED                      NM_000014.6   \n",
       "2    9606       2  REVIEWED                   NM_001347423.2   \n",
       "3    9606       2  REVIEWED                   NM_001347424.2   \n",
       "4    9606       2  REVIEWED                   NM_001347425.2   \n",
       "\n",
       "   RNA_nucleotide_gi protein_accession.version  protein_gi  \\\n",
       "0         1653960645               NP_570602.2    21071030   \n",
       "1         1779421316               NP_000005.3  1779421317   \n",
       "2         1677538580            NP_001334352.2  1677538581   \n",
       "3         1779421277            NP_001334353.2  1779421278   \n",
       "4         1779421327            NP_001334354.2  1779421328   \n",
       "\n",
       "  genomic_nucleotide_accession.version  genomic_nucleotide_gi  \\\n",
       "0                         NC_000019.10              568815579   \n",
       "1                         NC_000012.12              568815586   \n",
       "2                         NC_000012.12              568815586   \n",
       "3                         NC_000012.12              568815586   \n",
       "4                         NC_000012.12              568815586   \n",
       "\n",
       "   start_position_on_the_genomic_accession  \\\n",
       "0                                 58345182   \n",
       "1                                  9067707   \n",
       "2                                  9067707   \n",
       "3                                  9067707   \n",
       "4                                  9067707   \n",
       "\n",
       "   end_position_on_the_genomic_accession orientation  \\\n",
       "0                               58353491           -   \n",
       "1                                9116228           -   \n",
       "2                                9116228           -   \n",
       "3                                9116228           -   \n",
       "4                                9116228           -   \n",
       "\n",
       "                                assembly mature_peptide_accession.version  \\\n",
       "0  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "1  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "2  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "3  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "4  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "\n",
       "  mature_peptide_gi Symbol  \n",
       "0                 -   A1BG  \n",
       "1                 -    A2M  \n",
       "2                 -    A2M  \n",
       "3                 -    A2M  \n",
       "4                 -    A2M  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2refseq_filtered_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\"\n",
    "gene2refseq_filtered_df = pd.read_csv(gene2refseq_filtered_filepath, compression='gzip', comment='#', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "print(gene2refseq_filtered_df.shape)\n",
    "gene2refseq_filtered_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6828, 16)\n",
      "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)',\n",
      "       'Clinical significance (Last reviewed)', 'Review status', 'Accession',\n",
      "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
      "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
      "       'Canonical SPDI', 'Unnamed: 15'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Clinical significance (Last reviewed)</th>\n",
       "      <th>Review status</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>AlleleID(s)</th>\n",
       "      <th>dbSNP ID</th>\n",
       "      <th>Canonical SPDI</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_198576.4(AGRN):c.5909A&gt;G (p.Asn1970Ser)</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>N1869S, N1970S, N1993S</td>\n",
       "      <td>Neurodevelopmental disorder</td>\n",
       "      <td>Likely pathogenic(Last reviewed: Jun 3, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV001701897</td>\n",
       "      <td>1</td>\n",
       "      <td>989860</td>\n",
       "      <td>1</td>\n",
       "      <td>1054480</td>\n",
       "      <td>1701897</td>\n",
       "      <td>1694285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1054479:A:G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_031921.6(ATAD3B):c.2T&gt;A (p.Met1Lys)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>M1K</td>\n",
       "      <td>not provided</td>\n",
       "      <td>Likely pathogenic(Last reviewed: Oct 14, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV001710696</td>\n",
       "      <td>1</td>\n",
       "      <td>1447650</td>\n",
       "      <td>1</td>\n",
       "      <td>1512270</td>\n",
       "      <td>1710696</td>\n",
       "      <td>1709107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1512269:T:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001170535.3(ATAD3A):c.1338G&gt;T (p.Lys446Asn)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>K367N, K446N, K494N</td>\n",
       "      <td>not provided</td>\n",
       "      <td>Likely pathogenic(Last reviewed: May 19, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV001686723</td>\n",
       "      <td>1</td>\n",
       "      <td>1463075</td>\n",
       "      <td>1</td>\n",
       "      <td>1527695</td>\n",
       "      <td>1686723</td>\n",
       "      <td>1678840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1527694:G:T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001170535.3(ATAD3A):c.1361A&gt;G (p.Asn454Ser)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>N375S, N454S, N502S</td>\n",
       "      <td>Harel-Yoon syndrome</td>\n",
       "      <td>Pathogenic/Likely pathogenic(Last reviewed: Oc...</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>VCV001685550</td>\n",
       "      <td>1</td>\n",
       "      <td>1463098</td>\n",
       "      <td>1</td>\n",
       "      <td>1527718</td>\n",
       "      <td>1685550</td>\n",
       "      <td>1677567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1527717:A:G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001170535.3(ATAD3A):c.1583G&gt;A (p.Arg528Gln)</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>R449Q, R528Q, R576Q</td>\n",
       "      <td>Harel-Yoon syndrome</td>\n",
       "      <td>Likely pathogenic(Last reviewed: May 4, 2022)</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>VCV001679823</td>\n",
       "      <td>1</td>\n",
       "      <td>1464680</td>\n",
       "      <td>1</td>\n",
       "      <td>1529300</td>\n",
       "      <td>1679823</td>\n",
       "      <td>1671740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11:1529299:G:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Name Gene(s)  \\\n",
       "0      NM_198576.4(AGRN):c.5909A>G (p.Asn1970Ser)    AGRN   \n",
       "1          NM_031921.6(ATAD3B):c.2T>A (p.Met1Lys)  ATAD3A   \n",
       "2  NM_001170535.3(ATAD3A):c.1338G>T (p.Lys446Asn)  ATAD3A   \n",
       "3  NM_001170535.3(ATAD3A):c.1361A>G (p.Asn454Ser)  ATAD3A   \n",
       "4  NM_001170535.3(ATAD3A):c.1583G>A (p.Arg528Gln)  ATAD3A   \n",
       "\n",
       "           Protein change                 Condition(s)  \\\n",
       "0  N1869S, N1970S, N1993S  Neurodevelopmental disorder   \n",
       "1                     M1K                 not provided   \n",
       "2     K367N, K446N, K494N                 not provided   \n",
       "3     N375S, N454S, N502S          Harel-Yoon syndrome   \n",
       "4     R449Q, R528Q, R576Q          Harel-Yoon syndrome   \n",
       "\n",
       "               Clinical significance (Last reviewed)  \\\n",
       "0      Likely pathogenic(Last reviewed: Jun 3, 2022)   \n",
       "1     Likely pathogenic(Last reviewed: Oct 14, 2022)   \n",
       "2     Likely pathogenic(Last reviewed: May 19, 2022)   \n",
       "3  Pathogenic/Likely pathogenic(Last reviewed: Oc...   \n",
       "4      Likely pathogenic(Last reviewed: May 4, 2022)   \n",
       "\n",
       "                                       Review status     Accession  \\\n",
       "0                criteria provided, single submitter  VCV001701897   \n",
       "1                criteria provided, single submitter  VCV001710696   \n",
       "2                criteria provided, single submitter  VCV001686723   \n",
       "3  criteria provided, multiple submitters, no con...  VCV001685550   \n",
       "4                criteria provided, single submitter  VCV001679823   \n",
       "\n",
       "  GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \\\n",
       "0                1         989860                1        1054480   \n",
       "1                1        1447650                1        1512270   \n",
       "2                1        1463075                1        1527695   \n",
       "3                1        1463098                1        1527718   \n",
       "4                1        1464680                1        1529300   \n",
       "\n",
       "   VariationID AlleleID(s) dbSNP ID            Canonical SPDI  Unnamed: 15  \n",
       "0      1701897     1694285      NaN  NC_000001.11:1054479:A:G          NaN  \n",
       "1      1710696     1709107      NaN  NC_000001.11:1512269:T:A          NaN  \n",
       "2      1686723     1678840      NaN  NC_000001.11:1527694:G:T          NaN  \n",
       "3      1685550     1677567      NaN  NC_000001.11:1527717:A:G          NaN  \n",
       "4      1679823     1671740      NaN  NC_000001.11:1529299:G:A          NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clinvar_data_filepath = home_dir+\"data/clinvar/search_results/clinvar_HumanPathogenicMissenseVariants01012022To14022023.txt\"\n",
    "clinvar_data_filepath = home_dir+\"data/clinvar/search_results/clinvar_HumanLikelyPathogenicMissenseVariants01012022To14022023.txt\"\n",
    "clinvar_hum_patho_missense_variants_df = pd.read_csv(clinvar_data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "print(clinvar_hum_patho_missense_variants_df.shape)\n",
    "print(clinvar_hum_patho_missense_variants_df.columns)\n",
    "clinvar_hum_patho_missense_variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NAN rows removed: 15\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(6813, 16)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_hum_patho_missense_variants_df = filter_remove_null_nan_empty_entries(clinvar_hum_patho_missense_variants_df, a_col_name=\"Protein change\")\n",
    "clinvar_hum_patho_missense_variants_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6813, 34)\n",
      "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)',\n",
      "       'Clinical significance (Last reviewed)', 'Review status', 'Accession',\n",
      "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
      "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
      "       'Canonical SPDI', 'Unnamed: 15', 'RNA', 'tax_id', 'GeneID', 'status',\n",
      "       'RNA_nucleotide_accession.version', 'RNA_nucleotide_gi',\n",
      "       'protein_accession.version', 'protein_gi',\n",
      "       'genomic_nucleotide_accession.version', 'genomic_nucleotide_gi',\n",
      "       'start_position_on_the_genomic_accession',\n",
      "       'end_position_on_the_genomic_accession', 'orientation', 'assembly',\n",
      "       'mature_peptide_accession.version', 'mature_peptide_gi', 'Symbol',\n",
      "       'Canonical SPDI chr'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Name                                          NM_198576.4(AGRN):c.5909A>G (p.Asn1970Ser)\n",
       "Gene(s)                                                                             AGRN\n",
       "Protein change                                                    N1869S, N1970S, N1993S\n",
       "Condition(s)                                                 Neurodevelopmental disorder\n",
       "Clinical significance (Last reviewed)      Likely pathogenic(Last reviewed: Jun 3, 2022)\n",
       "Review status                                        criteria provided, single submitter\n",
       "Accession                                                                   VCV001701897\n",
       "GRCh37Chromosome                                                                       1\n",
       "GRCh37Location                                                                    989860\n",
       "GRCh38Chromosome                                                                       1\n",
       "GRCh38Location                                                                   1054480\n",
       "VariationID                                                                      1701897\n",
       "AlleleID(s)                                                                      1694285\n",
       "dbSNP ID                                                                             NaN\n",
       "Canonical SPDI                                                  NC_000001.11:1054479:A:G\n",
       "Unnamed: 15                                                                          NaN\n",
       "RNA                                                                          NM_198576.4\n",
       "tax_id                                                                              9606\n",
       "GeneID                                                                            375790\n",
       "status                                                                          REVIEWED\n",
       "RNA_nucleotide_accession.version                                             NM_198576.4\n",
       "RNA_nucleotide_gi                                                             1519244149\n",
       "protein_accession.version                                                    NP_940978.2\n",
       "protein_gi                                                                      54873613\n",
       "genomic_nucleotide_accession.version                                        NC_000001.11\n",
       "genomic_nucleotide_gi                                                          568815597\n",
       "start_position_on_the_genomic_accession                                          1020119\n",
       "end_position_on_the_genomic_accession                                            1056115\n",
       "orientation                                                                            +\n",
       "assembly                                           Reference GRCh38.p14 Primary Assembly\n",
       "mature_peptide_accession.version                                                       -\n",
       "mature_peptide_gi                                                                      -\n",
       "Symbol                                                                              AGRN\n",
       "Canonical SPDI chr                                                          NC_000001.11\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_hum_patho_missense_variants_df[\"RNA\"] = clinvar_hum_patho_missense_variants_df[\"Name\"].apply(lambda x: x.split(\"(\")[0]) # RNA transcript accession: NM_accession.version\n",
    "merged_df = clinvar_hum_patho_missense_variants_df.merge(gene2refseq_filtered_df, how=\"left\", left_on=\"RNA\", right_on=\"RNA_nucleotide_accession.version\") # mapping clinvar mRNA transcripts to protein accession\n",
    "\n",
    "# same gene can be present in X and Y. Removing chromosome X to Y wrongly mapped variations.\n",
    "merged_df[\"Canonical SPDI chr\"] = merged_df[\"Canonical SPDI\"].apply(lambda x:x.split(\":\")[0]) \n",
    "merged_df = merged_df[merged_df[\"Canonical SPDI chr\"] == merged_df[\"genomic_nucleotide_accession.version\"]] # not(!) in the condition will give wrongly mapped variations\n",
    "print(merged_df.shape)\n",
    "print(merged_df.columns)\n",
    "merged_df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(6813, 34)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nm_ids_not_mapped = set(clinvar_hum_patho_missense_variants_df[\"RNA\"].unique()) - set(merged_df[\"RNA\"].unique())\n",
    "print(nm_ids_not_mapped)\n",
    "df_removed_NNE_protChange = filter_remove_null_nan_empty_entries(merged_df, a_col_name=\"Protein change\")\n",
    "df_removed_NNE_protAccVersion = filter_remove_null_nan_empty_entries(df_removed_NNE_protChange, a_col_name=\"protein_accession.version\")\n",
    "df_removed_NNE_protAccVersion.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_filepath = home_dir+\"data/clinvar/filtered/\" + clinvar_data_filepath.split(\"/\")[-1]\n",
    "df_removed_NNE_protAccVersion.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
