{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import time\n",
    "import os\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_drop_duplicates, filter_remove_null_nan_empty_entries\n",
    "from utils.column_names_reader import get_col_names_from_gzip"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging downloaded missense variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_df(data_filepath):\n",
    "    df = pd.read_csv(data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "    # print(df.shape)\n",
    "    # df.head()\n",
    "    df = filter_drop_duplicates(df, a_col_name=\"snp_id\")\n",
    "    # print(df.shape)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1049000, 14)\n",
      "Number of duplicates rows removed: 11\n",
      "(1048989, 14)\n",
      "(1554500, 14)\n",
      "Number of duplicates rows removed: 174\n",
      "(1554326, 14)\n",
      "(1671500, 14)\n",
      "Number of duplicates rows removed: 0\n",
      "(1671500, 14)\n",
      "(1141322, 14)\n",
      "Number of duplicates rows removed: 60\n",
      "(1141262, 14)\n"
     ]
    }
   ],
   "source": [
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_0.txt\"\n",
    "df_0 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_1.txt\"\n",
    "df_1 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_2.txt\"\n",
    "df_2 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_3.txt\"\n",
    "df_3 = clean_df(data_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df_0, df_1, df_2, df_3], ignore_index=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initial cleaning of downloaded dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicates rows removed: 275727\n",
      "Number of NAN rows removed: 151569\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n"
     ]
    }
   ],
   "source": [
    "df_remove_dup = filter_drop_duplicates(df, a_col_name=\"snp_id\")\n",
    "df_remove_NNE = filter_remove_null_nan_empty_entries(df_remove_dup, a_col_name=\"variations\") # NNE: null, nan or empty variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4988781, 14)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# out_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_full.txt\"\n",
    "# df_remove_NNE.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)\n",
    "df_remove_NNE.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to MANE proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19120, 14)\n",
      "19062\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:1</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>NP_570602.2</td>\n",
       "      <td>ENST00000263100.8</td>\n",
       "      <td>ENSP00000263100.2</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>19</td>\n",
       "      <td>58345183</td>\n",
       "      <td>58353492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:2</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>NP_000005.3</td>\n",
       "      <td>ENST00000318602.12</td>\n",
       "      <td>ENSP00000323929.8</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>12</td>\n",
       "      <td>9067708</td>\n",
       "      <td>9115919</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:9</td>\n",
       "      <td>ENSG00000171428.15</td>\n",
       "      <td>HGNC:7645</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>NM_000662.8</td>\n",
       "      <td>NP_000653.3</td>\n",
       "      <td>ENST00000307719.9</td>\n",
       "      <td>ENSP00000307218.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18210109</td>\n",
       "      <td>18223689</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:10</td>\n",
       "      <td>ENSG00000156006.5</td>\n",
       "      <td>HGNC:7646</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>NM_000015.3</td>\n",
       "      <td>NP_000006.2</td>\n",
       "      <td>ENST00000286479.4</td>\n",
       "      <td>ENSP00000286479.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18391282</td>\n",
       "      <td>18401218</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:12</td>\n",
       "      <td>ENSG00000196136.18</td>\n",
       "      <td>HGNC:16</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>serpin family A member 3</td>\n",
       "      <td>NM_001085.5</td>\n",
       "      <td>NP_001076.2</td>\n",
       "      <td>ENST00000393078.5</td>\n",
       "      <td>ENSP00000376793.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>14</td>\n",
       "      <td>94612391</td>\n",
       "      <td>94624053</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #NCBI_GeneID        Ensembl_Gene    HGNC_ID    symbol  \\\n",
       "0     GeneID:1  ENSG00000121410.12     HGNC:5      A1BG   \n",
       "1     GeneID:2  ENSG00000175899.15     HGNC:7       A2M   \n",
       "2     GeneID:9  ENSG00000171428.15  HGNC:7645      NAT1   \n",
       "3    GeneID:10   ENSG00000156006.5  HGNC:7646      NAT2   \n",
       "4    GeneID:12  ENSG00000196136.18    HGNC:16  SERPINA3   \n",
       "\n",
       "                       name   RefSeq_nuc  RefSeq_prot         Ensembl_nuc  \\\n",
       "0    alpha-1-B glycoprotein  NM_130786.4  NP_570602.2   ENST00000263100.8   \n",
       "1     alpha-2-macroglobulin  NM_000014.6  NP_000005.3  ENST00000318602.12   \n",
       "2     N-acetyltransferase 1  NM_000662.8  NP_000653.3   ENST00000307719.9   \n",
       "3     N-acetyltransferase 2  NM_000015.3  NP_000006.2   ENST00000286479.4   \n",
       "4  serpin family A member 3  NM_001085.5  NP_001076.2   ENST00000393078.5   \n",
       "\n",
       "        Ensembl_prot  MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "0  ENSP00000263100.2  MANE Select         19   58345183  58353492          -  \n",
       "1  ENSP00000323929.8  MANE Select         12    9067708   9115919          -  \n",
       "2  ENSP00000307218.4  MANE Select          8   18210109  18223689          +  \n",
       "3  ENSP00000286479.3  MANE Select          8   18391282  18401218          +  \n",
       "4  ENSP00000376793.3  MANE Select         14   94612391  94624053          +  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2refseq_mapping_filepath = home_dir+\"data/refseq/MANE.GRCh38.v1.0.summary.txt.gz\" # \"symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\" # \"Symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/refseq/HumanLRGRefSeqGene.txt.gz\"  # \"Symbol\" is the Gene-symbol col\n",
    "\n",
    "gene_symbol_col_name = \"symbol\" if gene2refseq_mapping_filepath.__contains__(\"MANE\") else \"Symbol\"\n",
    "gene2refseq_mapping_df = pd.read_csv(gene2refseq_mapping_filepath, compression='gzip', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "\n",
    "print(gene2refseq_mapping_df.shape)\n",
    "print(len(gene2refseq_mapping_df[gene_symbol_col_name].unique())) # some gene can be mapped to multiple proteins\n",
    "gene2refseq_mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4987804, 29)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_missense_variants_df = df_remove_NNE\n",
    "dbsnp_missense_variants_df[\"gene_symbol\"] = dbsnp_missense_variants_df[\"genes\"].apply(lambda x: x.split(\",\")[0].split(\":\")[0])#HSFX4:101927685\"\n",
    "# gene2refseq_filtered_df[gene2refseq_filtered_df[\"Symbol\"]==\"ANK2\"] # 73 refseq transcripts and 73 protein isoforms\n",
    "merged_df = dbsnp_missense_variants_df.merge(gene2refseq_mapping_df, how=\"inner\", left_on=\"gene_symbol\", right_on=gene_symbol_col_name) # mapping clinvar mRNA transcripts to protein accession\n",
    "merged_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4987804, 30)\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 312738\n",
      "(4675066, 30)\n"
     ]
    }
   ],
   "source": [
    "def filter_protein_isoforms_to_canonical(x):\n",
    "    # return x[\"variations\"]\n",
    "    variations = x[\"variations\"].split(\",\")\n",
    "    # if len(variations) == 1: return \",\".join(variations)\n",
    "    new_variations = []\n",
    "    for var in variations:\n",
    "        if var.split(\":\")[0] == x[\"RefSeq_prot\"]:\n",
    "            new_variations.append(var)\n",
    "    return \",\".join(new_variations)\n",
    "\n",
    "# merged_df[[\"variations\", \"RefSeq_prot\"]].apply(xx, axis=1)\n",
    "merged_df[\"new_variations\"] = merged_df[[\"variations\", \"RefSeq_prot\"]].apply(filter_protein_isoforms_to_canonical, axis=1)\n",
    "print(merged_df.shape)\n",
    "merged_df_remove_NNE = filter_remove_null_nan_empty_entries(merged_df, a_col_name=\"new_variations\") # NNE: null, nan or empty variants\n",
    "print(merged_df_remove_NNE.shape)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_168666/2457036955.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  optim_df[\"snp_id\"] = optim_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4675066, 12)\n",
      "4646054\n",
      "rs753176482     2\n",
      "rs140002194     2\n",
      "rs141704244     2\n",
      "rs141536854     2\n",
      "rs141497053     2\n",
      "               ..\n",
      "rs1186734094    1\n",
      "rs1186176189    1\n",
      "rs1186159489    1\n",
      "rs1184408711    1\n",
      "rs11798010      1\n",
      "Name: snp_id, Length: 4646054, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>variations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2580070</th>\n",
       "      <td>rs753176482</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:71664422</td>\n",
       "      <td>NC_000002.12:71664421:C:G,NC_000002.12:7166442...</td>\n",
       "      <td>2015/04/01 22:25</td>\n",
       "      <td>2022/10/17 18:18</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NP_003485.1</td>\n",
       "      <td>NP_003485.1:p.Pro1681Ala,NP_003485.1:p.Pro1681Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2580071</th>\n",
       "      <td>rs753176482</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:71664422</td>\n",
       "      <td>NC_000002.12:71664421:C:G,NC_000002.12:7166442...</td>\n",
       "      <td>2015/04/01 22:25</td>\n",
       "      <td>2022/10/17 18:18</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NP_001124459.1</td>\n",
       "      <td>NP_001124459.1:p.Pro1720Ala,NP_001124459.1:p.P...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              snp_id           acc      chrpos  \\\n",
       "2580070  rs753176482  NC_000002.12  2:71664422   \n",
       "2580071  rs753176482  NC_000002.12  2:71664422   \n",
       "\n",
       "                                                      spdi       create_date  \\\n",
       "2580070  NC_000002.12:71664421:C:G,NC_000002.12:7166442...  2015/04/01 22:25   \n",
       "2580071  NC_000002.12:71664421:C:G,NC_000002.12:7166442...  2015/04/01 22:25   \n",
       "\n",
       "              update_date gene_symbol    HGNC_ID symbol       name  \\\n",
       "2580070  2022/10/17 18:18        DYSF  HGNC:3097   DYSF  dysferlin   \n",
       "2580071  2022/10/17 18:18        DYSF  HGNC:3097   DYSF  dysferlin   \n",
       "\n",
       "            RefSeq_prot                                         variations  \n",
       "2580070     NP_003485.1  NP_003485.1:p.Pro1681Ala,NP_003485.1:p.Pro1681Ser  \n",
       "2580071  NP_001124459.1  NP_001124459.1:p.Pro1720Ala,NP_001124459.1:p.P...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df_remove_NNE.columns\n",
    "columns = ['snp_id', 'acc', 'chrpos', 'spdi', 'create_date', 'update_date',\n",
    "           'gene_symbol', 'HGNC_ID', 'symbol', 'name', 'RefSeq_prot', 'new_variations']\n",
    "optim_df = merged_df_remove_NNE[columns]\n",
    "optim_df[\"snp_id\"] = optim_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n",
    "\n",
    "# an rs_id can be mapped to multiple proteins, because MANE sometimes maps single gene to multiple proteins.\n",
    "# the following is an example of that phenomena\n",
    "optim_df = optim_df.rename(columns={'new_variations': 'variations'})\n",
    "print(optim_df.shape)\n",
    "print(len(list(optim_df[\"snp_id\"].unique())))\n",
    "print(optim_df[\"snp_id\"].value_counts())\n",
    "optim_df[optim_df[\"snp_id\"]==\"rs753176482\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping with ALFA population frequency data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698', 'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702', 'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.9</td>\n",
       "      <td>144135212</td>\n",
       "      <td>rs1553120241</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>8560:5387</td>\n",
       "      <td>...</td>\n",
       "      <td>256:224</td>\n",
       "      <td>336:288</td>\n",
       "      <td>32:24</td>\n",
       "      <td>170:117</td>\n",
       "      <td>32:24</td>\n",
       "      <td>18:13</td>\n",
       "      <td>20:15</td>\n",
       "      <td>344:296</td>\n",
       "      <td>288:248</td>\n",
       "      <td>9432:6100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.9</td>\n",
       "      <td>144148243</td>\n",
       "      <td>rs2236566</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>5996:510</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>84:8</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>6080:518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.9</td>\n",
       "      <td>146267105</td>\n",
       "      <td>rs1553119693</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>37168:28800</td>\n",
       "      <td>...</td>\n",
       "      <td>56:44</td>\n",
       "      <td>1378:839</td>\n",
       "      <td>18:14</td>\n",
       "      <td>70:60</td>\n",
       "      <td>10:9</td>\n",
       "      <td>4836:3639</td>\n",
       "      <td>452:322</td>\n",
       "      <td>1414:861</td>\n",
       "      <td>66:53</td>\n",
       "      <td>44024:33749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.9</td>\n",
       "      <td>148488564</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>8552:0</td>\n",
       "      <td>...</td>\n",
       "      <td>256:0</td>\n",
       "      <td>338:0</td>\n",
       "      <td>32:0</td>\n",
       "      <td>170:0</td>\n",
       "      <td>32:0</td>\n",
       "      <td>16:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>346:0</td>\n",
       "      <td>288:0</td>\n",
       "      <td>9424:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>2701535</td>\n",
       "      <td>rs371068661</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>134:9</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>48:1</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>188:15</td>\n",
       "      <td>48:1</td>\n",
       "      <td>0:0</td>\n",
       "      <td>370:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3065495</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>6962:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2210:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>466:0</td>\n",
       "      <td>2294:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>10680:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3065497</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>6962:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2210:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>466:0</td>\n",
       "      <td>2294:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>10680:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3065498</td>\n",
       "      <td>rs981679567</td>\n",
       "      <td>C</td>\n",
       "      <td>G,T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>9688:0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>86:0,0</td>\n",
       "      <td>2784:0,0</td>\n",
       "      <td>146:0,0</td>\n",
       "      <td>610:0,0</td>\n",
       "      <td>26:0,0</td>\n",
       "      <td>98:0,0</td>\n",
       "      <td>496:0,0</td>\n",
       "      <td>2898:0,0</td>\n",
       "      <td>112:0,0</td>\n",
       "      <td>14048:0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3065499</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>6962:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2210:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>466:0</td>\n",
       "      <td>2294:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>10680:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>3065500</td>\n",
       "      <td>rs1407918474</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>9690:0</td>\n",
       "      <td>...</td>\n",
       "      <td>86:0</td>\n",
       "      <td>2784:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CHROM        POS            ID REF  ALT QUAL FILTER INFO  \\\n",
       "0        NC_000001.9  144135212  rs1553120241   G    A    .      .    .   \n",
       "1        NC_000001.9  144148243     rs2236566   G    T    .      .    .   \n",
       "2        NC_000001.9  146267105  rs1553119693   T    G    .      .    .   \n",
       "3        NC_000001.9  148488564             .   C    A    .      .    .   \n",
       "4       NC_000001.10    2701535   rs371068661   C    T    .      .    .   \n",
       "...              ...        ...           ...  ..  ...  ...    ...  ...   \n",
       "999995  NC_000001.11    3065495             .   G    T    .      .    .   \n",
       "999996  NC_000001.11    3065497             .   C    T    .      .    .   \n",
       "999997  NC_000001.11    3065498   rs981679567   C  G,T    .      .    .   \n",
       "999998  NC_000001.11    3065499             .   G    A    .      .    .   \n",
       "999999  NC_000001.11    3065500  rs1407918474   A    C    .      .    .   \n",
       "\n",
       "       FORMAT SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699  \\\n",
       "0       AN:AC    8560:5387  ...      256:224      336:288        32:24   \n",
       "1       AN:AC     5996:510  ...          0:0          0:0          0:0   \n",
       "2       AN:AC  37168:28800  ...        56:44     1378:839        18:14   \n",
       "3       AN:AC       8552:0  ...        256:0        338:0         32:0   \n",
       "4       AN:AC        134:9  ...          0:0         48:1          0:0   \n",
       "...       ...          ...  ...          ...          ...          ...   \n",
       "999995  AN:AC       6962:0  ...         84:0       2210:0        146:0   \n",
       "999996  AN:AC       6962:0  ...         84:0       2210:0        146:0   \n",
       "999997  AN:AC     9688:0,0  ...       86:0,0     2784:0,0      146:0,0   \n",
       "999998  AN:AC       6962:0  ...         84:0       2210:0        146:0   \n",
       "999999  AN:AC       9690:0  ...         86:0       2784:0        146:0   \n",
       "\n",
       "       SAMN10492700 SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703  \\\n",
       "0           170:117        32:24        18:13        20:15      344:296   \n",
       "1               0:0          0:0          0:0         84:8          0:0   \n",
       "2             70:60         10:9    4836:3639      452:322     1414:861   \n",
       "3             170:0         32:0         16:0         20:0        346:0   \n",
       "4               0:0          0:0          0:0       188:15         48:1   \n",
       "...             ...          ...          ...          ...          ...   \n",
       "999995        610:0         24:0         94:0        466:0       2294:0   \n",
       "999996        610:0         24:0         94:0        466:0       2294:0   \n",
       "999997      610:0,0       26:0,0       98:0,0      496:0,0     2898:0,0   \n",
       "999998        610:0         24:0         94:0        466:0       2294:0   \n",
       "999999        610:0         26:0         98:0        496:0       2898:0   \n",
       "\n",
       "       SAMN10492704 SAMN10492705  \n",
       "0           288:248    9432:6100  \n",
       "1               0:0     6080:518  \n",
       "2             66:53  44024:33749  \n",
       "3             288:0       9424:0  \n",
       "4               0:0       370:25  \n",
       "...             ...          ...  \n",
       "999995        108:0      10680:0  \n",
       "999996        108:0      10680:0  \n",
       "999997      112:0,0    14048:0,0  \n",
       "999998        108:0      10680:0  \n",
       "999999        112:0      14050:0  \n",
       "\n",
       "[1000000 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath = home_dir+\"data/ALFA_population_freq/freq.vcf.gz\"\n",
    "col_names = get_col_names_from_gzip(filepath, \"#CHROM\")\n",
    "print(col_names)\n",
    "alfa_popu_freq_df_chunk_iterator = pd.read_csv(filepath, compression='gzip', chunksize=1000000, comment=\"#\", names=col_names, delim_whitespace=False, sep=\"\\t\",)\n",
    "first_df = alfa_popu_freq_df_chunk_iterator.__next__()\n",
    "first_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4646054\n",
      "0 (5200, 21)\n",
      "1 (8667, 21)\n",
      "2 (13006, 21)\n",
      "3 (10625, 21)\n",
      "4 (11543, 21)\n",
      "5 (12827, 21)\n",
      "6 (8312, 21)\n",
      "7 (13537, 21)\n",
      "8 (5981, 21)\n",
      "9 (9720, 21)\n",
      "10 (9072, 21)\n",
      "11 (9952, 21)\n",
      "12 (12861, 21)\n",
      "13 (8517, 21)\n",
      "14 (2349, 21)\n",
      "15 (10570, 21)\n",
      "16 (2179, 21)\n",
      "17 (2070, 21)\n",
      "18 (5511, 21)\n",
      "19 (3642, 21)\n",
      "20 (1520, 21)\n",
      "21 (2519, 21)\n",
      "22 (3767, 21)\n",
      "23 (361, 21)\n",
      "24 (3408, 21)\n",
      "25 (3303, 21)\n",
      "26 (5676, 21)\n",
      "27 (6519, 21)\n",
      "28 (639, 21)\n",
      "29 (4112, 21)\n",
      "30 (1814, 21)\n",
      "31 (688, 21)\n",
      "32 (12630, 21)\n",
      "33 (7093, 21)\n",
      "34 (5982, 21)\n",
      "35 (3813, 21)\n",
      "36 (0, 21)\n",
      "37 (0, 21)\n",
      "38 (3822, 21)\n",
      "39 (7766, 21)\n",
      "40 (22367, 21)\n",
      "41 (17159, 21)\n",
      "42 (16528, 21)\n",
      "43 (12488, 21)\n",
      "44 (1776, 21)\n",
      "45 (4456, 21)\n",
      "46 (7628, 21)\n",
      "47 (4279, 21)\n",
      "48 (4476, 21)\n",
      "49 (7196, 21)\n",
      "50 (6156, 21)\n",
      "51 (3435, 21)\n",
      "52 (445, 21)\n",
      "53 (963, 21)\n",
      "54 (5332, 21)\n",
      "55 (9199, 21)\n",
      "56 (11319, 21)\n",
      "57 (8436, 21)\n",
      "58 (4118, 21)\n",
      "59 (4160, 21)\n",
      "60 (3690, 21)\n",
      "61 (3427, 21)\n",
      "62 (7464, 21)\n",
      "63 (12263, 21)\n",
      "64 (5571, 21)\n",
      "65 (5045, 21)\n",
      "66 (7062, 21)\n",
      "67 (2979, 21)\n",
      "68 (2618, 21)\n",
      "69 (8968, 21)\n",
      "70 (5499, 21)\n",
      "71 (1380, 21)\n",
      "72 (550, 21)\n",
      "73 (6242, 21)\n",
      "74 (235, 21)\n",
      "75 (3376, 21)\n",
      "76 (5629, 21)\n",
      "77 (2727, 21)\n",
      "78 (14541, 21)\n",
      "79 (5446, 21)\n",
      "80 (5465, 21)\n",
      "81 (2412, 21)\n",
      "82 (5211, 21)\n",
      "83 (1309, 21)\n",
      "84 (6020, 21)\n",
      "85 (4434, 21)\n",
      "86 (2644, 21)\n",
      "87 (238, 21)\n",
      "88 (6549, 21)\n",
      "89 (443, 21)\n",
      "90 (3987, 21)\n",
      "91 (3167, 21)\n",
      "92 (2250, 21)\n",
      "93 (6673, 21)\n",
      "94 (12304, 21)\n",
      "95 (741, 21)\n",
      "96 (638, 21)\n",
      "97 (2011, 21)\n",
      "98 (6720, 21)\n",
      "99 (1081, 21)\n",
      "100 (0, 21)\n",
      "101 (10169, 21)\n",
      "102 (6825, 21)\n",
      "103 (4368, 21)\n",
      "104 (2932, 21)\n",
      "105 (4898, 21)\n",
      "106 (6591, 21)\n",
      "107 (316, 21)\n",
      "108 (3457, 21)\n",
      "109 (1770, 21)\n",
      "110 (702, 21)\n",
      "111 (5298, 21)\n",
      "112 (7015, 21)\n",
      "113 (4064, 21)\n",
      "114 (1120, 21)\n",
      "115 (2148, 21)\n",
      "116 (979, 21)\n",
      "117 (1657, 21)\n",
      "118 (6548, 21)\n",
      "119 (340, 21)\n",
      "120 (4379, 21)\n",
      "121 (4576, 21)\n",
      "122 (3964, 21)\n",
      "123 (7086, 21)\n",
      "124 (5879, 21)\n",
      "125 (4262, 21)\n",
      "126 (21856, 21)\n",
      "127 (2485, 21)\n",
      "128 (3943, 21)\n",
      "129 (2777, 21)\n",
      "130 (4378, 21)\n",
      "131 (963, 21)\n",
      "132 (5029, 21)\n",
      "133 (7505, 21)\n",
      "134 (4007, 21)\n",
      "135 (5992, 21)\n",
      "136 (4040, 21)\n",
      "137 (3429, 21)\n",
      "138 (8428, 21)\n",
      "139 (10120, 21)\n",
      "140 (3126, 21)\n",
      "141 (4281, 21)\n",
      "142 (5253, 21)\n",
      "143 (8695, 21)\n",
      "144 (7067, 21)\n",
      "145 (7828, 21)\n",
      "146 (4160, 21)\n",
      "147 (2517, 21)\n",
      "148 (970, 21)\n",
      "149 (7835, 21)\n",
      "150 (5847, 21)\n",
      "151 (6344, 21)\n",
      "152 (2170, 21)\n",
      "153 (198, 21)\n",
      "154 (1598, 21)\n",
      "155 (1697, 21)\n",
      "156 (1581, 21)\n",
      "157 (2918, 21)\n",
      "158 (10318, 21)\n",
      "159 (5492, 21)\n",
      "160 (9543, 21)\n",
      "161 (19114, 21)\n",
      "162 (21682, 21)\n",
      "163 (7907, 21)\n",
      "164 (7604, 21)\n",
      "165 (181, 21)\n",
      "166 (3612, 21)\n",
      "167 (1910, 21)\n",
      "168 (2269, 21)\n",
      "169 (2016, 21)\n",
      "170 (1420, 21)\n",
      "171 (1322, 21)\n",
      "172 (329, 21)\n",
      "173 (109, 21)\n",
      "174 (767, 21)\n",
      "175 (326, 21)\n",
      "176 (649, 21)\n",
      "177 (4405, 21)\n",
      "178 (5670, 21)\n",
      "179 (584, 21)\n",
      "180 (1309, 21)\n",
      "181 (3146, 21)\n",
      "182 (8497, 21)\n",
      "183 (219, 21)\n",
      "184 (4525, 21)\n",
      "185 (11327, 21)\n",
      "186 (7057, 21)\n",
      "187 (9099, 21)\n",
      "188 (5548, 21)\n",
      "189 (2252, 21)\n",
      "190 (4820, 21)\n",
      "191 (4807, 21)\n",
      "192 (1122, 21)\n",
      "193 (4371, 21)\n",
      "194 (4317, 21)\n",
      "195 (2991, 21)\n",
      "196 (2976, 21)\n",
      "197 (506, 21)\n",
      "198 (2044, 21)\n",
      "199 (4334, 21)\n",
      "200 (2850, 21)\n",
      "201 (694, 21)\n",
      "202 (2293, 21)\n",
      "203 (3652, 21)\n",
      "204 (10785, 21)\n",
      "205 (2166, 21)\n",
      "206 (2430, 21)\n",
      "207 (11895, 21)\n",
      "208 (9543, 21)\n",
      "209 (10610, 21)\n",
      "210 (6133, 21)\n",
      "211 (5669, 21)\n",
      "212 (1500, 21)\n",
      "213 (3215, 21)\n",
      "214 (3812, 21)\n",
      "215 (774, 21)\n",
      "216 (1945, 21)\n",
      "217 (2236, 21)\n",
      "218 (526, 21)\n",
      "219 (0, 21)\n",
      "220 (1973, 21)\n",
      "221 (6732, 21)\n",
      "222 (2338, 21)\n",
      "223 (1313, 21)\n",
      "224 (4799, 21)\n",
      "225 (2210, 21)\n",
      "226 (6361, 21)\n",
      "227 (0, 21)\n",
      "228 (601, 21)\n",
      "229 (679, 21)\n",
      "230 (3160, 21)\n",
      "231 (5747, 21)\n",
      "232 (3875, 21)\n",
      "233 (5711, 21)\n",
      "234 (4220, 21)\n",
      "235 (4250, 21)\n",
      "236 (5988, 21)\n",
      "237 (3055, 21)\n",
      "238 (1235, 21)\n",
      "239 (929, 21)\n",
      "240 (3986, 21)\n",
      "241 (3074, 21)\n",
      "242 (3192, 21)\n",
      "243 (3942, 21)\n",
      "244 (4030, 21)\n",
      "245 (700, 21)\n",
      "246 (3131, 21)\n",
      "247 (4319, 21)\n",
      "248 (2746, 21)\n",
      "249 (2770, 21)\n",
      "250 (417, 21)\n",
      "251 (168, 21)\n",
      "252 (2050, 21)\n",
      "253 (2489, 21)\n",
      "254 (3587, 21)\n",
      "255 (1754, 21)\n",
      "256 (4174, 21)\n",
      "257 (5634, 21)\n",
      "258 (2666, 21)\n",
      "259 (379, 21)\n",
      "260 (2430, 21)\n",
      "261 (2785, 21)\n",
      "262 (1079, 21)\n",
      "263 (1653, 21)\n",
      "264 (1569, 21)\n",
      "265 (0, 21)\n",
      "266 (5058, 21)\n",
      "267 (6162, 21)\n",
      "268 (6263, 21)\n",
      "269 (2093, 21)\n",
      "270 (2412, 21)\n",
      "271 (2155, 21)\n",
      "272 (833, 21)\n",
      "273 (5497, 21)\n",
      "274 (548, 21)\n",
      "275 (339, 21)\n",
      "276 (801, 21)\n",
      "277 (328, 21)\n",
      "278 (2508, 21)\n",
      "279 (4109, 21)\n",
      "280 (7372, 21)\n",
      "281 (4136, 21)\n",
      "282 (2485, 21)\n",
      "283 (255, 21)\n",
      "284 (520, 21)\n",
      "285 (3786, 21)\n",
      "286 (2909, 21)\n",
      "287 (1219, 21)\n",
      "288 (2192, 21)\n",
      "289 (3008, 21)\n",
      "290 (5902, 21)\n",
      "291 (5491, 21)\n",
      "292 (5945, 21)\n",
      "293 (3449, 21)\n",
      "294 (1794, 21)\n",
      "295 (511, 21)\n",
      "296 (3215, 21)\n",
      "297 (3867, 21)\n",
      "298 (2125, 21)\n",
      "299 (1152, 21)\n",
      "300 (841, 21)\n",
      "301 (628, 21)\n",
      "302 (2224, 21)\n",
      "303 (3812, 21)\n",
      "304 (2495, 21)\n",
      "305 (2237, 21)\n",
      "306 (2646, 21)\n",
      "307 (1232, 21)\n",
      "308 (4149, 21)\n",
      "309 (6291, 21)\n",
      "310 (4335, 21)\n",
      "311 (6203, 21)\n",
      "312 (20693, 21)\n",
      "313 (2422, 21)\n",
      "314 (5295, 21)\n",
      "315 (9893, 21)\n",
      "316 (3093, 21)\n",
      "317 (3410, 21)\n",
      "318 (2824, 21)\n",
      "319 (513, 21)\n",
      "320 (2839, 21)\n",
      "321 (3893, 21)\n",
      "322 (2147, 21)\n",
      "323 (10898, 21)\n",
      "324 (9479, 21)\n",
      "325 (3187, 21)\n",
      "326 (1843, 21)\n",
      "327 (4090, 21)\n",
      "328 (4153, 21)\n",
      "329 (2435, 21)\n",
      "330 (2974, 21)\n",
      "331 (1091, 21)\n",
      "332 (8218, 21)\n",
      "333 (9355, 21)\n",
      "334 (29143, 21)\n",
      "335 (10089, 21)\n",
      "336 (5934, 21)\n",
      "337 (14322, 21)\n",
      "338 (4719, 21)\n",
      "339 (3521, 21)\n",
      "340 (4616, 21)\n",
      "341 (3133, 21)\n",
      "342 (4953, 21)\n",
      "343 (428, 21)\n",
      "344 (2253, 21)\n",
      "345 (364, 21)\n",
      "346 (3260, 21)\n",
      "347 (4915, 21)\n",
      "348 (1130, 21)\n",
      "349 (1667, 21)\n",
      "350 (3859, 21)\n",
      "351 (3516, 21)\n",
      "352 (4725, 21)\n",
      "353 (349, 21)\n",
      "354 (1770, 21)\n",
      "355 (3149, 21)\n",
      "356 (267, 21)\n",
      "357 (3600, 21)\n",
      "358 (5895, 21)\n",
      "359 (3407, 21)\n",
      "360 (5573, 21)\n",
      "361 (1755, 21)\n",
      "362 (1800, 21)\n",
      "363 (2055, 21)\n",
      "364 (3850, 21)\n",
      "365 (4418, 21)\n",
      "366 (4169, 21)\n",
      "367 (3674, 21)\n",
      "368 (3161, 21)\n",
      "369 (4331, 21)\n",
      "370 (4419, 21)\n",
      "371 (7047, 21)\n",
      "372 (3322, 21)\n",
      "373 (8319, 21)\n",
      "374 (1529, 21)\n",
      "375 (1431, 21)\n",
      "376 (4683, 21)\n",
      "377 (6951, 21)\n",
      "378 (5504, 21)\n",
      "379 (11815, 21)\n",
      "380 (1861, 21)\n",
      "381 (2747, 21)\n",
      "382 (826, 21)\n",
      "383 (1898, 21)\n",
      "384 (2470, 21)\n",
      "385 (5861, 21)\n",
      "386 (3894, 21)\n",
      "387 (4384, 21)\n",
      "388 (2472, 21)\n",
      "389 (2272, 21)\n",
      "390 (2730, 21)\n",
      "391 (4474, 21)\n",
      "392 (5016, 21)\n",
      "393 (5869, 21)\n",
      "394 (913, 21)\n",
      "395 (2311, 21)\n",
      "396 (473, 21)\n",
      "397 (261, 21)\n",
      "398 (3987, 21)\n",
      "399 (89, 21)\n",
      "400 (3968, 21)\n",
      "401 (7291, 21)\n",
      "402 (2339, 21)\n",
      "403 (4087, 21)\n",
      "404 (672, 21)\n",
      "405 (3984, 21)\n",
      "406 (6989, 21)\n",
      "407 (3655, 21)\n",
      "408 (7673, 21)\n",
      "409 (25132, 21)\n",
      "410 (4790, 21)\n",
      "411 (5709, 21)\n",
      "412 (1541, 21)\n",
      "413 (1377, 21)\n",
      "414 (3507, 21)\n",
      "415 (2819, 21)\n",
      "416 (2820, 21)\n",
      "417 (6618, 21)\n",
      "418 (4003, 21)\n",
      "419 (4255, 21)\n",
      "420 (3490, 21)\n",
      "421 (8118, 21)\n",
      "422 (8317, 21)\n",
      "423 (1004, 21)\n",
      "424 (13638, 21)\n",
      "425 (3339, 21)\n",
      "426 (2978, 21)\n",
      "427 (2913, 21)\n",
      "428 (5690, 21)\n",
      "429 (304, 21)\n",
      "430 (3610, 21)\n",
      "431 (3416, 21)\n",
      "432 (7649, 21)\n",
      "433 (1907, 21)\n",
      "434 (1808, 21)\n",
      "435 (5524, 21)\n",
      "436 (6925, 21)\n",
      "437 (6273, 21)\n",
      "438 (6146, 21)\n",
      "439 (3221, 21)\n",
      "440 (1342, 21)\n",
      "441 (5579, 21)\n",
      "442 (4208, 21)\n",
      "443 (2197, 21)\n",
      "444 (2869, 21)\n",
      "445 (1176, 21)\n",
      "446 (3541, 21)\n",
      "447 (1848, 21)\n",
      "448 (2641, 21)\n",
      "449 (1007, 21)\n",
      "450 (4168, 21)\n",
      "451 (2428, 21)\n",
      "452 (2279, 21)\n",
      "453 (2216, 21)\n",
      "454 (1185, 21)\n",
      "455 (860, 21)\n",
      "456 (3168, 21)\n",
      "457 (1511, 21)\n",
      "458 (2426, 21)\n",
      "459 (3632, 21)\n",
      "460 (5696, 21)\n",
      "461 (3140, 21)\n",
      "462 (2023, 21)\n",
      "463 (3322, 21)\n",
      "464 (1623, 21)\n",
      "465 (1151, 21)\n",
      "466 (3329, 21)\n",
      "467 (3419, 21)\n",
      "468 (2558, 21)\n",
      "469 (1051, 21)\n",
      "470 (4553, 21)\n",
      "471 (679, 21)\n",
      "472 (1623, 21)\n",
      "473 (4107, 21)\n",
      "474 (28563, 21)\n",
      "475 (3791, 21)\n",
      "476 (3161, 21)\n",
      "477 (4667, 21)\n",
      "478 (1008, 21)\n",
      "479 (857, 21)\n",
      "480 (4164, 21)\n",
      "481 (1635, 21)\n",
      "482 (6200, 21)\n",
      "483 (1759, 21)\n",
      "484 (308, 21)\n",
      "485 (2275, 21)\n",
      "486 (2261, 21)\n",
      "487 (14778, 21)\n",
      "488 (5632, 21)\n",
      "489 (1988, 21)\n",
      "490 (0, 21)\n",
      "491 (374, 21)\n",
      "492 (3381, 21)\n",
      "493 (3943, 21)\n",
      "494 (1957, 21)\n",
      "495 (4688, 21)\n",
      "496 (1852, 21)\n",
      "497 (2625, 21)\n",
      "498 (4219, 21)\n",
      "499 (2347, 21)\n",
      "500 (7582, 21)\n",
      "501 (6233, 21)\n",
      "502 (5275, 21)\n",
      "503 (3083, 21)\n",
      "504 (2167, 21)\n",
      "505 (6912, 21)\n",
      "506 (8885, 21)\n",
      "507 (2900, 21)\n",
      "508 (4176, 21)\n",
      "509 (7654, 21)\n",
      "510 (7371, 21)\n",
      "511 (15471, 21)\n",
      "512 (11971, 21)\n",
      "513 (21593, 21)\n",
      "514 (6667, 21)\n",
      "515 (1126, 21)\n",
      "516 (6031, 21)\n",
      "517 (498, 21)\n",
      "518 (3677, 21)\n",
      "519 (4065, 21)\n",
      "520 (4310, 21)\n",
      "521 (2827, 21)\n",
      "522 (4501, 21)\n",
      "523 (4045, 21)\n",
      "524 (3880, 21)\n",
      "525 (2456, 21)\n",
      "526 (1887, 21)\n",
      "527 (0, 21)\n",
      "528 (2691, 21)\n",
      "529 (6060, 21)\n",
      "530 (7223, 21)\n",
      "531 (1540, 21)\n",
      "532 (146, 21)\n",
      "533 (3254, 21)\n",
      "534 (2630, 21)\n",
      "535 (462, 21)\n",
      "536 (9839, 21)\n",
      "537 (10413, 21)\n",
      "538 (2216, 21)\n",
      "539 (4313, 21)\n",
      "540 (322, 21)\n",
      "541 (5072, 21)\n",
      "542 (4960, 21)\n",
      "543 (6253, 21)\n",
      "544 (9519, 21)\n",
      "545 (8791, 21)\n",
      "546 (14447, 21)\n",
      "547 (2126, 21)\n",
      "548 (708, 21)\n",
      "549 (3861, 21)\n",
      "550 (5613, 21)\n",
      "551 (4608, 21)\n",
      "552 (6184, 21)\n",
      "553 (4741, 21)\n",
      "554 (3364, 21)\n",
      "555 (4055, 21)\n",
      "556 (25873, 21)\n",
      "557 (8666, 21)\n",
      "558 (18279, 21)\n",
      "559 (7100, 21)\n",
      "560 (3067, 21)\n",
      "561 (6528, 21)\n",
      "562 (6831, 21)\n",
      "563 (1123, 21)\n",
      "564 (1071, 21)\n",
      "565 (1611, 21)\n",
      "566 (5185, 21)\n",
      "567 (4609, 21)\n",
      "568 (201, 21)\n",
      "569 (115, 21)\n",
      "570 (6116, 21)\n",
      "571 (9234, 21)\n",
      "572 (355, 21)\n",
      "573 (611, 21)\n",
      "574 (12022, 21)\n",
      "575 (11559, 21)\n",
      "576 (17564, 21)\n",
      "577 (28383, 21)\n",
      "578 (10659, 21)\n",
      "579 (8485, 21)\n",
      "580 (8381, 21)\n",
      "581 (5965, 21)\n",
      "582 (1910, 21)\n",
      "583 (2959, 21)\n",
      "584 (1884, 21)\n",
      "585 (2253, 21)\n",
      "586 (8374, 21)\n",
      "587 (775, 21)\n",
      "588 (2363, 21)\n",
      "589 (5590, 21)\n",
      "590 (6029, 21)\n",
      "591 (2496, 21)\n",
      "592 (5830, 21)\n",
      "593 (6813, 21)\n",
      "594 (11991, 21)\n",
      "595 (9011, 21)\n",
      "596 (3765, 21)\n",
      "597 (5280, 21)\n",
      "598 (3116, 21)\n",
      "599 (5985, 21)\n",
      "600 (5639, 21)\n",
      "601 (15624, 21)\n",
      "602 (5911, 21)\n",
      "603 (5086, 21)\n",
      "604 (1801, 21)\n",
      "605 (4006, 21)\n",
      "606 (1543, 21)\n",
      "607 (3996, 21)\n",
      "608 (3163, 21)\n",
      "609 (2897, 21)\n",
      "610 (0, 21)\n",
      "611 (1717, 21)\n",
      "612 (2940, 21)\n",
      "613 (4304, 21)\n",
      "614 (12863, 21)\n",
      "615 (16196, 21)\n",
      "616 (15778, 21)\n",
      "617 (15963, 21)\n",
      "618 (1064, 21)\n",
      "619 (3606, 21)\n",
      "620 (3427, 21)\n",
      "621 (5030, 21)\n",
      "622 (822, 21)\n",
      "623 (2363, 21)\n",
      "624 (3991, 21)\n",
      "625 (2018, 21)\n",
      "626 (2152, 21)\n",
      "627 (1661, 21)\n",
      "628 (3300, 21)\n",
      "629 (4304, 21)\n",
      "630 (5805, 21)\n",
      "631 (4950, 21)\n",
      "632 (5290, 21)\n",
      "633 (10966, 21)\n",
      "634 (6723, 21)\n",
      "635 (3509, 21)\n",
      "636 (9478, 21)\n",
      "637 (13823, 21)\n",
      "638 (413, 21)\n",
      "639 (3110, 21)\n",
      "640 (9907, 21)\n",
      "641 (0, 21)\n",
      "642 (4081, 21)\n",
      "643 (3777, 21)\n",
      "644 (5805, 21)\n",
      "645 (3804, 21)\n",
      "646 (6267, 21)\n",
      "647 (2892, 21)\n",
      "648 (3707, 21)\n",
      "649 (4675, 21)\n",
      "650 (4406, 21)\n",
      "651 (3367, 21)\n",
      "652 (4756, 21)\n",
      "653 (8, 21)\n",
      "654 (1338, 21)\n",
      "655 (421, 21)\n",
      "656 (176, 21)\n",
      "657 (518, 21)\n",
      "658 (1357, 21)\n",
      "659 (1533, 21)\n",
      "660 (3570, 21)\n",
      "661 (106, 21)\n",
      "662 (692, 21)\n",
      "663 (446, 21)\n",
      "664 (275, 21)\n",
      "665 (1518, 21)\n",
      "666 (3139, 21)\n",
      "667 (3646, 21)\n",
      "668 (4814, 21)\n",
      "669 (1021, 21)\n",
      "670 (4256, 21)\n",
      "671 (7126, 21)\n",
      "672 (8076, 21)\n",
      "673 (16185, 21)\n",
      "674 (4534, 21)\n",
      "675 (796, 21)\n",
      "676 (4656, 21)\n",
      "677 (3559, 21)\n",
      "678 (3485, 21)\n",
      "679 (234, 21)\n",
      "680 (3087, 21)\n",
      "681 (1396, 21)\n",
      "682 (6920, 21)\n",
      "683 (3791, 21)\n",
      "684 (4172, 21)\n",
      "685 (4462, 21)\n",
      "686 (7742, 21)\n",
      "687 (3048, 21)\n",
      "688 (5143, 21)\n",
      "689 (12092, 21)\n",
      "690 (5046, 21)\n",
      "691 (1935, 21)\n",
      "692 (284, 21)\n",
      "693 (2674, 21)\n",
      "694 (5693, 21)\n",
      "695 (8886, 21)\n",
      "696 (2075, 21)\n",
      "697 (4185, 21)\n",
      "698 (16382, 21)\n",
      "699 (2802, 21)\n",
      "700 (3566, 21)\n",
      "701 (3513, 21)\n",
      "702 (4168, 21)\n",
      "703 (6614, 21)\n",
      "704 (5676, 21)\n",
      "705 (1010, 21)\n",
      "706 (21149, 21)\n",
      "707 (13440, 21)\n",
      "708 (4097, 21)\n",
      "709 (8282, 21)\n",
      "710 (3060, 21)\n",
      "711 (4221, 21)\n",
      "712 (3419, 21)\n",
      "713 (8013, 21)\n",
      "714 (9274, 21)\n",
      "715 (3971, 21)\n",
      "716 (9868, 21)\n",
      "717 (9031, 21)\n",
      "718 (7618, 21)\n",
      "719 (6227, 21)\n",
      "720 (4471, 21)\n",
      "721 (10508, 21)\n",
      "722 (4681, 21)\n",
      "723 (566, 21)\n",
      "724 (2400, 21)\n",
      "725 (15578, 21)\n",
      "726 (30937, 21)\n",
      "727 (13298, 21)\n",
      "728 (264, 21)\n",
      "729 (1900, 21)\n",
      "730 (7409, 21)\n",
      "731 (5081, 21)\n",
      "732 (5158, 21)\n",
      "733 (10878, 21)\n",
      "734 (7234, 21)\n",
      "735 (7744, 21)\n",
      "736 (20486, 21)\n",
      "737 (0, 21)\n",
      "738 (3166, 21)\n",
      "739 (4911, 21)\n",
      "740 (2800, 21)\n",
      "741 (8832, 21)\n",
      "742 (3170, 21)\n",
      "743 (248, 21)\n",
      "744 (8506, 21)\n",
      "745 (14809, 21)\n",
      "746 (11130, 21)\n",
      "747 (6149, 21)\n",
      "748 (2676, 21)\n",
      "749 (1101, 21)\n",
      "750 (2650, 21)\n",
      "751 (8997, 21)\n",
      "752 (3640, 21)\n",
      "753 (18343, 21)\n",
      "754 (12656, 21)\n",
      "755 (17342, 21)\n",
      "756 (22263, 21)\n",
      "757 (9795, 21)\n",
      "758 (3214, 21)\n",
      "759 (13482, 21)\n",
      "760 (3497, 21)\n",
      "761 (1460, 21)\n",
      "762 (14684, 21)\n",
      "763 (3423, 21)\n",
      "764 (11856, 21)\n",
      "765 (21416, 21)\n",
      "766 (20171, 21)\n",
      "767 (9713, 21)\n",
      "768 (12430, 21)\n",
      "769 (820, 21)\n",
      "770 (5876, 21)\n",
      "771 (7482, 21)\n",
      "772 (11600, 21)\n",
      "773 (4715, 21)\n",
      "774 (4317, 21)\n",
      "775 (5618, 21)\n",
      "776 (19178, 21)\n",
      "777 (13644, 21)\n",
      "778 (15924, 21)\n",
      "779 (4050, 21)\n",
      "780 (2238, 21)\n",
      "781 (4792, 21)\n",
      "782 (5148, 21)\n",
      "783 (2126, 21)\n",
      "784 (319, 21)\n",
      "785 (6420, 21)\n",
      "786 (2136, 21)\n",
      "787 (1326, 21)\n",
      "788 (5580, 21)\n",
      "789 (3683, 21)\n",
      "790 (0, 21)\n",
      "791 (545, 21)\n",
      "792 (6182, 21)\n",
      "793 (3481, 21)\n",
      "794 (1802, 21)\n",
      "795 (2040, 21)\n",
      "796 (4093, 21)\n",
      "797 (3935, 21)\n",
      "798 (2338, 21)\n",
      "799 (1066, 21)\n",
      "800 (1572, 21)\n",
      "801 (2960, 21)\n",
      "802 (2791, 21)\n",
      "803 (21470, 21)\n",
      "804 (19202, 21)\n",
      "805 (16483, 21)\n",
      "806 (20263, 21)\n",
      "807 (21703, 21)\n",
      "808 (17442, 21)\n",
      "809 (22922, 21)\n",
      "810 (9896, 21)\n",
      "811 (4368, 21)\n",
      "812 (1669, 21)\n",
      "813 (5814, 21)\n",
      "814 (18554, 21)\n",
      "815 (22403, 21)\n",
      "816 (18207, 21)\n",
      "817 (21982, 21)\n",
      "818 (24428, 21)\n",
      "819 (23343, 21)\n",
      "820 (23317, 21)\n",
      "821 (24668, 21)\n",
      "822 (10345, 21)\n",
      "823 (7084, 21)\n",
      "824 (1713, 21)\n",
      "825 (1567, 21)\n",
      "826 (1769, 21)\n",
      "827 (3762, 21)\n",
      "828 (3933, 21)\n",
      "829 (5560, 21)\n",
      "830 (0, 21)\n",
      "831 (4790, 21)\n",
      "832 (12824, 21)\n",
      "833 (7829, 21)\n",
      "834 (3345, 21)\n",
      "835 (9694, 21)\n",
      "836 (6335, 21)\n",
      "837 (5686, 21)\n",
      "838 (1737, 21)\n",
      "839 (3991, 21)\n",
      "840 (4008, 21)\n",
      "841 (17976, 21)\n",
      "842 (553, 21)\n",
      "843 (1418, 21)\n",
      "844 (1296, 21)\n",
      "845 (317, 21)\n",
      "846 (0, 21)\n",
      "847 (1776, 21)\n",
      "848 (4540, 21)\n",
      "849 (7666, 21)\n",
      "850 (4986, 21)\n",
      "851 (2784, 21)\n",
      "852 (10674, 21)\n",
      "853 (8442, 21)\n",
      "854 (678, 21)\n",
      "855 (8318, 21)\n",
      "856 (10468, 21)\n",
      "857 (9119, 21)\n",
      "858 (5215, 21)\n",
      "859 (13839, 21)\n",
      "860 (2050, 21)\n",
      "861 (13363, 21)\n",
      "862 (12167, 21)\n",
      "863 (8298, 21)\n",
      "864 (6655, 21)\n",
      "865 (11541, 21)\n",
      "866 (2909, 21)\n",
      "867 (2611, 21)\n",
      "868 (3468, 21)\n",
      "869 (4586, 21)\n",
      "870 (2439, 21)\n",
      "871 (1944, 21)\n",
      "872 (2474, 21)\n",
      "873 (2142, 21)\n",
      "874 (3285, 21)\n",
      "875 (1187, 21)\n",
      "876 (9294, 21)\n",
      "877 (5249, 21)\n",
      "878 (1361, 21)\n",
      "879 (1050, 21)\n",
      "880 (1520, 21)\n",
      "881 (5497, 21)\n",
      "882 (1934, 21)\n",
      "883 (2162, 21)\n",
      "884 (714, 21)\n",
      "885 (1701, 21)\n",
      "886 (361, 21)\n",
      "887 (623, 21)\n",
      "888 (264, 21)\n",
      "889 (5072, 21)\n",
      "890 (3743, 21)\n",
      "891 (4134, 21)\n",
      "892 (1245, 21)\n",
      "893 (3252, 21)\n",
      "894 (1237, 21)\n",
      "895 (993, 21)\n",
      "896 (4467, 21)\n",
      "897 (3967, 21)\n",
      "898 (2386, 21)\n",
      "899 (533, 21)\n",
      "900 (1922, 21)\n",
      "901 (11871, 21)\n",
      "902 (3492, 21)\n",
      "903 (404, 21)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(4627886, 21)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_ids = list(optim_df[\"snp_id\"].unique())\n",
    "print(len(dbsnp_ids))\n",
    "\n",
    "chunk_list = []\n",
    "for i, alfa_popu_freq_df in enumerate(alfa_popu_freq_df_chunk_iterator):\n",
    "    chunk = alfa_popu_freq_df[alfa_popu_freq_df[\"ID\"].isin(dbsnp_ids)]\n",
    "    print(i, chunk.shape)\n",
    "    chunk_list.append(chunk)\n",
    "    # dbsnp_df = dbsnp_df.merge(filtered_alfa_popu_freq_df, how=\"left\", left_on=\"snp_id\", right_on=\"ID\") \n",
    "    # print(dbsnp_df.shape)\n",
    "    # if i==0: break\n",
    "\n",
    "filtered_alfa_popu_freq_df = pd.concat(chunk_list)    \n",
    "\n",
    "out_filepath = home_dir+\"data/ALFA_population_freq/freq_filtered.vcf.gz\"\n",
    "if os.path.exists(out_filepath): os.remove(out_filepath)\n",
    "filtered_alfa_popu_freq_df.to_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=1000000, index=False, mode=\"a\", header=True)\n",
    "filtered_alfa_popu_freq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4627886, 21)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4627886"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(filtered_alfa_popu_freq_df.shape)\n",
    "len(list(filtered_alfa_popu_freq_df[\"ID\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4656898, 33)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_with_prots_df = optim_df\n",
    "dbsnp_with_prots_popufreq_df = dbsnp_with_prots_df.merge(filtered_alfa_popu_freq_df, how=\"inner\", left_on=\"snp_id\", right_on=\"ID\") # mapping clinvar mRNA transcripts to protein accession\n",
    "dbsnp_with_prots_popufreq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4656898, 33)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_filepath = home_dir+\"data/ALFA_population_freq/dbsnp_with_prots_and_population_freq_mapping.vcf.gz\"\n",
    "if os.path.exists(out_filepath): os.remove(out_filepath)\n",
    "dbsnp_with_prots_popufreq_df.to_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=1000000, index=False, mode=\"a\", header=True)\n",
    "dbsnp_with_prots_popufreq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbsnp_with_prots_popufreq_df_iterator = pd.read_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=10000)\n",
    "dbsnp_with_prots_popufreq_first_df = dbsnp_with_prots_popufreq_df_iterator.__next__()\n",
    "out_filepath = home_dir+\"data/ALFA_population_freq/dbsnp_with_prots_and_population_freq_mapping_small.txt\"\n",
    "dbsnp_with_prots_popufreq_first_df.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'create_date', 'update_date',\n",
       "       'gene_symbol', 'HGNC_ID', 'symbol', 'name', 'RefSeq_prot', 'variations',\n",
       "       'CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
       "       'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698',\n",
       "       'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702',\n",
       "       'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_with_prots_popufreq_df.columns"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering common, rare and ultra-rare variants (deprecated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4577439, 30)\n"
     ]
    }
   ],
   "source": [
    "def filter_on_ALFA_mafs(all_mafs, maf_type):\n",
    "    # maf: minor allele frequency\n",
    "    alfa_mafs = [v.split(\":\")[1].strip() for v in all_mafs.split(\",\") if \"ALFA\" in v]\n",
    "    if len(alfa_mafs)>1: \n",
    "        # It is checked that there is only one ALFA maf for each snp_id, which could be multiple. ie GnomAD_exomes:C=0.000004/1,TOPMED:C=0.000008/2,ALFA:C=0./0,ALFA:A=0./0\n",
    "        # This error is for only next implementation\n",
    "        # This error can be commented out without any problem\n",
    "        raise NotImplementedError(\"A snp_id is mapped to multiple ALFA\")\n",
    "    alfa_maf= alfa_mafs[0]\n",
    "    freq = float(alfa_maf[alfa_maf.index(\"=\")+1:alfa_maf.index(\"/\")])\n",
    "    \n",
    "    # the following ranges are taken from https://www.ncbi.nlm.nih.gov/snp/docs/gsr/alfa/ALFA_20201027095038/\n",
    "    if maf_type==\"common\": return freq >= 0.01\n",
    "    elif maf_type==\"rare\": return 0.01 > freq >= 0.001\n",
    "    elif maf_type==\"ultrarare\": return 0.001 > freq >= 0.0\n",
    "\n",
    "\n",
    "maf_type=\"ultrarare\"\n",
    "out_df = merged_df_remove_NNE[merged_df_remove_NNE[\"mafs\"].apply(filter_on_ALFA_mafs, maf_type=maf_type)]\n",
    "print(out_df.shape)\n",
    "out_filepath = home_dir+f\"data/dbsnp/filtered/dbsnp_HumanMissenseALFAVariants_{maf_type}.txt\"\n",
    "out_df.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
