{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import time\n",
    "import os\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_drop_duplicates, filter_remove_null_nan_empty_entries\n",
    "from utils.column_names_reader import get_col_names_from_gzip"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging downloaded missense variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_df(data_filepath):\n",
    "    df = pd.read_csv(data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "    # print(df.shape)\n",
    "    # df.head()\n",
    "    df = filter_drop_duplicates(df, a_col_name=\"snp_id\")\n",
    "    # print(df.shape)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicates rows removed: 11\n",
      "Number of duplicates rows removed: 174\n",
      "Number of duplicates rows removed: 0\n",
      "Number of duplicates rows removed: 60\n"
     ]
    }
   ],
   "source": [
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_0.txt\"\n",
    "df_0 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_1.txt\"\n",
    "df_1 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_2.txt\"\n",
    "df_2 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_3.txt\"\n",
    "df_3 = clean_df(data_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5416077, 14)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "df = pd.concat([df_0, df_1, df_2, df_3], ignore_index=True)\n",
    "print(df.shape)\n",
    "print(df.columns)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initial cleaning of downloaded dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicates rows removed: 275727\n",
      "(5140350, 14)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n",
      "Number of NAN rows removed: 151569\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "(4988781, 14)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "df_remove_dup = filter_drop_duplicates(df, a_col_name=\"snp_id\")\n",
    "print(df_remove_dup.shape)\n",
    "print(df_remove_dup.columns)\n",
    "# snvs_with_no_NP = df_remove_dup[pd.isna(df_remove_dup[\"variations\"])]\n",
    "# snvs_with_no_NP.to_csv(home_dir+\"data/dbsnp/snvs_with_no_NP.csv\", index=False, header=True)\n",
    "df_remove_NNE = filter_remove_null_nan_empty_entries(df_remove_dup, a_col_name=\"variations\") # NNE: null, nan or empty variants\n",
    "print(df_remove_NNE.shape)\n",
    "print(df_remove_NNE.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                            UGGT1:56886\n",
       "1                             ITGA2:3673\n",
       "2          NECAB3:63941,C20orf144:128864\n",
       "3                           DPYSL5:56896\n",
       "4                             TMCC2:9911\n",
       "                       ...              \n",
       "5416072                         HBB:3043\n",
       "5416073                         LPL:4023\n",
       "5416074                         LPL:4023\n",
       "5416075                         LPL:4023\n",
       "5416076                    THSD7A:221981\n",
       "Name: genes, Length: 4988781, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_remove_NNE[\"genes\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to MANE proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19120, 14)\n",
      "(19062,)\n",
      "MANE Select           19062\n",
      "MANE Plus Clinical       58\n",
      "Name: MANE_status, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:1</td>\n",
       "      <td>ENSG00000121410.12</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>NM_130786.4</td>\n",
       "      <td>NP_570602.2</td>\n",
       "      <td>ENST00000263100.8</td>\n",
       "      <td>ENSP00000263100.2</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>19</td>\n",
       "      <td>58345183</td>\n",
       "      <td>58353492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:2</td>\n",
       "      <td>ENSG00000175899.15</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>NP_000005.3</td>\n",
       "      <td>ENST00000318602.12</td>\n",
       "      <td>ENSP00000323929.8</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>12</td>\n",
       "      <td>9067708</td>\n",
       "      <td>9115919</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:9</td>\n",
       "      <td>ENSG00000171428.15</td>\n",
       "      <td>HGNC:7645</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>NM_000662.8</td>\n",
       "      <td>NP_000653.3</td>\n",
       "      <td>ENST00000307719.9</td>\n",
       "      <td>ENSP00000307218.4</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18210109</td>\n",
       "      <td>18223689</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:10</td>\n",
       "      <td>ENSG00000156006.5</td>\n",
       "      <td>HGNC:7646</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>NM_000015.3</td>\n",
       "      <td>NP_000006.2</td>\n",
       "      <td>ENST00000286479.4</td>\n",
       "      <td>ENSP00000286479.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>8</td>\n",
       "      <td>18391282</td>\n",
       "      <td>18401218</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:12</td>\n",
       "      <td>ENSG00000196136.18</td>\n",
       "      <td>HGNC:16</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>serpin family A member 3</td>\n",
       "      <td>NM_001085.5</td>\n",
       "      <td>NP_001076.2</td>\n",
       "      <td>ENST00000393078.5</td>\n",
       "      <td>ENSP00000376793.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>14</td>\n",
       "      <td>94612391</td>\n",
       "      <td>94624053</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #NCBI_GeneID        Ensembl_Gene    HGNC_ID    symbol  \\\n",
       "0     GeneID:1  ENSG00000121410.12     HGNC:5      A1BG   \n",
       "1     GeneID:2  ENSG00000175899.15     HGNC:7       A2M   \n",
       "2     GeneID:9  ENSG00000171428.15  HGNC:7645      NAT1   \n",
       "3    GeneID:10   ENSG00000156006.5  HGNC:7646      NAT2   \n",
       "4    GeneID:12  ENSG00000196136.18    HGNC:16  SERPINA3   \n",
       "\n",
       "                       name   RefSeq_nuc  RefSeq_prot         Ensembl_nuc  \\\n",
       "0    alpha-1-B glycoprotein  NM_130786.4  NP_570602.2   ENST00000263100.8   \n",
       "1     alpha-2-macroglobulin  NM_000014.6  NP_000005.3  ENST00000318602.12   \n",
       "2     N-acetyltransferase 1  NM_000662.8  NP_000653.3   ENST00000307719.9   \n",
       "3     N-acetyltransferase 2  NM_000015.3  NP_000006.2   ENST00000286479.4   \n",
       "4  serpin family A member 3  NM_001085.5  NP_001076.2   ENST00000393078.5   \n",
       "\n",
       "        Ensembl_prot  MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "0  ENSP00000263100.2  MANE Select         19   58345183  58353492          -  \n",
       "1  ENSP00000323929.8  MANE Select         12    9067708   9115919          -  \n",
       "2  ENSP00000307218.4  MANE Select          8   18210109  18223689          +  \n",
       "3  ENSP00000286479.3  MANE Select          8   18391282  18401218          +  \n",
       "4  ENSP00000376793.3  MANE Select         14   94612391  94624053          +  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2refseq_mapping_filepath = home_dir+\"data/refseq/MANE.GRCh38.v1.0.summary.txt.gz\" # \"symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\" # \"Symbol\" is the Gene-symbol col\n",
    "# gene2refseq_mapping_filepath = home_dir+\"data/refseq/HumanLRGRefSeqGene.txt.gz\"  # \"Symbol\" is the Gene-symbol col\n",
    "\n",
    "gene_symbol_col_name = \"symbol\" if gene2refseq_mapping_filepath.__contains__(\"MANE\") else \"Symbol\"\n",
    "gene2refseq_mapping_df = pd.read_csv(gene2refseq_mapping_filepath, compression='gzip', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "\n",
    "print(gene2refseq_mapping_df.shape)\n",
    "print(gene2refseq_mapping_df[gene_symbol_col_name].unique().shape) # some gene can be mapped to multiple proteins\n",
    "print(gene2refseq_mapping_df[\"MANE_status\"].value_counts())\n",
    "gene2refseq_mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACNA1D         2\n",
      "CACNA1C         2\n",
      "DYNC2H1         2\n",
      "SLC39A14        2\n",
      "CLPB            2\n",
      "               ..\n",
      "ZNF432          1\n",
      "SAFB2           1\n",
      "DZIP3           1\n",
      "MARF1           1\n",
      "LOC122539214    1\n",
      "Name: symbol, Length: 19062, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>GeneID:8291</td>\n",
       "      <td>ENSG00000135636.16</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NM_003494.4</td>\n",
       "      <td>NP_003485.1</td>\n",
       "      <td>ENST00000258104.8</td>\n",
       "      <td>ENSP00000258104.3</td>\n",
       "      <td>MANE Plus Clinical</td>\n",
       "      <td>2</td>\n",
       "      <td>71453561</td>\n",
       "      <td>71686763</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5408</th>\n",
       "      <td>GeneID:8291</td>\n",
       "      <td>ENSG00000135636.16</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NM_001130987.2</td>\n",
       "      <td>NP_001124459.1</td>\n",
       "      <td>ENST00000410020.8</td>\n",
       "      <td>ENSP00000386881.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>2</td>\n",
       "      <td>71466699</td>\n",
       "      <td>71686763</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #NCBI_GeneID        Ensembl_Gene    HGNC_ID symbol       name  \\\n",
       "5407  GeneID:8291  ENSG00000135636.16  HGNC:3097   DYSF  dysferlin   \n",
       "5408  GeneID:8291  ENSG00000135636.16  HGNC:3097   DYSF  dysferlin   \n",
       "\n",
       "          RefSeq_nuc     RefSeq_prot        Ensembl_nuc       Ensembl_prot  \\\n",
       "5407     NM_003494.4     NP_003485.1  ENST00000258104.8  ENSP00000258104.3   \n",
       "5408  NM_001130987.2  NP_001124459.1  ENST00000410020.8  ENSP00000386881.3   \n",
       "\n",
       "             MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "5407  MANE Plus Clinical          2   71453561  71686763          +  \n",
       "5408         MANE Select          2   71466699  71686763          +  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gene2refseq_mapping_df[\"symbol\"].value_counts())\n",
    "gene2refseq_mapping_df[gene2refseq_mapping_df[\"symbol\"]==\"DYSF\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18534,)\n",
      "(4987804, 29)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'gene_symbol', '#NCBI_GeneID',\n",
      "       'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name', 'RefSeq_nuc',\n",
      "       'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot', 'MANE_status',\n",
      "       'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand'],\n",
      "      dtype='object')\n",
      "(18281,) (18281,)\n"
     ]
    }
   ],
   "source": [
    "dbsnp_missense_variants_df = df_remove_NNE.copy()\n",
    "dbsnp_missense_variants_df[\"gene_symbol\"] = dbsnp_missense_variants_df[\"genes\"].apply(lambda x: x.split(\",\")[0].split(\":\")[0])#HSFX4:101927685\"\n",
    "print(dbsnp_missense_variants_df[\"gene_symbol\"].unique().shape)\n",
    "# gene2refseq_filtered_df[gene2refseq_filtered_df[\"Symbol\"]==\"ANK2\"] # 73 refseq transcripts and 73 protein isoforms\n",
    "merged_df = dbsnp_missense_variants_df.merge(gene2refseq_mapping_df, how=\"inner\", left_on=\"gene_symbol\", right_on=gene_symbol_col_name) # mapping clinvar mRNA transcripts to protein accession\n",
    "print(merged_df.shape)\n",
    "print(merged_df.columns)\n",
    "print(merged_df[\"gene_symbol\"].unique().shape, merged_df[gene_symbol_col_name].unique().shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4987804, 30)\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 312738\n",
      "(4675066, 30)\n",
      "18326\n",
      "18268\n"
     ]
    }
   ],
   "source": [
    "def filter_protein_isoforms_to_canonical(x):\n",
    "    # return x[\"variations\"]\n",
    "    variations = x[\"variations\"].split(\",\")\n",
    "    # if len(variations) == 1: return \",\".join(variations)\n",
    "    new_variations = []\n",
    "    for var in variations:\n",
    "        if var.split(\":\")[0] == x[\"RefSeq_prot\"]:\n",
    "            new_variations.append(var)\n",
    "    return \",\".join(new_variations)\n",
    "\n",
    "# merged_df[[\"variations\", \"RefSeq_prot\"]].apply(xx, axis=1)\n",
    "merged_df[\"new_variations\"] = merged_df[[\"variations\", \"RefSeq_prot\"]].apply(filter_protein_isoforms_to_canonical, axis=1)\n",
    "print(merged_df.shape)\n",
    "merged_df_remove_NNE = filter_remove_null_nan_empty_entries(merged_df, a_col_name=\"new_variations\") # NNE: null, nan or empty variants\n",
    "print(merged_df_remove_NNE.shape)\n",
    "print(merged_df_remove_NNE[\"RefSeq_prot\"].unique().shape[0])\n",
    "print(merged_df_remove_NNE[\"gene_symbol\"].unique().shape[0])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'gene_symbol', '#NCBI_GeneID',\n",
      "       'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name', 'RefSeq_nuc',\n",
      "       'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot', 'MANE_status',\n",
      "       'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand', 'new_variations'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3379849/325403337.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  optim_df[\"snp_id\"] = optim_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4675066, 12)\n",
      "(4646054,)\n",
      "rs753176482     2\n",
      "rs140002194     2\n",
      "rs141704244     2\n",
      "rs141536854     2\n",
      "rs141497053     2\n",
      "               ..\n",
      "rs1186734094    1\n",
      "rs1186176189    1\n",
      "rs1186159489    1\n",
      "rs1184408711    1\n",
      "rs11798010      1\n",
      "Name: snp_id, Length: 4646054, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>variations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2580070</th>\n",
       "      <td>rs753176482</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:71664422</td>\n",
       "      <td>NC_000002.12:71664421:C:G,NC_000002.12:7166442...</td>\n",
       "      <td>2015/04/01 22:25</td>\n",
       "      <td>2022/10/17 18:18</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NP_003485.1</td>\n",
       "      <td>NP_003485.1:p.Pro1681Ala,NP_003485.1:p.Pro1681Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2580071</th>\n",
       "      <td>rs753176482</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:71664422</td>\n",
       "      <td>NC_000002.12:71664421:C:G,NC_000002.12:7166442...</td>\n",
       "      <td>2015/04/01 22:25</td>\n",
       "      <td>2022/10/17 18:18</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>HGNC:3097</td>\n",
       "      <td>DYSF</td>\n",
       "      <td>dysferlin</td>\n",
       "      <td>NP_001124459.1</td>\n",
       "      <td>NP_001124459.1:p.Pro1720Ala,NP_001124459.1:p.P...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              snp_id           acc      chrpos  \\\n",
       "2580070  rs753176482  NC_000002.12  2:71664422   \n",
       "2580071  rs753176482  NC_000002.12  2:71664422   \n",
       "\n",
       "                                                      spdi       create_date  \\\n",
       "2580070  NC_000002.12:71664421:C:G,NC_000002.12:7166442...  2015/04/01 22:25   \n",
       "2580071  NC_000002.12:71664421:C:G,NC_000002.12:7166442...  2015/04/01 22:25   \n",
       "\n",
       "              update_date gene_symbol    HGNC_ID symbol       name  \\\n",
       "2580070  2022/10/17 18:18        DYSF  HGNC:3097   DYSF  dysferlin   \n",
       "2580071  2022/10/17 18:18        DYSF  HGNC:3097   DYSF  dysferlin   \n",
       "\n",
       "            RefSeq_prot                                         variations  \n",
       "2580070     NP_003485.1  NP_003485.1:p.Pro1681Ala,NP_003485.1:p.Pro1681Ser  \n",
       "2580071  NP_001124459.1  NP_001124459.1:p.Pro1720Ala,NP_001124459.1:p.P...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(merged_df_remove_NNE.columns)\n",
    "columns = ['snp_id', 'acc', 'chrpos', 'spdi', 'create_date', 'update_date',\n",
    "           'gene_symbol', 'HGNC_ID', 'symbol', 'name', 'RefSeq_prot', 'new_variations']\n",
    "optim_df = merged_df_remove_NNE[columns]\n",
    "optim_df[\"snp_id\"] = optim_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n",
    "\n",
    "# an rs_id can be mapped to multiple proteins, because MANE sometimes maps single gene to multiple proteins.\n",
    "# the following is an example of that phenomena\n",
    "optim_df = optim_df.rename(columns={'new_variations': 'variations'})\n",
    "print(optim_df.shape)\n",
    "print(optim_df[\"snp_id\"].unique().shape)\n",
    "print(optim_df[\"snp_id\"].value_counts())\n",
    "optim_df[optim_df[\"snp_id\"]==\"rs753176482\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping with ALFA population frequency data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698', 'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702', 'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705']\n",
      "##fileformat=VCFv4.0\t\n",
      "\n",
      "##build_id=20201027095038\t\n",
      "\n",
      "##Population=https://www.ncbi.nlm.nih.gov/biosample/?term=GRAF-pop\t\n",
      "\n",
      "##FORMAT=<ID=AN,Number=1,Type=Integer,Description=\"Total allele count for the population, including REF\">\t\n",
      "\n",
      "##FORMAT=<ID=AC,Number=A,Type=Integer,Description=\"Allele count for each ALT allele for the population\">\t\n",
      "\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSAMN10492695\tSAMN10492696\tSAMN10492697\tSAMN10492698\tSAMN10492699\tSAMN10492700\tSAMN10492701\tSAMN10492702\tSAMN11605645\tSAMN10492703\tSAMN10492704\tSAMN10492705\n",
      "\n",
      "NC_000001.9\t144135212\trs1553120241\tG\tA\t.\t.\t.\tAN:AC\t8560:5387\t8:8\t256:224\t336:288\t32:24\t170:117\t32:24\t18:13\t20:15\t344:296\t288:248\t9432:6100\n",
      "\n",
      "NC_000001.9\t144148243\trs2236566\tG\tT\t.\t.\t.\tAN:AC\t5996:510\t0:0\t0:0\t0:0\t0:0\t0:0\t0:0\t0:0\t84:8\t0:0\t0:0\t6080:518\n",
      "\n",
      "NC_000001.9\t146267105\trs1553119693\tT\tG\t.\t.\t.\tAN:AC\t37168:28800\t36:22\t56:44\t1378:839\t18:14\t70:60\t10:9\t4836:3639\t452:322\t1414:861\t66:53\t44024:33749\n",
      "\n",
      "NC_000001.9\t148488564\t.\tC\tA\t.\t.\t.\tAN:AC\t8552:0\t8:0\t256:0\t338:0\t32:0\t170:0\t32:0\t16:0\t20:0\t346:0\t288:0\t9424:0\n",
      "\n",
      "NC_000001.10\t2701535\trs371068661\tC\tT\t.\t.\t.\tAN:AC\t134:9\t0:0\t0:0\t48:1\t0:0\t0:0\t0:0\t0:0\t188:15\t48:1\t0:0\t370:25\n",
      "\n"
     ]
    }
   ],
   "source": [
    "filepath = home_dir+\"data/ALFA_population_freq/freq.vcf.gz\"\n",
    "col_names = get_col_names_from_gzip(filepath, \"#CHROM\")\n",
    "print(col_names)\n",
    "alfa_popu_freq_df_chunk_iterator = pd.read_csv(filepath, compression='gzip', chunksize=1000, comment=\"#\", names=col_names, delim_whitespace=False, sep=\"\\t\")\n",
    "first_df = alfa_popu_freq_df_chunk_iterator.__next__()\n",
    "first_df\n",
    "\n",
    "import gzip\n",
    "with gzip.open(filepath, \"rt\") as file_handle:\n",
    "    for line_no, line in enumerate(file_handle):\n",
    "        print(line)\n",
    "        if line_no==10: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4646054\n",
      "0 (18099, 21)\n",
      "1 (5200, 21)\n",
      "2 (8667, 21)\n",
      "3 (13006, 21)\n",
      "4 (10625, 21)\n",
      "5 (11543, 21)\n",
      "6 (12827, 21)\n",
      "7 (8312, 21)\n",
      "8 (13537, 21)\n",
      "9 (5981, 21)\n",
      "10 (9720, 21)\n",
      "11 (9072, 21)\n",
      "12 (9952, 21)\n",
      "13 (12861, 21)\n",
      "14 (8517, 21)\n",
      "15 (2349, 21)\n",
      "16 (10570, 21)\n",
      "17 (2179, 21)\n",
      "18 (2070, 21)\n",
      "19 (5511, 21)\n",
      "20 (3642, 21)\n",
      "21 (1520, 21)\n",
      "22 (2519, 21)\n",
      "23 (3767, 21)\n",
      "24 (361, 21)\n",
      "25 (3408, 21)\n",
      "26 (3303, 21)\n",
      "27 (5676, 21)\n",
      "28 (6519, 21)\n",
      "29 (639, 21)\n",
      "30 (4112, 21)\n",
      "31 (1814, 21)\n",
      "32 (688, 21)\n",
      "33 (12630, 21)\n",
      "34 (7093, 21)\n",
      "35 (5982, 21)\n",
      "36 (3813, 21)\n",
      "37 (0, 21)\n",
      "38 (0, 21)\n",
      "39 (3822, 21)\n",
      "40 (7766, 21)\n",
      "41 (22367, 21)\n",
      "42 (17159, 21)\n",
      "43 (16528, 21)\n",
      "44 (12488, 21)\n",
      "45 (1776, 21)\n",
      "46 (4456, 21)\n",
      "47 (7628, 21)\n",
      "48 (4279, 21)\n",
      "49 (4476, 21)\n",
      "50 (7196, 21)\n",
      "51 (6156, 21)\n",
      "52 (3435, 21)\n",
      "53 (445, 21)\n",
      "54 (963, 21)\n",
      "55 (5332, 21)\n",
      "56 (9199, 21)\n",
      "57 (11319, 21)\n",
      "58 (8436, 21)\n",
      "59 (4118, 21)\n",
      "60 (4160, 21)\n",
      "61 (3690, 21)\n",
      "62 (3427, 21)\n",
      "63 (7464, 21)\n",
      "64 (12263, 21)\n",
      "65 (5571, 21)\n",
      "66 (5045, 21)\n",
      "67 (7062, 21)\n",
      "68 (2979, 21)\n",
      "69 (2618, 21)\n",
      "70 (8968, 21)\n",
      "71 (5499, 21)\n",
      "72 (1380, 21)\n",
      "73 (550, 21)\n",
      "74 (6242, 21)\n",
      "75 (235, 21)\n",
      "76 (3376, 21)\n",
      "77 (5629, 21)\n",
      "78 (2727, 21)\n",
      "79 (14541, 21)\n",
      "80 (5446, 21)\n",
      "81 (5465, 21)\n",
      "82 (2412, 21)\n",
      "83 (5211, 21)\n",
      "84 (1309, 21)\n",
      "85 (6020, 21)\n",
      "86 (4434, 21)\n",
      "87 (2644, 21)\n",
      "88 (238, 21)\n",
      "89 (6549, 21)\n",
      "90 (443, 21)\n",
      "91 (3987, 21)\n",
      "92 (3167, 21)\n",
      "93 (2250, 21)\n",
      "94 (6673, 21)\n",
      "95 (12304, 21)\n",
      "96 (741, 21)\n",
      "97 (638, 21)\n",
      "98 (2011, 21)\n",
      "99 (6720, 21)\n",
      "100 (1081, 21)\n",
      "101 (0, 21)\n",
      "102 (10169, 21)\n",
      "103 (6825, 21)\n",
      "104 (4368, 21)\n",
      "105 (2932, 21)\n",
      "106 (4898, 21)\n",
      "107 (6591, 21)\n",
      "108 (316, 21)\n",
      "109 (3457, 21)\n",
      "110 (1770, 21)\n",
      "111 (702, 21)\n",
      "112 (5298, 21)\n",
      "113 (7015, 21)\n",
      "114 (4064, 21)\n",
      "115 (1120, 21)\n",
      "116 (2148, 21)\n",
      "117 (979, 21)\n",
      "118 (1657, 21)\n",
      "119 (6548, 21)\n",
      "120 (340, 21)\n",
      "121 (4379, 21)\n",
      "122 (4576, 21)\n",
      "123 (3964, 21)\n",
      "124 (7086, 21)\n",
      "125 (5879, 21)\n",
      "126 (4262, 21)\n",
      "127 (21856, 21)\n",
      "128 (2485, 21)\n",
      "129 (3943, 21)\n",
      "130 (2777, 21)\n",
      "131 (4378, 21)\n",
      "132 (963, 21)\n",
      "133 (5029, 21)\n",
      "134 (7505, 21)\n",
      "135 (4007, 21)\n",
      "136 (5992, 21)\n",
      "137 (4040, 21)\n",
      "138 (3429, 21)\n",
      "139 (8428, 21)\n",
      "140 (10120, 21)\n",
      "141 (3126, 21)\n",
      "142 (4281, 21)\n",
      "143 (5253, 21)\n",
      "144 (8695, 21)\n",
      "145 (7067, 21)\n",
      "146 (7828, 21)\n",
      "147 (4160, 21)\n",
      "148 (2517, 21)\n",
      "149 (970, 21)\n",
      "150 (7835, 21)\n",
      "151 (5847, 21)\n",
      "152 (6344, 21)\n",
      "153 (2170, 21)\n",
      "154 (198, 21)\n",
      "155 (1598, 21)\n",
      "156 (1697, 21)\n",
      "157 (1581, 21)\n",
      "158 (2918, 21)\n",
      "159 (10318, 21)\n",
      "160 (5492, 21)\n",
      "161 (9543, 21)\n",
      "162 (19114, 21)\n",
      "163 (21682, 21)\n",
      "164 (7907, 21)\n",
      "165 (7604, 21)\n",
      "166 (181, 21)\n",
      "167 (3612, 21)\n",
      "168 (1910, 21)\n",
      "169 (2269, 21)\n",
      "170 (2016, 21)\n",
      "171 (1420, 21)\n",
      "172 (1322, 21)\n",
      "173 (329, 21)\n",
      "174 (109, 21)\n",
      "175 (767, 21)\n",
      "176 (326, 21)\n",
      "177 (649, 21)\n",
      "178 (4405, 21)\n",
      "179 (5670, 21)\n",
      "180 (584, 21)\n",
      "181 (1309, 21)\n",
      "182 (3146, 21)\n",
      "183 (8497, 21)\n",
      "184 (219, 21)\n",
      "185 (4525, 21)\n",
      "186 (11327, 21)\n",
      "187 (7057, 21)\n",
      "188 (9099, 21)\n",
      "189 (5548, 21)\n",
      "190 (2252, 21)\n",
      "191 (4820, 21)\n",
      "192 (4807, 21)\n",
      "193 (1122, 21)\n",
      "194 (4371, 21)\n",
      "195 (4317, 21)\n",
      "196 (2991, 21)\n",
      "197 (2976, 21)\n",
      "198 (506, 21)\n",
      "199 (2044, 21)\n",
      "200 (4334, 21)\n",
      "201 (2850, 21)\n",
      "202 (694, 21)\n",
      "203 (2293, 21)\n",
      "204 (3652, 21)\n",
      "205 (10785, 21)\n",
      "206 (2166, 21)\n",
      "207 (2430, 21)\n",
      "208 (11895, 21)\n",
      "209 (9543, 21)\n",
      "210 (10610, 21)\n",
      "211 (6133, 21)\n",
      "212 (5669, 21)\n",
      "213 (1500, 21)\n",
      "214 (3215, 21)\n",
      "215 (3812, 21)\n",
      "216 (774, 21)\n",
      "217 (1945, 21)\n",
      "218 (2236, 21)\n",
      "219 (526, 21)\n",
      "220 (0, 21)\n",
      "221 (1973, 21)\n",
      "222 (6732, 21)\n",
      "223 (2338, 21)\n",
      "224 (1313, 21)\n",
      "225 (4799, 21)\n",
      "226 (2210, 21)\n",
      "227 (6361, 21)\n",
      "228 (0, 21)\n",
      "229 (601, 21)\n",
      "230 (679, 21)\n",
      "231 (3160, 21)\n",
      "232 (5747, 21)\n",
      "233 (3875, 21)\n",
      "234 (5711, 21)\n",
      "235 (4220, 21)\n",
      "236 (4250, 21)\n",
      "237 (5988, 21)\n",
      "238 (3055, 21)\n",
      "239 (1235, 21)\n",
      "240 (929, 21)\n",
      "241 (3986, 21)\n",
      "242 (3074, 21)\n",
      "243 (3192, 21)\n",
      "244 (3942, 21)\n",
      "245 (4030, 21)\n",
      "246 (700, 21)\n",
      "247 (3131, 21)\n",
      "248 (4319, 21)\n",
      "249 (2746, 21)\n",
      "250 (2770, 21)\n",
      "251 (417, 21)\n",
      "252 (168, 21)\n",
      "253 (2050, 21)\n",
      "254 (2489, 21)\n",
      "255 (3587, 21)\n",
      "256 (1754, 21)\n",
      "257 (4174, 21)\n",
      "258 (5634, 21)\n",
      "259 (2666, 21)\n",
      "260 (379, 21)\n",
      "261 (2430, 21)\n",
      "262 (2785, 21)\n",
      "263 (1079, 21)\n",
      "264 (1653, 21)\n",
      "265 (1569, 21)\n",
      "266 (0, 21)\n",
      "267 (5058, 21)\n",
      "268 (6162, 21)\n",
      "269 (6263, 21)\n",
      "270 (2093, 21)\n",
      "271 (2412, 21)\n",
      "272 (2155, 21)\n",
      "273 (833, 21)\n",
      "274 (5497, 21)\n",
      "275 (548, 21)\n",
      "276 (339, 21)\n",
      "277 (801, 21)\n",
      "278 (328, 21)\n",
      "279 (2508, 21)\n",
      "280 (4109, 21)\n",
      "281 (7372, 21)\n",
      "282 (4136, 21)\n",
      "283 (2485, 21)\n",
      "284 (255, 21)\n",
      "285 (520, 21)\n",
      "286 (3786, 21)\n",
      "287 (2909, 21)\n",
      "288 (1219, 21)\n",
      "289 (2192, 21)\n",
      "290 (3008, 21)\n",
      "291 (5902, 21)\n",
      "292 (5491, 21)\n",
      "293 (5945, 21)\n",
      "294 (3449, 21)\n",
      "295 (1794, 21)\n",
      "296 (511, 21)\n",
      "297 (3215, 21)\n",
      "298 (3867, 21)\n",
      "299 (2125, 21)\n",
      "300 (1152, 21)\n",
      "301 (841, 21)\n",
      "302 (628, 21)\n",
      "303 (2224, 21)\n",
      "304 (3812, 21)\n",
      "305 (2495, 21)\n",
      "306 (2237, 21)\n",
      "307 (2646, 21)\n",
      "308 (1232, 21)\n",
      "309 (4149, 21)\n",
      "310 (6291, 21)\n",
      "311 (4335, 21)\n",
      "312 (6203, 21)\n",
      "313 (20693, 21)\n",
      "314 (2422, 21)\n",
      "315 (5295, 21)\n",
      "316 (9893, 21)\n",
      "317 (3093, 21)\n",
      "318 (3410, 21)\n",
      "319 (2824, 21)\n",
      "320 (513, 21)\n",
      "321 (2839, 21)\n",
      "322 (3893, 21)\n",
      "323 (2147, 21)\n",
      "324 (10898, 21)\n",
      "325 (9479, 21)\n",
      "326 (3187, 21)\n",
      "327 (1843, 21)\n",
      "328 (4090, 21)\n",
      "329 (4153, 21)\n",
      "330 (2435, 21)\n",
      "331 (2974, 21)\n",
      "332 (1091, 21)\n",
      "333 (8218, 21)\n",
      "334 (9355, 21)\n",
      "335 (29143, 21)\n",
      "336 (10089, 21)\n",
      "337 (5934, 21)\n",
      "338 (14322, 21)\n",
      "339 (4719, 21)\n",
      "340 (3521, 21)\n",
      "341 (4616, 21)\n",
      "342 (3133, 21)\n",
      "343 (4953, 21)\n",
      "344 (428, 21)\n",
      "345 (2253, 21)\n",
      "346 (364, 21)\n",
      "347 (3260, 21)\n",
      "348 (4915, 21)\n",
      "349 (1130, 21)\n",
      "350 (1667, 21)\n",
      "351 (3859, 21)\n",
      "352 (3516, 21)\n",
      "353 (4725, 21)\n",
      "354 (349, 21)\n",
      "355 (1770, 21)\n",
      "356 (3149, 21)\n",
      "357 (267, 21)\n",
      "358 (3600, 21)\n",
      "359 (5895, 21)\n",
      "360 (3407, 21)\n",
      "361 (5573, 21)\n",
      "362 (1755, 21)\n",
      "363 (1800, 21)\n",
      "364 (2055, 21)\n",
      "365 (3850, 21)\n",
      "366 (4418, 21)\n",
      "367 (4169, 21)\n",
      "368 (3674, 21)\n",
      "369 (3161, 21)\n",
      "370 (4331, 21)\n",
      "371 (4419, 21)\n",
      "372 (7047, 21)\n",
      "373 (3322, 21)\n",
      "374 (8319, 21)\n",
      "375 (1529, 21)\n",
      "376 (1431, 21)\n",
      "377 (4683, 21)\n",
      "378 (6951, 21)\n",
      "379 (5504, 21)\n",
      "380 (11815, 21)\n",
      "381 (1861, 21)\n",
      "382 (2747, 21)\n",
      "383 (826, 21)\n",
      "384 (1898, 21)\n",
      "385 (2470, 21)\n",
      "386 (5861, 21)\n",
      "387 (3894, 21)\n",
      "388 (4384, 21)\n",
      "389 (2472, 21)\n",
      "390 (2272, 21)\n",
      "391 (2730, 21)\n",
      "392 (4474, 21)\n",
      "393 (5016, 21)\n",
      "394 (5869, 21)\n",
      "395 (913, 21)\n",
      "396 (2311, 21)\n",
      "397 (473, 21)\n",
      "398 (261, 21)\n",
      "399 (3987, 21)\n",
      "400 (89, 21)\n",
      "401 (3968, 21)\n",
      "402 (7291, 21)\n",
      "403 (2339, 21)\n",
      "404 (4087, 21)\n",
      "405 (672, 21)\n",
      "406 (3984, 21)\n",
      "407 (6989, 21)\n",
      "408 (3655, 21)\n",
      "409 (7673, 21)\n",
      "410 (25132, 21)\n",
      "411 (4790, 21)\n",
      "412 (5709, 21)\n",
      "413 (1541, 21)\n",
      "414 (1377, 21)\n",
      "415 (3507, 21)\n",
      "416 (2819, 21)\n",
      "417 (2820, 21)\n",
      "418 (6618, 21)\n",
      "419 (4003, 21)\n",
      "420 (4255, 21)\n",
      "421 (3490, 21)\n",
      "422 (8118, 21)\n",
      "423 (8317, 21)\n",
      "424 (1004, 21)\n",
      "425 (13638, 21)\n",
      "426 (3339, 21)\n",
      "427 (2978, 21)\n",
      "428 (2913, 21)\n",
      "429 (5690, 21)\n",
      "430 (304, 21)\n",
      "431 (3610, 21)\n",
      "432 (3416, 21)\n",
      "433 (7649, 21)\n",
      "434 (1907, 21)\n",
      "435 (1808, 21)\n",
      "436 (5524, 21)\n",
      "437 (6925, 21)\n",
      "438 (6273, 21)\n",
      "439 (6146, 21)\n",
      "440 (3221, 21)\n",
      "441 (1342, 21)\n",
      "442 (5579, 21)\n",
      "443 (4208, 21)\n",
      "444 (2197, 21)\n",
      "445 (2869, 21)\n",
      "446 (1176, 21)\n",
      "447 (3541, 21)\n",
      "448 (1848, 21)\n",
      "449 (2641, 21)\n",
      "450 (1007, 21)\n",
      "451 (4168, 21)\n",
      "452 (2428, 21)\n",
      "453 (2279, 21)\n",
      "454 (2216, 21)\n",
      "455 (1185, 21)\n",
      "456 (860, 21)\n",
      "457 (3168, 21)\n",
      "458 (1511, 21)\n",
      "459 (2426, 21)\n",
      "460 (3632, 21)\n",
      "461 (5696, 21)\n",
      "462 (3140, 21)\n",
      "463 (2023, 21)\n",
      "464 (3322, 21)\n",
      "465 (1623, 21)\n",
      "466 (1151, 21)\n",
      "467 (3329, 21)\n",
      "468 (3419, 21)\n",
      "469 (2558, 21)\n",
      "470 (1051, 21)\n",
      "471 (4553, 21)\n",
      "472 (679, 21)\n",
      "473 (1623, 21)\n",
      "474 (4107, 21)\n",
      "475 (28563, 21)\n",
      "476 (3791, 21)\n",
      "477 (3161, 21)\n",
      "478 (4667, 21)\n",
      "479 (1008, 21)\n",
      "480 (857, 21)\n",
      "481 (4164, 21)\n",
      "482 (1635, 21)\n",
      "483 (6200, 21)\n",
      "484 (1759, 21)\n",
      "485 (308, 21)\n",
      "486 (2275, 21)\n",
      "487 (2261, 21)\n",
      "488 (14778, 21)\n",
      "489 (5632, 21)\n",
      "490 (1988, 21)\n",
      "491 (0, 21)\n",
      "492 (374, 21)\n",
      "493 (3381, 21)\n",
      "494 (3943, 21)\n",
      "495 (1957, 21)\n",
      "496 (4688, 21)\n",
      "497 (1852, 21)\n",
      "498 (2625, 21)\n",
      "499 (4219, 21)\n",
      "500 (2347, 21)\n",
      "501 (7582, 21)\n",
      "502 (6233, 21)\n",
      "503 (5275, 21)\n",
      "504 (3083, 21)\n",
      "505 (2167, 21)\n",
      "506 (6912, 21)\n",
      "507 (8885, 21)\n",
      "508 (2900, 21)\n",
      "509 (4176, 21)\n",
      "510 (7654, 21)\n",
      "511 (7371, 21)\n",
      "512 (15471, 21)\n",
      "513 (11971, 21)\n",
      "514 (21593, 21)\n",
      "515 (6667, 21)\n",
      "516 (1126, 21)\n",
      "517 (6031, 21)\n",
      "518 (498, 21)\n",
      "519 (3677, 21)\n",
      "520 (4065, 21)\n",
      "521 (4310, 21)\n",
      "522 (2827, 21)\n",
      "523 (4501, 21)\n",
      "524 (4045, 21)\n",
      "525 (3880, 21)\n",
      "526 (2456, 21)\n",
      "527 (1887, 21)\n",
      "528 (0, 21)\n",
      "529 (2691, 21)\n",
      "530 (6060, 21)\n",
      "531 (7223, 21)\n",
      "532 (1540, 21)\n",
      "533 (146, 21)\n",
      "534 (3254, 21)\n",
      "535 (2630, 21)\n",
      "536 (462, 21)\n",
      "537 (9839, 21)\n",
      "538 (10413, 21)\n",
      "539 (2216, 21)\n",
      "540 (4313, 21)\n",
      "541 (322, 21)\n",
      "542 (5072, 21)\n",
      "543 (4960, 21)\n",
      "544 (6253, 21)\n",
      "545 (9519, 21)\n",
      "546 (8791, 21)\n",
      "547 (14447, 21)\n",
      "548 (2126, 21)\n",
      "549 (708, 21)\n",
      "550 (3861, 21)\n",
      "551 (5613, 21)\n",
      "552 (4608, 21)\n",
      "553 (6184, 21)\n",
      "554 (4741, 21)\n",
      "555 (3364, 21)\n",
      "556 (4055, 21)\n",
      "557 (25873, 21)\n",
      "558 (8666, 21)\n",
      "559 (18279, 21)\n",
      "560 (7100, 21)\n",
      "561 (3067, 21)\n",
      "562 (6528, 21)\n",
      "563 (6831, 21)\n",
      "564 (1123, 21)\n",
      "565 (1071, 21)\n",
      "566 (1611, 21)\n",
      "567 (5185, 21)\n",
      "568 (4609, 21)\n",
      "569 (201, 21)\n",
      "570 (115, 21)\n",
      "571 (6116, 21)\n",
      "572 (9234, 21)\n",
      "573 (355, 21)\n",
      "574 (611, 21)\n",
      "575 (12022, 21)\n",
      "576 (11559, 21)\n",
      "577 (17564, 21)\n",
      "578 (28383, 21)\n",
      "579 (10659, 21)\n",
      "580 (8485, 21)\n",
      "581 (8381, 21)\n",
      "582 (5965, 21)\n",
      "583 (1910, 21)\n",
      "584 (2959, 21)\n",
      "585 (1884, 21)\n",
      "586 (2253, 21)\n",
      "587 (8374, 21)\n",
      "588 (775, 21)\n",
      "589 (2363, 21)\n",
      "590 (5590, 21)\n",
      "591 (6029, 21)\n",
      "592 (2496, 21)\n",
      "593 (5830, 21)\n",
      "594 (6813, 21)\n",
      "595 (11991, 21)\n",
      "596 (9011, 21)\n",
      "597 (3765, 21)\n",
      "598 (5280, 21)\n",
      "599 (3116, 21)\n",
      "600 (5985, 21)\n",
      "601 (5639, 21)\n",
      "602 (15624, 21)\n",
      "603 (5911, 21)\n",
      "604 (5086, 21)\n",
      "605 (1801, 21)\n",
      "606 (4006, 21)\n",
      "607 (1543, 21)\n",
      "608 (3996, 21)\n",
      "609 (3163, 21)\n",
      "610 (2897, 21)\n",
      "611 (0, 21)\n",
      "612 (1717, 21)\n",
      "613 (2940, 21)\n",
      "614 (4304, 21)\n",
      "615 (12863, 21)\n",
      "616 (16196, 21)\n",
      "617 (15778, 21)\n",
      "618 (15963, 21)\n",
      "619 (1064, 21)\n",
      "620 (3606, 21)\n",
      "621 (3427, 21)\n",
      "622 (5030, 21)\n",
      "623 (822, 21)\n",
      "624 (2363, 21)\n",
      "625 (3991, 21)\n",
      "626 (2018, 21)\n",
      "627 (2152, 21)\n",
      "628 (1661, 21)\n",
      "629 (3300, 21)\n",
      "630 (4304, 21)\n",
      "631 (5805, 21)\n",
      "632 (4950, 21)\n",
      "633 (5290, 21)\n",
      "634 (10966, 21)\n",
      "635 (6723, 21)\n",
      "636 (3509, 21)\n",
      "637 (9478, 21)\n",
      "638 (13823, 21)\n",
      "639 (413, 21)\n",
      "640 (3110, 21)\n",
      "641 (9907, 21)\n",
      "642 (0, 21)\n",
      "643 (4081, 21)\n",
      "644 (3777, 21)\n",
      "645 (5805, 21)\n",
      "646 (3804, 21)\n",
      "647 (6267, 21)\n",
      "648 (2892, 21)\n",
      "649 (3707, 21)\n",
      "650 (4675, 21)\n",
      "651 (4406, 21)\n",
      "652 (3367, 21)\n",
      "653 (4756, 21)\n",
      "654 (8, 21)\n",
      "655 (1338, 21)\n",
      "656 (421, 21)\n",
      "657 (176, 21)\n",
      "658 (518, 21)\n",
      "659 (1357, 21)\n",
      "660 (1533, 21)\n",
      "661 (3570, 21)\n",
      "662 (106, 21)\n",
      "663 (692, 21)\n",
      "664 (446, 21)\n",
      "665 (275, 21)\n",
      "666 (1518, 21)\n",
      "667 (3139, 21)\n",
      "668 (3646, 21)\n",
      "669 (4814, 21)\n",
      "670 (1021, 21)\n",
      "671 (4256, 21)\n",
      "672 (7126, 21)\n",
      "673 (8076, 21)\n",
      "674 (16185, 21)\n",
      "675 (4534, 21)\n",
      "676 (796, 21)\n",
      "677 (4656, 21)\n",
      "678 (3559, 21)\n",
      "679 (3485, 21)\n",
      "680 (234, 21)\n",
      "681 (3087, 21)\n",
      "682 (1396, 21)\n",
      "683 (6920, 21)\n",
      "684 (3791, 21)\n",
      "685 (4172, 21)\n",
      "686 (4462, 21)\n",
      "687 (7742, 21)\n",
      "688 (3048, 21)\n",
      "689 (5143, 21)\n",
      "690 (12092, 21)\n",
      "691 (5046, 21)\n",
      "692 (1935, 21)\n",
      "693 (284, 21)\n",
      "694 (2674, 21)\n",
      "695 (5693, 21)\n",
      "696 (8886, 21)\n",
      "697 (2075, 21)\n",
      "698 (4185, 21)\n",
      "699 (16382, 21)\n",
      "700 (2802, 21)\n",
      "701 (3566, 21)\n",
      "702 (3513, 21)\n",
      "703 (4168, 21)\n",
      "704 (6614, 21)\n",
      "705 (5676, 21)\n",
      "706 (1010, 21)\n",
      "707 (21149, 21)\n",
      "708 (13440, 21)\n",
      "709 (4097, 21)\n",
      "710 (8282, 21)\n",
      "711 (3060, 21)\n",
      "712 (4221, 21)\n",
      "713 (3419, 21)\n",
      "714 (8013, 21)\n",
      "715 (9274, 21)\n",
      "716 (3971, 21)\n",
      "717 (9868, 21)\n",
      "718 (9031, 21)\n",
      "719 (7618, 21)\n",
      "720 (6227, 21)\n",
      "721 (4471, 21)\n",
      "722 (10508, 21)\n",
      "723 (4681, 21)\n",
      "724 (566, 21)\n",
      "725 (2400, 21)\n",
      "726 (15578, 21)\n",
      "727 (30937, 21)\n",
      "728 (13298, 21)\n",
      "729 (264, 21)\n",
      "730 (1900, 21)\n",
      "731 (7409, 21)\n",
      "732 (5081, 21)\n",
      "733 (5158, 21)\n",
      "734 (10878, 21)\n",
      "735 (7234, 21)\n",
      "736 (7744, 21)\n",
      "737 (20486, 21)\n",
      "738 (0, 21)\n",
      "739 (3166, 21)\n",
      "740 (4911, 21)\n",
      "741 (2800, 21)\n",
      "742 (8832, 21)\n",
      "743 (3170, 21)\n",
      "744 (248, 21)\n",
      "745 (8506, 21)\n",
      "746 (14809, 21)\n",
      "747 (11130, 21)\n",
      "748 (6149, 21)\n",
      "749 (2676, 21)\n",
      "750 (1101, 21)\n",
      "751 (2650, 21)\n",
      "752 (8997, 21)\n",
      "753 (3640, 21)\n",
      "754 (18343, 21)\n",
      "755 (12656, 21)\n",
      "756 (17342, 21)\n",
      "757 (22263, 21)\n",
      "758 (9795, 21)\n",
      "759 (3214, 21)\n",
      "760 (13482, 21)\n",
      "761 (3497, 21)\n",
      "762 (1460, 21)\n",
      "763 (14684, 21)\n",
      "764 (3423, 21)\n",
      "765 (11856, 21)\n",
      "766 (21416, 21)\n",
      "767 (20171, 21)\n",
      "768 (9713, 21)\n",
      "769 (12430, 21)\n",
      "770 (820, 21)\n",
      "771 (5876, 21)\n",
      "772 (7482, 21)\n",
      "773 (11600, 21)\n",
      "774 (4715, 21)\n",
      "775 (4317, 21)\n",
      "776 (5618, 21)\n",
      "777 (19178, 21)\n",
      "778 (13644, 21)\n",
      "779 (15924, 21)\n",
      "780 (4050, 21)\n",
      "781 (2238, 21)\n",
      "782 (4792, 21)\n",
      "783 (5148, 21)\n",
      "784 (2126, 21)\n",
      "785 (319, 21)\n",
      "786 (6420, 21)\n",
      "787 (2136, 21)\n",
      "788 (1326, 21)\n",
      "789 (5580, 21)\n",
      "790 (3683, 21)\n",
      "791 (0, 21)\n",
      "792 (545, 21)\n",
      "793 (6182, 21)\n",
      "794 (3481, 21)\n",
      "795 (1802, 21)\n",
      "796 (2040, 21)\n",
      "797 (4093, 21)\n",
      "798 (3935, 21)\n",
      "799 (2338, 21)\n",
      "800 (1066, 21)\n",
      "801 (1572, 21)\n",
      "802 (2960, 21)\n",
      "803 (2791, 21)\n",
      "804 (21470, 21)\n",
      "805 (19202, 21)\n",
      "806 (16483, 21)\n",
      "807 (20263, 21)\n",
      "808 (21703, 21)\n",
      "809 (17442, 21)\n",
      "810 (22922, 21)\n",
      "811 (9896, 21)\n",
      "812 (4368, 21)\n",
      "813 (1669, 21)\n",
      "814 (5814, 21)\n",
      "815 (18554, 21)\n",
      "816 (22403, 21)\n",
      "817 (18207, 21)\n",
      "818 (21982, 21)\n",
      "819 (24428, 21)\n",
      "820 (23343, 21)\n",
      "821 (23317, 21)\n",
      "822 (24668, 21)\n",
      "823 (10345, 21)\n",
      "824 (7084, 21)\n",
      "825 (1713, 21)\n",
      "826 (1567, 21)\n",
      "827 (1769, 21)\n",
      "828 (3762, 21)\n",
      "829 (3933, 21)\n",
      "830 (5560, 21)\n",
      "831 (0, 21)\n",
      "832 (4790, 21)\n",
      "833 (12824, 21)\n",
      "834 (7829, 21)\n",
      "835 (3345, 21)\n",
      "836 (9694, 21)\n",
      "837 (6335, 21)\n",
      "838 (5686, 21)\n",
      "839 (1737, 21)\n",
      "840 (3991, 21)\n",
      "841 (4008, 21)\n",
      "842 (17976, 21)\n",
      "843 (553, 21)\n",
      "844 (1418, 21)\n",
      "845 (1296, 21)\n",
      "846 (317, 21)\n",
      "847 (0, 21)\n",
      "848 (1776, 21)\n",
      "849 (4540, 21)\n",
      "850 (7666, 21)\n",
      "851 (4986, 21)\n",
      "852 (2784, 21)\n",
      "853 (10674, 21)\n",
      "854 (8442, 21)\n",
      "855 (678, 21)\n",
      "856 (8318, 21)\n",
      "857 (10468, 21)\n",
      "858 (9119, 21)\n",
      "859 (5215, 21)\n",
      "860 (13839, 21)\n",
      "861 (2050, 21)\n",
      "862 (13363, 21)\n",
      "863 (12167, 21)\n",
      "864 (8298, 21)\n",
      "865 (6655, 21)\n",
      "866 (11541, 21)\n",
      "867 (2909, 21)\n",
      "868 (2611, 21)\n",
      "869 (3468, 21)\n",
      "870 (4586, 21)\n",
      "871 (2439, 21)\n",
      "872 (1944, 21)\n",
      "873 (2474, 21)\n",
      "874 (2142, 21)\n",
      "875 (3285, 21)\n",
      "876 (1187, 21)\n",
      "877 (9294, 21)\n",
      "878 (5249, 21)\n",
      "879 (1361, 21)\n",
      "880 (1050, 21)\n",
      "881 (1520, 21)\n",
      "882 (5497, 21)\n",
      "883 (1934, 21)\n",
      "884 (2162, 21)\n",
      "885 (714, 21)\n",
      "886 (1701, 21)\n",
      "887 (361, 21)\n",
      "888 (623, 21)\n",
      "889 (264, 21)\n",
      "890 (5072, 21)\n",
      "891 (3743, 21)\n",
      "892 (4134, 21)\n",
      "893 (1245, 21)\n",
      "894 (3252, 21)\n",
      "895 (1237, 21)\n",
      "896 (993, 21)\n",
      "897 (4467, 21)\n",
      "898 (3967, 21)\n",
      "899 (2386, 21)\n",
      "900 (533, 21)\n",
      "901 (1922, 21)\n",
      "902 (11871, 21)\n",
      "903 (3492, 21)\n",
      "904 (404, 21)\n"
     ]
    }
   ],
   "source": [
    "# do not run this unless you changed something before, it takes 1 hour\n",
    "alfa_popu_freq_df_chunk_iterator = pd.read_csv(filepath, compression='gzip', chunksize=1000000, comment=\"#\", names=col_names, delim_whitespace=False, sep=\"\\t\")\n",
    "\n",
    "dbsnp_ids = list(optim_df[\"snp_id\"].unique())\n",
    "print(len(dbsnp_ids))\n",
    "\n",
    "chunk_list = []\n",
    "for i, alfa_popu_freq_df in enumerate(alfa_popu_freq_df_chunk_iterator):\n",
    "    chunk = alfa_popu_freq_df[alfa_popu_freq_df[\"ID\"].isin(dbsnp_ids)]\n",
    "    print(i, chunk.shape)\n",
    "    chunk_list.append(chunk)\n",
    "    # dbsnp_df = dbsnp_df.merge(filtered_alfa_popu_freq_df, how=\"left\", left_on=\"snp_id\", right_on=\"ID\") \n",
    "    # print(dbsnp_df.shape)\n",
    "    # if i==0: break\n",
    "\n",
    "filtered_alfa_popu_freq_df = pd.concat(chunk_list)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4645985, 21)\n",
      "(4645985,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14405</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69116</td>\n",
       "      <td>rs757299236</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:1</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14410</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69134</td>\n",
       "      <td>rs781394307</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:9</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:5</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14414</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69149</td>\n",
       "      <td>rs1458521218</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:3</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:1</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14419</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69184</td>\n",
       "      <td>rs1231643173</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14422</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69200</td>\n",
       "      <td>rs1453226491</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:5</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:5</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              CHROM    POS            ID REF ALT QUAL FILTER INFO FORMAT  \\\n",
       "14405  NC_000001.11  69116   rs757299236   G   A    .      .    .  AN:AC   \n",
       "14410  NC_000001.11  69134   rs781394307   A   G    .      .    .  AN:AC   \n",
       "14414  NC_000001.11  69149  rs1458521218   T   A    .      .    .  AN:AC   \n",
       "14419  NC_000001.11  69184  rs1231643173   G   A    .      .    .  AN:AC   \n",
       "14422  NC_000001.11  69200  rs1453226491   T   C    .      .    .  AN:AC   \n",
       "\n",
       "      SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699 SAMN10492700  \\\n",
       "14405       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "14410       7618:9  ...         84:0       2708:0        146:0        610:5   \n",
       "14414       7618:3  ...         84:0       2708:0        146:1        610:0   \n",
       "14419       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "14422       7618:0  ...         84:0       2708:5        146:0        610:0   \n",
       "\n",
       "      SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703 SAMN10492704  \\\n",
       "14405         24:0         94:0        470:1       2816:0        108:0   \n",
       "14410         24:0         94:0        470:0       2816:0        108:0   \n",
       "14414         24:0         94:0        470:0       2816:0        108:0   \n",
       "14419         24:0         94:0        470:0       2816:0        108:0   \n",
       "14422         24:0         94:0        470:0       2816:5        108:0   \n",
       "\n",
       "      SAMN10492705  \n",
       "14405      11862:1  \n",
       "14410     11862:14  \n",
       "14414      11862:4  \n",
       "14419      11862:0  \n",
       "14422      11862:5  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is associated with the previous block of codes\n",
    "# out_filepath = home_dir+\"data/ALFA_population_freq/freq_filtered.vcf.gz\"\n",
    "# if os.path.exists(out_filepath): os.remove(out_filepath)\n",
    "# filtered_alfa_popu_freq_df.to_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=1000000, index=False, mode=\"a\", header=True)\n",
    "# print(filtered_alfa_popu_freq_df.shape)\n",
    "# print(filtered_alfa_popu_freq_df[\"ID\"].unique().shape)\n",
    "# filtered_alfa_popu_freq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4645985, 21)\n",
      "(4645985,)\n",
      "C                           1547270\n",
      "G                           1534441\n",
      "A                            783094\n",
      "T                            781174\n",
      "TA                                1\n",
      "CCACTTCTTTAATCACCACCACCA          1\n",
      "CCTT                              1\n",
      "ATAGCCTCTCCTCCTGTTCACGT           1\n",
      "AT                                1\n",
      "ATCT                              1\n",
      "Name: REF, dtype: int64\n",
      "A                           1137912\n",
      "T                           1135304\n",
      "G                            826882\n",
      "C                            823314\n",
      "A,T                          236409\n",
      "G,T                          154449\n",
      "A,C                          153408\n",
      "C,G                           69366\n",
      "A,G                           32040\n",
      "C,T                           31499\n",
      "A,G,T                         19538\n",
      "A,C,T                         19148\n",
      "A,C,G                          2887\n",
      "C,G,T                          2677\n",
      "T,A                             275\n",
      "T,G                             194\n",
      "A,T,G                           174\n",
      "A,T,C                           160\n",
      "C,A                              98\n",
      "G,C                              69\n",
      "G,A                              33\n",
      "G,T,A                            29\n",
      "T,C                              25\n",
      "C,T,G                            23\n",
      "G,A,T                            18\n",
      "A,G,C                            16\n",
      "C,T,A                            12\n",
      "C,A,T                             8\n",
      "G,C,T                             7\n",
      "C,A,G                             3\n",
      "C,G,A                             2\n",
      "T,A,C                             1\n",
      "AT                                1\n",
      "TCACTTCTTTGATCACCACCACGG          1\n",
      "T,G,A                             1\n",
      "GTGA                              1\n",
      "GTAGCCTCTCCTCCTGTTCACAC           1\n",
      "Name: ALT, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69116</td>\n",
       "      <td>rs757299236</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:1</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69134</td>\n",
       "      <td>rs781394307</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:9</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:5</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69149</td>\n",
       "      <td>rs1458521218</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:3</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:1</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69184</td>\n",
       "      <td>rs1231643173</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69200</td>\n",
       "      <td>rs1453226491</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:5</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:5</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM    POS            ID REF ALT QUAL FILTER INFO FORMAT  \\\n",
       "0  NC_000001.11  69116   rs757299236   G   A    .      .    .  AN:AC   \n",
       "1  NC_000001.11  69134   rs781394307   A   G    .      .    .  AN:AC   \n",
       "2  NC_000001.11  69149  rs1458521218   T   A    .      .    .  AN:AC   \n",
       "3  NC_000001.11  69184  rs1231643173   G   A    .      .    .  AN:AC   \n",
       "4  NC_000001.11  69200  rs1453226491   T   C    .      .    .  AN:AC   \n",
       "\n",
       "  SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699 SAMN10492700  \\\n",
       "0       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "1       7618:9  ...         84:0       2708:0        146:0        610:5   \n",
       "2       7618:3  ...         84:0       2708:0        146:1        610:0   \n",
       "3       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "4       7618:0  ...         84:0       2708:5        146:0        610:0   \n",
       "\n",
       "  SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703 SAMN10492704  \\\n",
       "0         24:0         94:0        470:1       2816:0        108:0   \n",
       "1         24:0         94:0        470:0       2816:0        108:0   \n",
       "2         24:0         94:0        470:0       2816:0        108:0   \n",
       "3         24:0         94:0        470:0       2816:0        108:0   \n",
       "4         24:0         94:0        470:0       2816:5        108:0   \n",
       "\n",
       "  SAMN10492705  \n",
       "0      11862:1  \n",
       "1     11862:14  \n",
       "2      11862:4  \n",
       "3      11862:0  \n",
       "4      11862:5  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_filepath = home_dir+\"data/ALFA_population_freq/freq_filtered.vcf.gz\"\n",
    "filtered_alfa_popu_freq_df = pd.read_csv(out_filepath, sep=\"\\t\")\n",
    "print(filtered_alfa_popu_freq_df.shape)\n",
    "print(filtered_alfa_popu_freq_df[\"ID\"].unique().shape)\n",
    "print(filtered_alfa_popu_freq_df[\"REF\"].value_counts())\n",
    "print(filtered_alfa_popu_freq_df[\"ALT\"].value_counts())\n",
    "filtered_alfa_popu_freq_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4674997, 33)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'create_date', 'update_date',\n",
      "       'gene_symbol', 'HGNC_ID', 'symbol', 'name', 'RefSeq_prot', 'variations',\n",
      "       'CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
      "       'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698',\n",
      "       'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702',\n",
      "       'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705'],\n",
      "      dtype='object')\n",
      "18326\n"
     ]
    }
   ],
   "source": [
    "dbsnp_with_prots_df = optim_df\n",
    "dbsnp_with_prots_popufreq_df = dbsnp_with_prots_df.merge(filtered_alfa_popu_freq_df, how=\"inner\", left_on=\"snp_id\", right_on=\"ID\") # mapping clinvar mRNA transcripts to protein accession\n",
    "print(dbsnp_with_prots_popufreq_df.shape)\n",
    "print(dbsnp_with_prots_popufreq_df.columns)\n",
    "print(dbsnp_with_prots_popufreq_df[\"RefSeq_prot\"].unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4674997, 33)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_filepath = home_dir+\"data/ALFA_population_freq/dbsnp_with_prots_and_population_freq_mapping.vcf.gz\"\n",
    "if os.path.exists(out_filepath): os.remove(out_filepath)\n",
    "dbsnp_with_prots_popufreq_df.to_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=1000000, index=False, mode=\"a\", header=True)\n",
    "dbsnp_with_prots_popufreq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dbsnp_with_prots_popufreq_df_iterator = pd.read_csv(out_filepath, sep=\"\\t\", compression='gzip', chunksize=10000)\n",
    "# dbsnp_with_prots_popufreq_first_df = dbsnp_with_prots_popufreq_df_iterator.__next__()\n",
    "# out_filepath = home_dir+\"data/ALFA_population_freq/dbsnp_with_prots_and_population_freq_mapping_small.txt\"\n",
    "# dbsnp_with_prots_popufreq_first_df.to_csv(out_filepath, sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1491004261</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128176888</td>\n",
       "      <td>NC_000002.12:128176887:T:C</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>HGNC:15663</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>UDP-glucose glycoprotein glucosyltransferase 1</td>\n",
       "      <td>...</td>\n",
       "      <td>86:0</td>\n",
       "      <td>2784:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1490185682</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128157261</td>\n",
       "      <td>NC_000002.12:128157260:T:C</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>HGNC:15663</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>UDP-glucose glycoprotein glucosyltransferase 1</td>\n",
       "      <td>...</td>\n",
       "      <td>86:0</td>\n",
       "      <td>2784:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1489586107</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128157294</td>\n",
       "      <td>NC_000002.12:128157293:A:G</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>HGNC:15663</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>UDP-glucose glycoprotein glucosyltransferase 1</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>354:0</td>\n",
       "      <td>18:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>4092:0</td>\n",
       "      <td>20:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>21382:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1487964835</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128138817</td>\n",
       "      <td>NC_000002.12:128138816:G:A</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>HGNC:15663</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>UDP-glucose glycoprotein glucosyltransferase 1</td>\n",
       "      <td>...</td>\n",
       "      <td>26:0</td>\n",
       "      <td>586:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>30:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>2276:0</td>\n",
       "      <td>594:0</td>\n",
       "      <td>56:0</td>\n",
       "      <td>8988:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1485620659</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:128133175</td>\n",
       "      <td>NC_000002.12:128133174:A:C</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 21:08</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>HGNC:15663</td>\n",
       "      <td>UGGT1</td>\n",
       "      <td>UDP-glucose glycoprotein glucosyltransferase 1</td>\n",
       "      <td>...</td>\n",
       "      <td>86:0</td>\n",
       "      <td>2784:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674992</th>\n",
       "      <td>rs201657622</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:144373466</td>\n",
       "      <td>NC_000001.11:144373465:C:T</td>\n",
       "      <td>2012/05/04 13:09</td>\n",
       "      <td>2022/10/12 17:08</td>\n",
       "      <td>PPIAL4E</td>\n",
       "      <td>HGNC:33997</td>\n",
       "      <td>PPIAL4E</td>\n",
       "      <td>peptidylprolyl isomerase A like 4E</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>8:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>4294:59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674993</th>\n",
       "      <td>rs200452456</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Y:23229611</td>\n",
       "      <td>NC_000024.10:23229610:A:T</td>\n",
       "      <td>2012/05/04 13:09</td>\n",
       "      <td>2022/10/13 18:31</td>\n",
       "      <td>DAZ2</td>\n",
       "      <td>HGNC:15964</td>\n",
       "      <td>DAZ2</td>\n",
       "      <td>deleted in azoospermia 2</td>\n",
       "      <td>...</td>\n",
       "      <td>10:0</td>\n",
       "      <td>260:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>12:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>1050:0</td>\n",
       "      <td>264:0</td>\n",
       "      <td>22:0</td>\n",
       "      <td>4262:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674994</th>\n",
       "      <td>rs142112856</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>4:9325189</td>\n",
       "      <td>NC_000004.12:9325188:A:G</td>\n",
       "      <td>2011/05/09 23:43</td>\n",
       "      <td>2022/10/13 04:44</td>\n",
       "      <td>USP17L24</td>\n",
       "      <td>HGNC:44453</td>\n",
       "      <td>USP17L24</td>\n",
       "      <td>ubiquitin specific peptidase 17 like family me...</td>\n",
       "      <td>...</td>\n",
       "      <td>84:35</td>\n",
       "      <td>2708:2323</td>\n",
       "      <td>146:34</td>\n",
       "      <td>610:37</td>\n",
       "      <td>24:10</td>\n",
       "      <td>94:7</td>\n",
       "      <td>470:138</td>\n",
       "      <td>2816:2427</td>\n",
       "      <td>108:45</td>\n",
       "      <td>11862:3689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674995</th>\n",
       "      <td>rs28613881</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>4:9363153</td>\n",
       "      <td>NC_000004.12:9363152:A:G</td>\n",
       "      <td>2005/05/24 13:24</td>\n",
       "      <td>2022/10/13 04:44</td>\n",
       "      <td>USP17L30</td>\n",
       "      <td>HGNC:44458</td>\n",
       "      <td>USP17L30</td>\n",
       "      <td>ubiquitin specific peptidase 17 like family me...</td>\n",
       "      <td>...</td>\n",
       "      <td>84:42</td>\n",
       "      <td>2708:1618</td>\n",
       "      <td>146:77</td>\n",
       "      <td>610:292</td>\n",
       "      <td>24:12</td>\n",
       "      <td>94:47</td>\n",
       "      <td>470:243</td>\n",
       "      <td>2816:1685</td>\n",
       "      <td>108:54</td>\n",
       "      <td>11862:6624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674996</th>\n",
       "      <td>rs11798010</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>X:49553056</td>\n",
       "      <td>NC_000023.11:49553055:T:C</td>\n",
       "      <td>2004/02/28 21:15</td>\n",
       "      <td>2022/10/17 03:51</td>\n",
       "      <td>GAGE12E</td>\n",
       "      <td>HGNC:31905</td>\n",
       "      <td>GAGE12E</td>\n",
       "      <td>G antigen 12E</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>6:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>20:6</td>\n",
       "      <td>6:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>108:40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4674997 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               snp_id           acc       chrpos                        spdi  \\\n",
       "0        rs1491004261  NC_000002.12  2:128176888  NC_000002.12:128176887:T:C   \n",
       "1        rs1490185682  NC_000002.12  2:128157261  NC_000002.12:128157260:T:C   \n",
       "2        rs1489586107  NC_000002.12  2:128157294  NC_000002.12:128157293:A:G   \n",
       "3        rs1487964835  NC_000002.12  2:128138817  NC_000002.12:128138816:G:A   \n",
       "4        rs1485620659  NC_000002.12  2:128133175  NC_000002.12:128133174:A:C   \n",
       "...               ...           ...          ...                         ...   \n",
       "4674992   rs201657622  NC_000001.11  1:144373466  NC_000001.11:144373465:C:T   \n",
       "4674993   rs200452456  NC_000024.10   Y:23229611   NC_000024.10:23229610:A:T   \n",
       "4674994   rs142112856  NC_000004.12    4:9325189    NC_000004.12:9325188:A:G   \n",
       "4674995    rs28613881  NC_000004.12    4:9363153    NC_000004.12:9363152:A:G   \n",
       "4674996    rs11798010  NC_000023.11   X:49553056   NC_000023.11:49553055:T:C   \n",
       "\n",
       "              create_date       update_date gene_symbol     HGNC_ID    symbol  \\\n",
       "0        2017/11/09 09:55  2022/10/12 21:08       UGGT1  HGNC:15663     UGGT1   \n",
       "1        2017/11/09 09:55  2022/10/12 21:08       UGGT1  HGNC:15663     UGGT1   \n",
       "2        2017/11/09 09:55  2022/10/12 21:08       UGGT1  HGNC:15663     UGGT1   \n",
       "3        2017/11/09 09:55  2022/10/12 21:08       UGGT1  HGNC:15663     UGGT1   \n",
       "4        2017/11/09 09:55  2022/10/12 21:08       UGGT1  HGNC:15663     UGGT1   \n",
       "...                   ...               ...         ...         ...       ...   \n",
       "4674992  2012/05/04 13:09  2022/10/12 17:08     PPIAL4E  HGNC:33997   PPIAL4E   \n",
       "4674993  2012/05/04 13:09  2022/10/13 18:31        DAZ2  HGNC:15964      DAZ2   \n",
       "4674994  2011/05/09 23:43  2022/10/13 04:44    USP17L24  HGNC:44453  USP17L24   \n",
       "4674995  2005/05/24 13:24  2022/10/13 04:44    USP17L30  HGNC:44458  USP17L30   \n",
       "4674996  2004/02/28 21:15  2022/10/17 03:51     GAGE12E  HGNC:31905   GAGE12E   \n",
       "\n",
       "                                                      name  ... SAMN10492697  \\\n",
       "0           UDP-glucose glycoprotein glucosyltransferase 1  ...         86:0   \n",
       "1           UDP-glucose glycoprotein glucosyltransferase 1  ...         86:0   \n",
       "2           UDP-glucose glycoprotein glucosyltransferase 1  ...          0:0   \n",
       "3           UDP-glucose glycoprotein glucosyltransferase 1  ...         26:0   \n",
       "4           UDP-glucose glycoprotein glucosyltransferase 1  ...         86:0   \n",
       "...                                                    ...  ...          ...   \n",
       "4674992                 peptidylprolyl isomerase A like 4E  ...          0:0   \n",
       "4674993                           deleted in azoospermia 2  ...         10:0   \n",
       "4674994  ubiquitin specific peptidase 17 like family me...  ...        84:35   \n",
       "4674995  ubiquitin specific peptidase 17 like family me...  ...        84:42   \n",
       "4674996                                      G antigen 12E  ...          0:0   \n",
       "\n",
       "        SAMN10492698 SAMN10492699  SAMN10492700 SAMN10492701 SAMN10492702  \\\n",
       "0             2784:0        146:0         610:0         26:0         98:0   \n",
       "1             2784:0        146:0         610:0         26:0         98:0   \n",
       "2               20:0        354:0          18:0          0:0          0:0   \n",
       "3              586:0          0:0           0:0         30:0          0:0   \n",
       "4             2784:0        146:0         610:0         26:0         98:0   \n",
       "...              ...          ...           ...          ...          ...   \n",
       "4674992          0:0          0:0           0:0          0:0          0:0   \n",
       "4674993        260:0          0:0           0:0         12:0          0:0   \n",
       "4674994    2708:2323       146:34        610:37        24:10         94:7   \n",
       "4674995    2708:1618       146:77       610:292        24:12        94:47   \n",
       "4674996          6:0          0:0           0:0          0:0          0:0   \n",
       "\n",
       "        SAMN11605645 SAMN10492703 SAMN10492704 SAMN10492705  \n",
       "0              496:0       2898:0        112:0      14050:0  \n",
       "1              496:0       2898:0        112:0      14050:0  \n",
       "2             4092:0         20:0          0:0      21382:1  \n",
       "3             2276:0        594:0         56:0       8988:1  \n",
       "4              496:0       2898:0        112:0      14050:0  \n",
       "...              ...          ...          ...          ...  \n",
       "4674992          8:0          0:0          0:0      4294:59  \n",
       "4674993       1050:0        264:0         22:0       4262:2  \n",
       "4674994      470:138    2816:2427       108:45   11862:3689  \n",
       "4674995      470:243    2816:1685       108:54   11862:6624  \n",
       "4674996         20:6          6:0          0:0       108:40  \n",
       "\n",
       "[4674997 rows x 33 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbsnp_with_prots_popufreq_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
