{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import pandas as pd\n",
    "from utils.column_names_reader import get_col_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10000, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>status</th>\n",
       "      <th>RNA_nucleotide_accession.version</th>\n",
       "      <th>RNA_nucleotide_gi</th>\n",
       "      <th>protein_accession.version</th>\n",
       "      <th>protein_gi</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>genomic_nucleotide_gi</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>assembly</th>\n",
       "      <th>mature_peptide_accession.version</th>\n",
       "      <th>mature_peptide_gi</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24</td>\n",
       "      <td>77267466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_011787424.1</td>\n",
       "      <td>500111419</td>\n",
       "      <td>NZ_CP104755.1</td>\n",
       "      <td>2310801391</td>\n",
       "      <td>0</td>\n",
       "      <td>1385</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>77267467</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_011787425.1</td>\n",
       "      <td>500111420</td>\n",
       "      <td>NZ_CP104755.1</td>\n",
       "      <td>2310801391</td>\n",
       "      <td>1406</td>\n",
       "      <td>2506</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>dnaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>77267468</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_261762234.1</td>\n",
       "      <td>2310800657</td>\n",
       "      <td>NZ_CP104755.1</td>\n",
       "      <td>2310801391</td>\n",
       "      <td>2703</td>\n",
       "      <td>3785</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>recF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24</td>\n",
       "      <td>77267469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_261762235.1</td>\n",
       "      <td>2310800658</td>\n",
       "      <td>NZ_CP104755.1</td>\n",
       "      <td>2310801391</td>\n",
       "      <td>3802</td>\n",
       "      <td>6219</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>gyrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>24</td>\n",
       "      <td>77267470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>WP_261762236.1</td>\n",
       "      <td>2310800659</td>\n",
       "      <td>NZ_CP104755.1</td>\n",
       "      <td>2310801391</td>\n",
       "      <td>6421</td>\n",
       "      <td>7467</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>N5094_RS00025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id    GeneID status RNA_nucleotide_accession.version RNA_nucleotide_gi  \\\n",
       "0      24  77267466    NaN                                -                 -   \n",
       "1      24  77267467    NaN                                -                 -   \n",
       "2      24  77267468    NaN                                -                 -   \n",
       "3      24  77267469    NaN                                -                 -   \n",
       "4      24  77267470    NaN                                -                 -   \n",
       "\n",
       "  protein_accession.version  protein_gi genomic_nucleotide_accession.version  \\\n",
       "0            WP_011787424.1   500111419                        NZ_CP104755.1   \n",
       "1            WP_011787425.1   500111420                        NZ_CP104755.1   \n",
       "2            WP_261762234.1  2310800657                        NZ_CP104755.1   \n",
       "3            WP_261762235.1  2310800658                        NZ_CP104755.1   \n",
       "4            WP_261762236.1  2310800659                        NZ_CP104755.1   \n",
       "\n",
       "   genomic_nucleotide_gi  start_position_on_the_genomic_accession  \\\n",
       "0             2310801391                                        0   \n",
       "1             2310801391                                     1406   \n",
       "2             2310801391                                     2703   \n",
       "3             2310801391                                     3802   \n",
       "4             2310801391                                     6421   \n",
       "\n",
       "   end_position_on_the_genomic_accession orientation assembly  \\\n",
       "0                                   1385           +        -   \n",
       "1                                   2506           +        -   \n",
       "2                                   3785           +        -   \n",
       "3                                   6219           +        -   \n",
       "4                                   7467           +        -   \n",
       "\n",
       "  mature_peptide_accession.version mature_peptide_gi         Symbol  \n",
       "0                                -                 -           dnaA  \n",
       "1                                -                 -           dnaN  \n",
       "2                                -                 -           recF  \n",
       "3                                -                 -           gyrB  \n",
       "4                                -                 -  N5094_RS00025  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inp_filepath = home_dir+\"data/gene/gene2refseq.gz\"\n",
    "col_names = get_col_names(inp_filepath, \"#tax_id\")\n",
    "gene2refseq_df_terator = pd.read_csv(inp_filepath, compression='gzip', comment='#', chunksize=10000, delim_whitespace=False, sep=\"\\t\", header=None, names=col_names)\n",
    "df = gene2refseq_df_terator.__next__()\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1650 (894, 16)\n",
      "1651 (2531, 16)\n",
      "1652 (2756, 16)\n",
      "1653 (2643, 16)\n",
      "1654 (2610, 16)\n",
      "1655 (2767, 16)\n",
      "1656 (2710, 16)\n",
      "1657 (2548, 16)\n",
      "1658 (2601, 16)\n",
      "1659 (2615, 16)\n",
      "1660 (2732, 16)\n",
      "1661 (2460, 16)\n",
      "1662 (2590, 16)\n",
      "1663 (2071, 16)\n",
      "1664 (2635, 16)\n",
      "1665 (2511, 16)\n",
      "1666 (2458, 16)\n",
      "1667 (2539, 16)\n",
      "1668 (2392, 16)\n",
      "1669 (2394, 16)\n",
      "1670 (2470, 16)\n",
      "1671 (2351, 16)\n",
      "1672 (2210, 16)\n",
      "1673 (2238, 16)\n",
      "1674 (1739, 16)\n",
      "1675 (1409, 16)\n",
      "1676 (690, 16)\n",
      "1677 (490, 16)\n",
      "1678 (156, 16)\n",
      "1679 (171, 16)\n",
      "1680 (68, 16)\n",
      "1681 (98, 16)\n",
      "1682 (13, 16)\n",
      "1683 (3, 16)\n",
      "1684 (6, 16)\n",
      "1685 (26, 16)\n",
      "1686 (54, 16)\n",
      "1687 (68, 16)\n",
      "1688 (21, 16)\n",
      "63738\n"
     ]
    }
   ],
   "source": [
    "# filtering gene2refseq on criteria defined below\n",
    "out_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\"\n",
    "if os.path.exists(out_filepath): os.remove(out_filepath)\n",
    "\n",
    "total_rows_filtered = 0\n",
    "for i, df in enumerate(gene2refseq_df_terator):\n",
    "\n",
    "    # keeping those that is associated with valid Protein-coding transcripts starting with \"NM_\"\n",
    "    # keeping those that is associated with valid complete genomic molecule starting with \"NC_\"\n",
    "    # keeping only those that associated with assymble GRCh38\n",
    "    mask = (df[\"RNA_nucleotide_accession.version\"].str.contains(\"NM_\")) & (df[\"genomic_nucleotide_accession.version\"].str.contains(\"NC_\")) & (df[\"assembly\"].str.contains(\"GRCh38\"))\n",
    "    df = df[mask]\n",
    "    \n",
    "    if not os.path.exists(out_filepath):\n",
    "        df.to_csv(out_filepath, mode=\"a\", compression='gzip', chunksize=10000, sep=\"\\t\", index=False, header=True)\n",
    "    else:\n",
    "        df.to_csv(out_filepath, mode=\"a\", compression='gzip', chunksize=10000, sep=\"\\t\", index=False, header=False)\n",
    "    total_rows_filtered += df.shape[0]\n",
    "    if df.shape[0] !=0: print(i, df.shape)\n",
    "    \n",
    "    # if i==200: break\n",
    "\n",
    "print(total_rows_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63738, 16)\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n",
      "Int64Index([], dtype='int64')\n"
     ]
    }
   ],
   "source": [
    "gene2refseq_filtered_filepath = home_dir+\"data/gene/gene2refseq_filtered.gz\"\n",
    "gene2refseq_filtered_df = pd.read_csv(gene2refseq_filtered_filepath, compression='gzip', comment='#', delim_whitespace=False, sep=\"\\t\")#, header=None, names=col_names)\n",
    "print(gene2refseq_filtered_df.shape)\n",
    "\n",
    "print(gene2refseq_filtered_df[pd.isna(gene2refseq_filtered_df[\"GeneID\"])].index)\n",
    "print(gene2refseq_filtered_df[pd.isnull(gene2refseq_filtered_df[\"GeneID\"])].index)\n",
    "print(gene2refseq_filtered_df[gene2refseq_filtered_df['GeneID'] == \"\"].index)\n",
    "print(gene2refseq_filtered_df[pd.isna(gene2refseq_filtered_df[\"Symbol\"])].index)\n",
    "print(gene2refseq_filtered_df[pd.isnull(gene2refseq_filtered_df[\"Symbol\"])].index)\n",
    "print(gene2refseq_filtered_df[gene2refseq_filtered_df['Symbol'] == \"\"].index)\n",
    "print(gene2refseq_filtered_df[pd.isna(gene2refseq_filtered_df[\"protein_accession.version\"])].index)\n",
    "print(gene2refseq_filtered_df[pd.isnull(gene2refseq_filtered_df[\"protein_accession.version\"])].index)\n",
    "print(gene2refseq_filtered_df[gene2refseq_filtered_df[\"protein_accession.version\"] == \"\"].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA_nucleotide_accession.version  protein_accession.version\n",
      "NM_005088.3                       NP_005079.2                  2\n",
      "NM_001379165.1                    NP_001366094.1               2\n",
      "NM_001379155.1                    NP_001366084.1               2\n",
      "NM_001379156.1                    NP_001366085.1               2\n",
      "NM_001161532.2                    NP_001155004.1               2\n",
      "                                                              ..\n",
      "NM_001320005.2                    NP_001306934.1               1\n",
      "NM_001320006.2                    NP_001306935.1               1\n",
      "NM_001320007.2                    NP_001306936.1               1\n",
      "NM_001320010.2                    NP_001306939.1               1\n",
      "NM_214711.4                       NP_999876.2                  1\n",
      "Length: 63675, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>status</th>\n",
       "      <th>RNA_nucleotide_accession.version</th>\n",
       "      <th>RNA_nucleotide_gi</th>\n",
       "      <th>protein_accession.version</th>\n",
       "      <th>protein_gi</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>genomic_nucleotide_gi</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>assembly</th>\n",
       "      <th>mature_peptide_accession.version</th>\n",
       "      <th>mature_peptide_gi</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>9606</td>\n",
       "      <td>1438</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001161532.2</td>\n",
       "      <td>1824163926</td>\n",
       "      <td>NP_001155004.1</td>\n",
       "      <td>238908521</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>568815575</td>\n",
       "      <td>1268813</td>\n",
       "      <td>1325217</td>\n",
       "      <td>+</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CSF2RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>9606</td>\n",
       "      <td>1438</td>\n",
       "      <td>REVIEWED</td>\n",
       "      <td>NM_001161532.2</td>\n",
       "      <td>1824163926</td>\n",
       "      <td>NP_001155004.1</td>\n",
       "      <td>238908521</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>568815574</td>\n",
       "      <td>1268813</td>\n",
       "      <td>1325217</td>\n",
       "      <td>+</td>\n",
       "      <td>Reference GRCh38.p14 Primary Assembly</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CSF2RA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tax_id  GeneID    status RNA_nucleotide_accession.version  \\\n",
       "3576    9606    1438  REVIEWED                   NM_001161532.2   \n",
       "3577    9606    1438  REVIEWED                   NM_001161532.2   \n",
       "\n",
       "      RNA_nucleotide_gi protein_accession.version  protein_gi  \\\n",
       "3576         1824163926            NP_001155004.1   238908521   \n",
       "3577         1824163926            NP_001155004.1   238908521   \n",
       "\n",
       "     genomic_nucleotide_accession.version  genomic_nucleotide_gi  \\\n",
       "3576                         NC_000023.11              568815575   \n",
       "3577                         NC_000024.10              568815574   \n",
       "\n",
       "      start_position_on_the_genomic_accession  \\\n",
       "3576                                  1268813   \n",
       "3577                                  1268813   \n",
       "\n",
       "      end_position_on_the_genomic_accession orientation  \\\n",
       "3576                                1325217           +   \n",
       "3577                                1325217           +   \n",
       "\n",
       "                                   assembly mature_peptide_accession.version  \\\n",
       "3576  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "3577  Reference GRCh38.p14 Primary Assembly                                -   \n",
       "\n",
       "     mature_peptide_gi  Symbol  \n",
       "3576                 -  CSF2RA  \n",
       "3577                 -  CSF2RA  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# same gene can be in the X and Y chromosome. Must keep them. After merging NM_ to NM_, one must remove wrongly mapped variants. See example in clinvar_mapping_gene2refseq.\n",
    "print(gene2refseq_filtered_df[[\"RNA_nucleotide_accession.version\", \"protein_accession.version\"]].value_counts())\n",
    "gene2refseq_filtered_df[gene2refseq_filtered_df[\"RNA_nucleotide_accession.version\"]==\"NM_001161532.2\"] # \n",
    "# gene2refseq_filtered_df.groupby(\"RNA_nucleotide_accession.version\").filter(lambda x: len(x) > 1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "831918974231b2fc7cbf226cf19505a68e47ade0e201ebff2df92e599b5c76db"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
