{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "home_dir = \"../../\"\n",
    "module_path = os.path.abspath(os.path.join(home_dir))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model_root_and_name_tuple_list = [(\"vespa_marquet\", \"vespal\"), (\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"bioembeddings_dallago\", \"prottrans_bert_bfd\")]\n",
    "\n",
    "model_root_and_name_tuple_list = [(\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"dbnsfp\", \"cadd\"), (\"dbnsfp\", \"mvp\"), (\"dbnsfp\", \"metarnn\"), (\"dbnsfp\", \"revel\"),\n",
    "                                  (\"tape_rao\", \"unirep\"), (\"tape_rao\", \"protbert\"), (\"sequnet_dunham\", \"sequnet\"), \n",
    "                                  (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"esm_rives\", \"esm1v_t33_650M_UR90S\"), (\"esm_rives\", \"esm2_t33_650M_UR50D\"),\n",
    "                                  (\"bioembeddings_dallago\", \"plus_rnn\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_bert_bfd\"), (\"bioembeddings_dallago\", \"prottrans_albert_bfd\"),\n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_xlnet_uniref100\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_t5_bfd\"), (\"bioembeddings_dallago\", \"prottrans_t5_uniref50\"), (\"bioembeddings_dallago\", \"prottrans_t5_xl_u50\"),\n",
    "                                  (\"vespa_marquet\", \"vespal\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CDD Conservation dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NPid         qPos\n",
      "NP_000005.3  129     1\n",
      "NP_056277.2  237     1\n",
      "             224     1\n",
      "             225     1\n",
      "             226     1\n",
      "                    ..\n",
      "NP_001703.2  263     1\n",
      "             264     1\n",
      "             265     1\n",
      "             266     1\n",
      "NP_999627.2  774     1\n",
      "Length: 4097087, dtype: int64\n",
      "(4097087, 36)\n",
      "Index(['qNo', 'NPid', 'accession', 'que', 'sub', 'ali', 'qPos', 'sPos', 'aPos',\n",
      "       'qPos_', 'sPos_', 'bitscore', 'A', 'D', 'C', 'E', 'F', 'G', 'H', 'I',\n",
      "       'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', 'cons',\n",
      "       'entropy', 'qcov', 'CScore'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "conservation_df = pd.read_csv(home_dir+\"data/cdd_conservationTable_18kHumanProts.csv.gzip\", compression='gzip', comment='#')\n",
    "conservation_df = conservation_df.drop_duplicates([\"NPid\", \"qPos\"], keep=\"first\")\n",
    "print(conservation_df[[\"NPid\", \"qPos\"]].value_counts())\n",
    "print(conservation_df.shape)\n",
    "print(conservation_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_conservation_df(df:pd.DataFrame):\n",
    "    # merge conservation_df with the input df\n",
    "    temp_conservation_df = conservation_df[[\"NPid\", \"qPos\", \"CScore\"]].drop_duplicates(keep=\"first\")\n",
    "    merged_df = pd.merge(df, temp_conservation_df, how=\"left\", left_on=[\"prot_acc_version\", \"prot_pos\"], right_on=[\"NPid\", \"qPos\"])\n",
    "    columns = list(df.columns)\n",
    "    columns.append(\"CScore\")\n",
    "    merged_df = merged_df[columns]\n",
    "    merged_df = merged_df.rename(columns={\"CScore\": \"conservation_pred\"})\n",
    "    return merged_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging pathogenicity analysis predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raw data: (10282, 12)\n",
      "Index(['clinvar_id', 'gene_symbol', 'gene_id', 'chrom_acc_version',\n",
      "       'chrom_pos', 'ref_allele', 'alt_allele', 'prot_acc_version', 'prot_pos',\n",
      "       'wt', 'mut', 'class'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 1434\n",
      "Pathogenic: 2308, Likely pathogenic: 4168, total: (6476, 12)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "clinvar_id  gene_symbol  gene_id  chrom_acc_version  chrom_pos  ref_allele  alt_allele  prot_acc_version  prot_pos  wt  mut  class            \n",
       "492         QDPR         5860.0   NC_000004.12       17509363   A           G           NP_000311.2       36        W   R    pathogenic           1\n",
       "1722630     G6PD         2539.0   NC_000023.11       154532389  C           A           NP_001346945.1    454       R   L    likely_pathogenic    1\n",
       "1722650     G6PD         2539.0   NC_000023.11       154536169  C           T           NP_001346945.1    44        A   T    pathogenic           1\n",
       "1722647     G6PD         2539.0   NC_000023.11       154533077  C           T           NP_001346945.1    306       G   S    pathogenic           1\n",
       "1722638     G6PD         2539.0   NC_000023.11       154532258  G           A           NP_001346945.1    463       R   C    likely_pathogenic    1\n",
       "                                                                                                                                                 ..\n",
       "1481925     NAGLU        4669.0   NC_000017.11       42537432   T           A           NP_000254.2       140       Y   N    likely_pathogenic    1\n",
       "1481897     AQP2         359.0    NC_000012.12       49954663   C           G           NP_000477.1       187       R   G    likely_pathogenic    1\n",
       "1481886     ETFDH        2110.0   NC_000004.12       158699113  A           G           NP_004444.2       367       N   D    likely_pathogenic    1\n",
       "1481857     ETFDH        2110.0   NC_000004.12       158695618  A           T           NP_004444.2       269       Q   L    likely_pathogenic    1\n",
       "2412792     ROR2         4920.0   NC_000009.12       91733419   C           G           NP_004551.2       214       G   R    likely_pathogenic    1\n",
       "Length: 6476, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_patho_and_likelypatho_SNVs\n",
    "patho_and_likelypatho_variants_df = get_patho_and_likelypatho_SNVs(home_dir)\n",
    "patho_and_likelypatho_variants_df.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (6476, 13)\n",
      "(6476, 13)\n",
      "dbnsfp polyphen2_HVAR (6476, 13)\n",
      "(6476, 14)\n",
      "dbnsfp cadd (6476, 13)\n",
      "(6476, 15)\n",
      "dbnsfp mvp (6476, 13)\n",
      "(6476, 16)\n",
      "dbnsfp metarnn (6476, 13)\n",
      "(6476, 17)\n",
      "dbnsfp revel (6476, 13)\n",
      "(6476, 18)\n",
      "tape_rao unirep (6476, 13)\n",
      "(6476, 19)\n",
      "tape_rao protbert (6476, 13)\n",
      "(6476, 20)\n",
      "sequnet_dunham sequnet (6472, 13)\n",
      "(6476, 21)\n",
      "esm_rives esm1b_t33_650M_UR50S (6476, 13)\n",
      "(6476, 22)\n",
      "esm_rives esm1v_t33_650M_UR90S (6476, 13)\n",
      "(6476, 23)\n",
      "esm_rives esm2_t33_650M_UR50D (6476, 13)\n",
      "(6476, 24)\n",
      "bioembeddings_dallago plus_rnn (6476, 13)\n",
      "(6476, 25)\n",
      "bioembeddings_dallago prottrans_bert_bfd (6476, 13)\n",
      "(6476, 26)\n",
      "bioembeddings_dallago prottrans_albert_bfd (6476, 13)\n",
      "(6476, 27)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (6476, 13)\n",
      "(6476, 28)\n",
      "bioembeddings_dallago prottrans_t5_bfd (6476, 13)\n",
      "(6476, 29)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (6476, 13)\n",
      "(6476, 30)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (6476, 13)\n",
      "(6476, 31)\n",
      "vespa_marquet vespal (5571, 13)\n",
      "(6476, 32)\n",
      "conservation_score\n",
      "(6476, 33)\n"
     ]
    }
   ],
   "source": [
    "task = \"patho_and_likelypatho\"\n",
    "\n",
    "merged_df = patho_and_likelypatho_variants_df.copy(deep=True)\n",
    "merge_col_list = list(patho_and_likelypatho_variants_df.columns)\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):    \n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}.csv\", sep=\"\\t\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    print(merged_df.shape)\n",
    "    # if i==5: break\n",
    "\n",
    "merged_df = merge_conservation_df(merged_df)  \n",
    "print(\"conservation_score\")  \n",
    "print(merged_df.shape)\n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}_analysis.csv\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging population frequency predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading data ...\n",
      "(2865836, 14)\n",
      "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
      "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
      "       'mut_poulation', 'wt_freq', 'mt_freq', 'class'],\n",
      "      dtype='object')\n",
      "NP_000242.1       1154\n",
      "NP_000526.2       1056\n",
      "NP_003493.1        898\n",
      "NP_000240.1        856\n",
      "NP_004351.1        830\n",
      "                  ... \n",
      "NP_001345345.1       1\n",
      "NP_001577.1          1\n",
      "NP_001005166.3       1\n",
      "NP_000992.1          1\n",
      "NP_001091888.1       1\n",
      "Name: prot_acc_version, Length: 15962, dtype: int64\n",
      "Common: 18167, rare: 28622, ultra-rare: 314013, singletons: 462444, zero-population: 2042590 and total: 2865836\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>chrom_acc_version</th>\n",
       "      <th>chrom_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>prot_acc_version</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>wt</th>\n",
       "      <th>mut</th>\n",
       "      <th>wt_population</th>\n",
       "      <th>mut_poulation</th>\n",
       "      <th>wt_freq</th>\n",
       "      <th>mt_freq</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1474861721</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>33659968</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_112509.3</td>\n",
       "      <td>187</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>10680</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Zero-population</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1473838200</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>33660296</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NP_112509.3</td>\n",
       "      <td>163</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>10680</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Zero-population</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1473404810</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>33657999</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_112509.3</td>\n",
       "      <td>369</td>\n",
       "      <td>R</td>\n",
       "      <td>C</td>\n",
       "      <td>14050</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Zero-population</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1469318731</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>33669380</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_112509.3</td>\n",
       "      <td>128</td>\n",
       "      <td>K</td>\n",
       "      <td>E</td>\n",
       "      <td>35244</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999972</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1469008200</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>33658753</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_112509.3</td>\n",
       "      <td>321</td>\n",
       "      <td>D</td>\n",
       "      <td>H</td>\n",
       "      <td>14050</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Zero-population</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2865831</th>\n",
       "      <td>rs201657622</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>144373466</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NP_001137504.2</td>\n",
       "      <td>40</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>4294</td>\n",
       "      <td>59</td>\n",
       "      <td>0.986446</td>\n",
       "      <td>0.013554</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2865832</th>\n",
       "      <td>rs200452456</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>23229611</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>NP_001376232.1</td>\n",
       "      <td>171</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>4262</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999531</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>Ultra-rare</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2865833</th>\n",
       "      <td>rs142112856</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>9325189</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_001229256.1</td>\n",
       "      <td>9</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>11862</td>\n",
       "      <td>3689</td>\n",
       "      <td>0.762781</td>\n",
       "      <td>0.237219</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2865834</th>\n",
       "      <td>rs28613881</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>9363153</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_001243796.1</td>\n",
       "      <td>9</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>11862</td>\n",
       "      <td>6624</td>\n",
       "      <td>0.641675</td>\n",
       "      <td>0.358325</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2865835</th>\n",
       "      <td>rs11798010</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>49553056</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>50</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>108</td>\n",
       "      <td>40</td>\n",
       "      <td>0.729730</td>\n",
       "      <td>0.270270</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2865836 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               snp_id chrom_acc_version  chrom_pos ref_allele alt_allele  \\\n",
       "0        rs1474861721      NC_000020.11   33659968          A          G   \n",
       "1        rs1473838200      NC_000020.11   33660296          C          T   \n",
       "2        rs1473404810      NC_000020.11   33657999          G          A   \n",
       "3        rs1469318731      NC_000020.11   33669380          T          C   \n",
       "4        rs1469008200      NC_000020.11   33658753          C          G   \n",
       "...               ...               ...        ...        ...        ...   \n",
       "2865831   rs201657622      NC_000001.11  144373466          C          T   \n",
       "2865832   rs200452456      NC_000024.10   23229611          A          T   \n",
       "2865833   rs142112856      NC_000004.12    9325189          A          G   \n",
       "2865834    rs28613881      NC_000004.12    9363153          A          G   \n",
       "2865835    rs11798010      NC_000023.11   49553056          T          C   \n",
       "\n",
       "        prot_acc_version  prot_pos wt mut  wt_population  mut_poulation  \\\n",
       "0            NP_112509.3       187  L   P          10680              0   \n",
       "1            NP_112509.3       163  A   T          10680              0   \n",
       "2            NP_112509.3       369  R   C          14050              0   \n",
       "3            NP_112509.3       128  K   E          35244              1   \n",
       "4            NP_112509.3       321  D   H          14050              0   \n",
       "...                  ...       ... ..  ..            ...            ...   \n",
       "2865831   NP_001137504.2        40  S   N           4294             59   \n",
       "2865832   NP_001376232.1       171  Y   F           4262              2   \n",
       "2865833   NP_001229256.1         9  R   G          11862           3689   \n",
       "2865834   NP_001243796.1         9  R   G          11862           6624   \n",
       "2865835   NP_001091888.1        50  C   R            108             40   \n",
       "\n",
       "          wt_freq   mt_freq            class  \n",
       "0        1.000000  0.000000  Zero-population  \n",
       "1        1.000000  0.000000  Zero-population  \n",
       "2        1.000000  0.000000  Zero-population  \n",
       "3        0.999972  0.000028        Singleton  \n",
       "4        1.000000  0.000000  Zero-population  \n",
       "...           ...       ...              ...  \n",
       "2865831  0.986446  0.013554           Common  \n",
       "2865832  0.999531  0.000469       Ultra-rare  \n",
       "2865833  0.762781  0.237219           Common  \n",
       "2865834  0.641675  0.358325           Common  \n",
       "2865835  0.729730  0.270270           Common  \n",
       "\n",
       "[2865836 rows x 14 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_population_freq_SNVs\n",
    "popu_freq_variants_df = get_population_freq_SNVs(home_dir=home_dir)\n",
    "popu_freq_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (2865836, 15)\n",
      "0 (2865836, 15)\n",
      "dbnsfp polyphen2_HVAR (2865836, 15)\n",
      "1 (2865836, 16)\n",
      "dbnsfp cadd (2865836, 15)\n",
      "2 (2865836, 17)\n",
      "dbnsfp mvp (2865836, 15)\n",
      "3 (2865836, 18)\n",
      "dbnsfp metarnn (2865836, 15)\n",
      "4 (2865836, 19)\n",
      "dbnsfp revel (2865836, 15)\n",
      "5 (2865836, 20)\n",
      "tape_rao unirep (2865836, 15)\n",
      "6 (2865836, 21)\n",
      "tape_rao protbert (2865836, 15)\n",
      "7 (2865836, 22)\n",
      "sequnet_dunham sequnet (2862363, 15)\n",
      "8 (2865836, 23)\n",
      "esm_rives esm1b_t33_650M_UR50S (2865836, 15)\n",
      "9 (2865836, 24)\n",
      "esm_rives esm1v_t33_650M_UR90S (2865836, 15)\n",
      "10 (2865836, 25)\n",
      "esm_rives esm2_t33_650M_UR50D (2865836, 15)\n",
      "11 (2865836, 26)\n",
      "bioembeddings_dallago plus_rnn (2865836, 15)\n",
      "12 (2865836, 27)\n",
      "bioembeddings_dallago prottrans_bert_bfd (2865836, 15)\n",
      "13 (2865836, 28)\n",
      "bioembeddings_dallago prottrans_albert_bfd (2865836, 15)\n",
      "14 (2865836, 29)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (2865836, 15)\n",
      "15 (2865836, 30)\n",
      "bioembeddings_dallago prottrans_t5_bfd (2865836, 15)\n",
      "16 (2865836, 31)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (2865836, 15)\n",
      "17 (2865836, 32)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (2865836, 15)\n",
      "18 (2865836, 33)\n",
      "vespa_marquet vespal (2433579, 15)\n",
      "19 (2865836, 34)\n",
      "conservation_score\n",
      "(2865836, 35)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
       "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
       "       'mut_poulation', 'wt_freq', 'mt_freq', 'class', 'sift_pred',\n",
       "       'polyphen2_HVAR_pred', 'cadd_pred', 'mvp_pred', 'metarnn_pred',\n",
       "       'revel_pred', 'unirep_pred', 'protbert_pred', 'sequnet_pred',\n",
       "       'esm1b_t33_650M_UR50S_pred', 'esm1v_t33_650M_UR90S_pred',\n",
       "       'esm2_t33_650M_UR50D_pred', 'plus_rnn_pred', 'prottrans_bert_bfd_pred',\n",
       "       'prottrans_albert_bfd_pred', 'prottrans_xlnet_uniref100_pred',\n",
       "       'prottrans_t5_bfd_pred', 'prottrans_t5_uniref50_pred',\n",
       "       'prottrans_t5_xl_u50_pred', 'vespal_pred', 'conservation_pred'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "task = \"popu_freq\"\n",
    "merged_df = popu_freq_variants_df.copy(deep=True)\n",
    "merge_col_list = list(popu_freq_variants_df.columns)\n",
    "# merge_col_list.remove(\"class\")\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}.csv\", sep=\"\\t\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    merged_df = merged_df.drop_duplicates(keep=\"first\")\n",
    "    print(i, merged_df.shape)\n",
    "    # if i==3: break\n",
    "\n",
    "merged_df = merge_conservation_df(merged_df)  \n",
    "print(\"conservation_score\")  \n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}_analysis.csv\", sep=\"\\t\", index=False, header=True)\n",
    "print(merged_df.shape)\n",
    "merged_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common: 18167, rare: 28622, ultra-rare: 314013, singletons: 462444, zero-population: 2042590 and total: (2865836, 35)\n"
     ]
    }
   ],
   "source": [
    "n_commnon = merged_df[merged_df[\"class\"]==\"Common\"].shape[0]\n",
    "n_rare = merged_df[merged_df[\"class\"]==\"Rare\"].shape[0]\n",
    "n_ultra_rare = merged_df[merged_df[\"class\"]==\"Ultra-rare\"].shape[0]\n",
    "n_singleton = merged_df[merged_df[\"class\"]==\"Singleton\"].shape[0]\n",
    "n_zero_population = merged_df[merged_df[\"class\"]==\"Zero-population\"].shape[0]\n",
    "print(f\"Common: {n_commnon}, rare: {n_rare}, ultra-rare: {n_ultra_rare}, singletons: {n_singleton}, zero-population: {n_zero_population} and total: {merged_df.shape}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging protein mutation dataset (PMD) predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading Protein Mutation Dataset (PMD) ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/ashehu/akabir4/projects/variant_effect_analysis/models/aa_common/data_loader.py:49: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pmd_df = pd.read_csv(home_dir+\"models/aa_common/datasets_pmd_analysis/pmd_data.tsv\", sep=\"\\t\") # PMD: protein mutation dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(66281, 32)\n",
      "Index(['mut_id', 'pmd_id', 'nr', 'crossref', 'uniprot_id', 'ensembl_id',\n",
      "       'taxid', 'protein', 'mut_PMD', 'mut_real', 'wt', 'mut', 'prot_pos',\n",
      "       'function_summarized', 'functional_effect', 'function', 'seq', 'snp_id',\n",
      "       'mrna_acc', 'mrna_ver', 'mrna_pos', 'allele', 'protein_acc',\n",
      "       'protein_ver', 'verified', 'chrom', 'chrom_pos', 'variation',\n",
      "       'variant_type', 'ref_allele', 'alt_allele', 'pmd_nr_id'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 10750\n",
      "(60656, 32)\n"
     ]
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_pmd_dataset\n",
    "pmd_variants_df = get_pmd_dataset(home_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (60656, 33)\n",
      "(60656, 33)\n",
      "dbnsfp polyphen2_HVAR (60656, 33)\n",
      "(60656, 34)\n",
      "dbnsfp cadd (60656, 33)\n",
      "(60656, 35)\n",
      "dbnsfp mvp (60656, 33)\n",
      "(60656, 36)\n",
      "dbnsfp metarnn (60656, 33)\n",
      "(60656, 37)\n",
      "dbnsfp revel (60656, 33)\n",
      "(60656, 38)\n",
      "tape_rao unirep (60656, 33)\n",
      "(60656, 39)\n",
      "tape_rao protbert (60656, 33)\n",
      "(60656, 40)\n",
      "sequnet_dunham sequnet (60347, 33)\n",
      "(60656, 41)\n",
      "esm_rives esm1b_t33_650M_UR50S (60656, 33)\n",
      "(60656, 42)\n",
      "esm_rives esm1v_t33_650M_UR90S (60656, 33)\n",
      "(60656, 43)\n",
      "esm_rives esm2_t33_650M_UR50D (60656, 33)\n",
      "(60656, 44)\n",
      "bioembeddings_dallago plus_rnn (60427, 33)\n",
      "(60656, 45)\n",
      "bioembeddings_dallago prottrans_bert_bfd (60427, 33)\n",
      "(60656, 46)\n",
      "bioembeddings_dallago prottrans_albert_bfd (60427, 33)\n",
      "(60656, 47)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (60427, 33)\n",
      "(60656, 48)\n",
      "bioembeddings_dallago prottrans_t5_bfd (60427, 33)\n",
      "(60656, 49)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (60427, 33)\n",
      "(60656, 50)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (60427, 33)\n",
      "(60656, 51)\n",
      "vespa_marquet vespa (60656, 33)\n",
      "(60656, 52)\n"
     ]
    }
   ],
   "source": [
    "task = \"pmd\"\n",
    "merged_df = pmd_variants_df.copy(deep=True)\n",
    "merge_col_list = list(pmd_variants_df.columns)\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    if model_name==\"vespal\": model_name=\"vespa\"\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}.tsv\", sep=\"\\t\")\n",
    "    models_pred_df = models_pred_df.drop_duplicates(keep=\"first\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    \n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    print(merged_df.shape)\n",
    "    # if i==5: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3097823/3390127692.py:1: DtypeWarning: Columns (11,12,13,16,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  geno2func = pd.read_csv(home_dir+f\"data/SNPdbe/geno2func.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1748930, 23)\n",
      "Index(['mut_id', 'md5', 'wt', 'pos', 'mt', 'in_dbSNP', 'in_SP', 'in_PMD',\n",
      "       'in_1KG', 'SNAP_status', 'SNAP_bin', 'SNAP_score', 'SNAP_ri',\n",
      "       'SNAP_acc', 'SIFT_bin', 'SIFT_score', 'PERC_wt', 'PERC_mt', 'PSSM_wt',\n",
      "       'PSSM_mt', 'PSIC_wt', 'PSIC_mt', 'pph2'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "geno2func = pd.read_csv(home_dir+f\"data/SNPdbe/geno2func.tsv\", sep=\"\\t\")\n",
    "geno2func.replace(\"\\\\N\", None, inplace=True)\n",
    "print(geno2func.shape)\n",
    "print(geno2func.columns)\n",
    "geno2func = geno2func[['mut_id', 'SIFT_score']] # if needed, add more columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60656, 53)\n",
      "Index(['mut_id', 'pmd_id', 'nr', 'crossref', 'uniprot_id', 'ensembl_id',\n",
      "       'taxid', 'protein', 'mut_PMD', 'mut_real', 'wt', 'mut', 'prot_pos',\n",
      "       'function_summarized', 'functional_effect', 'function', 'seq', 'snp_id',\n",
      "       'mrna_acc', 'mrna_ver', 'mrna_pos', 'allele', 'protein_acc',\n",
      "       'protein_ver', 'verified', 'chrom', 'chrom_pos', 'variation',\n",
      "       'variant_type', 'ref_allele', 'alt_allele', 'pmd_nr_id', 'sift_pred',\n",
      "       'polyphen2_HVAR_pred', 'cadd_pred', 'mvp_pred', 'metarnn_pred',\n",
      "       'revel_pred', 'unirep_pred', 'protbert_pred', 'sequnet_pred',\n",
      "       'esm1b_t33_650M_UR50S_pred', 'esm1v_t33_650M_UR90S_pred',\n",
      "       'esm2_t33_650M_UR50D_pred', 'plus_rnn_pred', 'prottrans_bert_bfd_pred',\n",
      "       'prottrans_albert_bfd_pred', 'prottrans_xlnet_uniref100_pred',\n",
      "       'prottrans_t5_bfd_pred', 'prottrans_t5_uniref50_pred',\n",
      "       'prottrans_t5_xl_u50_pred', 'vespa_pred', 'sift_snpdbe_pred'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "merged_df = pd.merge(left=merged_df, right=geno2func, how=\"left\", on=[\"mut_id\"])\n",
    "merged_df = merged_df.rename(columns={\"SIFT_score\": f\"sift_snpdbe_pred\"})\n",
    "print(merged_df.shape)\n",
    "print(merged_df.columns)\n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}.tsv\", sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
