{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "home_dir = \"../../\"\n",
    "module_path = os.path.abspath(os.path.join(home_dir))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "model_root_and_name_tuple_list = [(\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"dbnsfp\", \"cadd\"), (\"dbnsfp\", \"mvp\"), (\"dbnsfp\", \"metarnn\"), (\"dbnsfp\", \"revel\"),\n",
    "                                  (\"tape_rao\", \"unirep\"), (\"tape_rao\", \"protbert\"), (\"sequnet_dunham\", \"sequnet\"), \n",
    "                                  (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"esm_rives\", \"esm1v_t33_650M_UR90S\"), (\"esm_rives\", \"esm2_t33_650M_UR50D\"),\n",
    "                                  (\"bioembeddings_dallago\", \"plus_rnn\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_bert_bfd\"), (\"bioembeddings_dallago\", \"prottrans_albert_bfd\"),\n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_xlnet_uniref100\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_t5_bfd\"), (\"bioembeddings_dallago\", \"prottrans_t5_uniref50\"), (\"bioembeddings_dallago\", \"prottrans_t5_xl_u50\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CDD Conservation dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4104012, 36)\n",
      "Index(['qNo', 'NPid', 'accession', 'que', 'sub', 'ali', 'qPos', 'sPos', 'aPos',\n",
      "       'qPos_', 'sPos_', 'bitscore', 'A', 'D', 'C', 'E', 'F', 'G', 'H', 'I',\n",
      "       'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', 'cons',\n",
      "       'entropy', 'qcov', 'CScore'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qNo</th>\n",
       "      <th>NPid</th>\n",
       "      <th>accession</th>\n",
       "      <th>que</th>\n",
       "      <th>sub</th>\n",
       "      <th>ali</th>\n",
       "      <th>qPos</th>\n",
       "      <th>sPos</th>\n",
       "      <th>aPos</th>\n",
       "      <th>qPos_</th>\n",
       "      <th>...</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>cons</th>\n",
       "      <th>entropy</th>\n",
       "      <th>qcov</th>\n",
       "      <th>CScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>pfam18400</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>|</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.303030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X</td>\n",
       "      <td>0.492204</td>\n",
       "      <td>0.715113</td>\n",
       "      <td>0.507796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>pfam18400</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>|</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X</td>\n",
       "      <td>0.288488</td>\n",
       "      <td>0.715113</td>\n",
       "      <td>0.711512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>pfam18400</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>|</td>\n",
       "      <td>59</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>59</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.101010</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P</td>\n",
       "      <td>0.125695</td>\n",
       "      <td>0.715113</td>\n",
       "      <td>0.874305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>pfam18400</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>|</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>X</td>\n",
       "      <td>0.479242</td>\n",
       "      <td>0.715113</td>\n",
       "      <td>0.520758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>NP_064505.1</td>\n",
       "      <td>pfam18400</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>|</td>\n",
       "      <td>61</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010101</td>\n",
       "      <td>0.010101</td>\n",
       "      <td>0.050505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020202</td>\n",
       "      <td>L</td>\n",
       "      <td>0.352476</td>\n",
       "      <td>0.715113</td>\n",
       "      <td>0.647524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104007</th>\n",
       "      <td>961</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>pfam05831</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>|</td>\n",
       "      <td>107</td>\n",
       "      <td>103</td>\n",
       "      <td>108</td>\n",
       "      <td>107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X</td>\n",
       "      <td>0.554522</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.445478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104008</th>\n",
       "      <td>961</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>pfam05831</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>|</td>\n",
       "      <td>108</td>\n",
       "      <td>104</td>\n",
       "      <td>109</td>\n",
       "      <td>108</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P</td>\n",
       "      <td>0.308799</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.691201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104009</th>\n",
       "      <td>961</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>pfam05831</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td></td>\n",
       "      <td>109</td>\n",
       "      <td>105</td>\n",
       "      <td>110</td>\n",
       "      <td>109</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>E</td>\n",
       "      <td>0.278218</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.721782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104010</th>\n",
       "      <td>961</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>pfam05831</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>|</td>\n",
       "      <td>110</td>\n",
       "      <td>106</td>\n",
       "      <td>111</td>\n",
       "      <td>110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X</td>\n",
       "      <td>0.315743</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.684257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104011</th>\n",
       "      <td>961</td>\n",
       "      <td>NP_001091888.1</td>\n",
       "      <td>pfam05831</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>\\n</td>\n",
       "      <td>111</td>\n",
       "      <td>107</td>\n",
       "      <td>112</td>\n",
       "      <td>111</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>G</td>\n",
       "      <td>0.056891</td>\n",
       "      <td>0.905983</td>\n",
       "      <td>0.943109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4104012 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         qNo            NPid  accession que sub ali  qPos  sPos  aPos  qPos_  \\\n",
       "0          1     NP_064505.1  pfam18400   S   A   |    57     1     1     57   \n",
       "1          1     NP_064505.1  pfam18400   T   T   |    58     2     2     58   \n",
       "2          1     NP_064505.1  pfam18400   P   P   |    59     3     3     59   \n",
       "3          1     NP_064505.1  pfam18400   L   L   |    60     4     4     60   \n",
       "4          1     NP_064505.1  pfam18400   L   L   |    61     5     5     61   \n",
       "...      ...             ...        ...  ..  ..  ..   ...   ...   ...    ...   \n",
       "4104007  961  NP_001091888.1  pfam05831   T   M   |   107   103   108    107   \n",
       "4104008  961  NP_001091888.1  pfam05831   P   P   |   108   104   109    108   \n",
       "4104009  961  NP_001091888.1  pfam05831   E   E       109   105   110    109   \n",
       "4104010  961  NP_001091888.1  pfam05831   E   A   |   110   106   111    110   \n",
       "4104011  961  NP_001091888.1  pfam05831   G   G  \\n   111   107   112    111   \n",
       "\n",
       "         ...         R         S         T         V    W         Y  cons  \\\n",
       "0        ...  0.000000  0.303030  0.000000  0.000000  0.0  0.000000     X   \n",
       "1        ...  0.000000  0.030303  0.555556  0.000000  0.0  0.000000     X   \n",
       "2        ...  0.000000  0.101010  0.000000  0.000000  0.0  0.000000     P   \n",
       "3        ...  0.000000  0.000000  0.000000  0.050505  0.0  0.161616     X   \n",
       "4        ...  0.000000  0.010101  0.010101  0.050505  0.0  0.020202     L   \n",
       "...      ...       ...       ...       ...       ...  ...       ...   ...   \n",
       "4104007  ...  0.041667  0.000000  0.041667  0.125000  0.0  0.000000     X   \n",
       "4104008  ...  0.000000  0.041667  0.000000  0.000000  0.0  0.000000     P   \n",
       "4104009  ...  0.041667  0.000000  0.000000  0.000000  0.0  0.000000     E   \n",
       "4104010  ...  0.000000  0.000000  0.041667  0.000000  0.0  0.000000     X   \n",
       "4104011  ...  0.000000  0.000000  0.000000  0.000000  0.0  0.000000     G   \n",
       "\n",
       "          entropy      qcov    CScore  \n",
       "0        0.492204  0.715113  0.507796  \n",
       "1        0.288488  0.715113  0.711512  \n",
       "2        0.125695  0.715113  0.874305  \n",
       "3        0.479242  0.715113  0.520758  \n",
       "4        0.352476  0.715113  0.647524  \n",
       "...           ...       ...       ...  \n",
       "4104007  0.554522  0.905983  0.445478  \n",
       "4104008  0.308799  0.905983  0.691201  \n",
       "4104009  0.278218  0.905983  0.721782  \n",
       "4104010  0.315743  0.905983  0.684257  \n",
       "4104011  0.056891  0.905983  0.943109  \n",
       "\n",
       "[4104012 rows x 36 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conservation_df = pd.read_csv(home_dir+\"data/cdd_conservationTable_18kHumanProts.csv.gzip\", compression='gzip', comment='#')\n",
    "print(conservation_df.shape)\n",
    "print(conservation_df.columns)\n",
    "conservation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_conservation_df(df:pd.DataFrame):\n",
    "    # merge conservation_df with the input df\n",
    "    temp_conservation_df = conservation_df[[\"NPid\", \"qPos\", \"CScore\"]].drop_duplicates(keep=\"first\")\n",
    "    merged_df = pd.merge(df, temp_conservation_df, how=\"left\", left_on=[\"prot_acc_version\", \"prot_pos\"], right_on=[\"NPid\", \"qPos\"])\n",
    "    columns = list(df.columns)\n",
    "    columns.append(\"CScore\")\n",
    "    merged_df = merged_df[columns]\n",
    "    merged_df = merged_df.rename(columns={\"CScore\": \"conservation_pred\"})\n",
    "    return merged_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging pathogenicity analysis predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "likely_pathogenic raw data: (6668, 12)\n",
      "Index(['clinvar_id', 'gene_symbol', 'gene_id', 'chrom_acc_version',\n",
      "       'chrom_pos', 'ref_allele', 'alt_allele', 'prot_acc_version', 'prot_pos',\n",
      "       'wt', 'mut', 'class'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 1077\n",
      "(4168, 12)\n",
      "\n",
      "Log: Loading population freq variants dataset ...\n",
      "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
      "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
      "       'mut_poulation', 'wt_freq', 'mt_freq'],\n",
      "      dtype='object')\n",
      "0 (8336, 10)\n",
      "1 (8336, 10)\n",
      "2 (8336, 10)\n",
      "3 (8336, 10)\n",
      "4 (8336, 10)\n",
      "5 (8336, 10)\n",
      "6 (8336, 10)\n",
      "7 (8336, 10)\n",
      "8 (8336, 10)\n",
      "9 (8336, 10)\n"
     ]
    }
   ],
   "source": [
    "pathogenicity_type = \"likely_pathogenic\" # pathogenic, likely_pathogenic\n",
    "\n",
    "from models.aa_common.data_loader import get_pathogenicity_analysis_SNVs\n",
    "list_of_variants_df = get_pathogenicity_analysis_SNVs(home_dir=home_dir, pathogenicity_type=pathogenicity_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8327, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "1\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8325, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "2\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8328, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "3\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8328, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "4\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8329, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "5\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8329, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "6\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8330, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "7\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8326, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "8\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8329, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n",
      "9\n",
      "dbnsfp sift (8336, 11)\n",
      "(8336, 11)\n",
      "dbnsfp polyphen2_HVAR (8336, 11)\n",
      "(8336, 12)\n",
      "dbnsfp cadd (8336, 11)\n",
      "(8336, 13)\n",
      "dbnsfp mvp (8336, 11)\n",
      "(8336, 14)\n",
      "dbnsfp metarnn (8336, 11)\n",
      "(8336, 15)\n",
      "dbnsfp revel (8336, 11)\n",
      "(8336, 16)\n",
      "tape_rao unirep (8336, 11)\n",
      "(8336, 17)\n",
      "tape_rao protbert (8336, 11)\n",
      "(8336, 18)\n",
      "sequnet_dunham sequnet (8325, 11)\n",
      "(8336, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (8336, 11)\n",
      "(8336, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (8336, 11)\n",
      "(8336, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (8336, 11)\n",
      "(8336, 22)\n",
      "bioembeddings_dallago plus_rnn (8336, 11)\n",
      "(8336, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (8336, 11)\n",
      "(8336, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (8336, 11)\n",
      "(8336, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (8336, 11)\n",
      "(8336, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (8336, 11)\n",
      "(8336, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (8336, 11)\n",
      "(8336, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (8336, 11)\n",
      "(8336, 29)\n",
      "conservation_score\n",
      "(8336, 30)\n"
     ]
    }
   ],
   "source": [
    "for analysis_no, patho_variants_df in enumerate(list_of_variants_df):\n",
    "    print(analysis_no)\n",
    "    merged_df = patho_variants_df.copy(deep=True)\n",
    "    merge_col_list = list(patho_variants_df.columns)\n",
    "    \n",
    "    for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):    \n",
    "        models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{pathogenicity_type}/{str(analysis_no)}.csv\", sep=\"\\t\")\n",
    "        print(model_root, model_name, models_pred_df.shape)\n",
    "        merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "        merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "        print(merged_df.shape)\n",
    "        # if i==5: break\n",
    "    \n",
    "    # merged_df = merge_conservation_df(merged_df)  \n",
    "    # print(\"conservation_score\")  \n",
    "    print(merged_df.shape)\n",
    "    merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{pathogenicity_type}_analysis/{str(analysis_no)}.csv\", sep=\"\\t\", index=False, header=True)\n",
    "    # break"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging population frequency predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading data ...\n",
      "(407591, 13)\n",
      "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
      "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
      "       'mut_poulation', 'wt_freq', 'mt_freq'],\n",
      "      dtype='object')\n",
      "NP_001840.3       139\n",
      "NP_001277137.1    138\n",
      "NP_001291317.1    136\n",
      "NP_071375.1       135\n",
      "NP_005996.2       126\n",
      "                 ... \n",
      "NP_219479.2         1\n",
      "NP_001259000.1      1\n",
      "NP_006348.1         1\n",
      "NP_002361.1         1\n",
      "NP_057387.1         1\n",
      "Name: prot_acc_version, Length: 15871, dtype: int64\n",
      "Common: (18167), rare: (28622), ultra-rare: (314013), sampled-singletons: (46789) and total: 407591\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>chrom_acc_version</th>\n",
       "      <th>chrom_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>prot_acc_version</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>wt</th>\n",
       "      <th>mut</th>\n",
       "      <th>wt_population</th>\n",
       "      <th>mut_poulation</th>\n",
       "      <th>wt_freq</th>\n",
       "      <th>mt_freq</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs41288789</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>26944749</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_064519.2</td>\n",
       "      <td>512</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>218102</td>\n",
       "      <td>10302</td>\n",
       "      <td>0.954896</td>\n",
       "      <td>0.045104</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs34603401</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>9245386</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_004276.2</td>\n",
       "      <td>151</td>\n",
       "      <td>D</td>\n",
       "      <td>A</td>\n",
       "      <td>225918</td>\n",
       "      <td>32086</td>\n",
       "      <td>0.875638</td>\n",
       "      <td>0.124362</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115026899</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>210683774</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_758872.1</td>\n",
       "      <td>826</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>211392</td>\n",
       "      <td>6706</td>\n",
       "      <td>0.969252</td>\n",
       "      <td>0.030748</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs17260829</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>122451974</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_065806.1</td>\n",
       "      <td>225</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>301772</td>\n",
       "      <td>13333</td>\n",
       "      <td>0.957687</td>\n",
       "      <td>0.042313</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1438318937</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>54575019</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_001124389.2</td>\n",
       "      <td>214</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>11862</td>\n",
       "      <td>5489</td>\n",
       "      <td>0.683649</td>\n",
       "      <td>0.316351</td>\n",
       "      <td>Common</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407586</th>\n",
       "      <td>rs1311034485</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>71489095</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_542937.3</td>\n",
       "      <td>542</td>\n",
       "      <td>L</td>\n",
       "      <td>V</td>\n",
       "      <td>35428</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999972</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407587</th>\n",
       "      <td>rs751186776</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>12896958</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_000150.1</td>\n",
       "      <td>301</td>\n",
       "      <td>V</td>\n",
       "      <td>M</td>\n",
       "      <td>14528</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407588</th>\n",
       "      <td>rs200406718</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>76387303</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_001136073.1</td>\n",
       "      <td>291</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>41778</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407589</th>\n",
       "      <td>rs753278261</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>80432091</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_115709.3</td>\n",
       "      <td>228</td>\n",
       "      <td>K</td>\n",
       "      <td>E</td>\n",
       "      <td>14050</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407590</th>\n",
       "      <td>rs148950917</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>111592930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NP_001139580.1</td>\n",
       "      <td>69</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>8988</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>Singleton</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>407591 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              snp_id chrom_acc_version  chrom_pos ref_allele alt_allele  \\\n",
       "0         rs41288789      NC_000002.12   26944749          G          A   \n",
       "1         rs34603401      NC_000001.11    9245386          A          C   \n",
       "2        rs115026899      NC_000001.11  210683774          C          A   \n",
       "3         rs17260829      NC_000006.12  122451974          T          C   \n",
       "4       rs1438318937      NC_000019.10   54575019          A          G   \n",
       "...              ...               ...        ...        ...        ...   \n",
       "407586  rs1311034485      NC_000013.11   71489095          G          C   \n",
       "407587   rs751186776      NC_000019.10   12896958          G          A   \n",
       "407588   rs200406718      NC_000017.11   76387303          G          A   \n",
       "407589   rs753278261      NC_000010.11   80432091          A          G   \n",
       "407590   rs148950917      NC_000009.12  111592930          C          T   \n",
       "\n",
       "       prot_acc_version  prot_pos wt mut  wt_population  mut_poulation  \\\n",
       "0           NP_064519.2       512  A   T         218102          10302   \n",
       "1           NP_004276.2       151  D   A         225918          32086   \n",
       "2           NP_758872.1       826  G   V         211392           6706   \n",
       "3           NP_065806.1       225  S   G         301772          13333   \n",
       "4        NP_001124389.2       214  E   G          11862           5489   \n",
       "...                 ...       ... ..  ..            ...            ...   \n",
       "407586      NP_542937.3       542  L   V          35428              1   \n",
       "407587      NP_000150.1       301  V   M          14528              1   \n",
       "407588   NP_001136073.1       291  R   Q          41778              1   \n",
       "407589      NP_115709.3       228  K   E          14050              1   \n",
       "407590   NP_001139580.1        69  A   T           8988              1   \n",
       "\n",
       "         wt_freq   mt_freq      class  \n",
       "0       0.954896  0.045104     Common  \n",
       "1       0.875638  0.124362     Common  \n",
       "2       0.969252  0.030748     Common  \n",
       "3       0.957687  0.042313     Common  \n",
       "4       0.683649  0.316351     Common  \n",
       "...          ...       ...        ...  \n",
       "407586  0.999972  0.000028  Singleton  \n",
       "407587  0.999931  0.000069  Singleton  \n",
       "407588  0.999976  0.000024  Singleton  \n",
       "407589  0.999929  0.000071  Singleton  \n",
       "407590  0.999889  0.000111  Singleton  \n",
       "\n",
       "[407591 rows x 14 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_population_freq_SNVs\n",
    "popu_freq_variants_df = get_population_freq_SNVs(home_dir=home_dir)\n",
    "popu_freq_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (407591, 14)\n",
      "(407591, 15)\n",
      "dbnsfp polyphen2_HVAR (407591, 14)\n",
      "(407591, 16)\n",
      "dbnsfp cadd (407591, 14)\n",
      "(407591, 17)\n",
      "dbnsfp mvp (407591, 14)\n",
      "(407591, 18)\n",
      "dbnsfp metarnn (407591, 14)\n",
      "(407591, 19)\n",
      "dbnsfp revel (407591, 14)\n",
      "(407591, 20)\n",
      "tape_rao unirep (407591, 14)\n",
      "(407591, 21)\n",
      "tape_rao protbert (407591, 14)\n",
      "(407591, 22)\n",
      "sequnet_dunham sequnet (407098, 14)\n",
      "(407591, 23)\n",
      "esm_rives esm1b_t33_650M_UR50S (407591, 14)\n",
      "(407591, 24)\n",
      "esm_rives esm1v_t33_650M_UR90S (407591, 14)\n",
      "(407591, 25)\n",
      "esm_rives esm2_t33_650M_UR50D (407591, 14)\n",
      "(407591, 26)\n",
      "bioembeddings_dallago plus_rnn (407591, 14)\n",
      "(407591, 27)\n",
      "bioembeddings_dallago prottrans_bert_bfd (407591, 14)\n",
      "(407591, 28)\n",
      "bioembeddings_dallago prottrans_albert_bfd (407591, 14)\n",
      "(407591, 29)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (407591, 14)\n",
      "(407591, 30)\n",
      "bioembeddings_dallago prottrans_t5_bfd (407591, 14)\n",
      "(407591, 31)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (407591, 14)\n",
      "(407591, 32)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (407591, 14)\n",
      "(407591, 33)\n",
      "(407591, 33)\n"
     ]
    }
   ],
   "source": [
    "task = \"popu_freq\"\n",
    "merged_df = popu_freq_variants_df.copy(deep=True)\n",
    "merge_col_list = list(popu_freq_variants_df.columns)\n",
    "merge_col_list.remove(\"class\")\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}.csv\", sep=\"\\t\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    merged_df = merged_df.drop_duplicates(keep=\"first\")\n",
    "    print(merged_df.shape)\n",
    "    # if i==3: break\n",
    "\n",
    "# merged_df = merge_conservation_df(merged_df)  \n",
    "# print(\"conservation_score\")  \n",
    "print(merged_df.shape)\n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}_analysis_1.csv\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common: (18167), rare: (28622), ultra-rare: (314013) sampled-singletons: (46789) and total: 407591\n"
     ]
    }
   ],
   "source": [
    "n_commnon = merged_df[merged_df[\"class\"]==\"Common\"].shape[0]\n",
    "n_rare = merged_df[merged_df[\"class\"]==\"Rare\"].shape[0]\n",
    "n_ultra_rare = merged_df[merged_df[\"class\"]==\"Ultra-rare\"].shape[0]\n",
    "n_singleton = merged_df[merged_df[\"class\"]==\"Singleton\"].shape[0]\n",
    "print(f\"Common: ({n_commnon}), rare: ({n_rare}), ultra-rare: ({n_ultra_rare}) sampled-singletons: ({n_singleton}) and total: {merged_df.shape[0]}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging protein mutation dataset (PMD) predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading Protein Mutation Dataset (PMD) ...\n",
      "(55465, 15)\n",
      "Index(['mut_id', 'pmd_id', 'protein', 'mut_PMD', 'mut_real', 'function',\n",
      "       'taxid', 'function_e', 'function_e2', 'functional_effect', 'seq',\n",
      "       'protein_id', 'wt', 'mut', 'prot_pos'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 7677\n",
      "(51047, 15)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Knock-out    25116\n",
       "No effect    13297\n",
       "Effect       12634\n",
       "Name: functional_effect, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_pmd_analysis_dataset\n",
    "pmd_variants_df = get_pmd_analysis_dataset(home_dir=home_dir)\n",
    "pmd_variants_df[\"functional_effect\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tape_rao unirep (51047, 16)\n",
      "(51047, 16)\n",
      "tape_rao protbert (51047, 16)\n",
      "(51047, 17)\n",
      "sequnet_dunham sequnet (50994, 16)\n",
      "(51047, 18)\n",
      "esm_rives esm1b_t33_650M_UR50S (51047, 16)\n",
      "(51047, 19)\n",
      "esm_rives esm1v_t33_650M_UR90S (51047, 16)\n",
      "(51047, 20)\n",
      "esm_rives esm2_t33_650M_UR50D (51047, 16)\n",
      "(51047, 21)\n",
      "bioembeddings_dallago plus_rnn (51047, 16)\n",
      "(51047, 22)\n",
      "bioembeddings_dallago prottrans_bert_bfd (51047, 16)\n",
      "(51047, 23)\n",
      "bioembeddings_dallago prottrans_albert_bfd (51047, 16)\n",
      "(51047, 24)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (51047, 16)\n",
      "(51047, 25)\n",
      "bioembeddings_dallago prottrans_t5_bfd (51047, 16)\n",
      "(51047, 26)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (51047, 16)\n",
      "(51047, 27)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (51047, 16)\n",
      "(51047, 28)\n"
     ]
    }
   ],
   "source": [
    "task = \"pmd_analysis\"\n",
    "merged_df = pmd_variants_df.copy(deep=True)\n",
    "merge_col_list = list(pmd_variants_df.columns)\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}_(mt-wt).csv\", sep=\",\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    print(merged_df.shape)\n",
    "    # if i==5: break\n",
    "    \n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}.csv\", sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
