{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "home_dir = \"../../\"\n",
    "module_path = os.path.abspath(os.path.join(home_dir))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "model_root_and_name_tuple_list = [(\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"dbnsfp\", \"cadd\"), (\"dbnsfp\", \"mvp\"), (\"dbnsfp\", \"metarnn\"), (\"dbnsfp\", \"revel\"),\n",
    "                                  (\"tape_rao\", \"unirep\"), (\"tape_rao\", \"protbert\"), (\"sequnet_dunham\", \"sequnet\"), \n",
    "                                  (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"esm_rives\", \"esm1v_t33_650M_UR90S\"), (\"esm_rives\", \"esm2_t33_650M_UR50D\"),\n",
    "                                  (\"bioembeddings_dallago\", \"plus_rnn\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_bert_bfd\"), (\"bioembeddings_dallago\", \"prottrans_albert_bfd\"),\n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_xlnet_uniref100\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_t5_bfd\"), (\"bioembeddings_dallago\", \"prottrans_t5_uniref50\"), (\"bioembeddings_dallago\", \"prottrans_t5_xl_u50\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging pathogenicity analysis predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "likely_pathogenic raw data: (6668, 12)\n",
      "Index(['clinvar_id', 'gene_symbol', 'gene_id', 'chrom_acc_version',\n",
      "       'chrom_pos', 'ref_allele', 'alt_allele', 'prot_acc_version', 'prot_pos',\n",
      "       'wt', 'mut', 'class'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 1077\n",
      "(4168, 12)\n",
      "\n",
      "Log: Loading population freq variants dataset ...\n",
      "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
      "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
      "       'mut_poulation', 'wt_freq', 'mt_freq'],\n",
      "      dtype='object')\n",
      "0 (8336, 10)\n",
      "1 (8336, 10)\n",
      "2 (8336, 10)\n",
      "3 (8336, 10)\n",
      "4 (8336, 10)\n",
      "5 (8336, 10)\n",
      "6 (8336, 10)\n",
      "7 (8336, 10)\n",
      "8 (8336, 10)\n",
      "9 (8336, 10)\n"
     ]
    }
   ],
   "source": [
    "pathogenicity_type = \"likely_pathogenic\" # pathogenic, likely_pathogenic\n",
    "\n",
    "from models.aa_common.data_loader import get_pathogenicity_analysis_SNVs\n",
    "list_of_variants_df = get_pathogenicity_analysis_SNVs(home_dir=home_dir, pathogenicity_type=pathogenicity_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4582, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "1\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4580, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "2\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4579, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "3\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4576, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "4\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4580, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "5\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4580, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "6\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4579, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "7\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4579, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "8\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4582, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n",
      "9\n",
      "dbnsfp sift (4586, 11)\n",
      "(4586, 11)\n",
      "dbnsfp polyphen2_HVAR (4586, 11)\n",
      "(4586, 12)\n",
      "dbnsfp cadd (4586, 11)\n",
      "(4586, 13)\n",
      "dbnsfp mvp (4586, 11)\n",
      "(4586, 14)\n",
      "dbnsfp metarnn (4586, 11)\n",
      "(4586, 15)\n",
      "dbnsfp revel (4586, 11)\n",
      "(4586, 16)\n",
      "tape_rao unirep (4586, 11)\n",
      "(4586, 17)\n",
      "tape_rao protbert (4586, 11)\n",
      "(4586, 18)\n",
      "sequnet_dunham sequnet (4580, 11)\n",
      "(4586, 19)\n",
      "esm_rives esm1b_t33_650M_UR50S (4586, 11)\n",
      "(4586, 20)\n",
      "esm_rives esm1v_t33_650M_UR90S (4586, 11)\n",
      "(4586, 21)\n",
      "esm_rives esm2_t33_650M_UR50D (4586, 11)\n",
      "(4586, 22)\n",
      "bioembeddings_dallago plus_rnn (4586, 11)\n",
      "(4586, 23)\n",
      "bioembeddings_dallago prottrans_bert_bfd (4586, 11)\n",
      "(4586, 24)\n",
      "bioembeddings_dallago prottrans_albert_bfd (4586, 11)\n",
      "(4586, 25)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (4586, 11)\n",
      "(4586, 26)\n",
      "bioembeddings_dallago prottrans_t5_bfd (4586, 11)\n",
      "(4586, 27)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (4586, 11)\n",
      "(4586, 28)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (4586, 11)\n",
      "(4586, 29)\n"
     ]
    }
   ],
   "source": [
    "for analysis_no, patho_variants_df in enumerate(list_of_variants_df):\n",
    "    print(analysis_no)\n",
    "    merged_df = patho_variants_df.copy(deep=True)\n",
    "    merge_col_list = list(patho_variants_df.columns)\n",
    "    \n",
    "    for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):    \n",
    "        models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{pathogenicity_type}/{str(analysis_no)}.csv\", sep=\"\\t\")\n",
    "        print(model_root, model_name, models_pred_df.shape)\n",
    "        merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "        merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "        print(merged_df.shape)\n",
    "        # if i==5: break\n",
    "        \n",
    "    merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{pathogenicity_type}_analysis/{str(analysis_no)}.csv\", sep=\"\\t\", index=False, header=True)\n",
    "    # break"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging population frequency predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading data ...\n",
      "(95223, 13)\n",
      "Index(['snp_id', 'chrom_acc_version', 'chrom_pos', 'ref_allele', 'alt_allele',\n",
      "       'prot_acc_version', 'prot_pos', 'wt', 'mut', 'wt_population',\n",
      "       'mut_poulation', 'wt_freq', 'mt_freq'],\n",
      "      dtype='object')\n",
      "After combining common (18279), rare (29383) and sampled-singletons (48434), data: (95223, 13)\n",
      "NP_001277137.1    95\n",
      "NP_001291317.1    78\n",
      "NP_002115.2       74\n",
      "NP_112241.2       68\n",
      "NP_001075106.2    65\n",
      "                  ..\n",
      "NP_037408.2        1\n",
      "NP_066951.1        1\n",
      "NP_001013716.2     1\n",
      "NP_001316544.1     1\n",
      "NP_065171.2        1\n",
      "Name: prot_acc_version, Length: 14997, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>chrom_acc_version</th>\n",
       "      <th>chrom_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>prot_acc_version</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>wt</th>\n",
       "      <th>mut</th>\n",
       "      <th>wt_population</th>\n",
       "      <th>mut_poulation</th>\n",
       "      <th>wt_freq</th>\n",
       "      <th>mt_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs41288789</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>26944749</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_064519.2</td>\n",
       "      <td>512</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>218102</td>\n",
       "      <td>10302</td>\n",
       "      <td>0.954896</td>\n",
       "      <td>0.045104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs34603401</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>9245386</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_004276.2</td>\n",
       "      <td>151</td>\n",
       "      <td>D</td>\n",
       "      <td>A</td>\n",
       "      <td>225918</td>\n",
       "      <td>32086</td>\n",
       "      <td>0.875638</td>\n",
       "      <td>0.124362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115026899</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>210683774</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_758872.1</td>\n",
       "      <td>826</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>211392</td>\n",
       "      <td>6706</td>\n",
       "      <td>0.969252</td>\n",
       "      <td>0.030748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs17260829</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>122451974</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_065806.1</td>\n",
       "      <td>225</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>301772</td>\n",
       "      <td>13333</td>\n",
       "      <td>0.957687</td>\n",
       "      <td>0.042313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1438318937</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>54575019</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NP_001124389.2</td>\n",
       "      <td>214</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>11862</td>\n",
       "      <td>5489</td>\n",
       "      <td>0.683649</td>\n",
       "      <td>0.316351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95218</th>\n",
       "      <td>rs1404095987</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>143755938</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_848656.2</td>\n",
       "      <td>574</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>11862</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999916</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95219</th>\n",
       "      <td>rs1331626452</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>57328412</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NP_998763.2</td>\n",
       "      <td>317</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4470</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999776</td>\n",
       "      <td>0.000224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95220</th>\n",
       "      <td>rs751861106</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>44921715</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_001359108.1</td>\n",
       "      <td>287</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>35412</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999972</td>\n",
       "      <td>0.000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95221</th>\n",
       "      <td>rs771902243</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>28503161</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NP_059108.1</td>\n",
       "      <td>50</td>\n",
       "      <td>L</td>\n",
       "      <td>M</td>\n",
       "      <td>35424</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999972</td>\n",
       "      <td>0.000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95222</th>\n",
       "      <td>rs1433551223</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>11486860</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NP_620138.2</td>\n",
       "      <td>455</td>\n",
       "      <td>S</td>\n",
       "      <td>L</td>\n",
       "      <td>11862</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999916</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95223 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             snp_id chrom_acc_version  chrom_pos ref_allele alt_allele  \\\n",
       "0        rs41288789      NC_000002.12   26944749          G          A   \n",
       "1        rs34603401      NC_000001.11    9245386          A          C   \n",
       "2       rs115026899      NC_000001.11  210683774          C          A   \n",
       "3        rs17260829      NC_000006.12  122451974          T          C   \n",
       "4      rs1438318937      NC_000019.10   54575019          A          G   \n",
       "...             ...               ...        ...        ...        ...   \n",
       "95218  rs1404095987      NC_000007.14  143755938          T          C   \n",
       "95219  rs1331626452      NC_000019.10   57328412          G          C   \n",
       "95220   rs751861106      NC_000020.11   44921715          G          A   \n",
       "95221   rs771902243      NC_000008.11   28503161          C          T   \n",
       "95222  rs1433551223      NC_000019.10   11486860          G          A   \n",
       "\n",
       "      prot_acc_version  prot_pos wt mut  wt_population  mut_poulation  \\\n",
       "0          NP_064519.2       512  A   T         218102          10302   \n",
       "1          NP_004276.2       151  D   A         225918          32086   \n",
       "2          NP_758872.1       826  G   V         211392           6706   \n",
       "3          NP_065806.1       225  S   G         301772          13333   \n",
       "4       NP_001124389.2       214  E   G          11862           5489   \n",
       "...                ...       ... ..  ..            ...            ...   \n",
       "95218      NP_848656.2       574  N   S          11862              1   \n",
       "95219      NP_998763.2       317  G   A           4470              1   \n",
       "95220   NP_001359108.1       287  R   Q          35412              1   \n",
       "95221      NP_059108.1        50  L   M          35424              1   \n",
       "95222      NP_620138.2       455  S   L          11862              1   \n",
       "\n",
       "        wt_freq   mt_freq  \n",
       "0      0.954896  0.045104  \n",
       "1      0.875638  0.124362  \n",
       "2      0.969252  0.030748  \n",
       "3      0.957687  0.042313  \n",
       "4      0.683649  0.316351  \n",
       "...         ...       ...  \n",
       "95218  0.999916  0.000084  \n",
       "95219  0.999776  0.000224  \n",
       "95220  0.999972  0.000028  \n",
       "95221  0.999972  0.000028  \n",
       "95222  0.999916  0.000084  \n",
       "\n",
       "[95223 rows x 13 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_population_freq_SNVs\n",
    "popu_freq_variants_df = get_population_freq_SNVs(home_dir=home_dir)\n",
    "popu_freq_variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (95223, 14)\n",
      "(95223, 15)\n",
      "dbnsfp polyphen2_HVAR (95223, 14)\n",
      "(95223, 16)\n",
      "dbnsfp cadd (95223, 14)\n",
      "(95223, 17)\n",
      "dbnsfp mvp (95223, 14)\n",
      "(95223, 18)\n",
      "dbnsfp metarnn (95223, 14)\n",
      "(95223, 19)\n",
      "dbnsfp revel (95223, 14)\n",
      "(95223, 20)\n",
      "tape_rao unirep (95223, 14)\n",
      "(95223, 21)\n",
      "tape_rao protbert (95223, 14)\n",
      "(95223, 22)\n",
      "sequnet_dunham sequnet (95110, 14)\n",
      "(95223, 23)\n",
      "esm_rives esm1b_t33_650M_UR50S (95223, 14)\n",
      "(95223, 24)\n",
      "esm_rives esm1v_t33_650M_UR90S (95223, 14)\n",
      "(95223, 25)\n",
      "esm_rives esm2_t33_650M_UR50D (95223, 14)\n",
      "(95223, 26)\n",
      "bioembeddings_dallago plus_rnn (95223, 14)\n",
      "(95223, 27)\n",
      "bioembeddings_dallago prottrans_bert_bfd (95223, 14)\n",
      "(95223, 28)\n",
      "bioembeddings_dallago prottrans_albert_bfd (95223, 14)\n",
      "(95223, 29)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (95223, 14)\n",
      "(95223, 30)\n",
      "bioembeddings_dallago prottrans_t5_bfd (95223, 14)\n",
      "(95223, 31)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (95223, 14)\n",
      "(95223, 32)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (95223, 14)\n",
      "(95223, 33)\n"
     ]
    }
   ],
   "source": [
    "task = \"popu_freq\"\n",
    "merged_df = popu_freq_variants_df.copy(deep=True)\n",
    "merge_col_list = list(popu_freq_variants_df.columns)\n",
    "\n",
    "merged_df.loc[merged_df[\"mt_freq\"]>=.01, \"class\"] = \"Common\"\n",
    "merged_df.loc[(merged_df[\"mt_freq\"]<.01) & (merged_df[\"mt_freq\"]>=.001), \"class\"] = \"Rare\"\n",
    "merged_df.loc[(merged_df[\"mt_freq\"]<.001), \"class\"] = \"Extremely rare\"\n",
    "\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}_(mt-wt).csv\", sep=\"\\t\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    merged_df = merged_df.drop_duplicates(keep=\"first\")\n",
    "    print(merged_df.shape)\n",
    "    # if i==5: break\n",
    "    \n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}_analysis.csv\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging protein mutation dataset (PMD) predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading Protein Mutation Dataset (PMD) ...\n",
      "(55465, 15)\n",
      "Index(['mut_id', 'pmd_id', 'protein', 'mut_PMD', 'mut_real', 'function',\n",
      "       'taxid', 'function_e', 'function_e2', 'functional_effect', 'seq',\n",
      "       'protein_id', 'wt', 'mut', 'prot_pos'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 7677\n",
      "(51047, 15)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Knock-out    25116\n",
       "No effect    13297\n",
       "Effect       12634\n",
       "Name: functional_effect, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_pmd_analysis_dataset\n",
    "pmd_variants_df = get_pmd_analysis_dataset(home_dir=home_dir)\n",
    "pmd_variants_df[\"functional_effect\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tape_rao unirep (51047, 16)\n",
      "(51047, 16)\n",
      "tape_rao protbert (51047, 16)\n",
      "(51047, 17)\n",
      "sequnet_dunham sequnet (50994, 16)\n",
      "(51047, 18)\n",
      "esm_rives esm1b_t33_650M_UR50S (51047, 16)\n",
      "(51047, 19)\n",
      "esm_rives esm1v_t33_650M_UR90S (51047, 16)\n",
      "(51047, 20)\n",
      "esm_rives esm2_t33_650M_UR50D (51047, 16)\n",
      "(51047, 21)\n",
      "bioembeddings_dallago plus_rnn (51047, 16)\n",
      "(51047, 22)\n",
      "bioembeddings_dallago prottrans_bert_bfd (51047, 16)\n",
      "(51047, 23)\n",
      "bioembeddings_dallago prottrans_albert_bfd (51047, 16)\n",
      "(51047, 24)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (51047, 16)\n",
      "(51047, 25)\n",
      "bioembeddings_dallago prottrans_t5_bfd (51047, 16)\n",
      "(51047, 26)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (51047, 16)\n",
      "(51047, 27)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (51047, 16)\n",
      "(51047, 28)\n"
     ]
    }
   ],
   "source": [
    "task = \"pmd_analysis\"\n",
    "merged_df = pmd_variants_df.copy(deep=True)\n",
    "merge_col_list = list(pmd_variants_df.columns)\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}_(mt-wt).csv\", sep=\",\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    print(merged_df.shape)\n",
    "    # if i==5: break\n",
    "    \n",
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}.csv\", sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
