{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "home_dir = \"../../\"\n",
    "module_path = os.path.abspath(os.path.join(home_dir))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model_root_and_name_tuple_list = [(\"vespa_marquet\", \"vespal\"), (\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"bioembeddings_dallago\", \"prottrans_bert_bfd\")]\n",
    "\n",
    "model_root_and_name_tuple_list = [(\"dbnsfp\", \"sift\"), (\"dbnsfp\", \"polyphen2_HVAR\"), (\"dbnsfp\", \"cadd\"), (\"dbnsfp\", \"mvp\"), (\"dbnsfp\", \"metarnn\"), (\"dbnsfp\", \"revel\"),\n",
    "                                  (\"tape_rao\", \"unirep\"), (\"tape_rao\", \"protbert\"), (\"sequnet_dunham\", \"sequnet\"), \n",
    "                                  (\"esm_rives\", \"esm1b_t33_650M_UR50S\"), (\"esm_rives\", \"esm1v_t33_650M_UR90S\"), (\"esm_rives\", \"esm2_t33_650M_UR50D\"),\n",
    "                                  (\"bioembeddings_dallago\", \"plus_rnn\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_bert_bfd\"), (\"bioembeddings_dallago\", \"prottrans_albert_bfd\"),\n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_xlnet_uniref100\"), \n",
    "                                  (\"bioembeddings_dallago\", \"prottrans_t5_bfd\"), (\"bioembeddings_dallago\", \"prottrans_t5_uniref50\"), (\"bioembeddings_dallago\", \"prottrans_t5_xl_u50\"),\n",
    "                                  (\"vespa_marquet\", \"vespal\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CDD Conservation dataframe for PMD sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3689094, 36)\n",
      "NPid       qPos\n",
      "A000006_2  45      1\n",
      "A941326_1  366     1\n",
      "           479     1\n",
      "           480     1\n",
      "           481     1\n",
      "                  ..\n",
      "A900216_1  489     1\n",
      "           490     1\n",
      "           491     1\n",
      "           492     1\n",
      "R950334_1  84      1\n",
      "Length: 3687781, dtype: int64\n",
      "(3687781, 36)\n",
      "Index(['qNo', 'NPid', 'accession', 'que', 'sub', 'ali', 'qPos', 'sPos', 'aPos',\n",
      "       'qPos_', 'sPos_', 'bitscore', 'A', 'D', 'C', 'E', 'F', 'G', 'H', 'I',\n",
      "       'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', 'cons',\n",
      "       'entropy', 'qcov', 'CScore'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qNo</th>\n",
       "      <th>NPid</th>\n",
       "      <th>accession</th>\n",
       "      <th>que</th>\n",
       "      <th>sub</th>\n",
       "      <th>ali</th>\n",
       "      <th>qPos</th>\n",
       "      <th>sPos</th>\n",
       "      <th>aPos</th>\n",
       "      <th>qPos_</th>\n",
       "      <th>...</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>cons</th>\n",
       "      <th>entropy</th>\n",
       "      <th>qcov</th>\n",
       "      <th>CScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A000006_2</td>\n",
       "      <td>cd15353</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>|</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>F</td>\n",
       "      <td>0.164362</td>\n",
       "      <td>0.810241</td>\n",
       "      <td>0.835638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A000006_2</td>\n",
       "      <td>cd15353</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>|</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>V</td>\n",
       "      <td>0.164362</td>\n",
       "      <td>0.810241</td>\n",
       "      <td>0.835638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A000006_2</td>\n",
       "      <td>cd15353</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>|</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.810241</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>A000006_2</td>\n",
       "      <td>cd15353</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>|</td>\n",
       "      <td>48</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P</td>\n",
       "      <td>0.164362</td>\n",
       "      <td>0.810241</td>\n",
       "      <td>0.835638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A000006_2</td>\n",
       "      <td>cd15353</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>|</td>\n",
       "      <td>49</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>E</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.810241</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689089</th>\n",
       "      <td>973</td>\n",
       "      <td>A994631_1</td>\n",
       "      <td>smart00808</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>|</td>\n",
       "      <td>1100</td>\n",
       "      <td>122</td>\n",
       "      <td>123</td>\n",
       "      <td>1100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>D</td>\n",
       "      <td>0.236971</td>\n",
       "      <td>0.483696</td>\n",
       "      <td>0.763029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689090</th>\n",
       "      <td>973</td>\n",
       "      <td>A994631_1</td>\n",
       "      <td>smart00808</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td></td>\n",
       "      <td>1101</td>\n",
       "      <td>123</td>\n",
       "      <td>124</td>\n",
       "      <td>1101</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>X</td>\n",
       "      <td>0.370150</td>\n",
       "      <td>0.483696</td>\n",
       "      <td>0.629850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689091</th>\n",
       "      <td>973</td>\n",
       "      <td>A994631_1</td>\n",
       "      <td>smart00808</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>|</td>\n",
       "      <td>1102</td>\n",
       "      <td>124</td>\n",
       "      <td>125</td>\n",
       "      <td>1102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>V</td>\n",
       "      <td>0.094214</td>\n",
       "      <td>0.483696</td>\n",
       "      <td>0.905786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689092</th>\n",
       "      <td>973</td>\n",
       "      <td>A994631_1</td>\n",
       "      <td>smart00808</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>|</td>\n",
       "      <td>1103</td>\n",
       "      <td>125</td>\n",
       "      <td>126</td>\n",
       "      <td>1103</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.274916</td>\n",
       "      <td>0.483696</td>\n",
       "      <td>0.725084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689093</th>\n",
       "      <td>973</td>\n",
       "      <td>A994631_1</td>\n",
       "      <td>smart00808</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>|</td>\n",
       "      <td>1104</td>\n",
       "      <td>126</td>\n",
       "      <td>127</td>\n",
       "      <td>1104</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>R</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.483696</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3687781 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         qNo       NPid   accession que sub ali  qPos  sPos  aPos  qPos_  ...  \\\n",
       "0          1  A000006_2     cd15353   F   F   |    45     1     1     45  ...   \n",
       "1          1  A000006_2     cd15353   V   V   |    46     2     2     46  ...   \n",
       "2          1  A000006_2     cd15353   S   S   |    47     3     3     47  ...   \n",
       "3          1  A000006_2     cd15353   P   P   |    48     4     4     48  ...   \n",
       "4          1  A000006_2     cd15353   E   E   |    49     5     5     49  ...   \n",
       "...      ...        ...         ...  ..  ..  ..   ...   ...   ...    ...  ...   \n",
       "3689089  973  A994631_1  smart00808   D   D   |  1100   122   123   1100  ...   \n",
       "3689090  973  A994631_1  smart00808   I   V      1101   123   124   1101  ...   \n",
       "3689091  973  A994631_1  smart00808   V   V   |  1102   124   125   1102  ...   \n",
       "3689092  973  A994631_1  smart00808   Q   Q   |  1103   125   126   1103  ...   \n",
       "3689093  973  A994631_1  smart00808   R   R   |  1104   126   127   1104  ...   \n",
       "\n",
       "                R    S    T         V    W    Y  cons   entropy      qcov  \\\n",
       "0        0.000000  0.0  0.0  0.000000  0.0  0.0     F  0.164362  0.810241   \n",
       "1        0.000000  0.0  0.0  0.800000  0.0  0.0     V  0.164362  0.810241   \n",
       "2        0.000000  1.0  0.0  0.000000  0.0  0.0     S  0.000000  0.810241   \n",
       "3        0.000000  0.0  0.2  0.000000  0.0  0.0     P  0.164362  0.810241   \n",
       "4        0.000000  0.0  0.0  0.000000  0.0  0.0     E  0.000000  0.810241   \n",
       "...           ...  ...  ...       ...  ...  ...   ...       ...       ...   \n",
       "3689089  0.083333  0.0  0.0  0.000000  0.0  0.0     D  0.236971  0.483696   \n",
       "3689090  0.000000  0.0  0.0  0.500000  0.0  0.0     X  0.370150  0.483696   \n",
       "3689091  0.000000  0.0  0.0  0.916667  0.0  0.0     V  0.094214  0.483696   \n",
       "3689092  0.000000  0.0  0.0  0.000000  0.0  0.0     Q  0.274916  0.483696   \n",
       "3689093  1.000000  0.0  0.0  0.000000  0.0  0.0     R  0.000000  0.483696   \n",
       "\n",
       "           CScore  \n",
       "0        0.835638  \n",
       "1        0.835638  \n",
       "2        1.000000  \n",
       "3        0.835638  \n",
       "4        1.000000  \n",
       "...           ...  \n",
       "3689089  0.763029  \n",
       "3689090  0.629850  \n",
       "3689091  0.905786  \n",
       "3689092  0.725084  \n",
       "3689093  1.000000  \n",
       "\n",
       "[3687781 rows x 36 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conservation_df = pd.read_csv(home_dir+\"data/cdd_conservation/cdd_conservationTable_pmdSequences.csv.gz\", compression='gzip', comment='#')\n",
    "print(conservation_df.shape)\n",
    "conservation_df = conservation_df.drop_duplicates([\"NPid\", \"qPos\"], keep=\"first\") #qPos is 1-indexed\n",
    "print(conservation_df[[\"NPid\", \"qPos\"]].value_counts())\n",
    "print(conservation_df.shape)\n",
    "print(conservation_df.columns)\n",
    "conservation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_conservation_df(df:pd.DataFrame):\n",
    "    # merge conservation_df with the input df\n",
    "    temp_conservation_df = conservation_df[[\"NPid\", \"qPos\", \"CScore\"]].drop_duplicates(keep=\"first\")\n",
    "    merged_df = pd.merge(df, temp_conservation_df, how=\"left\", left_on=[\"pmd_nr_id\", \"prot_pos\"], right_on=[\"NPid\", \"qPos\"]) # prot-pos is 1-indexed\n",
    "    columns = list(df.columns)\n",
    "    columns.append(\"CScore\")\n",
    "    merged_df = merged_df[columns]\n",
    "    merged_df = merged_df.rename(columns={\"CScore\": \"conservation_pred\"})\n",
    "    return merged_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging protein mutation dataset (PMD) predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading Protein Mutation Dataset (PMD) ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/ashehu/akabir4/projects/variant_effect_analysis/models/aa_common/data_loader.py:49: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pmd_df = pd.read_csv(home_dir+\"models/aa_common/datasets_pmd_analysis/pmd_data.tsv\", sep=\"\\t\") # PMD: protein mutation dataset\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65901, 32)\n",
      "Index(['mut_id', 'pmd_id', 'nr', 'crossref', 'uniprot_id', 'ensembl_id',\n",
      "       'taxid', 'protein', 'mut_PMD', 'mut_real', 'wt', 'mut', 'prot_pos',\n",
      "       'function_summarized', 'functional_effect', 'function', 'seq', 'snp_id',\n",
      "       'mrna_acc', 'mrna_ver', 'mrna_pos', 'allele', 'protein_acc',\n",
      "       'protein_ver', 'verified', 'chrom', 'chrom_pos', 'variation',\n",
      "       'variant_type', 'ref_allele', 'alt_allele', 'pmd_nr_id'],\n",
      "      dtype='object')\n",
      "\n",
      "Log: excluding variants corresponding to proteins having seq-len>1022 ...\n",
      "\n",
      "Log: Loading combined fasta iterator ...\n",
      "#-protein sequences (seq-len<=1022): 10715\n",
      "(60306, 32)\n"
     ]
    }
   ],
   "source": [
    "from models.aa_common.data_loader import get_pmd_dataset\n",
    "pmd_variants_df = get_pmd_dataset(home_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbnsfp sift (60306, 33)\n",
      "(60306, 33)\n",
      "dbnsfp polyphen2_HVAR (60306, 33)\n",
      "(60306, 34)\n",
      "dbnsfp cadd (60306, 33)\n",
      "(60306, 35)\n",
      "dbnsfp mvp (60306, 33)\n",
      "(60306, 36)\n",
      "dbnsfp metarnn (60306, 33)\n",
      "(60306, 37)\n",
      "dbnsfp revel (60306, 33)\n",
      "(60306, 38)\n",
      "tape_rao unirep (60306, 33)\n",
      "(60306, 39)\n",
      "tape_rao protbert (60306, 33)\n",
      "(60306, 40)\n",
      "sequnet_dunham sequnet (60225, 33)\n",
      "(60306, 41)\n",
      "esm_rives esm1b_t33_650M_UR50S (60306, 33)\n",
      "(60306, 42)\n",
      "esm_rives esm1v_t33_650M_UR90S (60306, 33)\n",
      "(60306, 43)\n",
      "esm_rives esm2_t33_650M_UR50D (60306, 33)\n",
      "(60306, 44)\n",
      "bioembeddings_dallago plus_rnn (60306, 33)\n",
      "(60306, 45)\n",
      "bioembeddings_dallago prottrans_bert_bfd (60306, 33)\n",
      "(60306, 46)\n",
      "bioembeddings_dallago prottrans_albert_bfd (60306, 33)\n",
      "(60306, 47)\n",
      "bioembeddings_dallago prottrans_xlnet_uniref100 (60306, 33)\n",
      "(60306, 48)\n",
      "bioembeddings_dallago prottrans_t5_bfd (60306, 33)\n",
      "(60306, 49)\n",
      "bioembeddings_dallago prottrans_t5_uniref50 (60306, 33)\n",
      "(60306, 50)\n",
      "bioembeddings_dallago prottrans_t5_xl_u50 (60306, 33)\n",
      "(60306, 51)\n",
      "vespa_marquet vespa (60306, 33)\n",
      "(60306, 52)\n",
      "conservation_score\n",
      "(60306, 53)\n"
     ]
    }
   ],
   "source": [
    "task = \"pmd\"\n",
    "merged_df = pmd_variants_df.copy(deep=True)\n",
    "merge_col_list = list(pmd_variants_df.columns)\n",
    "\n",
    "for i, (model_root, model_name) in enumerate(model_root_and_name_tuple_list):\n",
    "    if model_name==\"vespal\": model_name=\"vespa\"\n",
    "    models_pred_df = pd.read_csv(home_dir+f\"models/{model_root}/outputs/{model_name}/{task}/preds_{model_name}.tsv\", sep=\"\\t\")\n",
    "    models_pred_df = models_pred_df.drop_duplicates(keep=\"first\")\n",
    "    print(model_root, model_name, models_pred_df.shape)\n",
    "    merged_df = pd.merge(left=merged_df, right=models_pred_df, how=\"left\", on=merge_col_list)\n",
    "    \n",
    "    merged_df = merged_df.rename(columns={\"pred\": f\"{model_name}_pred\"})\n",
    "    print(merged_df.shape)\n",
    "    # if i==2: break\n",
    "\n",
    "merged_df = merge_conservation_df(merged_df)  \n",
    "print(\"conservation_score\")  \n",
    "print(merged_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_34019/3390127692.py:1: DtypeWarning: Columns (11,12,13,16,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  geno2func = pd.read_csv(home_dir+f\"data/SNPdbe/geno2func.tsv\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1748930, 23)\n",
      "Index(['mut_id', 'md5', 'wt', 'pos', 'mt', 'in_dbSNP', 'in_SP', 'in_PMD',\n",
      "       'in_1KG', 'SNAP_status', 'SNAP_bin', 'SNAP_score', 'SNAP_ri',\n",
      "       'SNAP_acc', 'SIFT_bin', 'SIFT_score', 'PERC_wt', 'PERC_mt', 'PSSM_wt',\n",
      "       'PSSM_mt', 'PSIC_wt', 'PSIC_mt', 'pph2'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "geno2func = pd.read_csv(home_dir+f\"data/SNPdbe/geno2func.tsv\", sep=\"\\t\")\n",
    "geno2func.replace(\"\\\\N\", None, inplace=True)\n",
    "print(geno2func.shape)\n",
    "print(geno2func.columns)\n",
    "geno2func = geno2func[['mut_id', 'SIFT_score']] # if needed, add more columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60306, 54)\n",
      "Index(['mut_id', 'pmd_id', 'nr', 'crossref', 'uniprot_id', 'ensembl_id',\n",
      "       'taxid', 'protein', 'mut_PMD', 'mut_real', 'wt', 'mut', 'prot_pos',\n",
      "       'function_summarized', 'functional_effect', 'function', 'seq', 'snp_id',\n",
      "       'mrna_acc', 'mrna_ver', 'mrna_pos', 'allele', 'protein_acc',\n",
      "       'protein_ver', 'verified', 'chrom', 'chrom_pos', 'variation',\n",
      "       'variant_type', 'ref_allele', 'alt_allele', 'pmd_nr_id', 'sift_pred',\n",
      "       'polyphen2_HVAR_pred', 'cadd_pred', 'mvp_pred', 'metarnn_pred',\n",
      "       'revel_pred', 'unirep_pred', 'protbert_pred', 'sequnet_pred',\n",
      "       'esm1b_t33_650M_UR50S_pred', 'esm1v_t33_650M_UR90S_pred',\n",
      "       'esm2_t33_650M_UR50D_pred', 'plus_rnn_pred', 'prottrans_bert_bfd_pred',\n",
      "       'prottrans_albert_bfd_pred', 'prottrans_xlnet_uniref100_pred',\n",
      "       'prottrans_t5_bfd_pred', 'prottrans_t5_uniref50_pred',\n",
      "       'prottrans_t5_xl_u50_pred', 'vespa_pred', 'conservation_pred',\n",
      "       'sift_snpdbe_pred'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "merged_df = pd.merge(left=merged_df, right=geno2func, how=\"left\", on=[\"mut_id\"])\n",
    "merged_df = merged_df.rename(columns={\"SIFT_score\": \"sift_snpdbe_pred\"})\n",
    "print(merged_df.shape)\n",
    "print(merged_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "supervised_method_cols = ['sift_snpdbe_pred', 'polyphen2_HVAR_pred', 'cadd_pred', 'mvp_pred', 'metarnn_pred', 'revel_pred', 'sequnet_pred', 'vespa_pred']\n",
    "def get_n_supervised_methods_havings_prediction(row):\n",
    "    # print(row)\n",
    "    n = 0\n",
    "    for col in supervised_method_cols:\n",
    "        # print(pd.isna(row[col]))\n",
    "        if not pd.isna(row[col]):\n",
    "            n += 1\n",
    "    # print(n)\n",
    "    return n\n",
    "\n",
    "for i in range(merged_df.shape[0]):\n",
    "    merged_df.loc[i, \"n_methods_having_preds\"] = get_n_supervised_methods_havings_prediction(merged_df.loc[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv(home_dir+f\"models/aa_common/merged_predictions/{task}.tsv\", sep=\"\\t\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
