{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "home_dir = \"../\"\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "from utils.pandas_extented_filters import filter_remove_null_nan_empty_entries\n",
    "from utils.column_names_reader import get_col_names_from_gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging the downloaded dataset\n",
    "def clean_df(data_filepath):\n",
    "    df = pd.read_csv(data_filepath, delim_whitespace=False, sep=\"\\t\")\n",
    "    df = df.drop_duplicates(subset=\"snp_id\", keep=\"first\")\n",
    "    return df\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_0.txt\"\n",
    "df_0 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_1.txt\"\n",
    "df_1 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_2.txt\"\n",
    "df_2 = clean_df(data_filepath)\n",
    "\n",
    "data_filepath = home_dir+\"data/dbsnp/search_results/dbsnp_HumanMissenseALFAVariants_3.txt\"\n",
    "df_3 = clean_df(data_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5140350, 14)\n",
      "Number of NAN rows removed: 151569\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 0\n",
      "(4988781, 14)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# removing duplicate data points and variants with no protein variants\n",
    "df = pd.concat([df_0, df_1, df_2, df_3], ignore_index=True)\n",
    "df = df.drop_duplicates(subset=\"snp_id\", keep=\"first\")\n",
    "print(df.shape)\n",
    "df = filter_remove_null_nan_empty_entries(df, a_col_name=\"variations\") # NNE: null, nan or empty variants\n",
    "print(df.shape)\n",
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4988781\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "genes                                                                  \n",
       "TTN:7273,TTN-AS1:100506866                                                 9381\n",
       "MUC16:94025                                                                8341\n",
       "MUC4:4585                                                                  5782\n",
       "TTN:7273                                                                   5151\n",
       "AHNAK2:113146                                                              4726\n",
       "                                                                           ... \n",
       "CASP5:838,LOC124902742:124902742                                              1\n",
       "SLX1B:79008,BOLA2:552900,SLX1B-SULT1A4:100526831,BOLA2-SMG1P6:107282092       1\n",
       "RASSF7:8045,LMNTD2:256329,LMNTD2-AS1:692247                                   1\n",
       "CASP4:837,LOC124902813:124902813                                              1\n",
       "EDA:1896,MIR676:100500887                                                     1\n",
       "Length: 27296, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adding rs prefix to dbSNP-ids\n",
    "initially_cleaned_df = df.copy()\n",
    "initially_cleaned_df[\"snp_id\"] = initially_cleaned_df[\"snp_id\"].apply(lambda x: \"rs\"+str(x))\n",
    "dbsnp_ids = initially_cleaned_df[\"snp_id\"].unique().tolist()\n",
    "print(len(dbsnp_ids))\n",
    "initially_cleaned_df[[\"genes\"]].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading and filtering ALFA data as it is big"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do not run this unless you changed something before, it takes 1 hour\n",
    "# better use the following block\n",
    "# Variants reported on GRCh38, readme: https://www.ncbi.nlm.nih.gov/snp/docs/gsr/ftp_help/\n",
    "from utils.column_names_reader import get_col_names_from_gzip\n",
    "filepath = home_dir+\"data/ALFA_population_freq/freq.vcf.gz\"\n",
    "col_names = get_col_names_from_gzip(filepath, \"#CHROM\")\n",
    "alfa_df_iterator = pd.read_csv(filepath, compression='gzip', chunksize=1000000, comment=\"#\", names=col_names, delim_whitespace=False, sep=\"\\t\")\n",
    "# alfa_df_iterator.__next__() # first df \n",
    "\n",
    "chunk_list = []\n",
    "for i, alfa_df in enumerate(alfa_df_iterator):\n",
    "    chunk = alfa_df[alfa_df[\"ID\"].isin(dbsnp_ids)]\n",
    "    chunk_list.append(chunk)\n",
    "    print(i, chunk.shape)\n",
    "    # if i==0: break\n",
    "\n",
    "filtered_alfa_df = pd.concat(chunk_list)  \n",
    "print(filtered_alfa_df.shape)\n",
    "print(filtered_alfa_df.columns)\n",
    "filtered_alfa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT',\n",
      "       'SAMN10492695', 'SAMN10492696', 'SAMN10492697', 'SAMN10492698',\n",
      "       'SAMN10492699', 'SAMN10492700', 'SAMN10492701', 'SAMN10492702',\n",
      "       'SAMN11605645', 'SAMN10492703', 'SAMN10492704', 'SAMN10492705'],\n",
      "      dtype='object')\n",
      "(4988568, 21)\n",
      "(4988568,)\n"
     ]
    }
   ],
   "source": [
    "filtered_alfa_filepath = home_dir+\"data/temp_popu_freq_data/alfa_filtered.tsv\"\n",
    "# filtered_alfa_df.to_csv(filtered_alfa_filepath, index=False, sep=\"\\t\")\n",
    "filtered_alfa_df = pd.read_csv(filtered_alfa_filepath, sep=\"\\t\")\n",
    "print(filtered_alfa_df.columns)\n",
    "print(filtered_alfa_df.shape)\n",
    "print(filtered_alfa_df[\"ID\"].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>SAMN10492695</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492697</th>\n",
       "      <th>SAMN10492698</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69116</td>\n",
       "      <td>rs757299236</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:1</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69134</td>\n",
       "      <td>rs781394307</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:9</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:5</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69149</td>\n",
       "      <td>rs1458521218</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:3</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:1</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69184</td>\n",
       "      <td>rs1231643173</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:0</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:0</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>69200</td>\n",
       "      <td>rs1453226491</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AN:AC</td>\n",
       "      <td>7618:0</td>\n",
       "      <td>...</td>\n",
       "      <td>84:0</td>\n",
       "      <td>2708:5</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>24:0</td>\n",
       "      <td>94:0</td>\n",
       "      <td>470:0</td>\n",
       "      <td>2816:5</td>\n",
       "      <td>108:0</td>\n",
       "      <td>11862:5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM    POS            ID REF ALT QUAL FILTER INFO FORMAT  \\\n",
       "0  NC_000001.11  69116   rs757299236   G   A    .      .    .  AN:AC   \n",
       "1  NC_000001.11  69134   rs781394307   A   G    .      .    .  AN:AC   \n",
       "2  NC_000001.11  69149  rs1458521218   T   A    .      .    .  AN:AC   \n",
       "3  NC_000001.11  69184  rs1231643173   G   A    .      .    .  AN:AC   \n",
       "4  NC_000001.11  69200  rs1453226491   T   C    .      .    .  AN:AC   \n",
       "\n",
       "  SAMN10492695  ... SAMN10492697 SAMN10492698 SAMN10492699 SAMN10492700  \\\n",
       "0       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "1       7618:9  ...         84:0       2708:0        146:0        610:5   \n",
       "2       7618:3  ...         84:0       2708:0        146:1        610:0   \n",
       "3       7618:0  ...         84:0       2708:0        146:0        610:0   \n",
       "4       7618:0  ...         84:0       2708:5        146:0        610:0   \n",
       "\n",
       "  SAMN10492701 SAMN10492702 SAMN11605645 SAMN10492703 SAMN10492704  \\\n",
       "0         24:0         94:0        470:1       2816:0        108:0   \n",
       "1         24:0         94:0        470:0       2816:0        108:0   \n",
       "2         24:0         94:0        470:0       2816:0        108:0   \n",
       "3         24:0         94:0        470:0       2816:0        108:0   \n",
       "4         24:0         94:0        470:0       2816:5        108:0   \n",
       "\n",
       "  SAMN10492705  \n",
       "0      11862:1  \n",
       "1     11862:14  \n",
       "2      11862:4  \n",
       "3      11862:0  \n",
       "4      11862:5  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_alfa_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4988568, 35)\n",
      "4988568\n"
     ]
    }
   ],
   "source": [
    "snps_with_alfa_df = pd.merge(left=initially_cleaned_df, right=filtered_alfa_df, left_on=\"snp_id\", right_on=\"ID\", how=\"inner\")\n",
    "print(snps_with_alfa_df.shape)\n",
    "print(snps_with_alfa_df[\"snp_id\"].unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3275495, 35)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_all_mt_variants_zero_popu_count(row):\n",
    "    # print(row)\n",
    "    # raise\n",
    "    mt_popu_counts = row.SAMN10492705.split(\":\")[1].split(\",\")\n",
    "    for mt_popu_count in mt_popu_counts:\n",
    "        mt_popu_count = int(mt_popu_count)\n",
    "        if mt_popu_count != 0:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "snps_with_allmtvariants_popucount_zero_df = snps_with_alfa_df[snps_with_alfa_df.apply(is_all_mt_variants_zero_popu_count, axis=1)]\n",
    "snps_with_allmtvariants_popucount_zero_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1713073, 35)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_with_atleast1mtvariants_popucount_nonzero_df = snps_with_alfa_df[~snps_with_alfa_df.apply(is_all_mt_variants_zero_popu_count, axis=1)]\n",
    "snps_with_atleast1mtvariants_popucount_nonzero_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_with_allmtvariants_popucount_zero_df.to_csv(home_dir+\"data/temp_popufreq_data/snps_with_allmtvariants_popucount_zero.tsv\", index=False, sep=\"\\t\")\n",
    "snps_with_atleast1mtvariants_popucount_nonzero_df.to_csv(home_dir+\"data/temp_popufreq_data/snps_with_atleast1mtvariants_popucount_nonzero.tsv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping to MANE proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19120, 14)\n",
      "Index(['#NCBI_GeneID', 'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name',\n",
      "       'RefSeq_nuc', 'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot',\n",
      "       'MANE_status', 'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand'],\n",
      "      dtype='object')\n",
      "MANE Select           19062\n",
      "MANE Plus Clinical       58\n",
      "Name: MANE_status, dtype: int64\n",
      "CACNA1D         2\n",
      "CACNA1C         2\n",
      "DYNC2H1         2\n",
      "SLC39A14        2\n",
      "CLPB            2\n",
      "               ..\n",
      "ZNF432          1\n",
      "SAFB2           1\n",
      "DZIP3           1\n",
      "MARF1           1\n",
      "LOC122539214    1\n",
      "Name: symbol, Length: 19062, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#NCBI_GeneID</th>\n",
       "      <th>Ensembl_Gene</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>GeneID:776</td>\n",
       "      <td>ENSG00000157388.20</td>\n",
       "      <td>HGNC:1391</td>\n",
       "      <td>CACNA1D</td>\n",
       "      <td>calcium voltage-gated channel subunit alpha1 D</td>\n",
       "      <td>NM_000720.4</td>\n",
       "      <td>NP_000711.1</td>\n",
       "      <td>ENST00000288139.11</td>\n",
       "      <td>ENSP00000288139.3</td>\n",
       "      <td>MANE Plus Clinical</td>\n",
       "      <td>3</td>\n",
       "      <td>53494611</td>\n",
       "      <td>53813733</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>GeneID:776</td>\n",
       "      <td>ENSG00000157388.20</td>\n",
       "      <td>HGNC:1391</td>\n",
       "      <td>CACNA1D</td>\n",
       "      <td>calcium voltage-gated channel subunit alpha1 D</td>\n",
       "      <td>NM_001128840.3</td>\n",
       "      <td>NP_001122312.1</td>\n",
       "      <td>ENST00000350061.11</td>\n",
       "      <td>ENSP00000288133.5</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>3</td>\n",
       "      <td>53494611</td>\n",
       "      <td>53813733</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #NCBI_GeneID        Ensembl_Gene    HGNC_ID   symbol  \\\n",
       "542   GeneID:776  ENSG00000157388.20  HGNC:1391  CACNA1D   \n",
       "543   GeneID:776  ENSG00000157388.20  HGNC:1391  CACNA1D   \n",
       "\n",
       "                                               name      RefSeq_nuc  \\\n",
       "542  calcium voltage-gated channel subunit alpha1 D     NM_000720.4   \n",
       "543  calcium voltage-gated channel subunit alpha1 D  NM_001128840.3   \n",
       "\n",
       "        RefSeq_prot         Ensembl_nuc       Ensembl_prot  \\\n",
       "542     NP_000711.1  ENST00000288139.11  ENSP00000288139.3   \n",
       "543  NP_001122312.1  ENST00000350061.11  ENSP00000288133.5   \n",
       "\n",
       "            MANE_status GRCh38_chr  chr_start   chr_end chr_strand  \n",
       "542  MANE Plus Clinical          3   53494611  53813733          +  \n",
       "543         MANE Select          3   53494611  53813733          +  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# loading MANE data\n",
    "mane_df = pd.read_csv(home_dir+\"data/refseq/MANE.GRCh38.v1.0.summary.txt.gz\", compression='gzip', delim_whitespace=False, sep=\"\\t\") # MANE refseq-protein mapping\n",
    "print(mane_df.shape)\n",
    "print(mane_df.columns)\n",
    "\n",
    "print(mane_df[\"MANE_status\"].value_counts())\n",
    "print(mane_df[\"symbol\"].value_counts()) # some gene can be mapped to multiple proteins because of MANE select and Plus clinical.\n",
    "mane_df[mane_df[\"symbol\"]==\"CACNA1D\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100000\n",
      "200000\n",
      "300000\n",
      "400000\n",
      "500000\n",
      "600000\n",
      "700000\n",
      "800000\n",
      "900000\n",
      "1000000\n",
      "1100000\n",
      "1200000\n",
      "1300000\n",
      "1400000\n",
      "1500000\n",
      "1600000\n",
      "1700000\n",
      "(2055157, 37)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'CHROM', 'POS', 'ID', 'REF', 'ALT',\n",
      "       'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMN10492695', 'SAMN10492696',\n",
      "       'SAMN10492697', 'SAMN10492698', 'SAMN10492699', 'SAMN10492700',\n",
      "       'SAMN10492701', 'SAMN10492702', 'SAMN11605645', 'SAMN10492703',\n",
      "       'SAMN10492704', 'SAMN10492705', 'gene_symbol', 'gene_id'],\n",
      "      dtype='object')\n",
      "(1713073,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMN10492699</th>\n",
       "      <th>SAMN10492700</th>\n",
       "      <th>SAMN10492701</th>\n",
       "      <th>SAMN10492702</th>\n",
       "      <th>SAMN11605645</th>\n",
       "      <th>SAMN10492703</th>\n",
       "      <th>SAMN10492704</th>\n",
       "      <th>SAMN10492705</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1491003598</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269218</td>\n",
       "      <td>NC_000001.11:205269217:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>30:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>2276:1</td>\n",
       "      <td>594:0</td>\n",
       "      <td>56:0</td>\n",
       "      <td>8988:1</td>\n",
       "      <td>TMCC2</td>\n",
       "      <td>9911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1491003245</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2:105881549</td>\n",
       "      <td>NC_000002.12:105881548:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 20:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,intron_variant,missens...</td>\n",
       "      <td>...</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:1</td>\n",
       "      <td>NCK2</td>\n",
       "      <td>8440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1491001877</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11:4385428</td>\n",
       "      <td>NC_000011.10:4385427:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 13:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>26:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>496:0</td>\n",
       "      <td>2898:0</td>\n",
       "      <td>112:0</td>\n",
       "      <td>14050:1</td>\n",
       "      <td>TRIM21</td>\n",
       "      <td>6737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1491001822</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19:54576083</td>\n",
       "      <td>NC_000019.10:54576082:G:A,NC_000019.10:5457608...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/17 02:29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>...</td>\n",
       "      <td>146:0,0</td>\n",
       "      <td>610:0,0</td>\n",
       "      <td>56:0,0</td>\n",
       "      <td>98:0,0</td>\n",
       "      <td>2772:0,0</td>\n",
       "      <td>3492:0,0</td>\n",
       "      <td>168:0,0</td>\n",
       "      <td>23038:1,0</td>\n",
       "      <td>LILRA2</td>\n",
       "      <td>11027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1491000935</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16:21980608</td>\n",
       "      <td>NC_000016.10:21980607:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/16 22:57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>...</td>\n",
       "      <td>0:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>30:0</td>\n",
       "      <td>0:0</td>\n",
       "      <td>2276:0</td>\n",
       "      <td>594:0</td>\n",
       "      <td>56:0</td>\n",
       "      <td>8988:1</td>\n",
       "      <td>UQCRC2</td>\n",
       "      <td>7385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055152</th>\n",
       "      <td>rs334</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>NC_000011.10:5227001:T:A,NC_000011.10:5227001:...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/16 14:17</td>\n",
       "      <td>protective,other,pathogenic,benign,likely-beni...</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>498:11</td>\n",
       "      <td>626:0</td>\n",
       "      <td>56:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>7046:12</td>\n",
       "      <td>3340:10</td>\n",
       "      <td>168:0</td>\n",
       "      <td>44606:35</td>\n",
       "      <td>HBB</td>\n",
       "      <td>3043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055153</th>\n",
       "      <td>rs300</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19959376</td>\n",
       "      <td>NC_000008.11:19959375:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>benign,benign-likely-benign</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>1318:3</td>\n",
       "      <td>2530:2</td>\n",
       "      <td>1952:0</td>\n",
       "      <td>360:0</td>\n",
       "      <td>20152:9</td>\n",
       "      <td>8148:71</td>\n",
       "      <td>6814:0</td>\n",
       "      <td>278238:92</td>\n",
       "      <td>LPL</td>\n",
       "      <td>4023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055154</th>\n",
       "      <td>rs298</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19959349</td>\n",
       "      <td>NC_000008.11:19959348:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>146:0</td>\n",
       "      <td>610:0</td>\n",
       "      <td>56:0</td>\n",
       "      <td>98:0</td>\n",
       "      <td>3052:0</td>\n",
       "      <td>4606:1</td>\n",
       "      <td>168:0</td>\n",
       "      <td>30584:3</td>\n",
       "      <td>LPL</td>\n",
       "      <td>4023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055155</th>\n",
       "      <td>rs268</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NC_000008.11:19956017:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 11:06</td>\n",
       "      <td>conflicting-interpretations-of-pathogenicity,u...</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>818:9</td>\n",
       "      <td>1028:11</td>\n",
       "      <td>1850:0</td>\n",
       "      <td>280:2</td>\n",
       "      <td>16346:209</td>\n",
       "      <td>6136:28</td>\n",
       "      <td>6406:0</td>\n",
       "      <td>218762:3560</td>\n",
       "      <td>LPL</td>\n",
       "      <td>4023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055156</th>\n",
       "      <td>rs47</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7:11541494</td>\n",
       "      <td>NC_000007.14:11541493:T:A,NC_000007.14:1154149...</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2000/09/19 17:02</td>\n",
       "      <td>2022/10/14 07:37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>1454:1241</td>\n",
       "      <td>6976:6230</td>\n",
       "      <td>1928:1444</td>\n",
       "      <td>5206:3715</td>\n",
       "      <td>22038:17103</td>\n",
       "      <td>11372:10844</td>\n",
       "      <td>6598:4965</td>\n",
       "      <td>349794:267121</td>\n",
       "      <td>THSD7A</td>\n",
       "      <td>221981</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2055157 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               snp_id           acc       chrpos  \\\n",
       "0        rs1491003598  NC_000001.11  1:205269218   \n",
       "1        rs1491003245  NC_000002.12  2:105881549   \n",
       "2        rs1491001877  NC_000011.10   11:4385428   \n",
       "3        rs1491001822  NC_000019.10  19:54576083   \n",
       "4        rs1491000935  NC_000016.10  16:21980608   \n",
       "...               ...           ...          ...   \n",
       "2055152         rs334  NC_000011.10   11:5227002   \n",
       "2055153         rs300  NC_000008.11   8:19959376   \n",
       "2055154         rs298  NC_000008.11   8:19959349   \n",
       "2055155         rs268  NC_000008.11   8:19956018   \n",
       "2055156          rs47  NC_000007.14   7:11541494   \n",
       "\n",
       "                                                      spdi  tax_id snp_class  \\\n",
       "0                               NC_000001.11:205269217:A:G    9606       snv   \n",
       "1                               NC_000002.12:105881548:C:T    9606       snv   \n",
       "2                                 NC_000011.10:4385427:G:A    9606       snv   \n",
       "3        NC_000019.10:54576082:G:A,NC_000019.10:5457608...    9606       snv   \n",
       "4                                NC_000016.10:21980607:G:A    9606       snv   \n",
       "...                                                    ...     ...       ...   \n",
       "2055152  NC_000011.10:5227001:T:A,NC_000011.10:5227001:...    9606       snv   \n",
       "2055153                          NC_000008.11:19959375:A:G    9606       snv   \n",
       "2055154                          NC_000008.11:19959348:G:A    9606       snv   \n",
       "2055155                          NC_000008.11:19956017:A:G    9606       snv   \n",
       "2055156  NC_000007.14:11541493:T:A,NC_000007.14:1154149...    9606       snv   \n",
       "\n",
       "              create_date       update_date  \\\n",
       "0        2017/11/09 09:55  2022/10/12 18:08   \n",
       "1        2017/11/09 09:55  2022/10/12 20:41   \n",
       "2        2017/11/09 09:55  2022/10/16 13:47   \n",
       "3        2017/11/09 09:55  2022/10/17 02:29   \n",
       "4        2017/11/09 09:55  2022/10/16 22:57   \n",
       "...                   ...               ...   \n",
       "2055152  2000/09/19 17:02  2022/10/16 14:17   \n",
       "2055153  2000/09/19 17:02  2022/10/14 11:06   \n",
       "2055154  2000/09/19 17:02  2022/10/14 11:06   \n",
       "2055155  2000/09/19 17:02  2022/10/14 11:06   \n",
       "2055156  2000/09/19 17:02  2022/10/14 07:37   \n",
       "\n",
       "                                     clinical_significance  \\\n",
       "0                                                      NaN   \n",
       "1                                                      NaN   \n",
       "2                                                      NaN   \n",
       "3                                                      NaN   \n",
       "4                                                      NaN   \n",
       "...                                                    ...   \n",
       "2055152  protective,other,pathogenic,benign,likely-beni...   \n",
       "2055153                        benign,benign-likely-benign   \n",
       "2055154                                                NaN   \n",
       "2055155  conflicting-interpretations-of-pathogenicity,u...   \n",
       "2055156                                                NaN   \n",
       "\n",
       "                                                 fxn_class  ... SAMN10492699  \\\n",
       "0                 coding_sequence_variant,missense_variant  ...          0:0   \n",
       "1        coding_sequence_variant,intron_variant,missens...  ...        146:0   \n",
       "2                 missense_variant,coding_sequence_variant  ...        146:0   \n",
       "3        genic_downstream_transcript_variant,missense_v...  ...      146:0,0   \n",
       "4        genic_downstream_transcript_variant,missense_v...  ...          0:0   \n",
       "...                                                    ...  ...          ...   \n",
       "2055152           missense_variant,coding_sequence_variant  ...       498:11   \n",
       "2055153           missense_variant,coding_sequence_variant  ...       1318:3   \n",
       "2055154           missense_variant,coding_sequence_variant  ...        146:0   \n",
       "2055155           missense_variant,coding_sequence_variant  ...        818:9   \n",
       "2055156           missense_variant,coding_sequence_variant  ...    1454:1241   \n",
       "\n",
       "        SAMN10492700 SAMN10492701 SAMN10492702 SAMN11605645  SAMN10492703  \\\n",
       "0                0:0         30:0          0:0       2276:1         594:0   \n",
       "1              610:0         26:0         98:0        496:0        2898:0   \n",
       "2              610:0         26:0         98:0        496:0        2898:0   \n",
       "3            610:0,0       56:0,0       98:0,0     2772:0,0      3492:0,0   \n",
       "4                0:0         30:0          0:0       2276:0         594:0   \n",
       "...              ...          ...          ...          ...           ...   \n",
       "2055152        626:0         56:0         98:0      7046:12       3340:10   \n",
       "2055153       2530:2       1952:0        360:0      20152:9       8148:71   \n",
       "2055154        610:0         56:0         98:0       3052:0        4606:1   \n",
       "2055155      1028:11       1850:0        280:2    16346:209       6136:28   \n",
       "2055156    6976:6230    1928:1444    5206:3715  22038:17103   11372:10844   \n",
       "\n",
       "        SAMN10492704   SAMN10492705 gene_symbol gene_id  \n",
       "0               56:0         8988:1       TMCC2    9911  \n",
       "1              112:0        14050:1        NCK2    8440  \n",
       "2              112:0        14050:1      TRIM21    6737  \n",
       "3            168:0,0      23038:1,0      LILRA2   11027  \n",
       "4               56:0         8988:1      UQCRC2    7385  \n",
       "...              ...            ...         ...     ...  \n",
       "2055152        168:0       44606:35         HBB    3043  \n",
       "2055153       6814:0      278238:92         LPL    4023  \n",
       "2055154        168:0        30584:3         LPL    4023  \n",
       "2055155       6406:0    218762:3560         LPL    4023  \n",
       "2055156    6598:4965  349794:267121      THSD7A  221981  \n",
       "\n",
       "[2055157 rows x 37 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expanding SNPs from multiple genes to single gene column\n",
    "data = []\n",
    "for i, row in enumerate(snps_with_atleast1mtvariants_popucount_nonzero_df.itertuples(index=False)):\n",
    "    \n",
    "    for gene in row.genes.split(\",\"):\n",
    "        gene_symbol, gene_id = gene.split(\":\")\n",
    "\n",
    "        datam = row._asdict()\n",
    "        datam[\"gene_symbol\"] = gene_symbol\n",
    "        datam[\"gene_id\"] = int(gene_id)\n",
    "        data.append(datam)\n",
    "        # print(datam)\n",
    "\n",
    "    if i%100000==0: print(i)\n",
    "    # if i==5: break\n",
    "\n",
    "snps_with_alfa_expanded_on_genes_df = pd.DataFrame(data) \n",
    "print(snps_with_alfa_expanded_on_genes_df.shape)\n",
    "print(snps_with_alfa_expanded_on_genes_df.columns)\n",
    "print(snps_with_alfa_expanded_on_genes_df[\"snp_id\"].unique().shape)\n",
    "snps_with_alfa_expanded_on_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1864179, 51)\n",
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'CHROM', 'POS', 'ID', 'REF', 'ALT',\n",
      "       'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMN10492695', 'SAMN10492696',\n",
      "       'SAMN10492697', 'SAMN10492698', 'SAMN10492699', 'SAMN10492700',\n",
      "       'SAMN10492701', 'SAMN10492702', 'SAMN11605645', 'SAMN10492703',\n",
      "       'SAMN10492704', 'SAMN10492705', 'gene_symbol', 'gene_id',\n",
      "       '#NCBI_GeneID', 'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name',\n",
      "       'RefSeq_nuc', 'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot',\n",
      "       'MANE_status', 'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand'],\n",
      "      dtype='object')\n",
      "1696911\n",
      "18823\n"
     ]
    }
   ],
   "source": [
    "mane_filtered_snps_df = pd.merge(left=snps_with_alfa_expanded_on_genes_df, right=mane_df, left_on=\"gene_symbol\", right_on=\"symbol\", how=\"inner\")\n",
    "print(mane_filtered_snps_df.shape)\n",
    "print(mane_filtered_snps_df.columns)\n",
    "print(mane_filtered_snps_df[\"snp_id\"].unique().shape[0])\n",
    "print(mane_filtered_snps_df[\"gene_symbol\"].unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1864179, 52)\n",
      "Number of NAN rows removed: 0\n",
      "Number of NULL rows removed: 0\n",
      "Number of empty rows removed: 195936\n",
      "(1668243, 52)\n",
      "1656512\n",
      "18858\n",
      "18800\n"
     ]
    }
   ],
   "source": [
    "def filter_variants_on_canonical_proteins(x):\n",
    "    variations = x[\"variations\"].split(\",\")\n",
    "    new_variations = []\n",
    "    for var in variations:\n",
    "        if var.split(\":\")[0] == x[\"RefSeq_prot\"]:\n",
    "            new_variations.append(var)\n",
    "    return \",\".join(new_variations)\n",
    "\n",
    "mane_filtered_snps_df[\"mane_protein_variants\"] = mane_filtered_snps_df[[\"variations\", \"RefSeq_prot\"]].apply(filter_variants_on_canonical_proteins, axis=1)\n",
    "print(mane_filtered_snps_df.shape)\n",
    "mane_filtered_snps_df = filter_remove_null_nan_empty_entries(mane_filtered_snps_df, a_col_name=\"mane_protein_variants\") # NNE: null, nan or empty variants\n",
    "print(mane_filtered_snps_df.shape)\n",
    "print(mane_filtered_snps_df[\"snp_id\"].unique().shape[0])\n",
    "print(mane_filtered_snps_df[\"RefSeq_prot\"].unique().shape[0])\n",
    "print(mane_filtered_snps_df[\"gene_symbol\"].unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_with_alfa_mane_df = mane_filtered_snps_df.copy()\n",
    "snps_with_alfa_mane_df.to_csv(home_dir+\"data/temp_popufreq_data/snps_with_alfa_mane_raw.tsv\", index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['snp_id', 'acc', 'chrpos', 'spdi', 'tax_id', 'snp_class', 'create_date',\n",
      "       'update_date', 'clinical_significance', 'fxn_class', 'validated',\n",
      "       'genes', 'mafs', 'variations', 'CHROM', 'POS', 'ID', 'REF', 'ALT',\n",
      "       'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMN10492695', 'SAMN10492696',\n",
      "       'SAMN10492697', 'SAMN10492698', 'SAMN10492699', 'SAMN10492700',\n",
      "       'SAMN10492701', 'SAMN10492702', 'SAMN11605645', 'SAMN10492703',\n",
      "       'SAMN10492704', 'SAMN10492705', 'gene_symbol', 'gene_id',\n",
      "       '#NCBI_GeneID', 'Ensembl_Gene', 'HGNC_ID', 'symbol', 'name',\n",
      "       'RefSeq_nuc', 'RefSeq_prot', 'Ensembl_nuc', 'Ensembl_prot',\n",
      "       'MANE_status', 'GRCh38_chr', 'chr_start', 'chr_end', 'chr_strand',\n",
      "       'mane_protein_variants'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>spdi</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>create_date</th>\n",
       "      <th>update_date</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>...</th>\n",
       "      <th>RefSeq_nuc</th>\n",
       "      <th>RefSeq_prot</th>\n",
       "      <th>Ensembl_nuc</th>\n",
       "      <th>Ensembl_prot</th>\n",
       "      <th>MANE_status</th>\n",
       "      <th>GRCh38_chr</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>chr_strand</th>\n",
       "      <th>mane_protein_variants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1491003598</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269218</td>\n",
       "      <td>NC_000001.11:205269217:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Lys339Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1436854664</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205271828</td>\n",
       "      <td>NC_000001.11:205271827:T:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,synonymous_variant,mis...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Cys612Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1431352460</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205241637</td>\n",
       "      <td>NC_000001.11:205241636:A:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,coding_seque...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Met114Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1429019865</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205241752</td>\n",
       "      <td>NC_000001.11:205241751:T:C</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,coding_seque...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Leu152Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1423575191</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269047</td>\n",
       "      <td>NC_000001.11:205269046:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Pro282Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1415408852</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205241616</td>\n",
       "      <td>NC_000001.11:205241615:C:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,missense_var...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Gln107Lys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1410312003</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205241511</td>\n",
       "      <td>NC_000001.11:205241510:C:G</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genic_upstream_transcript_variant,missense_var...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Gln72Glu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1401358469</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205271242</td>\n",
       "      <td>NC_000001.11:205271241:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Ala602Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1375284370</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205269052</td>\n",
       "      <td>NC_000001.11:205269051:C:T</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Arg284Trp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1363717681</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1:205271876</td>\n",
       "      <td>NC_000001.11:205271875:G:A</td>\n",
       "      <td>9606</td>\n",
       "      <td>snv</td>\n",
       "      <td>2017/11/09 09:55</td>\n",
       "      <td>2022/10/12 18:08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>missense_variant,coding_sequence_variant,3_pri...</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_014858.4</td>\n",
       "      <td>NP_055673.2</td>\n",
       "      <td>ENST00000358024.8</td>\n",
       "      <td>ENSP00000350718.3</td>\n",
       "      <td>MANE Select</td>\n",
       "      <td>1</td>\n",
       "      <td>205227946</td>\n",
       "      <td>205273343</td>\n",
       "      <td>+</td>\n",
       "      <td>NP_055673.2:p.Val628Met</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          snp_id           acc       chrpos                        spdi  \\\n",
       "0   rs1491003598  NC_000001.11  1:205269218  NC_000001.11:205269217:A:G   \n",
       "3   rs1436854664  NC_000001.11  1:205271828  NC_000001.11:205271827:T:A   \n",
       "4   rs1431352460  NC_000001.11  1:205241637  NC_000001.11:205241636:A:G   \n",
       "5   rs1429019865  NC_000001.11  1:205241752  NC_000001.11:205241751:T:C   \n",
       "6   rs1423575191  NC_000001.11  1:205269047  NC_000001.11:205269046:C:T   \n",
       "7   rs1415408852  NC_000001.11  1:205241616  NC_000001.11:205241615:C:A   \n",
       "8   rs1410312003  NC_000001.11  1:205241511  NC_000001.11:205241510:C:G   \n",
       "9   rs1401358469  NC_000001.11  1:205271242  NC_000001.11:205271241:C:T   \n",
       "10  rs1375284370  NC_000001.11  1:205269052  NC_000001.11:205269051:C:T   \n",
       "12  rs1363717681  NC_000001.11  1:205271876  NC_000001.11:205271875:G:A   \n",
       "\n",
       "    tax_id snp_class       create_date       update_date  \\\n",
       "0     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "3     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "4     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "5     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "6     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "7     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "8     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "9     9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "10    9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "12    9606       snv  2017/11/09 09:55  2022/10/12 18:08   \n",
       "\n",
       "   clinical_significance                                          fxn_class  \\\n",
       "0                    NaN           coding_sequence_variant,missense_variant   \n",
       "3                    NaN  coding_sequence_variant,synonymous_variant,mis...   \n",
       "4                    NaN  genic_upstream_transcript_variant,coding_seque...   \n",
       "5                    NaN  genic_upstream_transcript_variant,coding_seque...   \n",
       "6                    NaN           missense_variant,coding_sequence_variant   \n",
       "7                    NaN  genic_upstream_transcript_variant,missense_var...   \n",
       "8                    NaN  genic_upstream_transcript_variant,missense_var...   \n",
       "9                    NaN           coding_sequence_variant,missense_variant   \n",
       "10                   NaN           coding_sequence_variant,missense_variant   \n",
       "12                   NaN  missense_variant,coding_sequence_variant,3_pri...   \n",
       "\n",
       "    ...   RefSeq_nuc  RefSeq_prot        Ensembl_nuc       Ensembl_prot  \\\n",
       "0   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "3   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "4   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "5   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "6   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "7   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "8   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "9   ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "10  ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "12  ...  NM_014858.4  NP_055673.2  ENST00000358024.8  ENSP00000350718.3   \n",
       "\n",
       "    MANE_status  GRCh38_chr  chr_start    chr_end chr_strand  \\\n",
       "0   MANE Select           1  205227946  205273343          +   \n",
       "3   MANE Select           1  205227946  205273343          +   \n",
       "4   MANE Select           1  205227946  205273343          +   \n",
       "5   MANE Select           1  205227946  205273343          +   \n",
       "6   MANE Select           1  205227946  205273343          +   \n",
       "7   MANE Select           1  205227946  205273343          +   \n",
       "8   MANE Select           1  205227946  205273343          +   \n",
       "9   MANE Select           1  205227946  205273343          +   \n",
       "10  MANE Select           1  205227946  205273343          +   \n",
       "12  MANE Select           1  205227946  205273343          +   \n",
       "\n",
       "      mane_protein_variants  \n",
       "0   NP_055673.2:p.Lys339Arg  \n",
       "3   NP_055673.2:p.Cys612Ser  \n",
       "4   NP_055673.2:p.Met114Val  \n",
       "5   NP_055673.2:p.Leu152Pro  \n",
       "6   NP_055673.2:p.Pro282Leu  \n",
       "7   NP_055673.2:p.Gln107Lys  \n",
       "8    NP_055673.2:p.Gln72Glu  \n",
       "9   NP_055673.2:p.Ala602Val  \n",
       "10  NP_055673.2:p.Arg284Trp  \n",
       "12  NP_055673.2:p.Val628Met  \n",
       "\n",
       "[10 rows x 52 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(snps_with_alfa_mane_df.columns)\n",
    "snps_with_alfa_mane_df.head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
